// @haxtheweb/haxcms-elements/lib/core/ui/app-hax-user-menu.js
import{LitElement as e,html as t,css as n}from"../../../../../lit/index.js";import{DDDSuper as o}from"../../../../d-d-d/d-d-d.js";import"../../../../simple-icon/lib/simple-icon-lite.js";export class AppHaxUserMenu extends(o(e)){static get tag(){return"app-hax-user-menu"}constructor(){super(),this.isOpen=!1,this.icon="account-circle",this.__boundKeydownHandler=this._handleKeydown.bind(this),this.__boundDocumentPointerDown=this._handleDocumentPointerDown.bind(this),this.__boundDocumentFocusIn=this._handleDocumentFocusIn.bind(this),this.addEventListener("keydown",this.__boundKeydownHandler)}connectedCallback(){super.connectedCallback(),globalThis.document.addEventListener("pointerdown",this.__boundDocumentPointerDown),globalThis.document.addEventListener("focusin",this.__boundDocumentFocusIn)}disconnectedCallback(){globalThis.document.removeEventListener("pointerdown",this.__boundDocumentPointerDown),globalThis.document.removeEventListener("focusin",this.__boundDocumentFocusIn),this.removeEventListener("keydown",this.__boundKeydownHandler),super.disconnectedCallback()}updated(e){e.forEach((e,t)=>{"isOpen"===t&&void 0!==e&&this.dispatchEvent(new CustomEvent("is-open-changed",{bubbles:!0,composed:!0,detail:{value:this.isOpen}}))})}static get properties(){return{isOpen:{type:Boolean,reflect:!0,attribute:"is-open"},icon:{type:String,reflect:!0}}}static get styles(){return[super.styles,n`
        :host {
          font-family: var(--ddd-font-primary, sans-serif);
          text-align: center;
          display: inline-block;
          margin: 0px;
          padding: 0px;
          position: relative;
        }

        .entireComponent {
          height: var(--ddd-spacing-16, 64px);
        }

        .menuToggle {
          cursor: pointer;
          height: var(--ddd-spacing-16, 64px);
        }

        .user-menu {
          display: none;
        }

        .user-menu.open {
          display: block;
          top: var(--ddd-spacing-16, 64px);
          right: 0px;
          position: absolute;
          border: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-slateGray, #666);
          background: light-dark(white, black);
          color: light-dark(black, white);
          border-radius: none;
          box-shadow: var(--ddd-boxShadow-lg);
          min-width: var(--ddd-spacing-30, 200px);
          z-index: 1000;
          overflow: hidden;
        }

        .user-menu.open ::slotted(*) {
          display: block;
          width: 100%;
          margin: 0;
          font-size: var(--ddd-font-size-3xs, 12px);
          text-align: left;
          font-family: var(--ddd-font-primary, sans-serif);
          background: transparent;
          text-decoration: none;
        }

        .user-menu.open .main-menu ::slotted(*:hover),
        .user-menu.open .main-menu ::slotted(*:active),
        .user-menu.open .main-menu ::slotted(*:focus) {
          background-color: light-dark(var(--ddd-theme-default-limestoneGray, #f5f5f5), var(--ddd-theme-default-coalyGray, #333));
        }

        :host([dark]) .user-menu.open .main-menu ::slotted(*:hover),
        :host([dark]) .user-menu.open .main-menu ::slotted(*:active),
        :host([dark]) .user-menu.open .main-menu ::slotted(*:focus),
        body.dark-mode .user-menu.open .main-menu ::slotted(*:hover),
        body.dark-mode .user-menu.open .main-menu ::slotted(*:active),
        body.dark-mode .user-menu.open .main-menu ::slotted(*:focus) {
          color: var(--ddd-theme-default-white, white);
        }

        .user-menu.open .post-menu ::slotted(*:hover),
        .user-menu.open .post-menu ::slotted(*:active),
        .user-menu.open .post-menu ::slotted(*:focus) {
          background-color: var(--ddd-theme-default-original87Pink, #e4007c);
        }

        .user-menu ::slotted(button) {
          cursor: pointer;
        }

        .user-menu ::slotted(*) simple-icon-lite {
          padding-right: var(--ddd-spacing-2, 8px);
        }

        .pre-menu,
        .post-menu {
          border-top: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-limestoneGray, #f5f5f5);
        }

        .pre-menu:first-child,
        .main-menu:first-child {
          border-top: none;
        }

        /* Keyboard focus indicators */
        .user-menu ::slotted(*:focus),
        .user-menu ::slotted(*[tabindex="0"]:focus) {
          outline: var(--ddd-border-sm, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: -2px;
        }
      `]}render(){return t`
      <div class="entireComponent">
        <div
          class="menuToggle"
          part="menuToggle"
          aria-expanded="${this.isOpen}"
          aria-haspopup="menu"
          aria-controls="user-menu-dropdown"
        >
          <slot name="menuButton"
            ><simple-icon-lite
              icon="${this.icon}"
              aria-hidden="true"
            ></simple-icon-lite
          ></slot>
        </div>

        <div
          id="user-menu-dropdown"
          class="user-menu ${this.isOpen?"open":""}"
          role="menu"
          aria-hidden="${!this.isOpen}"
        >
          <div class="pre-menu" role="group" aria-label="Menu Controls">
            <slot name="pre-menu"></slot>
          </div>
          <div class="main-menu" role="group" aria-label="Main Menu Items">
            <slot name="main-menu"></slot>
          </div>
          <div class="post-menu" role="group" aria-label="Account Actions">
            <slot name="post-menu"></slot>
          </div>
        </div>
      </div>
    `}_handleMenuToggleKeydown(e){"Enter"===e.key||" "===e.key?(e.preventDefault(),this._toggleMenu()):"Escape"===e.key&&this.isOpen&&(e.preventDefault(),this._closeMenu())}_handleKeydown(e){if(!this.isOpen)return;const t=this._getMenuItems(),n=this._getCurrentMenuItemIndex(t);switch(e.key){case"Escape":e.preventDefault(),this._closeMenu(),this._focusToggle();break;case"ArrowDown":e.preventDefault(),this._focusNextItem(t,n);break;case"ArrowUp":e.preventDefault(),this._focusPreviousItem(t,n);break;case"Home":e.preventDefault(),this._focusFirstItem(t);break;case"End":e.preventDefault(),this._focusLastItem(t);break;case"Tab":if(0===t.length)break;e.shiftKey?n<=0&&(e.preventDefault(),this._focusLastItem(t)):n!==t.length-1&&-1!==n||(e.preventDefault(),this._focusFirstItem(t))}}_toggleMenu(){this.isOpen=!this.isOpen,this.isOpen&&setTimeout(()=>{const e=this._getMenuItems();e.length>0&&e[0].focus()},0)}_closeMenu(){this.isOpen=!1}_handleDocumentPointerDown(e){if(!this.isOpen)return;-1===(e&&e.composedPath?e.composedPath():[]).indexOf(this)&&this._closeMenu()}_handleDocumentFocusIn(e){if(!this.isOpen)return;if(-1!==(e&&e.composedPath?e.composedPath():[]).indexOf(this))return;const t=this._getMenuItems();t.length>0&&t[0].focus()}_focusToggle(){const e=this.shadowRoot.querySelector(".menuToggle");e&&e.focus()}_getMenuItems(){const e=this.shadowRoot.querySelector(".user-menu");if(!e)return[];const t=e.querySelectorAll("slot"),n=[];return t.forEach(e=>{e.assignedElements().forEach(e=>{const t=e.matches('a, button, [tabindex="0"]')?[e]:e.querySelectorAll('a, button, [tabindex="0"]');n.push(...t)})}),n}_getCurrentMenuItemIndex(e){const t=this.shadowRoot.activeElement||document.activeElement;return e.indexOf(t)}_focusNextItem(e,t){const n=t<e.length-1?t+1:0;e[n]&&e[n].focus()}_focusPreviousItem(e,t){const n=t>0?t-1:e.length-1;e[n]&&e[n].focus()}_focusFirstItem(e){e[0]&&e[0].focus()}_focusLastItem(e){e[e.length-1]&&e[e.length-1].focus()}}customElements.define(AppHaxUserMenu.tag,AppHaxUserMenu);
