// @haxtheweb/haxcms-elements/lib/core/ui/haxcms-site-platform-ui.js
import{html as e,css as t}from"../../../../../lit/index.js";import{autorun as i,toJS as a}from"../../../../../mobx/dist/mobx.esm.js";import{DDD as s}from"../../../../d-d-d/d-d-d.js";import{HAXStore as l}from"../../../../hax-body/lib/hax-store.js";import{store as o}from"../haxcms-site-store.js";import{HAXCMSI18NMixin as r}from"../utils/HAXCMSI18NMixin.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";import"../../../../simple-icon/lib/simple-icon-lite.js";import"../../../../hax-body/lib/hax-text-editor-toolbar.js";const c=[{key:"addPage",label:"Add pages",icon:"hax:add-page",group:"CMS"},{key:"deletePage",label:"Delete pages",icon:"icons:delete",group:"CMS"},{key:"outlineDesigner",label:"Outline designer",icon:"hax:site-map",group:"CMS"},{key:"styleGuide",label:"Style guide",icon:"lrn:palette",group:"CMS"},{key:"insights",label:"Insights",icon:"hax:clipboard-pulse",group:"CMS"},{key:"manifest",label:"Site settings",icon:"hax:home-edit",group:"CMS"},{key:"pageBreak",label:"Edit Page Details",icon:"hax:page-edit",group:"CMS"},{key:"addBlock",label:"Add blocks",icon:"hax:add-brick",group:"HAX"},{key:"contentMap",label:"Page structure (content map)",icon:"hax:newspaper",group:"HAX"},{key:"viewSource",label:"View source",icon:"hax:html-code",group:"HAX"},{key:"onlineSearch",label:"Online search",icon:"hax:add-page",group:"HAX"}];class HAXCMSSitePlatformUI extends(r(s)){static get tag(){return"haxcms-site-platform-ui"}static get properties(){return{audience:{type:String},features:{type:Object},haxBlocks:{type:Object},allowedBlocks:{type:Object},blockFilter:{type:String,attribute:"block-filter"},busy:{type:Boolean,reflect:!0},pageCount:{type:Number,attribute:"page-count"},platformConfig:{type:Object}}}constructor(){super(),this.audience="expert",this.audienceOptions=[{value:"novice",label:"Novice"},{value:"expert",label:"Expert"}],this.features={},this.haxBlocks={},this.allowedBlocks=new Set,this.blockFilter="",this.busy=!1,this.pageCount=0,this.platformConfig={},this.__disposer=[],this.t=this.t||{},this.t={...this.t,title:"Skeleton platform settings",experienceLevel:"Experience level",cmsFeatures:"CMS features",editorFeatures:"Editor features",blocks:"Allowed blocks",filterBlocks:"Filter blocks",requiredTextNote:"Some text tags are required and always enabled (shown disabled).",selectAll:"Select all",deselectAll:"Deselect all",download:"Download skeleton",cancel:"Cancel",generating:"Generating skeleton…",generated:"Skeleton downloaded",largeSiteWarning:"This site has many pages. Skeleton generation may take a while."},this.registerLocalization({context:this,basePath:import.meta.url})}static get styles(){return[super.styles,t`
        :host {
          display: block;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary);
          min-width: 70vw;
          padding: var(--ddd-spacing-4);
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-white)
          );
          background: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-nittanyNavy)
          );
        }

        :host([busy]) {
          pointer-events: none;
          opacity: 0.8;
        }

        h2 {
          margin: 0 0 var(--ddd-spacing-4) 0;
          font-size: var(--ddd-font-size-m);
          font-weight: var(--ddd-font-weight-bold);
        }

        h4 {
          margin: 0;
          font-size: var(--ddd-font-size-xs);
          font-weight: var(--ddd-font-weight-bold);
        }

        .row {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-6);
        }

        .section {
          border: var(--ddd-border-sm);
          border-radius: var(--ddd-radius-md);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.15)
          );
          padding: var(--ddd-spacing-4);
        }

        .section-title {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-3);
        }

        .section-title h3 {
          margin: 0;
          font-size: var(--ddd-font-size-s);
          font-weight: var(--ddd-font-weight-bold);
        }

        fieldset {
          border: 0;
          padding: 0;
          margin: 0;
        }

        .check-grid {
          flex: 7;
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(260px, 1fr));
          gap: var(--ddd-spacing-3);
        }

        label.check {
          display: flex;
          align-items: flex-start;
          gap: var(--ddd-spacing-2);
          line-height: 1.3;
        }

        label.check input {
          margin-top: 2px;
        }

        .controls-container {
          display: flex;
        }

        .controls {
          flex: 1;
          display: inline-flex;
          gap: var(--ddd-spacing-2);
          align-items: center;
          flex-direction: column;
        }

        .controls simple-icon-button-lite {
          --simple-icon-height: var(--ddd-icon-4xs);
          --simple-icon-width: var(--ddd-icon-4xs);
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.2)
          );
        }

        select,
        input[type='text'] {
          font-family: var(--ddd-font-primary);
          font-size: var(--ddd-font-size-s);
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.25)
          );
          color: inherit;
        }

        .audience {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-2);
        }

        .audience-content {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
        }

        .audience label {
          font-weight: var(--ddd-font-weight-bold);
          font-size: var(--ddd-font-size-xs);
        }

        .blocks-list {
          max-height: 40vh;
          overflow: auto;
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-3);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.2)
          );
        }

        .block-category {
          padding: var(--ddd-spacing-3) 0;
          border-top: var(--ddd-border-xs);
        }

        .block-category:first-of-type {
          border-top: 0;
          padding-top: 0;
        }

        .block-category-title {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-2);
        }

        .block-category-note {
          margin: 0 0 var(--ddd-spacing-3) 0;
          font-size: var(--ddd-font-size-3xs);
          opacity: 0.9;
        }

        .blocks-meta {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-3);
        }

        .block-search {
          flex: 7;
        }

        .actions {
          position: sticky;
          bottom: 0;
          display: flex;
          justify-content: flex-end;
          gap: var(--ddd-spacing-3);
          padding-top: var(--ddd-spacing-4);
          margin-top: var(--ddd-spacing-4);
          background: linear-gradient(
            to top,
            light-dark(
              var(--ddd-theme-default-white),
              var(--ddd-theme-default-nittanyNavy)
            ) 60%,
            transparent
          );
        }

        button.action {
          font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-s);
          padding: var(--ddd-spacing-3) var(--ddd-spacing-5);
          border-radius: var(--ddd-radius-sm);
          border: var(--ddd-border-xs);
          cursor: pointer;
          background-color: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
        }

        button.action.secondary {
          background-color: var(--ddd-theme-default-discoveryCoral);
        }

        button.action:focus,
        button.action:hover {
          outline: 2px solid var(--ddd-theme-default-keystoneYellow);
        }

        .note {
          margin-top: var(--ddd-spacing-2);
          font-size: var(--ddd-font-size-xs);
          opacity: 0.9;
        }

        details {
          min-width: 100%;
          box-sizing: border-box;
        }

        summary .section-title {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
        }
      `]}connectedCallback(){super.connectedCallback(),i((e=>{const t=a(o.manifest);t&&t.items&&(this.pageCount=t.items.length),this.__disposer.push(e)})),i((e=>{this.features=a(l.platformConfig.features);const t=a(l.gizmoList);this.haxBlocks=t.filter((e=>!(e.meta&&(e.meta.inlineOnly||e.meta.hidden||e.meta.requiresParent)))),console.log(this.haxBlocks),this.__disposer.push(e)}))}disconnectedCallback(){for(;this.__disposer.length;){const e=this.__disposer.pop();e&&"function"==typeof e.dispose&&e.dispose()}super.disconnectedCallback()}render(){const t=c.filter((e=>"CMS"===e.group)),i=c.filter((e=>"HAX"===e.group)),a=this.haxBlocks?this.haxBlocks.filter((e=>!e.gizmoList)).length:0,s=this.haxBlocks?this.haxBlocks.filter((e=>l.requiredPrimitives.has(e.tag)||this.allowedBlocks.has(e.tag))).length:0;let o=new URL(`./assets/${this.audience.toLowerCase()}.png`,import.meta.url).href;return e`
      <h2>${this.t.title}</h2>

      <div class="row">
        <div class="section">
          <!-- <div class="audience"> -->
            <div class="audience-content">
            <div class="audience-selector">
            <label for="audience">${this.t.experienceLevel}</label>
            <select id="audience" @change="${this._audienceChanged}">
              ${this.audienceOptions.map((t=>e`<option value="${t.value}" ?selected=${this.audience===t.value}>
                  ${t.label}
                </option>`))}
            </select>
            </div>
            <img class="toolbar-img" src=${o}>
            </div>
            ${this.pageCount>20?e`<div class="note">${this.t.largeSiteWarning}</div>`:""}
          <!-- </div> -->
            </div>

        <details class="section" open>
          <summary class="section-title">
            <h3>${this.t.cmsFeatures}</h3>
          </summary>
          <div class="controls-container">
            <fieldset class="check-grid cms-container">
              ${t.map((e=>this._renderFeatureCheckbox(e)))}
            </fieldset>
            <div class="controls">
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.selectAll}"
                title="${this.t.selectAll}"
                data-category="cms"
                @click="${this._selectAll}"
              >Select All</simple-icon-button-lite>
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.deselectAll}"
                title="${this.t.deselectAll}"
                data-category="cms"
                @click="${this._deselectAll}"
              >Deselect All</simple-icon-button-lite>
            </div>
          </div>
        </details>

        <details class="section">
          <summary class="section-title">
            <h3>${this.t.editorFeatures}</h3>
          </summary>
          <div class="controls-container">
            <fieldset class="check-grid hax-container">
              ${i.map((e=>this._renderFeatureCheckbox(e)))}
            </fieldset>
            <div class="controls">
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.selectAll}"
                title="${this.t.selectAll}"
                data-category="hax"
                @click="${this._selectAll}"
              >Select All</simple-icon-button-lite>
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.deselectAll}"
                title="${this.t.deselectAll}"
                data-category="hax"
                @click="${this._deselectAll}"
              >Deselect All</simple-icon-button-lite>
            </div>
          </div>
        </details>

        <details class="section">
          <summary class="section-title">
            <h3>${this.t.blocks}</h3>
            </summary>

          <div class="blocks-meta">
            <div class="block-search">
              <label for="blockFilter">Filter: </label>
              <input
                id="blockFilter"
                type="text"
                .value=${this.blockFilter}
                @input=${this._blockFilterChanged}
              />
            </div>
            <div class="controls">
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.selectAll}"
                title="${this.t.selectAll}"
                data-category="all-blocks"
                @click="${this._selectAll}"
              >Select All</simple-icon-button-lite>
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.deselectAll}"
                title="${this.t.deselectAll}"
                data-category="all-blocks"
                @click="${this._deselectAll}"
              >Deselect All</simple-icon-button-lite>
            </div>
          </div>

          <div class="note">
            ${s} selected / ${a} available
            ${l.requiredPrimitives.size?e`<span> • ${this.t.requiredTextNote}</span>`:""}
          </div>

          <div class="blocks-list">
            ${this._groupBlocksByCategory(this.haxBlocks).map((t=>e`
                <details class="block-category">
                  <summary class="block-category-title">
                    <h4>${t.category}</h4>
                  </summary>
                  <div class="controls-container">
                    <fieldset class="check-grid ${t.category.toLowerCase()}-container">
                      ${t.blocks.map((e=>this._renderBlockCheckbox(e)))}
                    </fieldset>
                    <div class="controls">
                      <simple-icon-button-lite
                        icon="icons:select-all"
                        label="${this.t.selectAll}"
                        title="${this.t.selectAll}"
                        data-category=${t.category.toLowerCase()}
                        @click="${this._selectAll}"
                      >Select All</simple-icon-button-lite>
                      <simple-icon-button-lite
                        icon="icons:select-all"
                        label="${this.t.deselectAll}"
                        title="${this.t.deselectAll}"
                        data-category=${t.category.toLowerCase()}
                        @click="${this._deselectAll}"
                      >Deselect All</simple-icon-button-lite>
                    </div>
                  </div>
                </details>
              `))}
          </div>
          </details>
      </div>

      <div class="actions">
        <button class="action secondary" @click="${this._cancel}">
          ${this.t.cancel}
        </button>
        <button class="action" @click="${this._download}">
          ${this.busy?this.t.generating:this.t.download}
        </button>
      </div>
    `}_renderFeatureCheckbox(t){const i=o.platformAllows(t.key);return e`
      <label class="check">
        <input
          type="checkbox"
          data-key="${t.key}"
          .checked=${i}
          @change=${this._checkboxChanged}
        /> <simple-icon-lite icon=${t.icon}></simple-icon-lite>
        <span>${t.label}</span>
      </label>
    `}_renderBlockCheckbox(t){const i=!!l.requiredPrimitives.has(t.tag);return e`
      <label class="check">
        <input
          type="checkbox"
          data-tag="${t.tag}"
          .checked=${!t.platformRestricted}
          ?disabled=${i}
          @change=${this._checkboxChanged}
        />
        <simple-icon-lite icon="${t.icon}"></simple-icon-lite>
        <span>${t.title}</span>
      </label>
    `}_audienceChanged(e){const t=e&&e.target?e.target.value:"expert";this.audience=t}_checkboxChanged(e){const t=e.currentTarget;if(!t||"checkbox"!==t.type)return;const{key:i,tag:a}=t.dataset;if(i)return this.features={...this.features,[i]:e.target.checked},void console.log(this.features);if(a){if(l.requiredPrimitives.has(a))return;return e.target.checked?this.allowedBlocks.add(a):this.allowedBlocks.delete(a),this.requestUpdate(),void console.log(this.allowedBlocks)}}_blockFilterChanged(e){this.blockFilter=e&&e.target?e.target.value:""}_selectAll(e){const t=e&&e.target?e.target.getAttribute("data-category"):null;switch(t){case"cms":{const e={...this.features};c.filter((e=>"CMS"===e.group)).forEach((t=>{e[t.key]=!0})),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach((e=>{e.checked=!0}));break}case"hax":{const e={...this.features};c.filter((e=>"HAX"===e.group)).forEach((t=>{e[t.key]=!0})),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach((e=>{e.checked=!0}));break}case"all-blocks":this.allowedBlocks=new Set(this.haxBlocks.filter((e=>!l.requiredPrimitives.has(e.tag))).map((e=>e.tag)));this.shadowRoot.querySelector(".blocks-list").querySelectorAll('input[type="checkbox"]').forEach((e=>{e.checked=!0}));break;default:this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach((e=>{l.requiredPrimitives.has(e.dataset.tag)||(this.allowedBlocks.add(e.dataset.tag),e.checked=!0)})),this.requestUpdate();break}}_deselectAll(e){const t=e&&e.target?e.target.getAttribute("data-category"):null;switch(t){case"cms":{const e={...this.features};c.filter((e=>"CMS"===e.group)).forEach((t=>{e[t.key]=!1})),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach((e=>{e.checked=!1}));break}case"hax":{const e={...this.features};c.filter((e=>"HAX"===e.group)).forEach((t=>{e[t.key]=!1})),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach((e=>{e.checked=!1}));break}case"all-blocks":this.allowedBlocks=new Set([]);this.shadowRoot.querySelector(".blocks-list").querySelectorAll('input[type="checkbox"]').forEach((e=>{l.requiredPrimitives.has(e.dataset.tag)||(e.checked=!1)}));break;default:this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach((e=>{l.requiredPrimitives.has(e.dataset.tag)||(this.allowedBlocks.delete(e.dataset.tag),e.checked=!1)})),this.requestUpdate();break}}_compareCategories(e,t){const i=e||"Other",a=t||"Other",s=["Writing","Text"],l=["Other"];return s.includes(i)&&!s.includes(a)?-1:!s.includes(i)&&s.includes(a)||l.includes(i)&&!l.includes(a)?1:!l.includes(i)&&l.includes(a)?-1:s.includes(i)&&s.includes(a)?s.indexOf(i)-s.indexOf(a):i<a?-1:i>a?1:0}_groupBlocksByCategory(e){const t={};e.forEach((e=>{const i=l.requiredPrimitives.has(e.tag);let a="Other";Array.isArray(e.tags)&&e.tags.length>0&&(a=e.tags[0]),i&&"Other"===a&&(a="Writing"),t[a]||(t[a]=[]),t[a].push(e)}));const i=Object.keys(t);return i.sort(((e,t)=>this._compareCategories(e,t))),i.map((e=>({category:e,blocks:t[e]})))}_platformConfigForExport(){const e=Array.from(this.allowedBlocks||[]).filter((e=>!l.requiredPrimitives.has(e)));return{audience:this.audience,features:this.features,allowedBlocks:e.sort()}}async _download(){try{this.busy=!0;const{SiteSkeletonGenerator:e}=await import("../utils/site-skeleton-generator.js"),t=await e.generateFromCurrentSite(!0),i=this._platformConfigForExport();t._skeleton||(t._skeleton={}),t._skeleton.originalMetadata||(t._skeleton.originalMetadata={}),t._skeleton.originalMetadata.platform=i,t._skeleton.originalSettings||(t._skeleton.originalSettings={}),e.downloadSkeleton(t),l.toast(this.t.generated,3e3,{},"fit-bottom"),this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1}))}catch(e){console.error("Skeleton export failed:",e),l.toast(`Skeleton export failed: ${e.message}`,5e3,{},"fit-bottom")}this.busy=!1}_cancel(){this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1}))}}globalThis.customElements.define(HAXCMSSitePlatformUI.tag,HAXCMSSitePlatformUI);export{HAXCMSSitePlatformUI};
