// @haxtheweb/haxcms-elements/lib/core/ui/haxcms-site-platform-ui.js
import{html as e,css as t}from"../../../../../lit/index.js";import{autorun as i,toJS as s}from"../../../../../mobx/dist/mobx.esm.js";import{DDD as l}from"../../../../d-d-d/d-d-d.js";import{HAXStore as a}from"../../../../hax-body/lib/hax-store.js";import{store as o}from"../haxcms-site-store.js";import{HAXCMSI18NMixin as r}from"../utils/HAXCMSI18NMixin.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";import"../../../../simple-icon/lib/simple-icon-lite.js";import"../../../../hax-body/lib/hax-text-editor-toolbar.js";const d=["h1","h2","p","ul","ol"],n=new Set(d),c=["addPage","deletePage","outlineDesigner","styleGuide","insights","manifest","addBlock","contentMap","viewSource","onlineSearch","pageBreak"],h=[{key:"addPage",label:"Add pages",icon:"hax:add-page",group:"CMS"},{key:"deletePage",label:"Delete pages",icon:"icons:delete",group:"CMS"},{key:"outlineDesigner",label:"Outline designer",icon:"hax:site-map",group:"CMS"},{key:"styleGuide",label:"Style guide",icon:"lrn:palette",group:"CMS"},{key:"insights",label:"Insights",icon:"hax:clipboard-pulse",group:"CMS"},{key:"manifest",label:"Site settings",icon:"hax:home-edit",group:"CMS"},{key:"pageBreak",label:"Edit Page Details",icon:"hax:page-edit",group:"CMS"},{key:"addBlock",label:"Add blocks",icon:"hax:add-brick",group:"Editor"},{key:"contentMap",label:"Page structure (content map)",icon:"hax:newspaper",group:"Editor"},{key:"viewSource",label:"View source",icon:"hax:html-code",group:"Editor"},{key:"onlineSearch",label:"Online search",icon:"hax:add-page",group:"Editor"}];class HAXCMSSitePlatformUI extends(r(l)){static get tag(){return"haxcms-site-platform-ui"}static get properties(){return{audience:{type:String},features:{type:Object},allowedBlocks:{type:Object},blocks:{type:Array},blockFilter:{type:String,attribute:"block-filter"},busy:{type:Boolean,reflect:!0},pageCount:{type:Number,attribute:"page-count"},platformConfig:{type:Object},gizmoList:{type:Object}}}constructor(){super(),this.audience="expert",this.audienceList=[{value:"novice",label:"Novice"},{value:"expert",label:"Expert"}],this.featureList=c,this.features={},this.allowedBlocks=new Set,this.blocks=[],this.blockFilter="",this.busy=!1,this.pageCount=0,this.platformConfig={},this.gizmoList={},this.__disposer=[],this.__blocksInitialized=!1,this.__presetAllowedBlocks=null,this.t=this.t||{},this.t={...this.t,title:"Skeleton platform settings",experienceLevel:"Experience level",cmsFeatures:"CMS features",editorFeatures:"Editor features",blocks:"Allowed blocks",filterBlocks:"Filter blocks",requiredTextNote:"Some text tags are required and always enabled (shown disabled).",selectAll:"Select all",deselectAll:"Deselect all",download:"Download skeleton",cancel:"Cancel",generating:"Generating skeleton…",generated:"Skeleton downloaded",largeSiteWarning:"This site has many pages. Skeleton generation may take a while."},this.registerLocalization({context:this,basePath:import.meta.url})}static get styles(){return[super.styles,t`
        :host {
          display: block;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary);
          min-width: 70vw;
          padding: var(--ddd-spacing-4);
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-white)
          );
          background: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-nittanyNavy)
          );
        }

        :host([busy]) {
          pointer-events: none;
          opacity: 0.8;
        }

        h2 {
          margin: 0 0 var(--ddd-spacing-4) 0;
          font-size: var(--ddd-font-size-m);
          font-weight: var(--ddd-font-weight-bold);
        }

        h4 {
          margin: 0;
          font-size: var(--ddd-font-size-xs);
          font-weight: var(--ddd-font-weight-bold);
        }

        .row {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-6);
        }

        .section {
          border: var(--ddd-border-sm);
          border-radius: var(--ddd-radius-md);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.15)
          );
          padding: var(--ddd-spacing-4);
        }

        .section-title {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-3);
        }

        .section-title h3 {
          margin: 0;
          font-size: var(--ddd-font-size-s);
          font-weight: var(--ddd-font-weight-bold);
        }

        fieldset {
          border: 0;
          padding: 0;
          margin: 0;
        }

        .check-grid {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(260px, 1fr));
          gap: var(--ddd-spacing-3);
        }

        label.check {
          display: flex;
          align-items: flex-start;
          gap: var(--ddd-spacing-2);
          line-height: 1.3;
        }

        label.check input {
          margin-top: 2px;
        }

        .controls {
          display: inline-flex;
          gap: var(--ddd-spacing-2);
          align-items: center;
        }

        .controls simple-icon-button-lite {
          --simple-icon-height: var(--ddd-icon-4xs);
          --simple-icon-width: var(--ddd-icon-4xs);
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.2)
          );
        }

        select,
        input[type='text'] {
          font-family: var(--ddd-font-primary);
          font-size: var(--ddd-font-size-s);
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.25)
          );
          color: inherit;
        }

        .audience {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-2);
        }

        .audience-content {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
        }

        .audience label {
          font-weight: var(--ddd-font-weight-bold);
          font-size: var(--ddd-font-size-xs);
        }

        .blocks-list {
          max-height: 40vh;
          overflow: auto;
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-3);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.2)
          );
        }

        .block-category {
          padding: var(--ddd-spacing-3) 0;
          border-top: var(--ddd-border-xs);
        }

        .block-category:first-of-type {
          border-top: 0;
          padding-top: 0;
        }

        .block-category-title {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-2);
        }

        .block-category-note {
          margin: 0 0 var(--ddd-spacing-3) 0;
          font-size: var(--ddd-font-size-3xs);
          opacity: 0.9;
        }

        .blocks-meta {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-3);
        }

        .actions {
          position: sticky;
          bottom: 0;
          display: flex;
          justify-content: flex-end;
          gap: var(--ddd-spacing-3);
          padding-top: var(--ddd-spacing-4);
          margin-top: var(--ddd-spacing-4);
          background: linear-gradient(
            to top,
            light-dark(
              var(--ddd-theme-default-white),
              var(--ddd-theme-default-nittanyNavy)
            ) 60%,
            transparent
          );
        }

        button.action {
          font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-s);
          padding: var(--ddd-spacing-3) var(--ddd-spacing-5);
          border-radius: var(--ddd-radius-sm);
          border: var(--ddd-border-xs);
          cursor: pointer;
          background-color: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
        }

        button.action.secondary {
          background-color: var(--ddd-theme-default-discoveryCoral);
        }

        button.action:focus,
        button.action:hover {
          outline: 2px solid var(--ddd-theme-default-keystoneYellow);
        }

        .note {
          margin-top: var(--ddd-spacing-2);
          font-size: var(--ddd-font-size-xs);
          opacity: 0.9;
        }

        /*.editor-features {
          background-color: lightblue;
          border-left: 2px darkblue solid;
          padding-left: 4px;
        } */

        details {
          min-width: 100%;
          box-sizing: border-box;
        }

        summary .section-title {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
        }

        /* .toolbar-img {
          float: right;
        } */
      `]}connectedCallback(){super.connectedCallback(),i((e=>{const t=s(o.manifest);t&&t.items&&(this.pageCount=t.items.length),this.__disposer.push(e)})),i((e=>{s(o.platformConfig);this.__disposer.push(e)})),i((e=>{const t=s(a.gizmoList);this.gizmoList=t.filter((e=>!(e.meta&&(e.meta.inlineOnly||e.meta.hidden||e.meta.requiresParent)))),this.__disposer.push(e)}))}disconnectedCallback(){for(;this.__disposer.length;){const e=this.__disposer.pop();e&&"function"==typeof e.dispose&&e.dispose()}super.disconnectedCallback()}render(){const t=h.filter((e=>"CMS"===e.group)),i=h.filter((e=>"Editor"===e.group)),s=this.gizmoList?this.gizmoList.filter((e=>!e.gizmoList)).length:0,l=this.gizmoList?this.gizmoList.filter((e=>!e.locked&&this.allowedBlocks.has(e.tag))).length:0;let a=new URL(`./assets/${this.audience.toLowerCase()}.png`,import.meta.url).href;return e`
      <h2>${this.t.title}</h2>

      <div class="row">
        <div class="section">
          <!-- <div class="audience"> -->
            <div class="audience-content">
            <div class="audience-selector">
            <label for="audience">${this.t.experienceLevel}</label>
            <select id="audience" @change="${this._audienceChanged}">
              ${this._audienceOptions().map((t=>e`<option value="${t.value}" ?selected=${this.audience===t.value}>
                  ${t.label}
                </option>`))}
            </select>
            </div>
            <img class="toolbar-img" src=${a}>
            </div>
            ${this.pageCount>20?e`<div class="note">${this.t.largeSiteWarning}</div>`:""}
          <!-- </div> -->
            </div>

        <details class="section">
          <summary class="section-title">
            <h3>${this.t.cmsFeatures}</h3>
            <div class="controls">
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.selectAll}"
                title="${this.t.selectAll}"
                @click="${this._selectAllCmsFeatures}"
              ></simple-icon-button-lite>
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.deselectAll}"
                title="${this.t.deselectAll}"
                @click="${this._deselectAllCmsFeatures}"
              ></simple-icon-button-lite>
            </div>
          </summary>
          <fieldset class="check-grid">
            ${t.map((e=>this._renderFeatureCheckbox(e)))}
          </fieldset>
        </details>

        <details class="section">
          <summary class="section-title">
            <simple-icon-lite icon="hax:add-page"></simple-icon-lite> <h3>${this.t.editorFeatures}</h3>
            <div class="controls">
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.selectAll}"
                title="${this.t.selectAll}"
                @click="${this._selectAllEditorFeatures}"
              ></simple-icon-button-lite>
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.deselectAll}"
                title="${this.t.deselectAll}"
                @click="${this._deselectAllEditorFeatures}"
              ></simple-icon-button-lite>
            </div>
            </summary>
          <fieldset class="check-grid">
            ${i.map((e=>this._renderFeatureCheckbox(e)))}
          </fieldset>
            </details>

        <details class="section">
          <summary class="section-title">
            <h3>${this.t.blocks}</h3>
            <div class="controls">
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.selectAll}"
                title="${this.t.selectAll}"
                @click="${this._selectAllBlocks}"
              ></simple-icon-button-lite>
              <simple-icon-button-lite
                icon="icons:select-all"
                label="${this.t.deselectAll}"
                title="${this.t.deselectAll}"
                @click="${this._deselectAllBlocks}"
              ></simple-icon-button-lite>
            </div>
            </summary>

          <div class="blocks-meta">
            <label for="blockFilter">${this.t.filterBlocks}</label>
            <input
              id="blockFilter"
              type="text"
              .value=${this.blockFilter}
              @input=${this._blockFilterChanged}
            />
          </div>

          <div class="note">
            ${l} selected / ${s} available
            ${d.length?e`<span> • ${this.t.requiredTextNote}</span>`:""}
          </div>

          <div class="blocks-list">
            ${this._groupBlocksByCategory(this.gizmoList).map((t=>e`
                <details class="block-category">
                  <summary class="block-category-title">
                    <h4>${t.category}</h4>
                    <div class="controls">
                      <simple-icon-button-lite
                        icon="icons:select-all"
                        label="${this.t.selectAll}"
                        title="${this.t.selectAll}"
                        @click="${this._selectAllBlocksInCategory}"
                      ></simple-icon-button-lite>
                      <simple-icon-button-lite
                        icon="icons:select-all"
                        label="${this.t.deselectAll}"
                        title="${this.t.deselectAll}"
                        @click="${this._deselectAllBlocksInCategory}"
                      ></simple-icon-button-lite>
                    </div>
            </summary>
                  <fieldset class="check-grid">
                    ${t.blocks.map((e=>this._renderBlockCheckbox(e)))}
                  </fieldset>
                </details>
              `))}
          </div>
          </details>
      </div>

      <div class="actions">
        <button class="action secondary" @click="${this._cancel}">
          ${this.t.cancel}
        </button>
        <button class="action" @click="${this._download}">
          ${this.busy?this.t.generating:this.t.download}
        </button>
      </div>
    `}_renderFeatureCheckbox(t){const i=o.platformAllows(t.key);return e`
      <label class="check editor-features">
        <input
          type="checkbox"
          data-key="${t.key}"
          .checked=${i}
          @change=${this._featureChanged}
        /> <simple-icon-lite icon=${t.icon}></simple-icon-lite>
        <span>${t.label}</span>
      </label>
    `}_renderBlockCheckbox(t){const i=!!a.requiredPrimitives.has(t.tag);return e`
      <label class="check">
        <input
          type="checkbox"
          data-tag="${t.tag}"
          .checked=${!t.platformRestricted}
          ?disabled=${i}
          @change=${this._blockChanged}
        />
        <simple-icon-lite icon="${t.icon}"></simple-icon-lite>
        <span>${t.title}</span>
      </label>
    `}_audienceOptions(){return[{value:"novice",label:"Novice"},{value:"expert",label:"Expert"}]}_defaultFeatures(){const e={};return c.forEach((t=>{e[t]=!0})),e}_initFromManifestPlatform(){if(this.__initializedFromManifest)return;let e="expert";this.platformConfig&&this.platformConfig.audience&&(e=platformConfig.audience);let t=null;platform&&Array.isArray(platform.allowedBlocks)?t=platform.allowedBlocks:platform&&Array.isArray(platform.blocks)&&(t=platform.blocks),this.audience=e,this.features=nextFeatures,this.__presetAllowedBlocks=t&&Array.isArray(t)?new Set(t):null,this.__initializedFromManifest=!0}_refreshBlocksList(){const e=[];if(Object.keys(this.gizmoList).forEach((t=>{const i=this.gizmoList[t];if(i&&i.meta&&(i.meta.inlineOnly||i.meta.hidden||i.requiresParent))return;let s=t,l="Other",a=i.icon,o=[];const r=n.has(t);i&&"object"==typeof i&&(i.title&&(s=i.title),Array.isArray(i.tags)&&i.tags.length>0&&(o=i.tags,l=i.tags[0]||l)),r&&"Other"===l&&(l="Writing"),e.push({tag:t,title:s,category:l,icon:a,tags:o,locked:r})})),e.sort(((e,t)=>{const i=e.category||"Other",s=t.category||"Other",l=this._compareCategories(i,s);return 0!==l?l:e.locked&&t.locked?d.indexOf(e.tag)-d.indexOf(t.tag):e.locked?-1:t.locked?1:e.title<t.title?-1:e.title>t.title?1:0})),this.blocks=e,!this.__blocksInitialized){if(this.__presetAllowedBlocks&&this.__presetAllowedBlocks.size>0){const e=new Set;this.blocks.forEach((t=>{this.__presetAllowedBlocks.has(t.tag)&&e.add(t.tag)})),this.allowedBlocks=e}else this.allowedBlocks=new Set(this.blocks.map((e=>e.tag)));this.__blocksInitialized=!0}}_filteredBlocks(){if(!this.blocks||!this.blocks.length)return[];const e=(this.blockFilter||"").trim().toLowerCase();if(""===e)return this.blocks;const t=this.blocks.filter((e=>e.locked)),i=this.blocks.filter((t=>{const i=t.tag.toLowerCase(),s=(t.title||"").toLowerCase();return i.includes(e)||s.includes(e)})),s=new Set,l=[];return t.forEach((e=>{s.add(e.tag),l.push(e)})),i.forEach((e=>{s.has(e.tag)||(s.add(e.tag),l.push(e))})),l}_audienceChanged(e){const t=e&&e.target?e.target.value:"expert";this.audience=t}_featureChanged(e){const t=e&&e.target?e.target.getAttribute("data-key"):null;if(!t)return;const i=e.target.checked;this.features={...this.features,[t]:i}}_blockChanged(e){const t=e&&e.target?e.target.getAttribute("data-tag"):null;if(!t)return;if(n.has(t))return;const i=e.target.checked,s=new Set(this.allowedBlocks);i?s.add(t):s.delete(t),this.allowedBlocks=s}_blockFilterChanged(e){this.blockFilter=e&&e.target?e.target.value:""}_selectAllCmsFeatures(){const e={...this.features};h.filter((e=>"CMS"===e.group)).forEach((t=>{e[t.key]=!0})),this.features=e}_deselectAllCmsFeatures(){const e={...this.features};h.filter((e=>"CMS"===e.group)).forEach((t=>{e[t.key]=!1})),this.features=e}_selectAllEditorFeatures(){const e={...this.features};h.filter((e=>"Editor"===e.group)).forEach((t=>{e[t.key]=!0})),this.features=e}_deselectAllEditorFeatures(){const e={...this.features};h.filter((e=>"Editor"===e.group)).forEach((t=>{e[t.key]=!1})),this.features=e}_selectAllBlocks(){this.allowedBlocks=new Set(this.blocks.map((e=>e.tag)))}_deselectAllBlocks(){const e=new Set;this.blocks.forEach((t=>{t.locked&&e.add(t.tag)})),this.allowedBlocks=e}_selectAllBlocksInCategory(e){const t=e&&e.currentTarget?e.currentTarget.dataset.category:null;if(!t)return;const i=new Set(this.allowedBlocks);this.blocks.forEach((e=>{e.category!==t||e.locked||i.add(e.tag)})),this.allowedBlocks=i}_deselectAllBlocksInCategory(e){const t=e&&e.currentTarget?e.currentTarget.dataset.category:null;if(!t)return;const i=new Set(this.allowedBlocks);this.blocks.forEach((e=>{e.category!==t||e.locked||i.delete(e.tag)})),this.allowedBlocks=i}_compareCategories(e,t){const i=e||"Other",s=t||"Other",l=["Writing","Text"],a=["Other"];return l.includes(i)&&!l.includes(s)?-1:!l.includes(i)&&l.includes(s)||a.includes(i)&&!a.includes(s)?1:!a.includes(i)&&a.includes(s)?-1:l.includes(i)&&l.includes(s)?l.indexOf(i)-l.indexOf(s):i<s?-1:i>s?1:0}_groupBlocksByCategory(e){const t={};e.forEach((e=>{const i=n.has(e.tag);let s="Other";Array.isArray(e.tags)&&e.tags.length>0&&(s=e.tags[0]),i&&"Other"===s&&(s="Writing"),t[s]||(t[s]=[]),t[s].push(e)}));const i=Object.keys(t);return i.sort(((e,t)=>this._compareCategories(e,t))),i.map((e=>({category:e,blocks:t[e]})))}_platformConfigForExport(){const e={};c.forEach((t=>{e[t]=this.features&&!1!==this.features[t]}));const t=Array.from(this.allowedBlocks||[]).filter((e=>!n.has(e)));return{audience:this.audience||"expert",features:e,allowedBlocks:t.sort()}}async _download(){try{this.busy=!0;const{SiteSkeletonGenerator:e}=await import("../utils/site-skeleton-generator.js"),t=await e.generateFromCurrentSite(!0),i=this._platformConfigForExport();t._skeleton||(t._skeleton={}),t._skeleton.originalMetadata||(t._skeleton.originalMetadata={}),t._skeleton.originalMetadata.platform=i,t._skeleton.originalSettings||(t._skeleton.originalSettings={}),e.downloadSkeleton(t),a.toast(this.t.generated,3e3,{},"fit-bottom"),this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1}))}catch(e){console.error("Skeleton export failed:",e),a.toast(`Skeleton export failed: ${e.message}`,5e3,{},"fit-bottom")}this.busy=!1}_cancel(){this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1}))}}globalThis.customElements.define(HAXCMSSitePlatformUI.tag,HAXCMSSitePlatformUI);export{HAXCMSSitePlatformUI};
