// @haxtheweb/haxcms-elements/lib/core/ui/haxcms-site-platform-ui.js
import{html as e,css as t}from"../../../../../lit/index.js";import{autorun as i,toJS as a}from"../../../../../mobx/dist/mobx.esm.js";import{DDD as s}from"../../../../d-d-d/d-d-d.js";import{HAXStore as o}from"../../../../hax-body/lib/hax-store.js";import{store as l}from"../haxcms-site-store.js";import{HAXCMSI18NMixin as r}from"../utils/HAXCMSI18NMixin.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";import"../../../../simple-icon/lib/simple-icon-lite.js";import"../../../../hax-body/lib/hax-text-editor-toolbar.js";const d=[{key:"addPage",label:"Add pages",icon:"hax:add-page",group:"CMS"},{key:"saveAndEdit",label:"Save and continue",icon:"hax:add-page",group:"CMS"},{key:"deletePage",label:"Delete pages",icon:"icons:delete",group:"CMS"},{key:"outlineDesigner",label:"Outline designer",icon:"hax:site-map",group:"CMS"},{key:"styleGuide",label:"Style guide",icon:"lrn:palette",group:"CMS"},{key:"insights",label:"Reports",icon:"hax:graph",group:"CMS"},{key:"siteManifest",label:"Site settings",icon:"hax:home-edit",group:"CMS"},{key:"themeManifest",label:"Theme settings",icon:"hax:home-edit",group:"CMS"},{key:"authorManifest",label:"Author settings",icon:"hax:home-edit",group:"CMS"},{key:"seoManifest",label:"SEO settings",icon:"hax:home-edit",group:"CMS"},{key:"pageBreak",label:"Edit page details",icon:"hax:page-edit",group:"CMS"},{key:"addBlock",label:"Add blocks",icon:"hax:add-brick",group:"HAX"},{key:"popularGizmos",label:"Popular blocks section",icon:"hax:add-brick",group:"HAX"},{key:"recentGizmos",label:"Recent blocks section",icon:"hax:add-brick",group:"HAX"},{key:"contentMap",label:"Page content map",icon:"hax:newspaper",group:"HAX"},{key:"viewSource",label:"View source",icon:"hax:html-code",group:"HAX"},{key:"uploadMedia",label:"Upload media",icon:"hax:add-page",group:"HAX"},{key:"onlineMedia",label:"Online media search",icon:"hax:add-page",group:"HAX"},{key:"community",label:"Community outreach",icon:"hax:add-page",group:"HAX"}];class HAXCMSSitePlatformUI extends(r(s)){static get tag(){return"haxcms-site-platform-ui"}static get properties(){return{audience:{type:String},features:{type:Object},haxBlocks:{type:Array},allowedBlocks:{type:Object},blockFilter:{type:String,attribute:"block-filter"},busy:{type:Boolean,reflect:!0},pageCount:{type:Number,attribute:"page-count"},platformConfig:{type:Object},platformSettingsMode:{type:Boolean,attribute:"platform-settings-mode"}}}constructor(){super(),this.audience="expert",this.audienceOptions=[{value:"novice",label:"Novice"},{value:"expert",label:"Expert"}],this.features={},this.haxBlocks=[],this.allowedBlocks=new Set,this.blockFilter="",this.busy=!1,this.pageCount=0,this.platformConfig={},this.platformSettingsMode=!1,this.__disposer=[],this.t=this.t||{},this.t={...this.t,title:"Platform Features",experienceLevel:"Experience level",cmsFeatures:"CMS features",editorFeatures:"Editor features",blocks:"Blocks",filterBlocks:"Filter blocks",requiredTextNote:"Some text tags are required and always enabled (shown disabled).",selectAll:"Select all",deselectAll:"Deselect all",download:"Download skeleton",save:"Save",generating:"Generating skeleton…",generated:"Skeleton downloaded",largeSiteWarning:"This site has many pages. Skeleton generation may take a while."},this.registerLocalization({context:this,basePath:import.meta.url})}static get styles(){return[super.styles,t`
        :host {
          --haxcms-admin-panel-height: calc(
            var(--simple-modal-height, 85vh) -
              var(--simple-modal-titlebar-height, 80px) - var(--ddd-spacing-8, 32px)
          );
          display: flex;
          flex-direction: column;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary);
          min-width: 70vw;
          min-height: min(60vh, var(--haxcms-admin-panel-height));
          height: var(--haxcms-admin-panel-height);
          max-height: var(--haxcms-admin-panel-height);
          overflow: hidden;
          padding: var(--ddd-spacing-4);
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-white)
          );
          background: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
          flex-shrink: 0;
        }

        :host([busy]) {
          pointer-events: none;
          opacity: 0.8;
        }

        h2 {
          margin: 0 0 var(--ddd-spacing-4) 0;
          font-size: var(--ddd-font-size-m);
          font-weight: var(--ddd-font-weight-bold);
        }

        h4 {
          margin: 0;
          font-size: var(--ddd-font-size-xs);
          font-weight: var(--ddd-font-weight-bold);
        }
        .panel-shell {
          display: flex;
          flex-direction: column;
          flex: 1;
          min-height: 0;
        }
        .panel-scroll {
          display: flex;
          flex-direction: column;
          flex: 1;
          min-height: 0;
          overflow-y: auto;
          padding-right: var(--ddd-spacing-1);
        }

        .row {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-6);
        }

        .section {
          border: var(--ddd-border-sm);
          border-radius: var(--ddd-radius-md);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.15)
          );
          padding: var(--ddd-spacing-4);
        }

        .section-title {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-3);
        }

        .section-title h3 {
          margin: 0;
          font-size: var(--ddd-font-size-s);
          font-weight: var(--ddd-font-weight-bold);
        }

        fieldset {
          border: 0;
          padding: 0;
          margin: 0;
        }

        .check-grid {
          flex: 7;
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(260px, 1fr));
          gap: var(--ddd-spacing-3);
        }

        label.check {
          display: flex;
          align-items: flex-start;
          gap: var(--ddd-spacing-2);
          line-height: 1.3;
        }

        label.check input {
          margin-top: 2px;
        }

        .controls-container {
          display: flex;
        }

        .controls {
          flex: 1;
          display: inline-flex;
          gap: var(--ddd-spacing-2);
          align-items: center;
          flex-direction: column;
        }

        .controls simple-icon-button-lite {
          --simple-icon-height: var(--ddd-icon-4xs);
          --simple-icon-width: var(--ddd-icon-4xs);
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.2)
          );
        }

        select,
        input[type='text'] {
          font-family: var(--ddd-font-primary);
          font-size: var(--ddd-font-size-s);
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.25)
          );
          color: inherit;
        }

        .audience {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-2);
        }

        .audience-content {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
        }

        .audience label {
          font-weight: var(--ddd-font-weight-bold);
          font-size: var(--ddd-font-size-xs);
        }

        .blocks-list {
          max-height: 40vh;
          overflow: auto;
          border: var(--ddd-border-xs);
          border-radius: var(--ddd-radius-sm);
          padding: var(--ddd-spacing-3);
          background: light-dark(
            var(--ddd-theme-default-white),
            rgba(0, 0, 0, 0.2)
          );
        }

        .block-category {
          padding: var(--ddd-spacing-3) 0;
          border-top: var(--ddd-border-xs);
        }

        .block-category:first-of-type {
          border-top: 0;
          padding-top: 0;
        }

        .block-category-title {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-2);
        }

        .block-category-note {
          margin: 0 0 var(--ddd-spacing-3) 0;
          font-size: var(--ddd-font-size-3xs);
          opacity: 0.9;
        }

        .blocks-meta {
          display: flex;
          align-items: center;
          justify-content: space-between;
          gap: var(--ddd-spacing-3);
          margin-bottom: var(--ddd-spacing-3);
        }

        .block-search {
          flex: 7;
        }

        .actions {
          display: flex;
          justify-content: flex-end;
          gap: var(--ddd-spacing-3);
          padding-top: var(--ddd-spacing-4);
          margin-top: var(--ddd-spacing-4);
          border-top: var(--ddd-border-xs) solid
            light-dark(
              var(--ddd-theme-default-limestoneGray),
              var(--ddd-primary-5)
            );
          background: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
          flex-shrink: 0;
        }

        button.action {
          font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-s);
          padding: var(--ddd-spacing-3) var(--ddd-spacing-5);
          border-radius: var(--ddd-radius-sm);
          border: var(--ddd-border-xs);
          cursor: pointer;
          background-color: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
        }

        button.action.secondary {
          background-color: var(--ddd-theme-default-discoveryCoral);
        }

        button.action:focus,
        button.action:hover {
          outline: 2px solid var(--ddd-theme-default-keystoneYellow);
        }

        .note {
          margin-top: var(--ddd-spacing-2);
          font-size: var(--ddd-font-size-xs);
          opacity: 0.9;
        }

        details {
          min-width: 100%;
          box-sizing: border-box;
        }

        summary .section-title {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
        }
      `]}connectedCallback(){super.connectedCallback(),i(e=>{const t=a(l.manifest);t&&t.items&&(this.pageCount=t.items.length),this.__disposer.push(e)}),i(e=>{const t=a(o.platformConfig);if(this.platformConfig=t||{},t){t.audience&&(this.audience=t.audience),t.features?this.features=a(t.features):this.features={};const e=t.allowedBlocks;e&&(e instanceof Set?this.allowedBlocks=new Set(Array.from(e)):Array.isArray(e)&&(this.allowedBlocks=new Set(e)))}else this.features={};const i=a(o.gizmoList),s=Array.isArray(i)?i:[];this.haxBlocks=s.filter(e=>!(e.meta&&(e.meta.inlineOnly||e.meta.hidden||e.meta.requiresParent))),this.__disposer.push(e)})}disconnectedCallback(){for(;this.__disposer.length;){const e=this.__disposer.pop();e&&"function"==typeof e.dispose&&e.dispose()}super.disconnectedCallback()}render(){const t=d.filter(e=>"CMS"===e.group),i=d.filter(e=>"HAX"===e.group);let a=new URL(`./assets/${this.audience.toLowerCase()}.png`,import.meta.url).href;return e`
      <div class="panel-shell">
        <div class="panel-scroll">
          <h2>${this.t.title}</h2>

          <div class="row">
            ${this.platformSettingsMode?"":e`
                  <div class="section">
                    <div class="audience-content">
                      <div class="audience-selector">
                        <label for="audience">${this.t.experienceLevel}</label>
                        <select id="audience" @change="${this._audienceChanged}">
                          ${this.audienceOptions.map(t=>e`<option value="${t.value}" ?selected=${this.audience===t.value}>
                              ${t.label}
                            </option>`)}
                        </select>
                      </div>
                      <img
                        class="toolbar-img"
                        src=${a}
                        alt="${this.audience} editing toolbar preview"
                      />
                    </div>
                    ${this.pageCount>20?e`<div class="note">${this.t.largeSiteWarning}</div>`:""}
                  </div>
                `}

            <details class="section" open>
              <summary class="section-title">
                <h3>${this.t.cmsFeatures}</h3>
              </summary>
              <div class="controls-container">
                <fieldset class="check-grid cms-container">
                  ${t.map(e=>this._renderFeatureCheckbox(e))}
                </fieldset>
                <div class="controls">
                  <simple-icon-button-lite
                    icon="icons:select-all"
                    label="${this.t.selectAll}"
                    title="${this.t.selectAll}"
                    data-category="cms"
                    @click="${this._selectAll}"
                  >Select All</simple-icon-button-lite>
                  <simple-icon-button-lite
                    icon="icons:select-all"
                    label="${this.t.deselectAll}"
                    title="${this.t.deselectAll}"
                    data-category="cms"
                    @click="${this._deselectAll}"
                  >Deselect All</simple-icon-button-lite>
                </div>
              </div>
            </details>

            <details class="section">
              <summary class="section-title">
                <h3>${this.t.editorFeatures}</h3>
              </summary>
              <div class="controls-container">
                <fieldset class="check-grid hax-container">
                  ${i.map(e=>this._renderFeatureCheckbox(e))}
                </fieldset>
                <div class="controls">
                  <simple-icon-button-lite
                    icon="icons:select-all"
                    label="${this.t.selectAll}"
                    title="${this.t.selectAll}"
                    data-category="hax"
                    @click="${this._selectAll}"
                  >Select All</simple-icon-button-lite>
                  <simple-icon-button-lite
                    icon="icons:select-all"
                    label="${this.t.deselectAll}"
                    title="${this.t.deselectAll}"
                    data-category="hax"
                    @click="${this._deselectAll}"
                  >Deselect All</simple-icon-button-lite>
                </div>
              </div>
            </details>
          </div>
        </div>
        <div class="actions">
          <button
            class="action"
            @click="${this.platformSettingsMode?this._savePlatformSettings:this._download}"
          >
            ${this.busy?this.t.generating:this.platformSettingsMode?this.t.save:this.t.download}
          </button>
        </div>
      </div>
    `}_renderFeatureCheckbox(t){const i=l.platformAllows(t.key);return e`
      <label class="check">
        <input
          type="checkbox"
          data-key="${t.key}"
          .checked=${i}
          @change=${this._checkboxChanged}
        /> <simple-icon-lite icon=${t.icon}></simple-icon-lite>
        <span>${t.label}</span>
      </label>
    `}_renderBlockCheckbox(t){const i=!!o.requiredPrimitives.has(t.tag);return e`
      <label class="check">
        <input
          type="checkbox"
          data-tag="${t.tag}"
          .checked=${!t.platformRestricted}
          ?disabled=${i}
          @change=${this._checkboxChanged}
        />
        <simple-icon-lite icon="${t.icon}"></simple-icon-lite>
        <span>${t.title}</span>
      </label>
    `}_audienceChanged(e){const t=e&&e.target?e.target.value:"expert";this.audience=t}_checkboxChanged(e){const t=e.currentTarget;if(!t||"checkbox"!==t.type)return;const{key:i}=t.dataset;return i?(this.features={...this.features,[i]:e.target.checked},void console.log(this.features)):void 0}_selectAll(e){const t=e&&e.target?e.target.getAttribute("data-category"):null;switch(t){case"cms":{const e={...this.features};d.filter(e=>"CMS"===e.group).forEach(t=>{e[t.key]=!0}),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach(e=>{e.checked=!0});break}case"hax":{const e={...this.features};d.filter(e=>"HAX"===e.group).forEach(t=>{e[t.key]=!0}),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach(e=>{e.checked=!0});break}}}_deselectAll(e){const t=e&&e.target?e.target.getAttribute("data-category"):null;switch(t){case"cms":{const e={...this.features};d.filter(e=>"CMS"===e.group).forEach(t=>{e[t.key]=!1}),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach(e=>{e.checked=!1});break}case"hax":{const e={...this.features};d.filter(e=>"HAX"===e.group).forEach(t=>{e[t.key]=!1}),this.features=e,console.log(this.features);this.shadowRoot.querySelector(`.${t}-container`).querySelectorAll('input[type="checkbox"]').forEach(e=>{e.checked=!1});break}}}_compareCategories(e,t){const i=e||"Other",a=t||"Other",s=["Writing","Text"],o=["Other"];return s.includes(i)&&!s.includes(a)?-1:!s.includes(i)&&s.includes(a)||o.includes(i)&&!o.includes(a)?1:!o.includes(i)&&o.includes(a)?-1:s.includes(i)&&s.includes(a)?s.indexOf(i)-s.indexOf(a):i<a?-1:i>a?1:0}_groupBlocksByCategory(e){const t={};e.forEach(e=>{const i=o.requiredPrimitives.has(e.tag);let a="Other";Array.isArray(e.tags)&&e.tags.length>0&&(a=e.tags[0]),i&&"Other"===a&&(a="Text"),t[a]||(t[a]=[]),t[a].push(e)});const i=Object.keys(t);return i.sort((e,t)=>this._compareCategories(e,t)),i.map(e=>({category:e,blocks:t[e]}))}_platformConfigForExport(){const e=Array.from(this.allowedBlocks||[]).filter(e=>!o.requiredPrimitives.has(e));return{audience:this.audience,features:this.features,allowedBlocks:e.sort()}}async _download(){try{this.busy=!0;const{SiteSkeletonGenerator:e}=await import("../utils/site-skeleton-generator.js"),t=await e.generateFromCurrentSite(!0),i=this._platformConfigForExport();t.site||(t.site={}),t.site.platform=i,t._skeleton||(t._skeleton={}),t._skeleton.originalMetadata||(t._skeleton.originalMetadata={}),t._skeleton.originalMetadata.platform=i,t._skeleton.originalSettings||(t._skeleton.originalSettings={}),e.downloadSkeleton(t),o.toast(this.t.generated,3e3,{},"fit-bottom"),this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1}))}catch(e){console.error("Skeleton export failed:",e),o.toast(`Skeleton export failed: ${e.message}`,5e3,{},"fit-bottom")}this.busy=!1}async _savePlatformSettings(){try{this.busy=!0;const e=this._platformConfigForExport();this.dispatchEvent(new CustomEvent("haxcms-save-platform-settings",{bubbles:!0,composed:!0,cancelable:!0,detail:e})),this.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1}))}catch(e){console.error("Saving platform settings failed:",e),o.toast(`Saving platform settings failed: ${e.message}`,5e3,{},"fit-bottom")}this.busy=!1}}globalThis.customElements.define(HAXCMSSitePlatformUI.tag,HAXCMSSitePlatformUI);export{HAXCMSSitePlatformUI};
