// @lrnwebcomponents/lrn-assignment/lib/lrn-assignment-button.js
define(["exports", "../../../lit-element/lit-element.js", "../../../@polymer/paper-button/paper-button.js", "../../lrn-icons/lrn-icons.js", "../../../@polymer/iron-icon/iron-icon.js"], function (_exports, _litElement, _paperButton, _lrnIcons, _ironIcon) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrnAssignmentButton = void 0;

  function _templateObject4_868cd7a030c111eab3ab336d9c8a6f6e() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        a {\n          color: inherit;\n          text-decoration: inherit;\n        }\n        paper-button {\n          background: #b0e6f9;\n        }\n        paper-button[complete] {\n          background: #e7ffe7;\n        }\n        iron-icon {\n          margin-left: 8px;\n          opacity: 0.8;\n        }\n      "]);

    _templateObject4_868cd7a030c111eab3ab336d9c8a6f6e = function _templateObject4_868cd7a030c111eab3ab336d9c8a6f6e() {
      return data;
    };

    return data;
  }

  function _templateObject3_868cd7a030c111eab3ab336d9c8a6f6e() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <paper-button raised complete\n                >", " <iron-icon icon=\"lrn-icons:done\"></iron-icon\n              ></paper-button>\n            "]);

    _templateObject3_868cd7a030c111eab3ab336d9c8a6f6e = function _templateObject3_868cd7a030c111eab3ab336d9c8a6f6e() {
      return data;
    };

    return data;
  }

  function _templateObject2_868cd7a030c111eab3ab336d9c8a6f6e() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <paper-button raised open\n                >", " <iron-icon icon=\"lrn-icons:input\"></iron-icon\n              ></paper-button>\n            "]);

    _templateObject2_868cd7a030c111eab3ab336d9c8a6f6e = function _templateObject2_868cd7a030c111eab3ab336d9c8a6f6e() {
      return data;
    };

    return data;
  }

  function _templateObject_868cd7a030c111eab3ab336d9c8a6f6e() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <a href=\"", "\">\n        ", "\n        ", "\n      </a>\n    "]);

    _templateObject_868cd7a030c111eab3ab336d9c8a6f6e = function _templateObject_868cd7a030c111eab3ab336d9c8a6f6e() {
      return data;
    };

    return data;
  }

  var LrnAssignmentButton =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrnAssignmentButton, _LitElement);
    babelHelpers.createClass(LrnAssignmentButton, [{
      key: "render",

      /**
       * LitElement render
       */
      value: function render() {
        return (0, _litElement.html)(_templateObject_868cd7a030c111eab3ab336d9c8a6f6e(), this.url, this.open ? (0, _litElement.html)(_templateObject2_868cd7a030c111eab3ab336d9c8a6f6e(), this.title) : "", this.complete ? (0, _litElement.html)(_templateObject3_868cd7a030c111eab3ab336d9c8a6f6e(), this.title) : "");
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_868cd7a030c111eab3ab336d9c8a6f6e())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "lrn-assignment-button";
      }
    }]);

    function LrnAssignmentButton() {
      var _this;

      babelHelpers.classCallCheck(this, LrnAssignmentButton);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrnAssignmentButton).call(this));
      _this.open = false;
      _this.complete = false;
      return _this;
    }

    babelHelpers.createClass(LrnAssignmentButton, null, [{
      key: "properties",
      get: function get() {
        return {
          title: {
            type: String
          },
          url: {
            type: String
          },
          open: {
            type: Boolean
          },
          complete: {
            type: Boolean
          }
        };
      }
    }]);
    return LrnAssignmentButton;
  }(_litElement.LitElement);

  _exports.LrnAssignmentButton = LrnAssignmentButton;
  window.customElements.define(LrnAssignmentButton.tag, LrnAssignmentButton);
});
