// @haxtheweb/image-gallery/image-gallery.js
/**
 * Copyright 2026 haxtheweb
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as e,html as t,css as a}from"../../lit/index.js";import{DDD as i}from"../d-d-d/d-d-d.js";import{I18NMixin as r}from"../i18n-manager/lib/I18NMixin.js";import"../simple-icon/lib/simple-icon-button-lite.js";export class ImageGallery extends(r(i)){static get tag(){return"image-gallery"}constructor(){super(),this.edit=!1,this.activeIndex=0,this.mode="masonry",this.images=[],this._haxState=!1,this._newImageElement=null,this._newImageTransitionName=null,this.t=this.t||{},this.t={...this.t,gallery:"Gallery",previous:"Previous",next:"Next",editModeLabel:"Image gallery edit mode",openImage:"Open image"},this.registerLocalization({context:this,localesPath:new URL("./locales/image-gallery.ar.json",import.meta.url).href+"/../"}),this._observer=new MutationObserver(()=>{clearTimeout(this._debounceMutations),this._debounceMutations=setTimeout(()=>{this._updateImages()},100)})}connectedCallback(){super.connectedCallback(),this._observer.observe(this,{childList:!0,attributes:!0,subtree:!0,attributeFilter:["src","alt","source"]}),this.addEventListener("keydown",this._handleKeydown),this.addEventListener("drop",this._onDrop,!0)}disconnectedCallback(){this._observer.disconnect(),this.removeEventListener("keydown",this._handleKeydown),this.removeEventListener("drop",this._onDrop,!0),super.disconnectedCallback()}static get properties(){return{...super.properties,edit:{type:Boolean,reflect:!0},mode:{type:String,reflect:!0},activeIndex:{type:Number,reflect:!0,attribute:"active-index"},images:{type:Array},_haxState:{type:Boolean}}}setActiveIndex(e){this._setActiveIndex(e)}static get styles(){return[super.styles,a`
        :host {
          display: block;
          font-family: var(--ddd-font-navigation);
        }

        /* Edit mode */
        .edit-wrapper {
          border: var(--ddd-border-md);
          border-style: dashed;
          padding: var(--ddd-spacing-4);
          background-color: var(--ddd-theme-default-limestoneLight);
          box-sizing: border-box;
        }
        .edit-wrapper ::slotted(*) {
          display: block;
          width: 100%;
          margin-bottom: var(--ddd-spacing-2);
        }

        /* Grid mode */
        .grid-layout {
          display: grid;
          grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
          gap: var(--ddd-spacing-2);
        }

        .grid-item {
          position: relative;
          cursor: pointer;
          overflow: hidden;
          border-radius: var(--ddd-radius-sm);
          background-color: var(--ddd-theme-default-limestoneLight);
          aspect-ratio: 1;
          border: none;
          padding: 0;
        }

        .grid-item img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          transition: transform var(--ddd-duration-normal)
            var(--ddd-timing-ease);
        }

        .grid-item:hover img,
        .grid-item:focus img {
          transform: scale(1.05);
        }

        .grid-item:focus {
          outline: var(--ddd-border-lg);
          outline-color: var(--ddd-theme-primary);
          outline-offset: var(--ddd-spacing-1);
        }

        /* Masonry mode */
        .masonry-layout {
          column-count: 3;
          column-gap: var(--ddd-spacing-2);
        }

        @media (max-width: 900px) {
          .masonry-layout {
            column-count: 2;
          }
        }

        @media (max-width: 600px) {
          .masonry-layout {
            column-count: 1;
          }
        }

        .masonry-item {
          break-inside: avoid;
          margin-bottom: var(--ddd-spacing-2);
          border-radius: var(--ddd-radius-sm);
          overflow: hidden;
          background-color: var(--ddd-theme-default-limestoneLight);
          cursor: pointer;
          border: none;
          padding: 0;
          display: block;
          width: 100%;
        }

        .masonry-item img {
          width: 100%;
          display: block;
          transition: transform var(--ddd-duration-normal)
            var(--ddd-timing-ease);
        }

        .masonry-item:hover img,
        .masonry-item:focus img {
          transform: scale(1.05);
        }

        .masonry-item:focus {
          outline: var(--ddd-border-lg);
          outline-color: var(--ddd-theme-primary);
          outline-offset: var(--ddd-spacing-1);
        }

        /* Gallery mode */
        .gallery-layout {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-2);
          background: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
          border-radius: var(--ddd-radius-sm);
          overflow: hidden;
          padding: var(--ddd-spacing-2);
        }

        .gallery-main {
          position: relative;
          display: flex;
          align-items: center;
          justify-content: center;
          background: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-black)
          );
          min-height: 300px;
          border-radius: var(--ddd-radius-sm);
          overflow: hidden;
        }

        .gallery-main img {
          max-width: 100%;
          max-height: 60vh;
          object-fit: contain;
          display: block;
        }

        .gallery-caption {
          color: light-dark(
            var(--ddd-theme-default-black),
            var(--ddd-theme-default-white)
          );
          font-size: var(--ddd-font-size-s);
          font-family: var(--ddd-font-primary);
          text-align: center;
          padding: var(--ddd-spacing-2) var(--ddd-spacing-4);
          min-height: var(--ddd-spacing-6);
        }

        .gallery-nav-wrap {
          display: flex;
          justify-content: space-between;
          align-items: center;
          position: absolute;
          top: 50%;
          left: 0;
          right: 0;
          transform: translateY(-50%);
          pointer-events: none;
          z-index: 1;
          padding: 0 var(--ddd-spacing-2);
        }

        .gallery-nav-wrap simple-icon-button-lite {
          pointer-events: all;
          color: var(--ddd-theme-primary);
          --simple-icon-width: var(--ddd-icon-2xl);
          --simple-icon-height: var(--ddd-icon-2xl);
          height: var(--ddd-icon-2xl);
          width: var(--ddd-icon-2xl);
        }

        .gallery-thumbnails {
          display: flex;
          gap: var(--ddd-spacing-1);
          overflow-x: auto;
          padding: var(--ddd-spacing-1) 0;
          scrollbar-width: thin;
        }

        .gallery-thumbnail {
          flex: 0 0 auto;
          width: var(--ddd-icon-xl);
          height: var(--ddd-icon-xl);
          border: var(--ddd-border-sm);
          border-color: transparent;
          border-radius: var(--ddd-radius-sm);
          overflow: hidden;
          cursor: pointer;
          background: light-dark(
            var(--ddd-theme-default-limestoneLight),
            var(--ddd-theme-default-coalyGray)
          );
          transition: border-color var(--ddd-duration-fast)
            var(--ddd-timing-ease);
          padding: 0;
        }

        .gallery-thumbnail img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          display: block;
          pointer-events: none;
        }

        .gallery-thumbnail:hover,
        .gallery-thumbnail:focus {
          border-color: var(--ddd-theme-primary);
        }

        .gallery-thumbnail.active {
          border-color: var(--ddd-theme-primary);
        }

        .gallery-thumbnail:focus {
          outline: var(--ddd-border-lg);
          outline-color: var(--ddd-theme-primary);
          outline-offset: var(--ddd-spacing-1);
        }

        .slot-hidden {
          display: none;
        }

        /* Reduced motion */
        @media (prefers-reduced-motion: reduce) {
          .grid-item img,
          .masonry-item img {
            transition: none;
          }
          .gallery-nav-wrap,
          .gallery-thumbnail {
            transition: none;
          }
        }

        /* Dark mode */
        @media (prefers-color-scheme: dark) {
          .edit-wrapper {
            background-color: var(--ddd-theme-default-coalyGray);
          }
          .grid-item {
            background-color: var(--ddd-theme-default-coalyGray);
          }
          .masonry-item {
            background-color: var(--ddd-theme-default-coalyGray);
          }
        }
      `]}render(){if(this.edit)return t`
        <div class="edit-wrapper" aria-label="${this.t.editModeLabel}">
          <slot></slot>
        </div>
      `;switch(this.mode){case"masonry":return this._renderMasonry();case"gallery":return this._renderGallery();default:return this._renderGrid()}}_renderGrid(){return t`
      <div class="grid-layout">
        ${this.images.map((e,a)=>t`
            <button
              class="grid-item"
              style="${e.element===this._newImageElement?`view-transition-name: ${this._newImageTransitionName};`:""}"
              aria-label="${this.t.openImage}: ${e.alt||""}"
              title="${this.t.openImage}: ${e.alt||""}"
              @click="${t=>this._handleImageClick(t,e,a)}"
              @dblclick="${()=>this._handleDblClick(a)}"
              @keydown="${t=>{"Enter"!==t.key&&" "!==t.key||(t.preventDefault(),this._handleImageClick(t,e,a))}}"
            >
              <img src="${e.src}" alt="${e.alt}" loading="lazy" />
            </button>
          `)}
      </div>
      <div class="slot-hidden"><slot></slot></div>
    `}_renderMasonry(){return t`
      <div class="masonry-layout">
        ${this.images.map((e,a)=>t`
            <button
              class="masonry-item"
              style="${e.element===this._newImageElement?`view-transition-name: ${this._newImageTransitionName};`:""}"
              aria-label="${this.t.openImage}: ${e.alt||""}"
              title="${this.t.openImage}: ${e.alt||""}"
              @click="${t=>this._handleImageClick(t,e,a)}"
              @dblclick="${()=>this._handleDblClick(a)}"
            >
              <img src="${e.src}" alt="${e.alt}" loading="lazy" />
            </button>
          `)}
      </div>
      <div class="slot-hidden"><slot></slot></div>
    `}_renderGallery(){const e=this.images[this.activeIndex];return t`
      <div class="gallery-layout">
        <div class="gallery-main">
          ${e?t`
                <div class="gallery-nav-wrap">
                  <simple-icon-button-lite
                    id="prev-btn"
                    icon="hardware:keyboard-arrow-left"
                    label="${this.t.previous}"
                    title="${this.t.previous}"
                    ?disabled="${0===this.activeIndex}"
                    @click="${()=>this._setActiveIndex(this.activeIndex-1)}"
                  ></simple-icon-button-lite>
                  <simple-icon-button-lite
                    id="next-btn"
                    icon="hardware:keyboard-arrow-right"
                    label="${this.t.next}"
                    title="${this.t.next}"
                    ?disabled="${this.activeIndex===this.images.length-1}"
                    @click="${()=>this._setActiveIndex(this.activeIndex+1)}"
                  ></simple-icon-button-lite>
                </div>
                <img
                  src="${e.src}"
                  alt="${e.alt}"
                  @dblclick="${()=>this._handleDblClick(this.activeIndex)}"
                />
              `:t``}
        </div>
        ${e&&e.alt?t` <div class="gallery-caption">${e.alt}</div> `:t``}
        <div
          class="gallery-thumbnails"
          role="tablist"
          aria-label="${this.t.gallery}"
        >
          ${this.images.map((e,a)=>t`
              <button
                class="gallery-thumbnail ${a===this.activeIndex?"active":""}"
                style="${e.element===this._newImageElement?`view-transition-name: ${this._newImageTransitionName};`:""}"
                role="tab"
                aria-selected="${a===this.activeIndex?"true":"false"}"
                aria-label="${e.alt||""}"
                title="${e.alt||""}"
                @click="${e=>this._handleThumbnailClick(e,a)}"
                @dblclick="${()=>this._handleDblClick(a)}"
              >
                <img src="${e.src}" alt="" loading="lazy" />
              </button>
            `)}
        </div>
      </div>
      <div class="slot-hidden"><slot></slot></div>
    `}_updateImages(){const e=Array.from(this.children).filter(e=>"MEDIA-IMAGE"===e.tagName||"IMG"===e.tagName);this.activeIndex>=e.length&&(this.activeIndex=0),this.images=e.map(e=>{let t="",a="";return"MEDIA-IMAGE"===e.tagName?(t=e.source||e.getAttribute("source")||"",a=e.alt||e.getAttribute("alt")||""):"IMG"===e.tagName&&(t=e.getAttribute("src")||e.src||"",a=e.getAttribute("alt")||e.alt||""),{element:e,tagName:e.tagName,src:t,alt:a}})}_setActiveIndex(e){if(0===this.images.length)return void(this.activeIndex=0);const t=Math.max(0,Math.min(e,this.images.length-1));this.activeIndex=t}_openModal(e){const t=globalThis.document.createElement("div");t.style.display="flex",t.style.alignItems="center",t.style.justifyContent="center",t.style.width="100%",t.style.height="100%",t.style.overflow="hidden";const a=globalThis.document.createElement("img");a.src=e.src,a.alt=e.alt||"",a.style.maxWidth="100%",a.style.maxHeight="100%",a.style.objectFit="contain",a.style.display="block",t.appendChild(a);const i=new CustomEvent("simple-modal-show",{bubbles:!0,composed:!0,cancelable:!0,detail:{title:e.alt||"",elements:{content:t},styles:{"--simple-modal-width":"80vw","--simple-modal-height":"80vh","--simple-modal-min-width":"50vw","--simple-modal-min-height":"50vh","--simple-modal-content-padding":"0px"},invokedBy:this,clone:!1}});this.dispatchEvent(i)}_handleImageClick(e,t,a){this._setActiveIndex(a),this._haxState||this._openModal(t)}_handleThumbnailClick(e,t){this._setActiveIndex(t)}_handleKeydown(e){"gallery"!==this.mode||this.edit||("ArrowLeft"===e.key?(e.preventDefault(),this._setActiveIndex(this.activeIndex-1)):"ArrowRight"===e.key&&(e.preventDefault(),this._setActiveIndex(this.activeIndex+1)))}_getHaxStore(){if(globalThis.HaxStore&&globalThis.HaxStore.requestAvailability)try{return globalThis.HaxStore.requestAvailability()}catch(e){return null}return null}_onDrop(e){if(!this._haxState)return;const t=this._getHaxStore();if(!t||!t.__dragTarget)return;const a=t.__dragTarget,i=a.tagName;if("MEDIA-IMAGE"!==i&&"IMG"!==i)return e.stopPropagation(),e.stopImmediatePropagation(),e.preventDefault(),void(t.__dragTarget=null);if(!this.contains(a)){this._newImageElement=a,this._newImageTransitionName=`gallery-drop-${Date.now()}-${Math.random().toString(36).substring(2,9)}`;const doAppend=()=>{this.appendChild(a),this._updateImages()};if(!(globalThis.matchMedia&&globalThis.matchMedia("(prefers-reduced-motion: reduce)").matches)&&globalThis.document&&globalThis.document.startViewTransition){globalThis.document.startViewTransition(()=>{doAppend()}).finished.then(()=>{this._newImageElement=null,this._newImageTransitionName=null,this.requestUpdate()}).catch(()=>{this._newImageElement=null,this._newImageTransitionName=null,this.requestUpdate()})}else doAppend(),this._newImageElement=null,this._newImageTransitionName=null}e.stopPropagation(),e.stopImmediatePropagation(),e.preventDefault(),t.__dragTarget=null}async _handleDblClick(e){this._setActiveIndex(e);const t=this._getHaxStore();this._haxState&&t&&this.images[e]&&this.images[e].element&&(t.activeNode=this.images[e].element)}updated(e){super.updated&&super.updated(e),e.forEach((e,t)=>{"edit"===t&&(this.edit||!0!==e||this._updateImages())})}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),this._updateImages()}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}haxHooks(){return{inlineContextMenu:"haxinlineContextMenu",editModeChanged:"haxeditModeChanged",activeElementChanged:"haxactiveElementChanged",preProcessNodeToContent:"haxpreProcessNodeToContent"}}haxeditModeChanged(e){this._haxState=e;Array.from(this.children).filter(e=>"MEDIA-IMAGE"===e.tagName||"IMG"===e.tagName).forEach(t=>{t.haxeditModeChanged&&"function"==typeof t.haxeditModeChanged&&t.haxeditModeChanged(e)})}haxactiveElementChanged(e,t){t&&(this._haxState=t)}haxpreProcessNodeToContent(e){if(e&&e.tagName&&"image-gallery"===e.tagName.toLowerCase()){const t=e.cloneNode(!0);return t.removeAttribute("edit"),t}return e}haxinlineContextMenu(e){e.ceButtons=[{icon:"icons:open-in-new",callback:"haxBreakOutGallery",label:"Break items out of gallery"},{icon:"lrn:edit",callback:"haxToggleEdit",label:"Toggle edit mode"},{icon:"image:photo-camera",callback:"haxAddImage",label:"Add image"},{icon:"icons:delete",callback:"haxRemoveLastImage",label:"Remove last image"}]}haxToggleEdit(e){return this.edit=!this.edit,!0}haxAddImage(e){const t=globalThis.document.createElement("media-image");return t.setAttribute("source","https://dummyimage.com/300x200/000/fff"),t.setAttribute("alt","New image"),this.appendChild(t),!0}haxRemoveLastImage(e){const t=Array.from(this.children).filter(e=>"MEDIA-IMAGE"===e.tagName||"IMG"===e.tagName);return t.length>0&&t[t.length-1].remove(),!0}haxBreakOutGallery(e){if(!this.parentNode)return!1;let t;return Array.from(this.children).filter(e=>"MEDIA-IMAGE"===e.tagName||"IMG"===e.tagName).forEach(e=>{t=e.cloneNode(!0),this.getAttribute("slot")?t.setAttribute("slot",this.getAttribute("slot")):t.removeAttribute("slot"),this.parentNode.insertBefore(t,this)}),!!t&&(this.remove(),!0)}}globalThis.customElements.define(ImageGallery.tag,ImageGallery);
