// @lrnwebcomponents/lrndesign-sidenote/lrndesign-sidenote.js
define(["exports", "../../lit-element/lit-element.js", "../hax-body-behaviors/lib/HAXWiring.js", "../../@polymer/iron-icons/iron-icons.js", "../../@polymer/iron-icon/iron-icon.js"], function (_exports, _litElement, _HAXWiring, _ironIcons, _ironIcon) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LrndesignSidenote = void 0;

  function _templateObject_c39e4a80d55811e9a4251fbb0db90cd4() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <style>\n        :host {\n          display: block;\n          --container-bg-color: lightgray;\n          --container-text-color: black;\n          --container-padding: 16px;\n          --container-outset: 0;\n          @apply --host-styles;\n        }\n\n        #container {\n          display: block;\n          background: var(--container-bg-color);\n          color: var(--container-text-color);\n          padding: var(--container-padding);\n          margin-left: -var(--container-outset);\n          @apply --container-styles;\n        }\n\n        #header {\n          display: flex;\n          align-items: center;\n          @apply --container-header;\n        }\n\n        #icon {\n          margin-right: 8px;\n          @apply --icon-styles;\n        }\n\n        #label {\n          font-size: 20.8px;\n          margin: 12.8px 0;\n          flex: 1 1 auto;\n          @apply --label-styles;\n        }\n      </style>\n      <div id=\"container\">\n        <div id=\"header\">\n          <iron-icon id=\"icon\" icon=", "></iron-icon>\n          <div id=\"label\">", "</div>\n        </div>\n        <slot></slot>\n      </div>\n    "]);

    _templateObject_c39e4a80d55811e9a4251fbb0db90cd4 = function _templateObject_c39e4a80d55811e9a4251fbb0db90cd4() {
      return data;
    };

    return data;
  }

  /**
  `lrndesign-sidenote`
  A basic side note
  
  * @demo demo/index.html
  */
  var LrndesignSidenote =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(LrndesignSidenote, _LitElement);
    babelHelpers.createClass(LrndesignSidenote, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(LrndesignSidenote.prototype), "connectedCallback", this).call(this);
        this.HAXWiring = new _HAXWiring.HAXWiring();
        this.HAXWiring.setup(LrndesignSidenote.haxProperties, LrndesignSidenote.tag, this);
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          label: {
            type: String
          },
          icon: {
            type: String
          },
          bgColor: {
            type: String
          }
        };
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return {
          canScale: false,
          canPosition: true,
          canEditSource: false,
          gizmo: {
            title: "Side-Note",
            description: "A .",
            icon: "icons:bookmark",
            color: "blue",
            groups: [""],
            meta: {
              author: "LRNWebComponents"
            }
          },
          settings: {
            quick: [{
              property: "label",
              title: "Label",
              description: "The label of the sidenote.",
              inputMethod: "textfield",
              icon: "editor:title"
            }],
            configure: [{
              property: "icon",
              title: "Icon",
              description: "The icon of the sidenote.",
              inputMethod: "iconpicker",
              options: ["icons:announcement", "icons:book", "icons:bookmark", "icons:check-circle", "icons:feedback", "icons:thumb-down", "icons:thumb-up", "icons:warning"]
            }, {
              property: "bgColor",
              title: "Color",
              description: "The background color of the sidenote.",
              inputMethod: "colorpicker",
              icon: "editor:format-color-fill"
            }],
            advanced: []
          }
        };
      }
    }]);

    function LrndesignSidenote() {
      var _this;

      babelHelpers.classCallCheck(this, LrndesignSidenote);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(LrndesignSidenote).call(this));
      _this.label = "";
      _this.icon = "";
      _this.bgColor = "";
      return _this;
    }

    babelHelpers.createClass(LrndesignSidenote, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_c39e4a80d55811e9a4251fbb0db90cd4(), this.icon, this.label);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "lrndesign-sidenote";
      }
    }]);
    return LrndesignSidenote;
  }(_litElement.LitElement);

  _exports.LrndesignSidenote = LrndesignSidenote;
  customElements.define("lrndesign-sidenote", LrndesignSidenote);
});
