// @haxtheweb/haxcms-elements/lib/ui-components/utilities/site-available-themes.js
import{html as e,css as t}from"../../../../../lit/index.js";import{DDD as i}from"../../../../d-d-d/d-d-d.js";import{I18NMixin as a}from"../../../../i18n-manager/lib/I18NMixin.js";import"../../../../simple-icon/lib/simple-icon-lite.js";import"../../../../simple-icon/lib/simple-icons.js";export class SiteAvailableThemes extends(a(i)){static get tag(){return"site-available-themes"}constructor(){super(),this.themes=[],this.loading=!0,this.error=null,this.viewMode="gallery",this.showDetails=!0,this.showAllThemes=!1,this.columns=3,this.currentTheme="",this.t=this.t||{},this.registerLocalization({context:this,localesPath:new URL("../../../../../../locales/site-available-themes/",import.meta.url).href+"/",locales:["en"]})}static get properties(){return{...super.properties,themes:{type:Array},loading:{type:Boolean},error:{type:String},viewMode:{type:String,attribute:"view-mode"},showDetails:{type:Boolean,attribute:"show-details"},showAllThemes:{type:Boolean,attribute:"show-all-themes"},columns:{type:Number},currentTheme:{type:String,attribute:"current-theme"}}}static get styles(){return[super.styles,t`
        :host {
          display: block;
          --theme-gallery-gap: var(--ddd-spacing-4);
          --theme-gallery-border-radius: var(--ddd-radius-xs);
          --theme-gallery-border: var(--ddd-border-xs);
          --theme-card-width: 180px;
          --theme-card-height: 300px;
          --theme-card-preview-height: 240px;
        }
        .toggle-button simple-icon-lite {
          --simple-icon-width: var(--ddd-icon-4xs);
          --simple-icon-height: var(--ddd-icon-4xs);
        }

        .loading,
        .error {
          text-align: center;
          padding: var(--ddd-spacing-8);
          color: var(--ddd-theme-default-coalyGray);
        }

        .error {
          color: var(--ddd-theme-default-original87Pink);
          background: var(--ddd-theme-default-opportunityGreen);
          border-radius: var(--theme-gallery-border-radius);
          padding: var(--ddd-spacing-4);
        }

        .gallery-header {
          display: flex;
          justify-content: space-between;
          align-items: center;
          margin-bottom: var(--ddd-spacing-4);
          padding: var(--ddd-spacing-2) 0;
          border-bottom: var(--theme-gallery-border);
        }

        .gallery-title {
          margin: 0;
          color: var(--ddd-theme-default-coalyGray);
          font-size: var(--ddd-font-size-s);
          font-family: var(--ddd-font-navigation);
        }

        .view-toggle {
          display: flex;
          gap: var(--ddd-spacing-2);
        }

        .toggle-button {
          background: var(--ddd-theme-default-white);
          border: var(--theme-gallery-border);
          border-radius: var(--theme-gallery-border-radius);
          padding: var(--ddd-spacing-2) var(--ddd-spacing-3);
          cursor: pointer;
          transition: all 0.2s ease;
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          font-size: var(--ddd-font-size-4xs);
          font-family: var(--ddd-font-navigation);
        }

        .toggle-button:hover,
        .toggle-button.active {
          background: var(--ddd-theme-default-skyBlue);
          color: var(--ddd-theme-default-white);
        }

        .theme-count {
          color: var(--ddd-theme-default-coalyGray);
          font-size: var(--ddd-font-size-4xs);
        }

        /* Gallery View */
        .gallery-grid {
          display: grid;
          grid-template-columns: repeat(var(--columns, 3), 1fr);
          gap: var(--theme-gallery-gap);
        }

        .theme-card {
          border: var(--theme-gallery-border);
          border-radius: var(--theme-gallery-border-radius);
          overflow: hidden;
          background: var(--ddd-theme-default-white);
          transition:
            transform 0.2s ease,
            box-shadow 0.2s ease;
          position: relative;
          width: 100%;
          max-width: var(--theme-card-width);
          min-height: var(--theme-card-height);
          margin: 0 auto;
        }

        .theme-thumbnail img {
          width: 85%;
          height: 85%;
          object-fit: cover;
          object-position: left;
          display: block;
        }

        .theme-card:hover {
          transform: translateY(-2px);
          box-shadow: var(--ddd-boxShadow-md);
          border-color: var(--ddd-theme-default-beaverBlue);
        }

        .theme-card.current {
          border-color: var(--ddd-theme-default-skyBlue);
          border-width: 2px;
        }

        .theme-thumbnail {
          width: 100%;
          height: var(--theme-card-preview-height);
          position: relative;
          display: flex;
          align-items: center;
          justify-content: center;
          border-bottom: var(--theme-gallery-border);
          background: var(--ddd-theme-default-white);
        }

        .current-badge {
          position: absolute;
          top: var(--ddd-spacing-2);
          right: var(--ddd-spacing-2);
          background: var(--ddd-theme-default-keystoneYellow);
          color: var(--ddd-theme-default-coalyGray);
          padding: var(--ddd-spacing-1) var(--ddd-spacing-2);
          border-radius: var(--ddd-radius-xs);
          font-size: var(--ddd-font-size-5xs);
          font-family: var(--ddd-font-navigation);
          z-index: 1;
        }

        .theme-info {
          padding: var(--ddd-spacing-2);
        }

        .theme-name {
          font-weight: var(--ddd-font-weight-medium);
          color: var(--ddd-theme-default-coalyGray);
          margin: 0 0 var(--ddd-spacing-1) 0;
          font-size: var(--ddd-font-size-4xs);
          font-family: var(--ddd-font-navigation);
        }

        .theme-description {
          font-size: var(--ddd-font-size-5xs);
          color: var(--ddd-theme-default-slateGray);
          margin: 0 0 var(--ddd-spacing-1) 0;
        }

        .theme-element {
          font-size: var(--ddd-font-size-5xs);
          color: var(--ddd-theme-default-limestoneGray);
          font-family: var(--ddd-font-navigation);
          margin-bottom: var(--ddd-spacing-1);
        }

        .preview-button {
          background: var(--ddd-theme-default-link);
          color: var(--ddd-theme-default-white);
          border: var(--ddd-border-xs);
          border-color: transparent;
          border-radius: var(--theme-gallery-border-radius);
          padding: var(--ddd-spacing-2) var(--ddd-spacing-4);
          cursor: pointer;
          font-size: var(--ddd-font-size-5xs);
          display: inline-flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          transition: all 0.2s ease;
          font-family: var(--ddd-font-navigation);
        }

        .preview-button:hover {
          background: var(--ddd-theme-default-beaverBlue);
        }

        .preview-button:disabled {
          background: var(--ddd-theme-default-limestoneGray);
          cursor: not-allowed;
        }
        .preview-button simple-icon-lite {
          --simple-icon-width: var(--ddd-icon-5xs);
          --simple-icon-height: var(--ddd-icon-5xs);
        }

        /* Table View */
        .theme-table {
          width: 100%;
          border-collapse: collapse;
          border: var(--theme-gallery-border);
          border-radius: var(--theme-gallery-border-radius);
          overflow: hidden;
        }

        .theme-table th,
        .theme-table td {
          padding: var(--ddd-spacing-4);
          text-align: left;
          border-bottom: var(--theme-gallery-border);
        }

        .theme-table th {
          background: var(--ddd-theme-default-limestoneLight);
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
          font-size: var(--ddd-font-size-xs);
        }

        .theme-table tr:hover {
          background: var(--ddd-theme-default-limestoneLight);
        }

        .theme-table tr.current {
          background: var(--ddd-theme-default-skyMaxlight);
        }

        .table-thumbnail {
          width: 90px;
          height: 60px;
          border: var(--ddd-border-xs);
          border-color: var(--ddd-theme-default-limestoneGray);
          border-radius: var(--ddd-radius-xs);
          overflow: hidden;
          background: var(--ddd-theme-default-white);
        }

        .table-thumbnail img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          display: block;
        }

        .table-name {
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
        }

        .table-element {
          font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-5xs);
          color: var(--ddd-theme-default-limestoneGray);
        }

        .current-label {
          color: var(--ddd-theme-default-link);
          font-size: var(--ddd-font-size-5xs);
          font-family: var(--ddd-font-navigation);
        }

        /* Responsive */
        @media (max-width: 768px) {
          .gallery-grid {
            grid-template-columns: repeat(2, 1fr);
          }

          .gallery-header {
            flex-direction: column;
            gap: var(--ddd-spacing-2);
            align-items: flex-start;
          }

          .theme-table {
            font-size: var(--ddd-font-size-xs);
          }

          .table-thumbnail {
            width: 72px;
            height: 46px;
          }
        }

        @media (max-width: 480px) {
          .gallery-grid {
            grid-template-columns: 1fr;
          }
        }
      `]}updated(e){super.updated(e),e.has("columns")&&this.style.setProperty("--columns",this.columns)}connectedCallback(){super.connectedCallback(),this.loadThemes(),this.detectCurrentTheme()}_isTruthy(e){return!0===e||"true"===e||1===e||"1"===e}_resolveThemeThumbnail(e=""){if(!e||"string"!=typeof e)return"";if(0===e.indexOf("@haxtheweb/")){const t="/@haxtheweb/",i=import.meta.url.indexOf(t);if(-1!==i){return`${import.meta.url.substring(0,i+t.length)}${e.replace("@haxtheweb/","")}`}let a="";if(globalThis.WCAutoloadBasePath&&"string"==typeof globalThis.WCAutoloadBasePath?a=globalThis.WCAutoloadBasePath:globalThis.WCGlobalBasePath&&"string"==typeof globalThis.WCGlobalBasePath&&(a=globalThis.WCGlobalBasePath),a)return"/"!==a.charAt(a.length-1)&&(a+="/"),`${a}${e}`;return new URL(`../../../../${e}`,import.meta.url).href}return e}_getVisibleThemes(){return this.themes.filter(e=>!!e&&(e.element===this.currentTheme||(!!this.showAllThemes||!(e.hidden||e.terrible||e.legacy))))}async loadThemes(){this.loading=!0,this.error=null;try{const e=new URL("../../themes.json",import.meta.url),t=await fetch(e);if(!t.ok)throw new Error(`HTTP error! status: ${t.status}`);const i=await t.json();this.themes=Object.entries(i).map(([e,t])=>{const i=t&&"object"==typeof t?t:{element:e},a=i.element||e,r="number"==typeof i.priority?i.priority:Number(i.priority),d=`@haxtheweb/haxcms-elements/lib/theme-screenshots/theme-${a}-thumb.jpg`,l=i.thumbnail||d,s=i.screenshot||i.thumbnail||d;return{key:e,...i,element:a,name:i.name||a,description:i.description||"",hidden:this._isTruthy(i.hidden),terrible:this._isTruthy(i.terrible)||0===e.indexOf("terrible"),legacy:this._isTruthy(i.legacy)||this._isTruthy(i.deprecated)||this._isTruthy(i.isLegacy),priority:Number.isFinite(r)?r:0,thumbnailPath:this._resolveThemeThumbnail(l),fullImagePath:this._resolveThemeThumbnail(s)}}).sort((e,t)=>e.priority!==t.priority?e.priority-t.priority:e.name.localeCompare(t.name)),this.loading=!1}catch(e){this.error=`Failed to load themes: ${e.message}`,this.loading=!1,console.error("Error loading themes:",e)}}detectCurrentTheme(){globalThis.HAXCMS&&globalThis.HAXCMS.instance&&globalThis.HAXCMS.instance.store&&globalThis.HAXCMS.instance.store.manifest&&globalThis.HAXCMS.instance.store.manifest.metadata&&globalThis.HAXCMS.instance.store.manifest.metadata.theme&&globalThis.HAXCMS.instance.store.manifest.metadata.theme.element&&(this.currentTheme=globalThis.HAXCMS.instance.store.manifest.metadata.theme.element)}async switchTheme(e){if(globalThis.HAXCMS&&globalThis.HAXCMS.setTheme)try{await globalThis.HAXCMS.setTheme(e),this.currentTheme=e,this.dispatchEvent(new CustomEvent("theme-changed",{bubbles:!0,composed:!0,detail:{themeElement:e}}))}catch(e){console.error("Failed to switch theme:",e)}else console.warn("HAXCMS theme switching not available")}setViewMode(e){this.viewMode=e}renderGalleryView(){const t=this._getVisibleThemes();return 0===t.length?e`<div class="loading">No themes available for this view.</div>`:e`
      <div class="gallery-grid">
        ${t.map(t=>e`
            <div
              class="theme-card ${t.element===this.currentTheme?"current":""}"
            >
              ${t.element===this.currentTheme?e` <div class="current-badge">Active</div> `:""}
              <div class="theme-thumbnail">
                <img
                  src="${t.thumbnailPath||t.fullImagePath}"
                  alt=""
                  loading="lazy"
                  decoding="async"
                />
              </div>
              ${this.showDetails?e`
                    <div class="theme-info">
                      <h3 class="theme-name">${t.name}</h3>
                      ${t.description?e`
                            <p class="theme-description">
                              ${t.description}
                            </p>
                          `:""}
                      <div class="theme-element">&lt;${t.element}&gt;</div>
                      <button
                        class="preview-button"
                        type="button"
                        aria-label="${t.element===this.currentTheme?`${t.name} theme is active`:`Apply ${t.name} theme`}"
                        ?disabled="${t.element===this.currentTheme}"
                        @click="${()=>this.switchTheme(t.element)}"
                      >
                        <simple-icon-lite icon="visibility"></simple-icon-lite>
                        ${t.element===this.currentTheme?"Active":"Apply"}
                      </button>
                    </div>
                  `:""}
            </div>
          `)}
      </div>
    `}renderTableView(){const t=this._getVisibleThemes();return 0===t.length?e`<div class="loading">No themes available for this view.</div>`:e`
      <table class="theme-table">
        <thead>
          <tr>
            <th>Preview</th>
            <th>Name</th>
            <th>Element</th>
            ${this.showDetails?e`<th>Description</th>`:""}
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          ${t.map(t=>e`
              <tr
                class="${t.element===this.currentTheme?"current":""}"
              >
                <td>
                  <div class="table-thumbnail">
                    <img
                      src="${t.thumbnailPath||t.fullImagePath}"
                      alt=""
                      loading="lazy"
                      decoding="async"
                    />
                  </div>
                </td>
                <td>
                  <div class="table-name">${t.name}</div>
                  ${t.element===this.currentTheme?e` <div class="current-label">Active</div> `:""}
                </td>
                <td>
                  <div class="table-element">&lt;${t.element}&gt;</div>
                </td>
                ${this.showDetails?e`<td>${t.description||""}</td>`:""}
                <td>
                  <button
                    class="preview-button"
                    type="button"
                    aria-label="${t.element===this.currentTheme?`${t.name} theme is active`:`Apply ${t.name} theme`}"
                    ?disabled="${t.element===this.currentTheme}"
                    @click="${()=>this.switchTheme(t.element)}"
                  >
                    <simple-icon-lite icon="visibility"></simple-icon-lite>
                    ${t.element===this.currentTheme?"Active":"Apply"}
                  </button>
                </td>
              </tr>
            `)}
        </tbody>
      </table>
    `}render(){if(this.loading)return e`<div class="loading">Loading themes...</div>`;if(this.error)return e`<div class="error">${this.error}</div>`;const t=this._getVisibleThemes(),i=this.showAllThemes?`${t.length} themes shown`:`${t.length} themes shown (app-hax defaults)`;return e`
      <div class="gallery-header">
        <div>
          <h2 class="gallery-title">Available HAXcms Themes</h2>
          <div class="theme-count">${i}</div>
        </div>
        <div class="view-toggle">
          <button
            class="toggle-button ${"gallery"===this.viewMode?"active":""}"
            type="button"
            aria-pressed="${"gallery"===this.viewMode?"true":"false"}"
            @click="${()=>this.setViewMode("gallery")}"
          >
            <simple-icon-lite icon="view-module"></simple-icon-lite>
            Gallery
          </button>
          <button
            class="toggle-button ${"table"===this.viewMode?"active":""}"
            type="button"
            aria-pressed="${"table"===this.viewMode?"true":"false"}"
            @click="${()=>this.setViewMode("table")}"
          >
            <simple-icon-lite icon="view-list"></simple-icon-lite>
            Table
          </button>
        </div>
      </div>

      ${"gallery"===this.viewMode?this.renderGalleryView():this.renderTableView()}
    `}static get haxProperties(){return{canScale:!1,canPosition:!1,canEditSource:!0,contentEditable:!1,gizmo:{title:"Available Themes",description:"Display available HAXcms themes with live preview switching",icon:"image:collections",color:"purple",tags:["Other","Theme","Gallery","HAXcms","Developer","Documentation"],handles:[],meta:{author:"HAXTheWeb team",owner:"The Pennsylvania State University"}},settings:{configure:[{property:"viewMode",title:"View Mode",description:"Choose how to display the themes",inputMethod:"select",options:{gallery:"Gallery",table:"Table"}},{property:"showDetails",title:"Show Details",description:"Show theme descriptions and additional information",inputMethod:"boolean"},{property:"showAllThemes",title:"Show Hidden and Legacy Themes",description:"Show hidden, legacy, and terrible themes in addition to dashboard defaults",inputMethod:"boolean"},{property:"columns",title:"Columns (Gallery)",description:"Number of columns in gallery view",inputMethod:"number",min:1,max:6}]},demoSchema:[{tag:"site-available-themes",properties:{viewMode:"gallery",showDetails:!0,showAllThemes:!1,columns:3},content:""}]}}}globalThis.customElements.define(SiteAvailableThemes.tag,SiteAvailableThemes);
