// @lrnwebcomponents/lrndesign-abbreviation/lrndesign-abbreviation.js
define(["exports","../../@polymer/polymer/polymer-legacy.js","../hax-body-behaviors/lib/HAXWiring.js","../schema-behaviors/schema-behaviors.js","../../@polymer/paper-tooltip/paper-tooltip.js"],function(_exports,_polymerLegacy,_HAXWiring,_schemaBehaviors,_paperTooltip){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.LrndesignAbbreviation=void 0;/**
 * Copyright 2018 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */ /**
 * `lrndesign-abbreviation`
 * `A wrapper to make a cleaner abbreviation deign element`
 *
 * @demo demo/index.html
 */let LrndesignAbbreviation=(0,_polymerLegacy.Polymer)({_template:_polymerLegacy.html`
    <style>
      :host {
        display: inline-block;
      }
      abbr {
        transition: 0.6s all ease;
        padding: 2px 4px;
        font-style: italic;
        background-color: var(--abbreviation-bg, #f9f9f9);
        text-underline-position: under;
        text-decoration: underline double;
        cursor: help;
        outline: var(--abbreviation-selection, #ffff33);
        @apply --abbreviation-main;
      }
      abbr:focus,
      abbr:active,
      abbr:hover {
        text-decoration: none;
        background-color: var(--abbreviation-selection, #ffff33);
        @apply --abbreviation-hover;
      }
      abbr::-moz-selection,
      abbr::selection {
        text-decoration: none;
        background-color: var(--abbreviation-selection, #ffff33);
        @apply --abbreviation-selection;
      }
    </style>
    <abbr tabindex="0" title$="[[phrase]]" id="abbr">[[abbr]]</abbr>
    <paper-tooltip for="abbr" position="top" offset="2" animation-delay="300"
      >[[phrase]]</paper-tooltip
    >
  `,is:"lrndesign-abbreviation",behaviors:[HAXBehaviors.PropertiesBehaviors,SchemaBehaviors.Schema],properties:{/**
     * Abbreviation text.
     */abbr:{type:String,reflectToAttribute:!0,notify:!0},/**
     * The thing the abbreviation represents.
     */phrase:{type:String,reflectToAttribute:!0,notify:!0}},/**
   * Attached to the DOM, now fire.
   */attached:function(){// Establish hax property binding
let props={canScale:!1,canPosition:!1,canEditSource:!1,gizmo:{title:"Abbreviation",description:"Simple abbreviation with tooltip of full word",icon:"editor:title",color:"grey",groups:["Instructional","Term"],handles:[{type:"inline",text:"text"}],meta:{author:"LRNWebComponents"}},settings:{quick:[{property:"abbr",title:"Abbreviation",description:"Abbreviation word",inputMethod:"textfield",icon:"editor:title"},{property:"phrase",title:"Phrase",description:"The phrase / original words",inputMethod:"textfield",icon:"editor:title"}],configure:[{property:"abbr",title:"Abbreviation",description:"Abbreviation word",inputMethod:"textfield",icon:"editor:title"},{property:"phrase",title:"Phrase",description:"The phrase / original words",inputMethod:"textfield",icon:"editor:title"}],advanced:[]}};this.setHaxProperties(props)}});_exports.LrndesignAbbreviation=LrndesignAbbreviation});
