// @lrnwebcomponents/map-menu/lib/map-menu-builder.js
define(["exports", "../../../lit/index.js", "./map-menu-submenu.js", "./map-menu-item.js"], function (_exports, _index, _mapMenuSubmenu, _mapMenuItem) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.MapMenuBuilder = void 0;
  var _templateObject_111ad450e8e111ed8789699b2ad5335a, _templateObject2_111ad450e8e111ed8789699b2ad5335a, _templateObject3_111ad450e8e111ed8789699b2ad5335a, _templateObject4_111ad450e8e111ed8789699b2ad5335a, _templateObject5_111ad450e8e111ed8789699b2ad5335a;
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  var MapMenuBuilder = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(MapMenuBuilder, _LitElement);
    var _super = _createSuper(MapMenuBuilder);
    function MapMenuBuilder() {
      var _this;
      babelHelpers.classCallCheck(this, MapMenuBuilder);
      _this = _super.call(this);
      _this.items = [];
      return _this;
    }
    /**
     * LitElement life cycle - render
     */
    babelHelpers.createClass(MapMenuBuilder, [{
      key: "render",
      value: function render() {
        var _this2 = this;
        return (0, _index.html)(_templateObject_111ad450e8e111ed8789699b2ad5335a || (_templateObject_111ad450e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "])), this.items ? this.items.map(function (item) {
          return (0, _index.html)(_templateObject2_111ad450e8e111ed8789699b2ad5335a || (_templateObject2_111ad450e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n              ", "\n            "])), item.children.length > 0 ? (0, _index.html)(_templateObject3_111ad450e8e111ed8789699b2ad5335a || (_templateObject3_111ad450e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n                    <map-menu-submenu\n                      itemtitle=\"", "\"\n                      id=\"", "\"\n                      url=\"", "\"\n                      icon=\"", "\"\n                      icon-label=\"", "\"\n                      avatar-label=\"", "\"\n                      selected=\"", "\"\n                      ?published=\"", "\"\n                    >\n                      <map-menu-builder\n                        .items=\"", "\"\n                        selected=\"", "\"\n                      ></map-menu-builder>\n                    </map-menu-submenu>\n                  "])), item.title, item.id, item.slug, item.metadata && item.metadata.icon ? item.metadata.icon : null, item.metadata && item.metadata.pageType ? item.metadata.pageType : "", item.metadata && item.metadata.avatarLabel ? item.metadata.avatarLabel : "", _this2.selected, item.metadata.published, item.children, _this2.selected) : (0, _index.html)(_templateObject4_111ad450e8e111ed8789699b2ad5335a || (_templateObject4_111ad450e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n                    <map-menu-item\n                      itemtitle=\"", "\"\n                      id=\"", "\"\n                      url=\"", "\"\n                      icon=\"", "\"\n                      icon-label=\"", "\"\n                      selected=\"", "\"\n                      ?published=\"", "\"\n                      ?locked=\"", "\"\n                      status=\"", "\"\n                    ></map-menu-item>\n                  "])), item.title, item.id, item.slug, item.metadata && item.metadata.icon ? item.metadata.icon : "", item.metadata && item.metadata.pageType ? item.metadata.pageType : "", _this2.selected, item.metadata.published, item.metadata.locked, item.metadata.status));
        }) : "");
      }
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject5_111ad450e8e111ed8789699b2ad5335a || (_templateObject5_111ad450e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "map-menu-builder";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          items: {
            type: Array
          },
          selected: {
            type: String
          }
        };
      }
    }]);
    return MapMenuBuilder;
  }(_index.LitElement);
  _exports.MapMenuBuilder = MapMenuBuilder;
  customElements.define(MapMenuBuilder.tag, MapMenuBuilder);
});
