// @haxtheweb/simple-modal/simple-modal.js
/**
 * Copyright 2019 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{LitElement as e,html as t,css as i}from"../../lit/index.js";import"../simple-icon/lib/simple-icons.js";import"../simple-icon/lib/simple-icon-lite.js";import"../simple-icon/lib/simple-icon-button-lite.js";import{DDDSuper as o}from"../d-d-d/d-d-d.js";const l=["--simple-modal-resize","--simple-modal-width","--simple-modal-z-index","--simple-modal-height","--simple-modal-min-width","--simple-modal-min-height","--simple-modal-max-width","--simple-modal-max-height","--simple-modal-titlebar-color","--simple-modal-titlebar-height","--simple-modal-titlebar-line-height","--simple-modal-titlebar-background","--simple-modal-titlebar-padding","--simple-modal-header-color","--simple-modal-header-background","--simple-modal-header-padding","--simple-modal-content-container-color","--simple-modal-content-container-background","--simple-modal-content-padding","--simple-modal-buttons-color","--simple-modal-buttons-background","--simple-modal-buttons-padding","--simple-modal-button-color","--simple-modal-button-background"];class SimpleModal extends e{static get styles(){return[i`
        :host {
          display: block;
        }

        :host([hidden]) {
          display: none;
        }

        #titlebar {
          margin-top: 0;
          padding: var(
            --simple-modal-titlebar-padding,
            0px var(--ddd-spacing-4)
          );
          display: flex;
          align-items: center;
          justify-content: space-between;
          background-color: light-dark(black, var(--ddd-theme-default-coalyGray));
          height: var(--simple-modal-titlebar-height, unset);
          line-height: var(--simple-modal-titlebar-line-height, unset);
          font-family: var(--ddd-font-navigation);
        }
        #simple-modal-title {
          margin: 0;
          display: inline-flex;
          align-items: center;
          gap: var(--ddd-spacing-2);
          flex-wrap: wrap;
          line-height: 1;
        }
        .title-inline {
          display: inline-flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          flex-wrap: wrap;
          color: white;
           font-family: var(--ddd-font-navigation);
          font-size: var(--ddd-font-size-m);
          font-weight: var(--ddd-font-weight-bold);
        }
        .title-inline span {
          transform: translateY(2px);
          display: inline-block; 
        }
        .title-icon{
          --simple-icon-color: currentColor;
          --simple-icon-width: 40px;
          --simple-icon-height: 40px;
          display: inline-flex; 
          align-items: center;  
          vertical-align: middle;
        }


        .breadcrumb-icon {
          --simple-icon-color: currentColor;
          --simple-icon-width: 1em;
          --simple-icon-height: 1em;
          display: inline-flex; 
          align-items: center;  
          vertical-align: middle;
        }
        .breadcrumbs {
          display: inline-flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          flex-wrap: wrap;
        }
        .breadcrumb-separator {
          display: inline-flex;
          align-items: center;
          font-size: var(--ddd-font-size-xs);
          color: white;
          font-family: var(--ddd-font-navigation);
        }
        .breadcrumb-button,
        .breadcrumb-current {
          display: inline-flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
           color: white;
          font: inherit;
          line-height: inherit;
          border: 0;
          background: transparent;
          padding: 0;
          margin: 0;
          text-align: left;
          font-family: var(--ddd-font-navigation);
        }
        .breadcrumb-button {
          cursor: pointer;
        }
        .breadcrumb-button:hover,
        .breadcrumb-button:focus-visible {
          text-decoration: underline;
          outline: none;
        }

        #headerbar {
          margin: 0;
          padding: var(--simple-modal-header-padding, 0px var(--ddd-spacing-4));
        }

        #close {
          top: 0;
          border: var(--simple-modal-titlebar-button-border, none);
          padding: var(
            --simple-modal-titlebar-button-padding,
            var(--ddd-spacing-3) 0
          );
          min-width: unset;
          text-transform: none;
          background-color: transparent;
          color: white;
          --simple-icon-width: var(--ddd-icon-sm);
          --simple-icon-height: var(--ddd-icon-sm);
        }

        #close:hover,
        #close:focus-visible {
          color: var(--ddd-theme-default-skyBlue);
          opacity: 0.7;
        }

        #close simple-icon-lite {
          --simple-icon-height: var(
            --simple-modal-titlebar-icon-height,
            --ddd-icon-3xs
          );
          --simple-icon-width: var(
            --simple-modal-titlebar-icon-width,
            --ddd-icon-3xs
          );
        }

        #simple-modal-content {
          flex-grow: 1;
          padding: var(
            --simple-modal-content-padding,
            var(--ddd-spacing-2) var(--ddd-spacing-4) var(--ddd-spacing-4)
          );
          margin: 0;
          color: var(
            --simple-modal-content-container-color,
            var(--ddd-theme-primary)
          );
          background-color: light-dark(white, black);
        }

        .buttons {
          padding: 0;
          padding: var(--simple-modal-buttons-padding, 0);
          margin: var(--ddd-spacing-2);
          background-color: light-dark(white, black);
        }

        .buttons ::slotted(*) {
          padding: 0 var(--ddd-spacing-4) var(--ddd-spacing-4);
          margin: 0;
          color: var(--simple-modal-button-color, --simple-modal-buttons-color);
          background-color: light-dark(white, black);
        }
        web-dialog {
          --dialog-border-radius: var(--ddd-radius-sm);
          z-index: var(--simple-modal-z-index, 1) !important;
          padding: 0;
        }
        web-dialog::part(dialog) {
          background-color: light-dark(white, black);
          min-height: var(--simple-modal-min-height, unset);
          min-width: var(--simple-modal-min-width, unset);
          z-index: var(--simple-modal-z-index, 1000);
          resize: var(--simple-modal-resize, unset);
          padding: 0;
          --dialog-height: var(--simple-modal-height, auto);
          --dialog-width: var(--simple-modal-width, 75vw);
          --dialog-max-width: var(--simple-modal-max-width, 100vw);
          --dialog-max-height: var(--simple-modal-max-height, 100vh);
        }
        web-dialog.style-scope.simple-modal {
          display: none !important;
        }
        web-dialog[open].style-scope.simple-modal {
          display: flex !important;
          position: fixed !important;
          margin: auto;
        }
        :host([resize="none"]) web-dialog[open].style-scope.simple-modal,
        :host([resize="horizontal"]) web-dialog[open].style-scope.simple-modal {
          top: calc(50% - var(--simple-modal-height, auto) / 2);
        }
        :host([resize="none"]) web-dialog[open].style-scope.simple-modal,
        :host([resize="vertical"]) web-dialog[open].style-scope.simple-modal {
          left: calc(50% - var(--simple-modal-width, 75vw) / 2);
        }

        .full {
          background-color: light-dark(white, black);
        }

        div.empty {
          background-color: transparent;
        }
      `]}render(){return t` <web-dialog
      id="dialog"
      center
      role="dialog"
      part="dialog"
      aria-describedby="simple-modal-content"
      aria-label="${this._getAriaLabel(this.title)}"
      aria-labelledby="${this._getAriaLabelledby(this.title)}"
      aria-modal="true"
      ?open="${this.opened}"
      @open="${this.open}"
      @close="${this.close}"
    >
      <div id="titlebar" part="titlebar" class=${this.title?"full":"empty"}>
        <h3 id="simple-modal-title" ?hidden="${!this.title}" part="title" class=${this.title?"hr-vert":""}>
          ${this._renderTitle()}
        </h3>
        <div></div>
        ${!this.modal||this.showClose?t`<simple-icon-button-lite
                id="close"
                dark
                icon="${this.closeIcon}"
                @click="${this.close}"
                label="${this.closeLabel}"
                part="close"
              >
              </simple-icon-button-lite>`:""}
      </div>
      <h5 id="headerbar" part="headerbar"><slot name="header"></slot></h4>
      <p id="simple-modal-content" part="content">
        <slot name="content"></slot>
      </p>
      <slot name="custom" part="custom"></slot>
      <div class="buttons" part="buttons">
        <slot name="buttons"></slot>
      </div>
    </web-dialog>`}static get properties(){return{...super.properties,title:{type:String},opened:{type:Boolean,reflect:!0},closeLabel:{attribute:"close-label",type:String},closeIcon:{type:String,attribute:"close-icon"},invokedBy:{type:Object},modal:{type:Boolean},showClose:{type:Boolean,attribute:"show-close"},mode:{type:String,reflect:!0},titleIcon:{type:String,attribute:"title-icon"},breadcrumbs:{type:Array}}}static get tag(){return"simple-modal"}constructor(){super(),this.windowControllers=new AbortController,this.title="",this.opened=!1,this.closeLabel="Close",this.closeIcon="close",this.modal=!1,this.showClose=!1,this.titleIcon="",this.breadcrumbs=[]}firstUpdated(){import("../../web-dialog/index.js").then(e=>{setTimeout(()=>{this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.backgroundColor="var(--ddd-theme-default-potential70)"},0)})}updated(e){e.forEach((e,t)=>{"opened"==t&&this._openedChanged(this[t])})}connectedCallback(){super.connectedCallback(),this.close=this.close.bind(this),this.showEvent=this.showEvent.bind(this),setTimeout(()=>{globalThis.addEventListener("simple-modal-hide",this.close,{signal:this.windowControllers.signal}),globalThis.addEventListener("simple-modal-show",this.showEvent,{signal:this.windowControllers.signal})},0)}disconnectedCallback(){this.windowControllers.abort(),super.disconnectedCallback()}showEvent(e){globalThis.dispatchEvent(new CustomEvent("simple-toast-hide",{bubbles:!0,composed:!0,cancelable:!1,detail:!1})),this.opened?(this.innerHTML="",setTimeout(()=>{this.show(e.detail.title,e.detail.mode,e.detail.elements,e.detail.invokedBy,e.detail.id,e.detail.modalClass,e.detail.styles,e.detail.clone,e.detail.modal,e.detail.showClose,e.detail.titleIcon,e.detail.breadcrumbs)},0)):this.show(e.detail.title,e.detail.mode,e.detail.elements,e.detail.invokedBy,e.detail.id,e.detail.modalClass,e.detail.styles,e.detail.clone,e.detail.modal,e.detail.showClose,e.detail.titleIcon,e.detail.breadcrumbs)}show(e,t,i,o,a=null,s=null,d=null,n=!1,r=!1,c=!1,m=null,h=[]){let p;this.invokedBy=o,this.modal=r,this.showClose=c,this.title=e,this.mode=t,this.titleIcon=m||"",this.breadcrumbs=Array.isArray(h)?h:[];let b=["header","content","buttons","custom"];for(var u in a?this.setAttribute("id",a):this.removeAttribute("id"),this.setAttribute("style",""),d&&l.forEach(e=>{this.style.setProperty(e,d[e]||"inherit")}),s?this.setAttribute("class",s):this.removeAttribute("class"),b)i[b[u]]&&(p=n?i[b[u]].cloneNode(!0):i[b[u]],p.setAttribute("slot",b[u]),this.appendChild(p));this.opened=!0}close(){document.body.style.overflow="",document.documentElement.style.overflow="",this.opened=!1,globalThis.ShadyCSS&&!globalThis.ShadyCSS.nativeShadow&&(this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.position="relative")}open(){this.opened=!0;const e=this.shadowRoot.querySelector("web-dialog");this.modal?(e.$backdrop.removeEventListener("click",e.onBackdropClick),e.removeEventListener("keydown",e.onKeyDown,{capture:!0})):(e.$backdrop.addEventListener("click",e.onBackdropClick),e.addEventListener("keydown",e.onKeyDown,{capture:!0,passive:!0})),globalThis.ShadyCSS&&!globalThis.ShadyCSS.nativeShadow&&(this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.position="fixed",this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.top=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.bottom=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.left=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.right=0)}_openedChanged(e){if(void 0===e||e){if(e){document.body.style.overflow="hidden",this.querySelectorAll("[dialog-dismiss]").forEach(e=>{e.addEventListener("click",e=>{const t=new CustomEvent("simple-modal-dismissed",{bubbles:!0,composed:!0,cancelable:!0,detail:{opened:!1,invokedBy:this.invokedBy}});this.dispatchEvent(t),this.close()})}),this.querySelectorAll("[dialog-confirm]").forEach(e=>{e.addEventListener("click",e=>{const t=new CustomEvent("simple-modal-confirmed",{composed:!0,bubbles:!0,cancelable:!0,detail:{opened:!1,invokedBy:this.invokedBy}});this.dispatchEvent(t),this.close()})});const e=new CustomEvent("simple-modal-opened",{bubbles:!0,composed:!0,cancelable:!0,detail:{opened:!0,invokedBy:this.invokedBy}});this.dispatchEvent(e),setTimeout(()=>{this._focusModalContent()},0)}}else{for(document.body.style.overflow="",document.documentElement.style.overflow="",this.title="",this.titleIcon="",this.breadcrumbs=[];null!==this.firstChild;)this.removeChild(this.firstChild);this.invokedBy&&setTimeout(()=>{this.invokedBy.focus()},500),this.showClose=!1;const e=new CustomEvent("simple-modal-closed",{bubbles:!0,cancelable:!0,detail:{opened:!1,invokedBy:this.invokedBy}});this.dispatchEvent(e)}}_focusModalContent(e=0){if(!this.opened)return;const t=this.querySelector(["[slot='content'] [autofocus]","[slot='content'] button:not([disabled])","[slot='content'] [href]","[slot='content'] input:not([disabled]):not([type='hidden'])","[slot='content'] select:not([disabled])","[slot='content'] textarea:not([disabled])","[slot='content'] [tabindex]:not([tabindex='-1'])","[slot='buttons'] button:not([disabled])","[slot='buttons'] [tabindex]:not([tabindex='-1'])"].join(", "));if(t&&"function"==typeof t.focus)return void t.focus();const i=this.shadowRoot.querySelector("#close");if(i&&i.shadowRoot){const e=i.shadowRoot.querySelector("button");if(e&&"function"==typeof e.focus)return void e.focus()}const o=this.shadowRoot.querySelector("#dialog");if(o&&"function"==typeof o.focus)return o.hasAttribute("tabindex")||o.setAttribute("tabindex","-1"),void o.focus();e<20&&setTimeout(()=>{this._focusModalContent(e+1)},25)}_getAriaLabelledby(e){return e?"simple-modal-title":null}_getAriaLabel(e){return e?null:"Modal Dialog"}_renderTitle(){return Array.isArray(this.breadcrumbs)&&this.breadcrumbs.length>0?this._renderBreadcrumbs():this.titleIcon?t`<span class="title-inline">
        <simple-icon-lite
          class="title-icon"
          icon="${this.titleIcon}"
          aria-hidden="true"
        ></simple-icon-lite>
        <span>${this.title}</span>
      </span>`:this.title}_renderBreadcrumbs(){const e=this.breadcrumbs.filter(e=>e&&e.label&&"string"==typeof e.label);if(0===e.length)return this.title;const i=e.length-1;return t`<nav class="breadcrumbs" aria-label="Modal breadcrumb">
      ${e.map((e,o)=>{const l=e.icon||(o===i&&this.titleIcon?this.titleIcon:""),a=this._isBreadcrumbClickable(e,o,i)?t`<button
              type="button"
              class="breadcrumb-button"
              @click="${t=>this._handleBreadcrumbClick(e,o,t)}"
            >
              ${l?t`<simple-icon-lite
                    class="breadcrumb-icon"
                    icon="${l}"
                    aria-hidden="true"
                  ></simple-icon-lite>`:""}
              <span>${e.label}</span>
            </button>`:t`<span class="breadcrumb-current" aria-current="page">
              ${l?t`<simple-icon-lite
                    class="breadcrumb-icon"
                    icon="${l}"
                    aria-hidden="true"
                  ></simple-icon-lite>`:""}
              <span>${e.label}</span>
            </span>`;return t`${o>0?t`<span class="breadcrumb-separator" aria-hidden="true"
                >&gt;</span
              >`:""}${a}`})}
    </nav>`}_isBreadcrumbClickable(e,t,i){return!(t>=i)&&(!1!==e.clickable&&!(!e.action&&"function"!=typeof e.onClick))}_handleBreadcrumbClick(e,t,i){i&&(i.preventDefault(),i.stopPropagation());const o={breadcrumb:e,index:t,breadcrumbs:this.breadcrumbs,title:this.title};e&&"function"==typeof e.onClick&&e.onClick(o),this.dispatchEvent(new CustomEvent("simple-modal-breadcrumb-click",{bubbles:!0,composed:!0,cancelable:!0,detail:o}))}}globalThis.customElements.define(SimpleModal.tag,SimpleModal);export{SimpleModal,l as SimpleModalCssVars};globalThis.SimpleModal=globalThis.SimpleModal||{},globalThis.SimpleModal.requestAvailability=()=>(!globalThis.SimpleModal.instance&&globalThis.document&&globalThis.document.body&&(globalThis.SimpleModal.instance=globalThis.document.createElement("simple-modal"),globalThis.document.body.appendChild(globalThis.SimpleModal.instance)),globalThis.SimpleModal.instance);export const SimpleModalStore=globalThis.SimpleModal.requestAvailability();
