// @haxtheweb/haxcms-elements/lib/core/utils/haxcms-views-render-utility.js
/**
 * Copyright 2025 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{html as e,css as r}from"../../../../../lit/index.js";import{unsafeHTML as t}from"../../../../../lit/directives/unsafe-html.js";export const safeString=e=>null==e?"":String(e);export const isEmptyValue=e=>null==e||(Array.isArray(e)?e.length<1:"string"==typeof e&&""===e.trim());export const valueAtPath=(e,r)=>{if(!r)return"";const t=String(r).split(".");let a=e;for(let e=0;e<t.length;e++){if(!a||"object"!=typeof a||!Object.prototype.hasOwnProperty.call(a,t[e]))return"";a=a[t[e]]}return a};export const recordTitle=e=>{const r=["title","name","tag","path","slug","id","machineName"];for(let t=0;t<r.length;t++){const a=valueAtPath(e,r[t]);if(!isEmptyValue(a))return safeString(a)}return"No results"};export const recordDescription=e=>{const r=["description","summary","body","metadata.description","mimetype"];for(let t=0;t<r.length;t++){const a=valueAtPath(e,r[t]);if(!isEmptyValue(a))return safeString(a)}return""};export const recordBody=e=>{const r=["body","content","description"];for(let t=0;t<r.length;t++){const a=valueAtPath(e,r[t]);if(!isEmptyValue(a))return safeString(a)}return""};export const recordBodyIsHtml=e=>{const r=recordBody(e).trim();return!!r&&(-1!==r.indexOf("<")&&-1!==r.indexOf(">"))};export const stripHtml=e=>safeString(e).replace(/<[^>]*>/g," ").replace(/\s+/g," ").trim();export const recordTag=e=>{const r=["tag","haxElementSchema.tag","instances.0.haxElementSchema.tag"];for(let t=0;t<r.length;t++){const a=safeString(valueAtPath(e,r[t])).trim();if(a)return a.toLowerCase()}return""};export const recordElementSchema=e=>{let r=valueAtPath(e,"haxElementSchema");return Array.isArray(r)&&r.length>0&&(r=r[0]),r&&"object"==typeof r||!e||"object"!=typeof e||(r=valueAtPath(e,"instances.0.haxElementSchema")),!r||"object"!=typeof r||Array.isArray(r)?null:r.tag&&""!==safeString(r.tag).trim()?r:null};export const isRenderableTag=e=>{const r=safeString(e).trim().toLowerCase();if(!r)return!1;return!!Object.prototype.hasOwnProperty.call({img:!0,video:!0,audio:!0,iframe:!0},r)||/^[a-z][a-z0-9]*-[a-z0-9-]*$/.test(r)};export const attributeValue=e=>null==e||!1===e?"":!0===e?"__BOOLEAN_TRUE__":"object"==typeof e?"":safeString(e).replace(/&/g,"&amp;").replace(/"/g,"&quot;").trim();export const attributeStringFromProperties=e=>{if(!e||"object"!=typeof e||Array.isArray(e))return"";const r=Object.keys(e);let t="";for(let a=0;a<r.length;a++){const i=safeString(r[a]).trim();if(!i)continue;const n=attributeValue(e[i]);n&&(t+="__BOOLEAN_TRUE__"!==n?` ${i}="${n}"`:` ${i}`)}return t};export const renderRecordElementPreview=r=>{const a=recordElementSchema(r),i=a&&a.tag?safeString(a.tag).trim():recordTag(r);if(!isRenderableTag(i))return null;const n=a&&a.properties&&"object"==typeof a.properties?a.properties:{},o=a&&"string"==typeof a.content?a.content:"",l=attributeStringFromProperties(n);return e`<div class="record-element-preview">
    ${t(`<${i}${l}>${o}</${i}>`)}
  </div>`};export const recordUrl=(e,r={})=>{const t=r.resolveHref||(e=>e),a=["links.page","slug","url","fullUrl","path","metadata.path","links.self","links.content"];for(let r=0;r<a.length;r++){const i=safeString(valueAtPath(e,a[r])).trim();if(i)return t(i)}return""};export const looksLikeImageSource=e=>{const r=safeString(e).trim().toLowerCase();return!!r&&(!!(r.endsWith(".png")||r.endsWith(".jpg")||r.endsWith(".jpeg")||r.endsWith(".gif")||r.endsWith(".webp")||r.endsWith(".svg")||r.endsWith(".avif")||r.endsWith(".bmp"))||(-1!==r.indexOf(".png?")||-1!==r.indexOf(".jpg?")||-1!==r.indexOf(".jpeg?")||-1!==r.indexOf(".gif?")||-1!==r.indexOf(".webp?")||-1!==r.indexOf(".svg?")||-1!==r.indexOf(".avif?")||-1!==r.indexOf(".bmp?")))};export const recordImage=(e,r={})=>{const t=r.resolveHref||(e=>e),a=["image","src","properties.src","properties.image","haxElementSchema.properties.src","haxElementSchema.properties.image","instances.0.haxElementSchema.properties.src","instances.0.haxElementSchema.properties.image","metadata.image","logo","metadata.site.logo","screenshot","preview","thumbnail","poster","haxElementSchema.properties.poster","instances.0.haxElementSchema.properties.poster","url","fullUrl","path","metadata.path"];for(let r=0;r<a.length;r++){const i=safeString(valueAtPath(e,a[r])).trim();if(i){const e=t(i);if(looksLikeImageSource(e))return e}}return""};export const recordIcon=e=>{const r=["icon","metadata.icon","pageType"];for(let t=0;t<r.length;t++){const a=safeString(valueAtPath(e,r[t])).trim();if(a)return a}return"hax:module"};export const recordTags=e=>{const r=valueAtPath(e,"tags");if(Array.isArray(r))return r.map(e=>safeString(e).trim()).filter(e=>e);if("string"==typeof r)return r.split(",").map(e=>e.trim()).filter(e=>e);const t=valueAtPath(e,"metadata.tags");return Array.isArray(t)?t.map(e=>safeString(e).trim()).filter(e=>e):"string"==typeof t?t.split(",").map(e=>e.trim()).filter(e=>e):[]};export const looksLikeVideoSource=e=>{const r=safeString(e).trim().toLowerCase();return!!r&&(-1!==r.indexOf("youtube.com/")||-1!==r.indexOf("youtu.be/")||-1!==r.indexOf("vimeo.com/")||!!(r.endsWith(".mp4")||r.endsWith(".webm")||r.endsWith(".ogg")||r.endsWith(".m3u8")||-1!==r.indexOf(".mp4?")||-1!==r.indexOf(".webm?")||-1!==r.indexOf(".ogg?")||-1!==r.indexOf(".m3u8?")))};export const videoSource=(e,r={})=>{const t=r.resolveHref||(e=>e),a=["source","src","video","properties.source","properties.src","haxElementSchema.properties.source","haxElementSchema.properties.src","haxElementSchema.properties.url","instances.0.haxElementSchema.properties.source","instances.0.haxElementSchema.properties.src","instances.0.haxElementSchema.properties.url","url","fullUrl","path","metadata.path","links.self","links.page","links.content"],i=recordTag(e);for(let r=0;r<a.length;r++){const n=a[r],o=safeString(valueAtPath(e,a[r])).trim();if(!o)continue;const l=t(o);if(looksLikeVideoSource(l))return l;if(i&&-1!==i.indexOf("video")&&0!==n.indexOf("links.")&&-1===l.indexOf("/x/api/"))return l}return""};export const displayValue=e=>{if(null==e||""===e)return"—";if("boolean"==typeof e)return e?"true":"false";if("number"==typeof e)return String(e);if(Array.isArray(e)){if(e.some(e=>e&&"object"==typeof e))try{return JSON.stringify(e,null,2)}catch(e){}return e.map(e=>safeString(e)).join(", ")}if("object"==typeof e)try{return JSON.stringify(e,null,2)}catch(r){return safeString(e)}return safeString(e)};export const shortValue=(e,r=240)=>{const t=safeString(e);return t.length<=r?t:`${t.substring(0,r)}…`};export const tableCellValue=(e,r)=>{const t=displayValue(r);return-1!==safeString(e).toLowerCase().indexOf("haxelementschema")?t:shortValue(t,180)};export const tableColumns=(e,r)=>{if(Array.isArray(r)&&r.length>0)return r;if(!Array.isArray(e)||e.length<1)return["id","title","slug"];const t=e[0];return t&&"object"==typeof t?Object.keys(t).slice(0,6):["value"]};export const recordTimelineDate=e=>{const r=["generatedAt","dateCreated","updated","created","metadata.updated","metadata.created"];for(let t=0;t<r.length;t++){const a=safeString(valueAtPath(e,r[t])).trim();if(!a)continue;const i=new Date(a);if(!Number.isNaN(i.getTime()))return i.toISOString()}return""};export const timelineEvents=e=>Array.isArray(e)?e.map(e=>{const r=recordTimelineDate(e),t=r?new Date(r).toLocaleDateString():"Record",a=recordDescription(e)||recordBody(e)||"",i=shortValue(safeString(a).replace(/<[^>]*>/g," ").replace(/\s+/g," ").trim(),420);return{heading:`${t} — ${recordTitle(e)}`,details:i,imagesrc:recordImage(e),imagealt:`${recordTitle(e)} image`}}).filter(e=>e.heading&&e.details):[];export const numericValue=e=>{if("number"==typeof e&&Number.isFinite(e))return e;const r=safeString(e).trim();if(!r)return null;const t=Number(r);return!Number.isNaN(t)&&Number.isFinite(t)?t:null};export const chartDescriptor=(e,r,t)=>{if(!Array.isArray(e)||e.length<1)return null;const a=e[0],i=Array.isArray(r)?r:[],n=(()=>{const e={},walk=(r,t="")=>{const a=Object.keys(r||{});for(let i=0;i<a.length;i++){const n=a[i],o=r[n],l=t?`${t}.${n}`:n;o&&"object"==typeof o&&!Array.isArray(o)?walk(o,l):e[l]=o}};return walk(a||{}),e})(),o=i.length>0?i:Object.keys(n);let l="";for(let r=0;r<o.length;r++){const t=o[r];for(let r=0;r<e.length;r++){if(null!==numericValue(valueAtPath(e[r],t))){l=t;break}}if(l)break}if(!l)return null;let d="";for(let e=0;e<o.length;e++){const r=o[e];if(r===l)continue;if(safeString(valueAtPath(a,r)).trim()){d=r;break}}const s=[],c=[],p=Math.min(e.length,24);for(let r=0;r<p;r++){const t=e[r],a=d?safeString(valueAtPath(t,d)).trim():recordTitle(t),i=numericValue(valueAtPath(t,l));s.push(a||recordTitle(t)),c.push(null===i?0:i)}return{valueField:l,labelField:d,data:{labels:s,series:[c]}}};export const treeRecords=e=>{if(!Array.isArray(e)||e.length<1)return[];const r={},t=e.map((e,t)=>{const a=safeString(e.id||e.slug||e.path||e.tag||`${t+1}`).trim()||`${t+1}`;return r[a]=!0,{id:a,parent:safeString(e.parent).trim(),title:recordTitle(e),slug:recordUrl(e),metadata:{icon:recordIcon(e),pageType:safeString(e.pageType||""),published:!0}}});return t.forEach(e=>{e.parent&&Object.prototype.hasOwnProperty.call(r,e.parent)||(e.parent=null)}),t};export const renderRecordBody=(r,a=240)=>{const i=recordBody(r);return i?recordBodyIsHtml(r)?e`<div class="record-html">${t(i)}</div>`:e`<p>${shortValue(i,a)}</p>`:e``};export const renderRecordPrimary=(r,t=240,a={})=>{const i=videoSource(r,a);if(i)return e`
      <div class="record-media">
        <video-player
          source="${i}"
          media-title="${recordTitle(r)}"
        ></video-player>
      </div>
    `;const n=recordImage(r,a);if(n)return e`
      <img
        class="record-image"
        src="${n}"
        alt="${recordTitle(r)} preview"
        loading="lazy"
        decoding="async"
      />
    `;const o=renderRecordElementPreview(r);return o||renderRecordBody(r,t)};export const titleFromParameterName=e=>safeString(e).replace(/[._-]/g," ").replace(/\s+/g," ").trim().replace(/\b\w/g,e=>e.toUpperCase());export const renderCollectionPreview=(r,t={})=>Array.isArray(r)?e`
    <collection-list>
      ${r.map(r=>{const a=recordTags(r),i=recordUrl(r,t),n=recordBodyIsHtml(r),o=renderRecordElementPreview(r);return e`
          <collection-item
            line1="${recordTitle(r)}"
            line2="${n?"":shortValue(recordDescription(r),180)}"
            url="${i}"
            image="${recordImage(r,t)}"
            icon="${recordIcon(r)}"
            tags="${a.join(",")}"
            saturate
          >
            ${o||renderRecordBody(r,220)}
          </collection-item>
        `})}
    </collection-list>
  `:e``;export const renderGridPreview=(r,t={})=>Array.isArray(r)?e`
    <div class="card-grid">
      ${r.map(r=>{const a=recordUrl(r,t);return e`
          <accent-card no-border>
            <div slot="heading">${recordTitle(r)}</div>
            <div slot="subheading">
              ${recordTags(r).slice(0,5).join(", ")}
            </div>
            <div slot="content">
              ${renderRecordPrimary(r,240,t)}
              ${a?e`<a
                    class="card-link"
                    href="${a}"
                    target="_blank"
                    rel="noopener"
                    >Open record</a
                  >`:e``}
            </div>
          </accent-card>
        `})}
    </div>
  `:e``;export const renderCardsPreview=(e,r={})=>renderGridPreview(e,r);export const renderBulletedListPreview=(r,t={})=>Array.isArray(r)?e`
    <ul class="result-list">
      ${r.map(r=>{const a=recordUrl(r,t);return e`
          <li class="result-list-item">
            <h4 class="result-title">
              ${a?e`<a href="${a}" target="_blank" rel="noopener"
                    >${recordTitle(r)}</a
                  >`:recordTitle(r)}
            </h4>
            ${renderRecordPrimary(r,360,t)}
          </li>
        `})}
    </ul>
  `:e``;export const renderContentPreview=(r,t={})=>Array.isArray(r)?e`
    <div class="content-records">
      ${r.map(r=>{const a=recordUrl(r,t);return e`
          <article class="content-record">
            <h3 class="content-record-title">
              ${a?e`<a href="${a}" target="_blank" rel="noopener"
                    >${recordTitle(r)}</a
                  >`:recordTitle(r)}
            </h3>
            ${renderRecordPrimary(r,720,t)}
          </article>
        `})}
    </div>
  `:e``;export const renderTablePreview=(r,t={})=>{if(!Array.isArray(r))return e``;const a=tableColumns(r,t.selectedFields);return e`
    <div class="table-scroll">
      <editable-table-display
        bordered
        condensed
        column-header
        responsive
        sort
        striped
        scroll
      >
        <table>
          <thead>
            <tr>
              ${a.map(r=>e`<th>${titleFromParameterName(r)}</th>`)}
            </tr>
          </thead>
          <tbody>
            ${r.map(r=>e`
                <tr>
                  ${a.map(t=>{const a=valueAtPath(r,t);return e`<td title="${displayValue(a)}">
                      ${tableCellValue(t,a)}
                    </td>`})}
                </tr>
              `)}
          </tbody>
        </table>
      </editable-table-display>
    </div>
  `};export const renderCarouselPreview=(r,t={})=>{if(!Array.isArray(r))return e``;return r.every(e=>{const r=videoSource(e,t),a=recordImage(e,t);return!r&&!!a})?e`
      <div class="media-gallery">
        ${r.map(r=>{const a=recordImage(r,t),i=recordUrl(r,t);return e`
            <div class="media-gallery-item">
              ${i?e`<a href="${i}" target="_blank" rel="noopener"
                    ><img
                      src="${a}"
                      alt="${recordTitle(r)} preview"
                      loading="lazy"
                      decoding="async"
                  /></a>`:e`<img
                    src="${a}"
                    alt="${recordTitle(r)} preview"
                    loading="lazy"
                    decoding="async"
                  />`}
              <span class="media-gallery-caption"
                >${recordTitle(r)}</span
              >
            </div>
          `})}
      </div>
    `:e`
    <div class="carousel-shell">
      <play-list navigation pagination>
        ${r.map(r=>{const a=videoSource(r,t);if(a)return e`
              <video-player
                source="${a}"
                media-title="${recordTitle(r)}"
              ></video-player>
            `;const i=recordUrl(r,t),n=recordImage(r,t),o=renderRecordElementPreview(r);return n?e`
              <figure>
                <img
                  class="record-image"
                  src="${n}"
                  alt="${recordTitle(r)} preview"
                  loading="lazy"
                  decoding="async"
                />
                <figcaption>
                  ${i?e`<a
                        href="${i}"
                        target="_blank"
                        rel="noopener"
                        >${recordTitle(r)}</a
                      >`:recordTitle(r)}
                </figcaption>
              </figure>
            `:o?e`${o}`:e`
            <accent-card>
              <div slot="heading">${recordTitle(r)}</div>
              <div slot="content">
                ${renderRecordBody(r,300)}
                ${i?e`<a href="${i}" target="_blank" rel="noopener"
                      >Open record</a
                    >`:e``}
              </div>
            </accent-card>
          `})}
      </play-list>
    </div>
  `};export const renderAccordionPreview=(r,t={})=>Array.isArray(r)?e`
    <div class="accordion-preview">
      <a11y-collapse-group>
        ${r.map((r,a)=>e`
            <a11y-collapse
              heading-button
              heading="${recordTitle(r)}"
              ?expanded="${0===a}"
            >
              ${renderRecordPrimary(r,420,t)}
            </a11y-collapse>
          `)}
      </a11y-collapse-group>
    </div>
  `:e``;export const renderTabsPreview=(r,t={})=>Array.isArray(r)?e`
    <div class="tabs-preview">
      <a11y-tabs full-width>
        ${r.map((r,a)=>e`
            <a11y-tab
              id="views-preview-tab-${a}"
              label="${recordTitle(r)}"
            >
              ${renderRecordPrimary(r,520,t)}
            </a11y-tab>
          `)}
      </a11y-tabs>
    </div>
  `:e``;export const renderTimelinePreview=(r,t={})=>{if(!Array.isArray(r))return e``;const a=timelineEvents(r);return!a||a.length<1?e`<p class="empty-preview">
      No date-oriented records available for timeline rendering.
    </p>`:e`
    <div class="timeline-shell">
      <lrndesign-timeline
        timeline-title="${titleFromParameterName(t.selectedEntity||"")} timeline"
        .events="${a}"
      ></lrndesign-timeline>
    </div>
  `};export const renderChartPreview=(r,t={})=>{if(!Array.isArray(r))return e``;const a=chartDescriptor(r,t.selectedFields,t.selectedEntity);return a?e`
    <div class="chart-shell">
      <p class="note">
        Charting
        <strong>${titleFromParameterName(a.valueField)}</strong
        >${a.labelField?e` by
              <strong
                >${titleFromParameterName(a.labelField)}</strong
              >`:e``}
      </p>
      <lrndesign-bar
        .data="${a.data}"
        chart-title="${titleFromParameterName(t.selectedEntity||"")} chart"
        show-table
      ></lrndesign-bar>
    </div>
  `:e`<p class="empty-preview">
      No numeric field detected for chart rendering.
    </p>`};export const renderTreePreview=(r,t={})=>{if(!Array.isArray(r))return e``;const a=treeRecords(r);return!a||a.length<1?e`<p class="empty-preview">
      No hierarchical records available for tree rendering.
    </p>`:e`
    <div class="tree-shell">
      <map-menu .data="${a}" auto-scroll></map-menu>
    </div>
  `};export const renderImageGalleryPreview=(r,t={})=>{if(!Array.isArray(r))return e``;const a=r.filter(e=>!!recordImage(e,t));return a.length<1?e`<p class="empty-preview">No image records available for gallery rendering.</p>`:e`
    <image-gallery mode="masonry">
      ${a.map(r=>{const a=recordImage(r,t);return e`<img src="${a}" alt="${recordTitle(r)} preview" loading="lazy" decoding="async" />`})}
    </image-gallery>
  `};export const renderMediaPlaylistPreview=(r,t={})=>{if(!Array.isArray(r))return e``;const a=r.filter(e=>!!videoSource(e,t)||!!recordImage(e,t));return a.length<1?e`<p class="empty-preview">No media records available for playlist rendering.</p>`:e`
    <media-playlist>
      ${a.map(r=>{const a=videoSource(r,t);if(a)return e`<video-player source="${a}" media-title="${recordTitle(r)}"></video-player>`;const i=recordImage(r,t);return e`<img src="${i}" alt="${recordTitle(r)} preview" loading="lazy" decoding="async" />`})}
    </media-playlist>
  `};export const renderPreview=(r,t,a={})=>{if(!Array.isArray(r))return e``;switch("playlist"===t?"carousel":"cards"===t?"grid":t){case"table":return renderTablePreview(r,a);case"grid":return renderGridPreview(r,a);case"carousel":return renderCarouselPreview(r,a);case"image-gallery":return renderImageGalleryPreview(r,a);case"media-playlist":return renderMediaPlaylistPreview(r,a);case"list":return renderBulletedListPreview(r,a);case"content":return renderContentPreview(r,a);case"accordion":return renderAccordionPreview(r,a);case"tabs":return renderTabsPreview(r,a);case"timeline":return renderTimelinePreview(r,a);case"chart":return renderChartPreview(r,a);case"tree":return renderTreePreview(r,a);default:return renderCollectionPreview(r,a)}};export const RENDERER_OPTIONS=[{value:"collection",text:"Collection"},{value:"table",text:"Table"},{value:"grid",text:"Grid"},{value:"carousel",text:"Carousel"},{value:"image-gallery",text:"Image Gallery"},{value:"media-playlist",text:"Media Playlist"},{value:"list",text:"List"},{value:"content",text:"Content"},{value:"accordion",text:"Accordion"},{value:"tabs",text:"Tabs"},{value:"timeline",text:"Timeline"},{value:"chart",text:"Chart"},{value:"tree",text:"Tree"}];export const viewsRendererStyles=r`
  .table-scroll {
    width: 100%;
    overflow-x: auto;
    -webkit-overflow-scrolling: touch;
  }
  .table-scroll table {
    min-width: 600px;
    width: 100%;
    border-collapse: collapse;
  }
  .table-scroll th,
  .table-scroll td {
    text-align: left;
    vertical-align: top;
    padding: var(--ddd-spacing-2);
    border-bottom: var(--ddd-border-xs) solid
      light-dark(
        var(--ddd-theme-default-limestoneGray),
        var(--ddd-primary-5)
      );
    font-size: var(--ddd-font-size-5xs);
  }
  .card-grid {
    display: grid;
    grid-template-columns: repeat(2, minmax(0, 1fr));
    gap: var(--ddd-spacing-3);
  }
  .card-grid accent-card {
    width: 100%;
  }
  .card-link {
    font-size: var(--ddd-font-size-4xs);
  }
  .record-html {
    font-size: var(--ddd-font-size-5xs);
    line-height: var(--ddd-lh-150);
    overflow-wrap: anywhere;
  }
  .record-image {
    display: block;
    width: 100%;
    max-height: 220px;
    object-fit: cover;
    border-radius: var(--ddd-radius-sm);
    margin-bottom: var(--ddd-spacing-2);
    border: var(--ddd-border-xs) solid
      light-dark(
        var(--ddd-theme-default-limestoneGray),
        var(--ddd-primary-5)
      );
  }
  .record-media {
    display: block;
    width: 100%;
    margin-bottom: var(--ddd-spacing-2);
  }
  .record-media video-player {
    display: block;
    width: 100%;
  }
  .record-element-preview {
    border: var(--ddd-border-xs) solid
      light-dark(
        var(--ddd-theme-default-limestoneGray),
        var(--ddd-primary-5)
      );
    border-radius: var(--ddd-radius-sm);
    padding: var(--ddd-spacing-2);
    margin-bottom: var(--ddd-spacing-2);
    overflow: auto;
  }
  .record-element-preview > * {
    max-width: 100%;
  }
  .result-list {
    margin: 0;
    padding-left: var(--ddd-spacing-4);
    display: flex;
    flex-direction: column;
    gap: var(--ddd-spacing-3);
  }
  .result-list-item {
    margin: 0;
    font-size: var(--ddd-font-size-5xs);
  }
  .result-title {
    margin: 0 0 var(--ddd-spacing-1) 0;
    font-size: var(--ddd-font-size-4xs);
  }
  .content-records {
    display: flex;
    flex-direction: column;
    gap: var(--ddd-spacing-4);
  }
  .content-record {
    border-bottom: var(--ddd-border-xs) solid
      light-dark(
        var(--ddd-theme-default-limestoneGray),
        var(--ddd-primary-5)
      );
    padding-bottom: var(--ddd-spacing-3);
  }
  .content-record:last-child {
    border-bottom: 0;
    padding-bottom: 0;
  }
  .content-record-title {
    margin: 0 0 var(--ddd-spacing-2) 0;
    font-size: var(--ddd-font-size-s);
  }
  .carousel-shell play-list,
  .playlist-shell play-list {
    display: block;
  }
  .media-gallery {
    display: grid;
    grid-template-columns: repeat(2, minmax(0, 1fr));
    gap: var(--ddd-spacing-3);
  }
  .media-gallery-item {
    border: var(--ddd-border-xs) solid
      light-dark(
        var(--ddd-theme-default-limestoneGray),
        var(--ddd-primary-5)
      );
    border-radius: var(--ddd-radius-sm);
    overflow: hidden;
    background: light-dark(
      var(--ddd-theme-default-white),
      rgba(255, 255, 255, 0.03)
    );
  }
  .media-gallery-item img {
    display: block;
    width: 100%;
    height: 220px;
    object-fit: cover;
  }
  .media-gallery-caption {
    display: block;
    padding: var(--ddd-spacing-2);
    font-size: var(--ddd-font-size-5xs);
  }
  .accordion-preview {
    display: block;
  }
  .accordion-preview a11y-collapse {
    --a11y-collapse-margin: 0 0 var(--ddd-spacing-2) 0;
  }
  .tabs-preview a11y-tabs {
    --a11y-tabs-content-padding: var(--ddd-spacing-3);
    --a11y-tabs-button-padding: var(--ddd-spacing-2);
  }
  .timeline-shell lrndesign-timeline {
    display: block;
  }
  .chart-shell lrndesign-bar {
    display: block;
    min-height: 280px;
  }
  .tree-shell {
    min-height: 320px;
    border: var(--ddd-border-xs) solid
      light-dark(
        var(--ddd-theme-default-limestoneGray),
        var(--ddd-primary-5)
      );
    border-radius: var(--ddd-radius-sm);
    overflow: hidden;
  }
  .tree-shell map-menu {
    height: 320px;
    --map-menu-gap: var(--ddd-spacing-2);
  }
  @media screen and (max-width: 1024px) {
    .card-grid {
      grid-template-columns: repeat(1, minmax(0, 1fr));
    }
    .media-gallery {
      grid-template-columns: repeat(1, minmax(0, 1fr));
    }
  }
`;export default viewsRendererStyles;
