// @lrnwebcomponents/rich-text-editor/rich-text-editor.js
define(["exports","../../@polymer/polymer/polymer-element.js","../hax-body-behaviors/lib/HAXWiring.js","../../@polymer/iron-a11y-keys/iron-a11y-keys.js","./lib/rich-text-editor-styles.js","./lib/rich-text-editor-toolbar.js"],function(_exports,_polymerElement,_HAXWiring,_ironA11yKeys,_richTextEditorStyles,_richTextEditorToolbar){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.RichTextEditor=void 0;/**
 * Copyright 2018 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */ /**
 * `rich-text-editor`
 * `a standalone rich text editor`
 *
 * @microcopy - language worth noting:
 *  -
 *
 * @customElement
 * @polymer
 * @demo demo/index.html demo
 * @demo demo/config.html custom configuration
 */class RichTextEditor extends _polymerElement.PolymerElement{// render function
static get template(){return _polymerElement.html`
      <style>
        :host([hidden]) {
          display: none;
        }
        :host {
          display: block;
          min-height: 20px;
          cursor: pointer;
          @apply --rich-text-editor-content;
        }
        :host([contenteditable="true"]) {
          border: var(--rich-text-editor-border);
          border-top: none;
          overflow: hidden;
          @apply --rich-text-editor-content-edit;
        }
        :host([contenteditable="true"]):empty:before {
          content: attr(placeholder);
          display: block;
          @apply --rich-text-editor-content-placeholder;
        }
        span {
          background-color: blue;
        }
        .rich-text-editor-selection {
          background-color: var(--rich-text-editor-selection-bg);
          @apply --rich-text-editor-content-selection;
        }
      </style>
      <style include="rich-text-editor-styles"></style>
      <slot></slot>
    `}// haxProperty definition
static get haxProperties(){return{canScale:!0,canPosition:!0,canEditSource:!1,gizmo:{title:"Rich text-editor",description:"a standalone rich text editor",icon:"icons:android",color:"green",groups:["Text"],handles:[{type:"todo:read-the-docs-for-usage"}],meta:{author:"nikkimk",owner:"Penn State University"}},settings:{quick:[],configure:[{property:"title",description:"",inputMethod:"textfield",required:!1,icon:"icons:android"}],advanced:[]}}}// properties available to the custom element for data binding
static get properties(){return{/**
       * The editableElement element for the editor.
       */editorId:{name:"editableElement",type:"Object",value:null}}}/**
   * Store the tag name to make it easier to obtain directly.
   * @notice function name must be here for tooling to operate correctly
   */static get tag(){return"rich-text-editor"}/**
   * life cycle, element is afixed to the DOM
   */connectedCallback(){super.connectedCallback();let style=document.createElement("style");style.setAttribute("is","custom-style");style.setAttribute("include","rich-text-editor-styles");document.head.append(style)}/**
   * ready
   */ready(){super.ready();//find an editor by id
let id=document.querySelector("rich-text-editor-toolbar#"+this.editorId),editor=null!==id?id:document.querySelector("rich-text-editor-toolbar");if(null===editor){editor=document.createElement("rich-text-editor-toolbar");this.parentNode.appendChild(editor)}editor.addEditableRegion(this)}/**
   * life cycle, element is removed from the DOM
   */ //disconnectedCallback() {}
}_exports.RichTextEditor=RichTextEditor;window.customElements.define(RichTextEditor.tag,RichTextEditor)});
