// @haxtheweb/app-hax/app-hax.js
import{css as e,html as t,unsafeCSS as a}from"../../lit/index.js";import{toJS as o,autorun as i}from"../../mobx/dist/mobx.esm.js";import{localStorageSet as d,localStorageGet as r}from"../utils/utils.js";import"../simple-tooltip/simple-tooltip.js";import{SimpleColors as n}from"../simple-colors/simple-colors.js";import{store as s}from"./lib/v2/AppHaxStore.js";import{I18NMixin as l}from"../i18n-manager/lib/I18NMixin.js";import{AppHaxAPI as p}from"./lib/v2/AppHaxBackendAPI.js";import{SimpleTourManager as c}from"../simple-popover/lib/simple-tour.js";import{SuperDaemonInstance as h}from"../super-daemon/super-daemon.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import"../simple-colors-shared-styles/simple-colors-shared-styles.js";import"./lib/v2/AppHaxRouter.js";import"./lib/v2/app-hax-label.js";import"../haxcms-elements/lib/core/ui/app-hax-top-bar.js";import{SimpleTourFinder as m}from"../simple-popover/lib/SimpleTourFinder.js";import"./lib/v2/app-hax-use-case.js";import"./lib/v2/app-hax-use-case-filter.js";import"./lib/v2/app-hax-search-results.js";import"./lib/v2/app-hax-scroll-button.js";new URL("./lib/assets/images/Logout.svg",import.meta.url).href;function darkToggle(e){e.matches?s.darkMode=!0:s.darkMode=!1}function soundToggle(){s.soundStatus=!o(s.soundStatus),d("app-hax-soundStatus",o(s.soundStatus)),s.appEl.playSound("click")}export class AppHax extends(l(m(n))){static get tag(){return"app-hax"}_openExternalLink(e){globalThis.open(e,"_blank")}async _haxStoreContribute(e,t,a=null){let o="";if("merlin"==e){var i=`[${e}] New command request from HAX daemon`;o=`Location: ${globalThis.location.href}\nMerlin command: ${a}\nWhat did you want merlin to do?\n`}else{i=`[${e}] User report from HAX daemon`;if(o=`Location: ${globalThis.location.href}\nBrowser: ${navigator.userAgent}\nOS: ${navigator.userAgentData.platform} - ${navigator.deviceMemory}GB RAM - ${navigator.hardwareConcurrency} cores\nScreen: ${globalThis.screen.width}x${globalThis.screen.height}\nWindow size: ${globalThis.innerWidth}x${globalThis.innerHeight}\n`,navigator.getBattery){o+=`Battery: ${100*(await globalThis.navigator.getBattery()).level}%\n`}navigator.connection&&globalThis.navigator.connection.effectiveType&&(o+=`Connection: ${navigator.connection.effectiveType}\n`),o+=("feature"==e?"Your idea:":"Bug you experienced:")+"\n"}globalThis.open(`https://github.com/haxtheweb/issues/issues/new?assignees=&labels=${t}&template=issue-report.md&title=${i}&body=${encodeURIComponent(o)}`,"_blank")}playSound(e="coin2"){return new Promise((t=>{if(s.soundStatus&&s.appReady){let a=["click","click2","coin","coin2","hit","success"].includes(e)?e:"hit";this.audio=new Audio(new URL(`./lib/assets/sounds/${a}.mp3`,import.meta.url).href),this.audio.volume=.2,this.audio.onended=e=>{t()},this.audio.play(),setTimeout((()=>{t()}),1e3)}else t()}))}_tokenRefreshFailed(e){globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),setTimeout((()=>{this.reset(!0)}),100)}connectedCallback(){super.connectedCallback(),globalThis.matchMedia("(prefers-color-scheme: dark)").addEventListener("change",darkToggle,{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-logged-in",this._jwtLoggedIn.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-login-refresh-error",this._tokenRefreshFailed.bind(this),{signal:this.windowControllers.signal})}goToLocation(e){globalThis.location=e}disconnectedCallback(){this.windowControllers.abort(),super.disconnectedCallback()}constructor(){super(),this.unlockComingSoon=!1,this.unlockTerrible=!1,this.__loginModalOpen=!1,this.t=this.t||{},this.t={...this.t,selectPage:"Select page",backToSiteList:"Back to site list",listMySites:"List my sites",cancel:"Cancel",editDetails:"Page details",add:"Add",editSettings:"Edit settings",source:"Source",viewSource:"View source",findMedia:"Find media",undo:"Undo",redo:"Redo",media:"Media",outline:"Outline",blocks:"Blocks",addBlock:"Add block",addPage:"Add page",addChildPage:"Add child page",clonePage:"Clone page",delete:"Delete page",siteSettings:"Site settings",close:"Close",settings:"Settings",edit:"Edit",configureBlock:"Configure block",configure:"Configure",save:"Save",home:"Home",startNewJourney:"Start new journey",newJourney:"New Journey",accountInfo:"Account Info",outlineDesigner:"Site Outline",pageOutline:"Structure",more:"More",siteActions:"Site actions",insights:"Insights dashboard",merlin:"Merlin",summonMerlin:"Summon Merlin",logOut:"Log out",menu:"Menu",showMore:"More",userAccess:"User Access"},void 0!==globalThis.speechSynthesis&&(globalThis.SpeechRecognition||globalThis.webkitSpeechRecognition||globalThis.mozSpeechRecognition||globalThis.msSpeechRecognition||globalThis.oSpeechRecognition)&&(h.voiceSearch=!0),h.icon="hax:wizard-hat",h.appendContext("*"),h.allowedCallback=()=>!(!o(s.appReady)||!o(s.isLoggedIn)),h.defineOption({title:"Tour of top menu buttons",icon:"help",tags:["Help","ui","tour"],priority:-1e3,value:{target:this,method:"helpClick",args:[]},eventName:"super-daemon-element-method",path:"HAX/app/tour",context:["*"]}),h.defineOption({title:"Unlock hidden features",icon:"hax:hax2022",tags:["Developer","features","hidden"],value:{target:this,method:"fireUnlocked",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/unlockAll",context:[">"]}),h.defineOption({title:"Unlock terrible 2000s themes",icon:"hax:table-multiple",tags:["Developer","terrible","2000","tables"],value:{target:this,method:"fireTerrible",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/terrible2000s",context:[">"]}),h.defineOption({title:"Go to site",icon:"hax:hax2022",tags:["Sites","Administration","change"],eventName:"super-daemon-run-program",path:"HAX/action/goToSite",value:{name:"Go to site",program:async(e,t)=>{let a=[];return o(s.manifest.items).forEach((async t=>{(""==e||t.metadata.site&&t.metadata.site.name&&t.metadata.site.name.includes(e))&&a.push({title:t.title,icon:t.metadata.theme&&t.metadata.theme.variables&&t.metadata.theme.variables.icon?t.metadata.theme.variables.icon:"hax:hax2022",tags:["site",t.description],value:{target:this,method:"goToLocation",args:[t.slug]},eventName:"super-daemon-element-method",context:["*","HAX/action/goToSite/"+t.metadata.site.name],path:"HAX/action/goToSite/"+t.metadata.site.name})})),a}},context:["*"]}),h.defineOption({title:"Join our Community",icon:"hax:discord",priority:-100,tags:["community","discord","chat","help"],value:{target:this,method:"_openExternalLink",args:["https://bit.ly/hax-discord"]},eventName:"super-daemon-element-method",path:"HAX/community/join",context:["logged-in","CMS","HAX","*"]}),h.defineOption({title:"User Tutorials",icon:"hax:hax2022",priority:-1e3,tags:["Documentation","community","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/tutorials"]},eventName:"super-daemon-element-method",path:"HAX/community/tutorials",context:["logged-in","CMS","HAX","*"]}),h.defineOption({title:"User Documentation",icon:"hax:hax2022",tags:["Documentation","community","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/documentation"]},eventName:"super-daemon-element-method",path:"HAX/community/documentation",context:["logged-in","CMS","HAX","*"]}),h.defineOption({title:"HAX Teaching Excellence",icon:"hax:hax2022",tags:["Ontology","community","pedagogy","documentation","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/ontology"]},eventName:"super-daemon-element-method",path:"HAX/community/pedagogy",context:["logged-in","CMS","HAX","*"]}),h.defineOption({title:"Bug / issue",icon:"mdi-social:github-circle",tags:["Bug report","github","git","community","issue queue"],value:{target:this,method:"_haxStoreContribute",args:["bug","POP,bug"]},eventName:"super-daemon-element-method",path:"HAX/community/contribute",context:["logged-in","CMS","HAX","*"]}),h.defineOption({title:"Idea / Feature request",icon:"mdi-social:github-circle",tags:["Feature request","idea","github","git","community","issue queue"],value:{target:this,method:"_haxStoreContribute",args:["feature","POP,enhancement"]},context:["logged-in","CMS","HAX","*"],eventName:"super-daemon-element-method",path:"HAX/community/contribute"}),this.windowControllers=new AbortController,this.__tour=c,this.__tour.registerNewTour({key:"hax",name:"HAX top menu",style:"\n      simple-popover-manager::part(simple-popover) {\n        max-width: 250px;\n        font-family: sans-serif;\n      }\n      simple-popover-manager button {\n        font-family: sans-serif;\n        font-size: 12px;\n        margin: 0px 2px;\n        color: var(--simple-colors-default-theme-grey-12);\n      }\n      simple-popover-manager p {\n        font-family: sans-serif;\n        padding: 0;\n        margin: 0;\n        width: 250px;\n        font-size: 10px;\n        line-height: 20px;\n      }\n      simple-popover-manager h1 {\n        font-family: sans-serif;\n        margin: 0;\n        font-size: 12px;\n        width: 250px;\n        padding: 0;\n      }\n      simple-popover-manager::part(simple-popover-body),\n      simple-popover-manager::part(simple-popover-heading) {\n        color: black;\n        background-color: white;\n        font-family: sans-serif;\n      }\n      body.dark-mode simple-popover-manager::part(simple-popover-body),\n      body.dark-mode simple-popover-manager::part(simple-popover-heading) {\n        color: white;\n        background-color: black;\n        font-family: sans-serif;\n      }\n      body.dark-mode simple-popover-manager simple-icon-button-lite {\n        color: white;\n        background-color: black;\n        font-family: sans-serif;\n      }\n      "}),this.tourName="hax",i((()=>{const e=o(s.activeItem);e&&e.title&&s.setPageTitle(e.title)})),i((()=>{this.siteReady=o(s.siteReady)})),i((()=>{const e=o(s.badDevice);!1===e?import("../rpg-character/rpg-character.js"):!0===e&&globalThis.document.body.classList.add("bad-device")})),this.userMenuOpen=!1,this.courses=[],this.activeItem={},this.phrases={new:["What's ya name?","HAX to the moon","Welcome to the Jungle","We like to party","Build something awesome","Everything is awesome!","Everything is cool","When you're part of the team","When you're living our dream","Welcome to the up-side-down"],return:["Welcome back, take 2?","That wasn't very long","Stranger thiings have happened","Student driven platform","Faculty centered platform","Instructional designer influenced platform","Free, Open, Community driven","One brick at a time..","Sup?","You again? Awesome!","Let's do this","There can only be one ring...","There is another","Fancy that, I love HAX and you show up"]},this.isNewUser=null,this.basePath="/",this.searchTerm="",this.appMode="",this.soundIcon="",this.store=s,this.sound=new Audio,i((()=>{this.isNewUser=o(s.isNewUser),this.isNewUser&&"create"!==o(s.appMode)&&"404"!==o(s.appMode)&&(s.appMode="create",setTimeout((()=>{s.createSiteSteps=!0}),0))})),i((()=>{this.userName=o(s.user.name)})),i((()=>{this.appMode=o(s.appMode)})),i((()=>{this.searchTerm=o(s.searchTerm)})),i((async()=>{const e=o(s.location);e&&e.route&&(e.route.step?(s.appMode="create",s.createSiteSteps=!0):e.route.slug?(this.reset(),setTimeout((()=>{globalThis.location=e.route.slug}),0)):"404"===e.route.name?(s.createSiteSteps=!1,s.appMode="404",setTimeout((()=>{s.toast("the page miss.. it burns!!!",3e3,{fire:!0,walking:!0})}),500)):"home"!==e.route.name&&"search"!==e.route.name||(s.appMode="home",s.createSiteSteps=!1))})),i((()=>{s.routes.length>0&&null===s.location&&(s.location=o(s.routes[0]))})),null===r("app-hax-darkMode",null)&&(s.darkMode=globalThis.matchMedia("(prefers-color-scheme: dark)").matches),i((()=>{d("app-hax-darkMode",o(s.darkMode)),requestAnimationFrame((()=>{o(s.darkMode)?(globalThis.document.body.classList.add("dark-mode"),s.toast("I'm ascared of the dark",2e3,{fire:!0}),this.dark=!0,h.dark=!0,h.toastInstance.darkMode=!0):(globalThis.document.body.classList.remove("dark-mode"),s.toast("Sunny day it is",2e3,{hat:"random"}),this.dark=!1,h.dark=!1,h.toastInstance.darkMode=!1)}))})),i((()=>{const e=o(s.appMode);e&&(globalThis.document.body.classList.remove("app-hax-search"),globalThis.document.body.classList.remove("app-hax-create"),globalThis.document.body.classList.remove("app-hax-404"),globalThis.document.body.classList.remove("app-hax-home"),globalThis.document.body.classList.add(`app-hax-${e}`))}))}static get properties(){return{...super.properties,unlockComingSoon:{type:Boolean},unlockTerrible:{type:Boolean},courses:{type:Array},userName:{type:String},activeItem:{type:Object},soundIcon:{type:String},searchTerm:{type:String},appMode:{type:String},isNewUser:{type:Boolean},phrases:{type:Object},userMenuOpen:{type:Boolean},siteReady:{type:Boolean},basePath:{type:String,attribute:"base-path"},token:{type:String}}}reset(e=!1,t=!1){try{globalThis.localStorage.removeItem("app-hax-step"),globalThis.localStorage.removeItem("app-hax-site"),t&&(globalThis.localStorage.removeItem("jwt"),s.jwt=null,globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0}))),e&&(document.querySelector("base")?globalThis.location=globalThis.document.querySelector("base").href:globalThis.location.reload())}catch(e){}}fireTerrible(){this.unlockTerrible=!0,s.appEl.playSound("coin").then((()=>{s.appEl.playSound("coin2").then((()=>{s.appEl.playSound("success").then((()=>{h.merlinSpeak("Enjoy these early 2000s table based layouts. May they remind you how never to web, again.")}))}))}))}fireUnlocked(){this.unlockComingSoon=!0,s.appEl.playSound("coin").then((()=>{s.appEl.playSound("coin2").then((()=>{s.appEl.playSound("success").then((()=>{h.merlinSpeak("Unbelievable! You, (Subject Name), must be the pride of (Subject Hometown). Enjoy all locked features as a boon!")}))}))}))}logout(){globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),this.closeMenu(),this.__logoutUserAction=!0}_jwtLoggedIn(e){!1===e.detail&&this.__logoutUserAction&&(this.__logoutUserAction=!1,this.__loginModalOpen=!1,setTimeout((()=>{this.reset(!0)}),100))}login(){this.__loginModalOpen||(this.__loginModalOpen=!0,import("./lib/v2/app-hax-site-login.js").then((()=>{const e=globalThis.document.createElement("app-hax-site-login");if(this.querySelector('[slot="externalproviders"]')){const t=this.querySelector('[slot="externalproviders"]').cloneNode(!0);e.appendChild(t)}import("../simple-modal/simple-modal.js").then((()=>{setTimeout((()=>{this.dispatchEvent(new CustomEvent("simple-modal-show",{bubbles:!0,cancelable:!0,composed:!0,detail:{title:"Login to HAX",elements:{content:e},modal:!0,styles:{"--simple-modal-titlebar-background":"var(--ddd-theme-default-nittanyNavy, #001e44)","--simple-modal-titlebar-color":"var(--ddd-theme-default-white, white)","--simple-modal-width":"90vw","--simple-modal-max-width":"var(--ddd-spacing-32, 480px)","--simple-modal-min-width":"300px","--simple-modal-z-index":"1000","--simple-modal-height":"auto","--simple-modal-min-height":"300px","--simple-modal-max-height":"80vh","--simple-modal-titlebar-height":"80px","--simple-modal-border-radius":"var(--ddd-radius-md, 8px)","--simple-modal-background":"var(--ddd-theme-default-white, white)","--simple-modal-box-shadow":"var(--ddd-boxShadow-xl)"}}}))}),0)}))})))}static get styles(){return[super.styles,e`
        :host {
          overflow: hidden;
          display: block;
          max-width: 100%;
          font-family: var(--ddd-font-primary, sans-serif);
          padding-left: var(--ddd-spacing-5, 20px);
          padding-right: var(--ddd-spacing-5, 20px);
        }
        .contentSection {
          display: flex;
          align-items: flex-start;
          justify-content: flex-start;
          gap: var(--ddd-spacing-12, 48px);
          width: 100%;
          margin: 0;
          margin-top: var(--ddd-spacing-5, 20px);
          padding: 0;
          box-sizing: border-box;
        }
        .leftSection,
        .rightSection {
          display: flex;
          flex-direction: column;
          flex: 1 1 0;
        }
        .leftSection {
          width: 240px;
          min-width: 200px;
          max-width: 260px;
          margin-left: 0;
          padding-top: 0;
          box-sizing: border-box;
          align-self: flex-start;
        }
        .rightSection {
          flex: 1;
          min-width: 0;
          box-sizing: border-box;
          display: flex;
          flex-direction: column;
          overflow: visible;
        }
        .template-results {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(220px, 1fr));
          width: 100%;
          min-height: 330px;
          box-sizing: border-box;
          gap: var(--ddd-spacing-2, 16px);
        }
        #returnToSection {
          width: 100%;
        }
        #returnToSection app-hax-search-results {
          width: 100%;
          min-height: 280px;
          box-sizing: border-box;
          height: 300px;
        }
        :host(:not([show-filter])) app-hax-search-results {
          width: 100%;
        }

        h2,
        .returnTo h2,
        .startNew h2 {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-l, 24px);
          color: var(--app-hax-accent-color, var(--accent-color));
          margin: 0 0 var(--ddd-spacing-4, 16px) 0;
        }
        .startNew,
        .returnTo {
          position: relative;
          display: flex;
          flex-direction: column;
          justify-content: flex-start;
          align-items: flex-start;
          margin: 0;
        }
        .upper-filter {
          margin-bottom: var(--ddd-spacing-4, 16px);
          position: relative;
          display: inline-block;
          width: 100%;
        }
        input[type="text"] {
          width: 100%;
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-2, 8px)
            var(--ddd-spacing-2, 8px) var(--ddd-spacing-8, 32px);
          font-size: var(--ddd-font-size-xs, 12px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid);
          border-color: var(--ddd-theme-default-slateGray, #666);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-coalyGray, #222);
          transition: all 0.2s ease;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          margin: 0;
          min-height: var(--ddd-spacing-8, 32px);
        }
        :host([dark]) input[type="text"],
        body.dark-mode input[type="text"] {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        input[type="text"]:focus {
          border: var(--ddd-border-md, 2px solid);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          background: var(--ddd-theme-default-white, white);
          outline: none;
        }
        :host([dark]) input[type="text"]:focus,
        body.dark-mode input[type="text"]:focus {
          background: var(--ddd-theme-default-coalyGray, #333);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
        }
        .search-icon {
          position: absolute;
          left: var(--ddd-spacing-2, 8px);
          top: 50%;
          transform: translateY(-50%);
          font-size: var(--ddd-font-size-xs, 14px);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          pointer-events: none;
          z-index: 1;
          --simple-icon-width: var(--ddd-icon-3xs, 20px);
          --simple-icon-height: var(--ddd-icon-3xs, 20px);
        }
        :host([dark]) .search-icon,
        body.dark-mode .search-icon {
          color: var(--ddd-theme-default-white, white);
        }
        .filter {
          position: relative;
          top: 0;
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-5, 20px);
          background: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-lg, 12px);
          box-shadow: var(--ddd-boxShadow-lg);
          border: var(--ddd-border-xs, 1px solid);
          border-color: var(--ddd-theme-default-slateGray, #666);
          padding: var(--ddd-spacing-6, 24px);
          margin-top: 0;
          margin-bottom: 0;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          transition: box-shadow 0.2s ease;
        }
        :host([dark]) .filter,
        body.dark-mode .filter {
          background: var(--ddd-theme-default-coalyGray, #222);
          border-color: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
        }
        .filter:hover {
          box-shadow: var(--ddd-boxShadow-xl);
        }
        .filterButtons {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-3, 12px);
          margin-top: 0;
          border: none;
          padding: 0;
          margin: 0;
        }
        .filter-btn {
          display: flex;
          align-items: center;
          justify-content: flex-start;
          gap: var(--ddd-spacing-1, 4px);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid) transparent;
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          cursor: pointer;
          box-shadow: var(--ddd-boxShadow-sm);
          transition: all 0.2s ease;
          outline: none;
          min-height: var(--ddd-spacing-7, 28px);
          text-align: left;
        }
        :host([dark]) .filter-btn,
        body.dark-mode .filter-btn {
          background: var(--ddd-theme-default-slateGray, #444);
          color: var(--ddd-theme-default-white, white);
        }
        .filter-btn.active,
        .filter-btn:active {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          box-shadow: var(--ddd-boxShadow-md);
        }
        :host([dark]) .filter-btn.active,
        :host([dark]) .filter-btn:active,
        body.dark-mode .filter-btn.active,
        body.dark-mode .filter-btn:active {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          border-color: var(--ddd-theme-default-white, white);
        }
        .filter-btn:hover,
        .filter-btn:focus {
          background: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
          transform: translateY(-1px);
        }
        .filter-btn:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        :host([dark]) .filter-btn:hover,
        :host([dark]) .filter-btn:focus,
        body.dark-mode .filter-btn:hover,
        body.dark-mode .filter-btn:focus {
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        .filter-btn.active:hover,
        .filter-btn.active:focus {
            background: var(--ddd-theme-default-keystoneYellow, #ffd100);
            color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        :host([dark]) .filter-btn.active:hover,
        :host([dark]) .filter-btn.active:focus,
        body.dark-mode .filter-btn.active:hover,
        body.dark-mode .filter-btn.active:focus {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        .filter-btn .icon {
          font-size: var(--ddd-font-size-3xs, 12px);
          color: inherit;
          display: flex;
          align-items: center;
          flex-shrink: 0;
          width: var(--ddd-icon-3xs, 20px);
          height: var(--ddd-icon-3xs, 20px);
        }
        .filter-btn .icon simple-icon-lite {
          color: inherit;
          --simple-icon-width: var(--ddd-icon-3xs, 20px);
          --simple-icon-height: var(--ddd-icon-3xs, 20px);
        }
        .filter-btn.active .icon,
        .filter-btn.active .icon simple-icon-lite {
          color: inherit;
        }
        :host([dark]) .filter-btn.active .icon,
        :host([dark]) .filter-btn.active .icon simple-icon-lite,
        body.dark-mode .filter-btn.active .icon,
        body.dark-mode .filter-btn.active .icon simple-icon-lite {
          color: inherit;
        }
        .filter-btn:hover .icon simple-icon-lite,
        .filter-btn:focus .icon simple-icon-lite {
          color: inherit;
        }
        input[type="checkbox"] {
          display: none;
        }
        .sort-control {
          margin-top: var(--ddd-spacing-2, 8px);
          margin-bottom: var(--ddd-spacing-3, 12px);
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-1, 4px);
        }
        .sort-label {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-3xs, 11px);
          color: var(--ddd-theme-default-coalyGray, #222);
        }
        :host([dark]) .sort-label,
        body.dark-mode .sort-label {
          color: var(--ddd-theme-default-limestoneGray, #f5f5f5);
        }
        .sort-select {
          width: 100%;
          padding: var(--ddd-spacing-1, 4px) var(--ddd-spacing-2, 8px);
          font-size: var(--ddd-font-size-3xs, 11px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-slateGray, #666);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-coalyGray, #222);
          font-family: var(--ddd-font-primary, sans-serif);
          box-sizing: border-box;
        }
        :host([dark]) .sort-select,
        body.dark-mode .sort-select {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        .sort-select:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        .reset-button {
          margin-top: var(--ddd-spacing-1, 4px);
          background: var(--ddd-theme-default-original87Pink, #e4007c);
          border: var(--ddd-border-xs, 1px solid) transparent;
          color: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-sm, 4px);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--ddd-spacing-1, 4px);
          box-shadow: var(--ddd-boxShadow-sm);
          cursor: pointer;
          transition: all 0.2s ease;
          min-height: var(--ddd-spacing-7, 28px);
        }
        .reset-button:hover,
        .reset-button:focus {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
          transform: translateY(-1px);
        }
        .reset-button:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        :host([dark]) .reset-button,
        body.dark-mode .reset-button {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
        }
        :host([dark]) .reset-button:hover,
        :host([dark]) .reset-button:focus,
        body.dark-mode .reset-button:hover,
        body.dark-mode .reset-button:focus {
          background: var(--ddd-theme-default-original87Pink, #e4007c);
        }
        .collapseFilter {
          display: none;
        }

        /* Visually hidden content for screen readers */
        .visually-hidden {
          position: absolute !important;
          width: 1px !important;
          height: 1px !important;
          padding: 0 !important;
          margin: -1px !important;
          overflow: hidden !important;
          clip: rect(0, 0, 0, 0) !important;
          white-space: nowrap !important;
          border: 0 !important;
        }

        /* Loading and fallback messages */
        .loading-message,
        .no-results {
          grid-column: 1 / -1;
          text-align: center;
          padding: var(--ddd-spacing-8, 32px);
          font-size: var(--ddd-font-size-s, 16px);
          color: var(--ddd-theme-default-slateGray, #666);
          font-family: var(--ddd-font-primary, sans-serif);
        }
        :host([dark]) .loading-message,
        :host([dark]) .no-results,
        body.dark-mode .loading-message,
        body.dark-mode .no-results {
          color: var(--ddd-theme-default-limestoneGray, #ccc);
        }
        .fallback-message {
          grid-column: 1 / -1;
          display: flex;
          flex-direction: column;
          align-items: center;
          gap: var(--ddd-spacing-4, 16px);
        }
        .fallback-message .no-results {
          padding: var(--ddd-spacing-4, 16px) 0;
        }
        .fallback-message app-hax-use-case {
          max-width: 300px;
        }

        /* 780px - 601px */
        @media (max-width: 780px) {
          :host {
            padding-left: var(--ddd-spacing-5, 20px);
            padding-right: var(--ddd-spacing-5, 20px);
          }
          .contentSection {
            display: block;
          }
          .leftSection, .rightSection {
            width: 100%;
            max-width: 100%;
            position: relative;
          }
          .leftSection {
            margin-bottom: var(--ddd-spacing-4, 16px);
          }
          :host([show-filter]) .filter {
            display: flex;
            width: 250px;
            max-width: 250px;
          }
          :host .collapseFilter {
            display: flex;
          }
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-m, 20px);
          }
          .template-results {
            grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
            gap: var(--ddd-spacing-3, 12px);
          }
        }

        /* 600px - 481px*/
        @media (max-width: 600px) {
          .leftSection {
            margin-bottom: var(--ddd-spacing-3, 12px);
          }
          :host([show-filter]) .filter {
            width: 200px;
          }
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-s, 18px);
          }
          .template-results {
            grid-template-columns: 1fr;
            gap: var(--ddd-spacing-2, 8px);
          }
        }

        /* 480px - 0px*/
        @media (max-width: 480px) {
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-s, 16px);
            margin-bottom: var(--ddd-spacing-3, 12px);
          }
          #returnToSection app-hax-search-results {
            min-width: 100%;
          }
        }
        
        .no-results {
          font-size: var(--ddd-font-size-s, 16px);
          color: light-dark(
            var(--ddd-theme-default-coalyGray, #222),
            var(--ddd-theme-default-white, white)
          );
        }
      `]}helpClick(){s.appEl.playSound("coin2"),this.__tour.startTour("hax")}updated(e){super.updated&&super.updated(e),e.forEach(((e,t)=>{["basePath"].includes(t)&&this[t]&&(s.AppHaxAPI[t]=this[t]),["token"].includes(t)&&this[t]&&(s[t]=this[t])}))}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),s.AppHaxAPI&&this.basePath&&(s.AppHaxAPI.basePath=this.basePath),import("./lib/v2/app-hax-steps.js"),import("./lib/v2/app-hax-site-button.js"),import("../../wired-elements/lib/wired-button.js"),import("./lib/v2/app-hax-toast.js"),import("./lib/v2/app-hax-wired-toggle.js"),import("./lib/v2/app-hax-search-bar.js"),import("./lib/v2/app-hax-search-results.js"),import("./lib/v2/app-hax-user-menu.js"),import("./lib/v2/app-hax-user-menu-button.js"),this.dispatchEvent(new CustomEvent("app-hax-loaded",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),s.appReady=!0,i((()=>{o(s.appReady)?globalThis.document.body.classList.add("app-loaded"):globalThis.document.body.classList.remove("app-loaded")})),s.appEl=this,i((()=>{s.activeItem&&(this.activeItem=o(s.activeItem))})),i((()=>{this.soundIcon=o(s.soundStatus)?new URL("./lib/assets/images/FullVolume.svg",import.meta.url).href:new URL("./lib/assets/images/Silence.svg",import.meta.url).href,o(s.soundStatus)?s.toast("You can hear me now? Good.",2e3,{hat:"random"}):s.toast("Sound off.. hey! HELLO!?!",2e3,{fire:!0})})),i((async()=>{if(o(s.appReady)&&s.AppHaxAPI&&o(s.isLoggedIn)&&o(s.appSettings)&&o(s.refreshSiteList)){const e=await p.makeCall("getSitesList");s.manifest=e.data}else o(s.appReady)&&!o(s.isLoggedIn)&&this.login()}))}toggleMenu(){this.userMenuOpen=!this.userMenuOpen,s.appEl.playSound("click")}closeMenu(){this.userMenuOpen=!1}openMerlin(){s.appEl.playSound("click"),h.open()}render(){return t`<app-hax-router></app-hax-router>
      <header>
        <app-hax-top-bar>
          <slot name="app-header-pre" slot="left"></slot>
          <a
            id="home"
            title="${this.t.home}"
            href="home"
            tabindex="-1"
            slot="left"
          >
            <simple-icon-lite
              id="hlogo"
              icon="hax:hax2022"
              tabindex="0"
              class="haxLogo"
              title="${this.t.home}"
              data-simple-tour-stop
              data-stop-title="data-label"
              data-label="${this.t.home}"
            >
              <div data-stop-content slot="tour" style="display:none;">
                You can come back to this home screen whenever you click this!
              </div>
            </simple-icon-lite>
          </a>
          <simple-tooltip for="hlogo" position="right" slot="left"
            >${this.t.home}</simple-tooltip
          >
          <simple-toolbar-button
            icon="hax:wizard-hat"
            label="${this.t.merlin}"
            voice-command="${this.t.merlin}"
            icon-position="left"
            id="merlin"
            @click="${this.openMerlin}"
            slot="center"
            data-event="super-daemon"
            show-text-label
          ></simple-toolbar-button>
          <wired-button
            elevation="1"
            slot="right"
            class="soundToggle"
            id="soundtb"
            @click="${soundToggle}"
            data-simple-tour-stop
            data-stop-title="data-label"
            data-label="Sound"
            aria-label="Toggle sound effects ${this.soundIcon.includes("FullVolume")?"off":"on"}"
            aria-pressed="${this.soundIcon.includes("FullVolume")}"
          >
            <span class="wired-button-label">Toggle sound effects</span>
            <simple-icon-lite
              src="${this.soundIcon}"
              loading="lazy"
              decoding="async"
              alt="${this.soundIcon.includes("FullVolume")?"Sound enabled":"Sound disabled"}"
            >
            </simple-icon-lite>
            <div slot="tour" data-stop-content>
              Not a fan of the (awesome) sound effects? You can mute them if you
              prefer.
            </div>
          </wired-button>
          <simple-tooltip for="soundtb" position="bottom" slot="right"
            >Toggle sound</simple-tooltip
          >
          <app-hax-wired-toggle id="wt" slot="right"></app-hax-wired-toggle>
          <simple-tooltip for="wt" position="bottom" slot="right"
            >Toggle dark mode</simple-tooltip
          >
          <app-hax-user-menu
            slot="right"
            id="user-menu"
            ?is-open="${this.userMenuOpen}"
            data-simple-tour-stop
            data-stop-title="data-label"
            data-label="Menu"
          >
            <div slot="tour" data-stop-content>
              You want to log out and be someone else? Create a new site? Click
              your character. Your character is unique to you!
            </div>
            <button
              @click="${this.toggleMenu}"
              class="topbar-character"
              slot="menuButton"
              id="tbchar"
              title="System menu"
              aria-label="User menu for ${this.userName}"
              aria-expanded="${this.userMenuOpen}"
              aria-haspopup="menu"
            >
              <rpg-character
                seed="${this.userName}"
                width="68"
                height="68"
                hat="${this.userMenuOpen?"edit":"none"}"
                alt="Avatar for ${this.userName}"
                role="img"
              ></rpg-character>
              <span class="characterbtn-name" aria-hidden="true"
                >${this.userName}</span
              >
            </button>
            <a
              slot="main-menu"
              title="${this.t.listMySites}"
              href="home"
              tabindex="-1"
            >
              <app-hax-user-menu-button
                icon="hax:hax2022"
                label="${this.t.listMySites}"
                part="listMySites"
              ></app-hax-user-menu-button>
            </a>
            <!-- <app-hax-user-menu-button
              slot="main-menu"
              icon="face"
              label="Account Info"
            ></app-hax-user-menu-button> -->
            <app-hax-user-menu-button
              slot="post-menu"
              class="logout"
              label="Logout"
              @click=${this.logout}
            ></app-hax-user-menu-button>
          </app-hax-user-menu>
        </app-hax-top-bar>
      </header>
      <main @click="${this.closeMenu}">
        <confetti-container id="confetti">
          <div class="label">
            <app-hax-label heading-level="1">
              HAX Site Dashboard
              <div slot="subtitle">Let's build something awesome!</div>
            </app-hax-label>
          </div>
          <section class="content">
            <div class="start-journey">
              <app-hax-use-case-filter></app-hax-use-case-filter>
            </div>
          </section>
        </confetti-container>
      </main>`}siteReadyToGo(e){import("../multiple-choice/lib/confetti-container.js").then((e=>{setTimeout((()=>{this.shadowRoot.querySelector("#confetti").setAttribute("popped","")}),0)})),e.detail&&(s.siteReady=!0)}}customElements.define(AppHax.tag,AppHax),globalThis.AppHax=globalThis.AppHax||{},globalThis.AppHax.requestAvailability=()=>(!globalThis.AppHax.instance&&globalThis.document&&(globalThis.AppHax.instance=globalThis.document.querySelector(AppHax.tag)),globalThis.AppHax.instance);
