// @haxtheweb/app-hax/app-hax.js
import{css as e,html as t,unsafeCSS as a}from"../../lit/index.js";import{toJS as i,autorun as o}from"../../mobx/dist/mobx.esm.js";import{localStorageSet as s,localStorageGet as n}from"../utils/utils.js";import"../simple-tooltip/simple-tooltip.js";import"../simple-icon/lib/simple-icon-lite.js";import"../simple-icon/lib/simple-icons.js";import{DDD as r}from"../d-d-d/d-d-d.js";import{store as d}from"./lib/v2/AppHaxStore.js";import{I18NMixin as l}from"../i18n-manager/lib/I18NMixin.js";import{AppHaxAPI as h}from"./lib/v2/AppHaxBackendAPI.js";import{SuperDaemonInstance as p}from"../super-daemon/super-daemon.js";import"../simple-toolbar/lib/simple-toolbar-button.js";import{ResponsiveUtilityBehaviors as m}from"../responsive-utility/lib/responsive-utility-behaviors.js";import"./lib/v2/AppHaxRouter.js";import"../haxcms-elements/lib/core/ui/app-hax-top-bar.js";import"./lib/v2/app-hax-use-case.js";import"./lib/v2/app-hax-use-case-filter.js";import"./lib/v2/app-hax-search-results.js";import"./lib/v2/app-hax-scroll-button.js";function darkToggle(e){e.matches?d.darkMode=!0:d.darkMode=!1}function soundToggle(){d.soundStatus=!i(d.soundStatus),s("app-hax-soundStatus",i(d.soundStatus)),d.appEl.playSound("click")}const c={dashboard:{label:"Settings",icon:"settings",route:""},status:{label:"Status",icon:"hax:graph",route:"status"},skeletons:{label:"Skeletons",icon:"hax:site-map",route:"skeletons"},themes:{label:"Themes",icon:"lrn:palette",route:"themes"},blocks:{label:"Blocks",icon:"hax:blocks",route:"blocks"},integrations:{label:"Integrations",icon:"communication:vpn-key",route:"integrations"},media:{label:"Media",icon:"image:image",route:"media"},"style-guide":{label:"Style guide",icon:"editor:format-color-text",route:"style-guide"},"custom-code":{label:"JS/CSS",icon:"code",route:"js-css"}},u={dashboard:"dashboard",status:"status",skeletons:"skeletons",themes:"themes",blocks:"blocks",integrations:"integrations",media:"media","style-guide":"style-guide","custom-code":"custom-code","js-css":"custom-code"};export class AppHax extends(l(m(r))){static get tag(){return"app-hax"}_syncResponsiveStoreState(){"string"==typeof this.responsiveSize&&""!==this.responsiveSize&&d.responsiveSize!==this.responsiveSize&&(d.responsiveSize=this.responsiveSize)}_openExternalLink(e){globalThis.open(e,"_blank")}async _haxStoreContribute(e,t,a=null){let i="";if("merlin"==e){var o=`[${e}] New command request from HAX daemon`;i=`Location: ${globalThis.location.href}\nMerlin command: ${a}\nWhat did you want merlin to do?\n`}else{o=`[${e}] User report from HAX daemon`;if(i=`Location: ${globalThis.location.href}\nBrowser: ${navigator.userAgent}\nOS: ${navigator.userAgentData.platform} - ${navigator.deviceMemory}GB RAM - ${navigator.hardwareConcurrency} cores\nScreen: ${globalThis.screen.width}x${globalThis.screen.height}\nWindow size: ${globalThis.innerWidth}x${globalThis.innerHeight}\n`,navigator.getBattery){i+=`Battery: ${100*(await globalThis.navigator.getBattery()).level}%\n`}navigator.connection&&globalThis.navigator.connection.effectiveType&&(i+=`Connection: ${navigator.connection.effectiveType}\n`),i+=("feature"==e?"Your idea:":"Bug you experienced:")+"\n"}globalThis.open(`https://github.com/haxtheweb/issues/issues/new?assignees=&labels=${t}&template=issue-report.md&title=${o}&body=${encodeURIComponent(i)}`,"_blank")}playSound(e="coin2"){return new Promise(t=>{if(d.soundStatus&&d.appReady){let a=["click","click2","coin","coin2","hit","success"].includes(e)?e:"hit";this.audio=new Audio(new URL(`./lib/assets/sounds/${a}.mp3`,import.meta.url).href),this.audio.volume=.1,this.audio.onended=e=>{t()},this.audio.play(),setTimeout(()=>{t()},1e3)}else t()})}_tokenRefreshFailed(e){globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),setTimeout(()=>{this.reset(!0)},100)}connectedCallback(){super.connectedCallback(),globalThis.matchMedia("(prefers-color-scheme: dark)").addEventListener("change",darkToggle,{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-logged-in",this._jwtLoggedIn.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("jwt-login-refresh-error",this._tokenRefreshFailed.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("super-daemon-close",this._handleSuperDaemonClose.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("popstate",this._adminRoutePopState.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("simple-modal-closed",this._simpleModalClosed.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("haxcms-system-settings-panel-changed",this._systemSettingsPanelChanged.bind(this),{signal:this.windowControllers.signal}),globalThis.addEventListener("simple-modal-breadcrumb-click",this._simpleModalBreadcrumbClick.bind(this),{signal:this.windowControllers.signal})}goToLocation(e){globalThis.location=e}disconnectedCallback(){this.windowControllers.abort(),super.disconnectedCallback()}constructor(){super(),this.rpgWalk=!1,this.unlockComingSoon=!1,this.unlockTerrible=!1,this.__loginModalOpen=!1,this.t=this.t||{},this.t={...this.t,selectPage:"Select page",listMySites:"List my sites",cancel:"Cancel",editDetails:"Page details",add:"Add",editSettings:"Edit settings",source:"Source",viewSource:"View source",findMedia:"Find media",undo:"Undo",redo:"Redo",media:"Media",outline:"Outline",blocks:"Blocks",addBlock:"Add block",addPage:"Add page",addChildPage:"Add child page",clonePage:"Clone page",delete:"Delete page",siteSettings:"Site Settings",close:"Close",settings:"Settings",edit:"Edit",configureBlock:"Configure block",configure:"Configure",save:"Save",home:"Home",startNewJourney:"Start new journey",newJourney:"New Journey",accountInfo:"Account Info",outlineDesigner:"Site Outline",pageOutline:"Structure",more:"More",siteActions:"Site actions",insights:"Insights",merlin:"Merlin",summonMerlin:"Summon Merlin",logOut:"Log out",menu:"Menu",showMore:"More",userAccess:"User Access"},void 0!==globalThis.speechSynthesis&&(globalThis.SpeechRecognition||globalThis.webkitSpeechRecognition||globalThis.mozSpeechRecognition||globalThis.msSpeechRecognition||globalThis.oSpeechRecognition)&&(p.voiceSearch=!0),p.icon="hax:wizard-hat",p.appendContext("*"),p.allowedCallback=()=>!(!i(d.appReady)||!i(d.isLoggedIn)),p.defineOption({title:"Unlock hidden features",icon:"hax:hax2022",tags:["Developer","features","hidden"],value:{target:this,method:"fireUnlocked",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/unlockAll",context:[">"]}),p.defineOption({title:"Show terrible themes",icon:"icons:warning",tags:["Developer","themes","terrible"],value:{target:this,method:"fireTerrible",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/showTerribleThemes",context:[">"]}),p.defineOption({title:"Show hidden templates",icon:"icons:visibility",tags:["Developer","templates","hidden"],value:{target:this,method:"fireHiddenTemplates",args:[]},eventName:"super-daemon-element-method",path:">developer/hax/showHiddenTemplates",context:[">"]}),p.defineOption({title:"Go to site",icon:"hax:hax2022",tags:["Sites","Administration","change"],eventName:"super-daemon-run-program",path:"HAX/action/goToSite",value:{name:"Go to site",program:async(e,t)=>{let a=[];return i(d.manifest.items).forEach(async t=>{(""==e||t.metadata.site&&t.metadata.site.name&&t.metadata.site.name.includes(e))&&a.push({title:t.title,icon:t.metadata.theme&&t.metadata.theme.variables&&t.metadata.theme.variables.icon?t.metadata.theme.variables.icon:"hax:hax2022",tags:["site",t.description],value:{target:this,method:"goToLocation",args:[t.slug]},eventName:"super-daemon-element-method",context:["*","HAX/action/goToSite/"+t.metadata.site.name],path:"HAX/action/goToSite/"+t.metadata.site.name})}),a}},context:["*"]}),p.defineOption({title:"System Settings",icon:"settings",tags:["Administration","settings","system","dashboard"],value:{target:this,method:"openSystemSettingsFromMerlin",args:[]},eventName:"super-daemon-element-method",path:"HAX/action/systemSettings",context:["system-settings"]}),p.defineOption({title:"Join our Community",icon:"hax:discord",priority:-100,tags:["community","discord","chat","help"],value:{target:this,method:"_openExternalLink",args:["https://discord.gg/VVcAcCeZQ"]},eventName:"super-daemon-element-method",path:"HAX/community/join",context:["logged-in","CMS","HAX","*"]}),p.defineOption({title:"User Tutorials",icon:"hax:hax2022",priority:-1e3,tags:["Documentation","community","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/tutorials"]},eventName:"super-daemon-element-method",path:"HAX/community/tutorials",context:["logged-in","CMS","HAX","*"]}),p.defineOption({title:"User Documentation",icon:"hax:hax2022",tags:["Documentation","community","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/documentation"]},eventName:"super-daemon-element-method",path:"HAX/community/documentation",context:["logged-in","CMS","HAX","*"]}),p.defineOption({title:"HAX Teaching Excellence",icon:"hax:hax2022",tags:["Ontology","community","pedagogy","documentation","help"],value:{target:this,method:"_openExternalLink",args:["https://haxtheweb.org/ontology"]},eventName:"super-daemon-element-method",path:"HAX/community/pedagogy",context:["logged-in","CMS","HAX","*"]}),p.defineOption({title:"Bug / issue",icon:"mdi-social:github-circle",tags:["Bug report","github","git","community","issue queue"],value:{target:this,method:"_haxStoreContribute",args:["bug","POP,bug"]},eventName:"super-daemon-element-method",path:"HAX/community/contribute",context:["logged-in","CMS","HAX","*"]}),p.defineOption({title:"Idea / Feature request",icon:"mdi-social:github-circle",tags:["Feature request","idea","github","git","community","issue queue"],value:{target:this,method:"_haxStoreContribute",args:["feature","POP,enhancement"]},context:["logged-in","CMS","HAX","*"],eventName:"super-daemon-element-method",path:"HAX/community/contribute"}),this.windowControllers=new AbortController,o(()=>{const e=i(d.activeItem);e&&e.title&&d.setPageTitle(e.title)}),o(()=>{this.siteReady=i(d.siteReady)}),o(()=>{const e=i(d.badDevice);!1===e?import("../rpg-character/rpg-character.js"):!0===e&&globalThis.document.body.classList.add("bad-device")}),this.userMenuOpen=!1,this.hasHeaderPre=!1,this.isHAXiamDeployment=!1,this.__currentAdminRoutePath=null,this.__systemSettingsModalOpen=!1,this.__systemSettingsActivePanel="dashboard",this.__systemSettingsDialog=null,this.courses=[],this.activeItem={},this.phrases={new:["What's ya name?","HAX to the moon","Welcome to the Jungle","We like to party","Build something awesome","Everything is awesome!","Everything is cool","When you're part of the team","When you're living our dream","Welcome to the up-side-down"],return:["Welcome back, take 2?","That wasn't very long","Stranger thiings have happened","Student driven platform","Faculty centered platform","Instructional designer influenced platform","Free, Open, Community driven","One brick at a time..","Sup?","You again? Awesome!","Let's do this","There can only be one ring...","There is another","Fancy that, I love HAX and you show up"]},this.isNewUser=null,this.basePath="/",this.searchTerm="",this.appMode="",this.soundIcon="",this.store=d,this.sound=new Audio,o(()=>{this.isNewUser=i(d.isNewUser),this.isNewUser&&"create"!==i(d.appMode)&&"404"!==i(d.appMode)&&(d.appMode="create",setTimeout(()=>{d.createSiteSteps=!0},0))}),o(()=>{this.userName=i(d.user.name)}),o(()=>{this.appMode=i(d.appMode)}),o(()=>{this.searchTerm=i(d.searchTerm)}),o(()=>{const e=i(d.appSettings);this.isHAXiamDeployment=this._isHAXiamDeployment(e),this._syncSystemSettingsMerlinContext(e)}),o(async()=>{const e=i(d.location);e&&e.route&&(e.route.step?(d.appMode="create",d.createSiteSteps=!0):e.route.slug?(this.reset(),setTimeout(()=>{globalThis.location=e.route.slug},0)):"404"===e.route.name?(d.createSiteSteps=!1,d.appMode="404",setTimeout(()=>{d.toast("the page miss.. it burns!!!",3e3,{fire:!0,walking:!0})},500)):"home"!==e.route.name&&"search"!==e.route.name||(d.appMode="home",d.createSiteSteps=!1))}),o(()=>{d.routes.length>0&&null===d.location&&(d.location=i(d.routes[0]))}),null===n("app-hax-darkMode",null)&&(d.darkMode=globalThis.matchMedia("(prefers-color-scheme: dark)").matches),o(()=>{s("app-hax-darkMode",i(d.darkMode)),requestAnimationFrame(()=>{i(d.darkMode)?(globalThis.document.body.classList.add("dark-mode"),d.toast("I'm ascared of the dark",2e3,{fire:!0}),this.dark=!0,p.dark=!0,p.toastInstance.darkMode=!0):(globalThis.document.body.classList.remove("dark-mode"),d.toast("Sunny day it is",2e3,{hat:"random"}),this.dark=!1,p.dark=!1,p.toastInstance.darkMode=!1)})}),o(()=>{const e=i(d.appMode);e&&(globalThis.document.body.classList.remove("app-hax-search"),globalThis.document.body.classList.remove("app-hax-create"),globalThis.document.body.classList.remove("app-hax-404"),globalThis.document.body.classList.remove("app-hax-home"),globalThis.document.body.classList.add(`app-hax-${e}`))})}static get properties(){return{...super.properties,rpgWalk:{type:Boolean},unlockComingSoon:{type:Boolean},unlockTerrible:{type:Boolean},courses:{type:Array},userName:{type:String},activeItem:{type:Object},soundIcon:{type:String},searchTerm:{type:String},appMode:{type:String},isNewUser:{type:Boolean},phrases:{type:Object},userMenuOpen:{type:Boolean},isHAXiamDeployment:{type:Boolean,attribute:"is-haxiam-deployment"},hasHeaderPre:{type:Boolean,reflect:!0,attribute:"has-header-pre"},siteReady:{type:Boolean},basePath:{type:String,attribute:"base-path"},token:{type:String}}}reset(e=!1,t=!1){try{globalThis.localStorage.removeItem("app-hax-step"),globalThis.localStorage.removeItem("app-hax-site"),t&&(globalThis.localStorage.removeItem("jwt"),d.jwt=null,globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0}))),e&&(document.querySelector("base")?globalThis.location=globalThis.document.querySelector("base").href:globalThis.location.reload())}catch(e){}}fireTerrible(){this.unlockTerrible=!0,d.showTerribleTemplates=!0,d.appEl.playSound("coin").then(()=>{d.appEl.playSound("coin2").then(()=>{d.appEl.playSound("success").then(()=>{p.merlinSpeak("Enjoy these early 2000s table based layouts. May they remind you how never to web, again.")})})})}fireHiddenTemplates(){d.showHiddenTemplates=!0,d.appEl.playSound("success").then(()=>{p.merlinSpeak("Hidden templates revealed. With great power comes great questionable design decisions.")})}fireUnlocked(){this.unlockComingSoon=!0,d.appEl.playSound("coin").then(()=>{d.appEl.playSound("coin2").then(()=>{d.appEl.playSound("success").then(()=>{p.merlinSpeak("Unbelievable! You, (Subject Name), must be the pride of (Subject Hometown). Enjoy all locked features as a boon!")})})})}logout(){globalThis.dispatchEvent(new CustomEvent("jwt-login-logout",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),this.closeMenu(),this.__logoutUserAction=!0}_jwtLoggedIn(e){!1===e.detail&&this.__logoutUserAction&&(this.__logoutUserAction=!1,this.__loginModalOpen=!1,setTimeout(()=>{this.reset(!0)},100))}login(){this.__loginModalOpen||(this.__loginModalOpen=!0,import("./lib/v2/app-hax-site-login.js").then(()=>{const e=globalThis.document.createElement("app-hax-site-login");if(this.querySelector('[slot="externalproviders"]')){const t=this.querySelector('[slot="externalproviders"]').cloneNode(!0);e.appendChild(t)}import("../simple-modal/simple-modal.js").then(()=>{setTimeout(()=>{this.dispatchEvent(new CustomEvent("simple-modal-show",{bubbles:!0,cancelable:!0,composed:!0,detail:{title:"Login to HAX",elements:{content:e},modal:!0,styles:{"--simple-modal-titlebar-background":"var(--ddd-theme-default-nittanyNavy, #001e44)","--simple-modal-titlebar-color":"var(--ddd-theme-default-white, white)","--simple-modal-width":"90vw","--simple-modal-max-width":"var(--ddd-spacing-32, 480px)","--simple-modal-min-width":"300px","--simple-modal-z-index":"10000","--simple-modal-backdrop-background":"rgba(0, 0, 0, 0.88)","--simple-modal-height":"auto","--simple-modal-min-height":"300px","--simple-modal-max-height":"80vh","--simple-modal-titlebar-height":"80px","--simple-modal-border-radius":"var(--ddd-radius-md, 8px)","--simple-modal-background":"var(--ddd-theme-default-white, white)","--simple-modal-box-shadow":"var(--ddd-boxShadow-xl)"}}}))},0)})}))}static get styles(){return[super.styles,e`
        :host {
          overflow: visible;
          display: block;
          max-width: 100%;
          font-family: var(--ddd-font-primary, sans-serif);
        }
        app-hax-top-bar::part(top-bar) {
          grid-template-columns: minmax(0, 1fr) auto minmax(0, 1fr);
          overflow: visible;
        }

        :host([dark]) app-hax-top-bar,
        body.dark-mode app-hax-top-bar {
          --bg-color: #000;
          --accent-color: #fff;
        }

        .left-group {
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-2, 8px);
          height: var(--top-bar-height, 64px);
          justify-content: flex-start;
        }
        slot[name="app-header-pre"] {
          display: inline-flex;
          align-items: center;
          justify-content: center;
          height: var(--top-bar-height, 64px);
        }
        slot[name="app-header-pre"]::slotted(*) {
          color: var(--ddd-theme-default-white, #fff);
          background-color: var(--ddd-primary-2, #1e407c);
          height: var(--top-bar-height, 64px);
          width: var(--top-bar-height, 64px);
          min-height: var(--top-bar-height, 64px);
          min-width: var(--top-bar-height, 64px);
          flex: 0 0 var(--top-bar-height, 64px);
          box-sizing: border-box;
          display: inline-flex;
          align-items: center;
          justify-content: center;
          text-decoration: none;
          --simple-icon-height: var(--ddd-icon-2xl, 48px);
          --simple-icon-width: var(--ddd-icon-2xl, 48px);
        }

        #home {
          color: var(--ddd-theme-default-white, #fff);
          background-color: var(--ddd-primary-2, #1e407c);
          height: var(--top-bar-height, 64px);
          width: var(--top-bar-height, 64px);
          display: flex;
          align-items: center;
          justify-content: center;
          text-decoration: none;
        }

        #home:hover,
        #home:focus,
        #home:active {
          color: var(--ddd-theme-default-white, #fff);
        }

        #home:focus-visible,
        .topbar-character:focus-visible,
        #tbchar:focus-visible {
          outline: var(--ddd-border-sm, 2px solid)
            var(--ddd-theme-default-skyBlue);
          outline-offset: 2px;
        }

        .haxLogo {
          --simple-icon-height: var(--ddd-icon-md, 48px);
          --simple-icon-width: var(--ddd-icon-md, 48px);
        }

        .right-group {
          align-items: center;
          justify-content: flex-end;
          display: flex;
          gap: var(--ddd-spacing-2, 8px);
          height: var(--top-bar-height, 64px);
        }

        .topbar-character {
          cursor: pointer;
          border: none;
          background-color: transparent;
          min-height: var(--top-bar-height, 64px);
          display: inline-flex;
          align-items: center;
          justify-content: center;
          color: light-dark(
            var(--ddd-theme-default-coalyGray, #222),
            var(--ddd-theme-default-white, #fff)
          );
          padding: 0;
          margin: 0;
          line-height: 0;
        }

        .topbar-character rpg-character {
          --simple-toolbar-border-radius: var(--ddd-radius-md);
          border-radius: var(--ddd-radius-md);
          background-color: var(--simple-colors-default-theme-grey-1);
          border: var(--ddd-border-sm);
          border-color: var(--ddd-theme-default-limestoneGray);
          padding: 0;
          margin: 0 8px 0 4px;
          height: var(--ddd-spacing-11, 44px);
          width: var(--ddd-spacing-11, 44px);
          display: block;
        }

        .topbar-character:hover rpg-character,
        .topbar-character:focus rpg-character,
        .topbar-character:focus-visible rpg-character {
          border-color: var(--simple-colors-default-theme-grey-12);
          background-color: var(--simple-colors-default-theme-grey-2);
        }

        .characterbtn-name {
          margin-left: var(--ddd-spacing-2, 8px);
          font-size: var(--ddd-font-size-3xs, 12px);
          vertical-align: bottom;
          line-height: var(--top-bar-height, 64px);
          overflow: hidden;
          text-overflow: ellipsis;
          min-height: var(--top-bar-height, 64px);
          word-break: break-all;
        }
        .ops-panel {
          justify-content: space-around;
          display: flex;
          align-items: center;
          padding: var(--ddd-spacing-1, 4px) 0px;
          gap: var(--ddd-spacing-2, 8px);
        }
        .soundToggle {
          position: relative;
          display: inline-flex;
          vertical-align: top;
          min-width: var(--ddd-spacing-12, 48px);
          min-height: var(--ddd-spacing-12, 48px);
          align-items: center;
          justify-content: center;
        }
        .soundToggle simple-icon-lite {
          --simple-icon-height: var(--ddd-icon-xs, 24px);
          --simple-icon-width: var(--ddd-icon-xs, 24px);
        }

        .characterbtn-menu-icon {
          --simple-icon-height: var(--ddd-icon-4xs, 16px);
          --simple-icon-width: var(--ddd-icon-4xs, 16px);
          margin-left: var(--ddd-spacing-1, 4px);
          display: inline-flex;
          vertical-align: top;
          margin-top: 24px;
          color: currentColor;
          transform: rotate(0deg);
          transition: transform var(--ddd-duration-fast, 150ms)
            var(--ddd-timing-ease, ease);
        }

        .characterbtn-menu-icon.rotated {
          transform: rotate(-90deg);
        }

        div[slot="center"] {
          display: flex;
          align-items: center;
          justify-content: center;
          height: var(--top-bar-height, 64px);
          width: 100%;
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          color: light-dark(
            var(--ddd-theme-default-coalyGray, #222),
            var(--ddd-theme-default-white, #fff)
          );
        }
                
        .template-results {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(200px, 0fr));
          width: 100%;
          min-height: 330px;
          box-sizing: border-box;
          gap: var(--ddd-spacing-2, 16px);
        }
        #returnToSection {
          width: 100%;
          padding-left: var(--ddd-spacing-5, 20px);
          padding-right: var(--ddd-spacing-5, 20px);
        }
        #returnToSection app-hax-search-results {
          width: 100%;
          min-height: 280px;
          box-sizing: border-box;
          height: 300px;
        }
        :host(:not([show-filter])) app-hax-search-results {
          width: 100%;
        }

        h2,
        .returnTo h2,
        .startNew h2 {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-l, 24px);
          color: var(--app-hax-accent-color, var(--accent-color));
          margin: 0 0 var(--ddd-spacing-4, 16px) 0;
        }
        .startNew,
        .returnTo {
          position: relative;
          display: flex;
          flex-direction: column;
          justify-content: flex-start;
          align-items: flex-start;
          margin: 0;
        }
        .upper-filter {
          padding-left: var(--ddd-spacing-5, 20px);
          padding-right: var(--ddd-spacing-5, 20px);
          margin-bottom: var(--ddd-spacing-4, 16px);
          position: relative;
          display: inline-block;
          width: 100%;
        }
        input[type="text"] {
          width: 100%;
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-2, 8px)
            var(--ddd-spacing-2, 8px) var(--ddd-spacing-8, 32px);
          font-size: var(--ddd-font-size-xs, 12px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid);
          border-color: var(--ddd-theme-default-slateGray, #666);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-coalyGray, #222);
          transition: all 0.2s ease;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          margin: 0;
          min-height: var(--ddd-spacing-8, 32px);
        }
        :host([dark]) input[type="text"],
        body.dark-mode input[type="text"] {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        input[type="text"]:focus {
          border: var(--ddd-border-md, 2px solid);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          background: var(--ddd-theme-default-white, white);
          outline: none;
        }
        :host([dark]) input[type="text"]:focus,
        body.dark-mode input[type="text"]:focus {
          background: var(--ddd-theme-default-coalyGray, #333);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
        }
        .search-icon {
          position: absolute;
          left: var(--ddd-spacing-2, 8px);
          top: 50%;
          transform: translateY(-50%);
          font-size: var(--ddd-font-size-xs, 14px);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          pointer-events: none;
          z-index: 1;
          --simple-icon-width: var(--ddd-icon-3xs, 20px);
          --simple-icon-height: var(--ddd-icon-3xs, 20px);
        }
        :host([dark]) .search-icon,
        body.dark-mode .search-icon {
          color: var(--ddd-theme-default-white, white);
        }
        .filter {
          position: relative;
          top: 0;
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-5, 20px);
          background: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-lg, 12px);
          box-shadow: var(--ddd-boxShadow-lg);
          border: var(--ddd-border-xs, 1px solid);
          border-color: var(--ddd-theme-default-slateGray, #666);
          padding: var(--ddd-spacing-6, 24px);
          margin-top: 0;
          margin-bottom: 0;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          transition: box-shadow 0.2s ease;
        }
        :host([dark]) .filter,
        body.dark-mode .filter {
          background: var(--ddd-theme-default-coalyGray, #222);
          border-color: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
        }
        .filter:hover {
          box-shadow: var(--ddd-boxShadow-xl);
        }
        .filter-btn {
          display: flex;
          align-items: center;
          justify-content: flex-start;
          gap: var(--ddd-spacing-1, 4px);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid) transparent;
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          cursor: pointer;
          box-shadow: var(--ddd-boxShadow-sm);
          transition: all 0.2s ease;
          outline: none;
          min-height: var(--ddd-spacing-7, 28px);
          text-align: left;
        }
        :host([dark]) .filter-btn,
        body.dark-mode .filter-btn {
          background: var(--ddd-theme-default-slateGray, #444);
          color: var(--ddd-theme-default-white, white);
        }
        .filter-btn.active,
        .filter-btn:active {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          box-shadow: var(--ddd-boxShadow-md);
        }
        :host([dark]) .filter-btn.active,
        :host([dark]) .filter-btn:active,
        body.dark-mode .filter-btn.active,
        body.dark-mode .filter-btn:active {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          border-color: var(--ddd-theme-default-white, white);
        }
        .filter-btn:hover,
        .filter-btn:focus {
          background: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
          transform: translateY(-1px);
        }
        .filter-btn:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        :host([dark]) .filter-btn:hover,
        :host([dark]) .filter-btn:focus,
        body.dark-mode .filter-btn:hover,
        body.dark-mode .filter-btn:focus {
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        .filter-btn.active:hover,
        .filter-btn.active:focus {
            background: var(--ddd-theme-default-keystoneYellow, #ffd100);
            color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        :host([dark]) .filter-btn.active:hover,
        :host([dark]) .filter-btn.active:focus,
        body.dark-mode .filter-btn.active:hover,
        body.dark-mode .filter-btn.active:focus {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        .filter-btn .icon {
          font-size: var(--ddd-font-size-3xs, 12px);
          color: inherit;
          display: flex;
          align-items: center;
          flex-shrink: 0;
          width: var(--ddd-icon-3xs, 20px);
          height: var(--ddd-icon-3xs, 20px);
        }
        .filter-btn .icon simple-icon-lite {
          color: inherit;
          --simple-icon-width: var(--ddd-icon-3xs, 20px);
          --simple-icon-height: var(--ddd-icon-3xs, 20px);
        }
        .filter-btn.active .icon,
        .filter-btn.active .icon simple-icon-lite {
          color: inherit;
        }
        :host([dark]) .filter-btn.active .icon,
        :host([dark]) .filter-btn.active .icon simple-icon-lite,
        body.dark-mode .filter-btn.active .icon,
        body.dark-mode .filter-btn.active .icon simple-icon-lite {
          color: inherit;
        }
        .filter-btn:hover .icon simple-icon-lite,
        .filter-btn:focus .icon simple-icon-lite {
          color: inherit;
        }
        input[type="checkbox"] {
          display: none;
        }
        .sort-control {
          margin-top: var(--ddd-spacing-2, 8px);
          margin-bottom: var(--ddd-spacing-3, 12px);
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-1, 4px);
        }
        .sort-label {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-3xs, 11px);
          color: var(--ddd-theme-default-coalyGray, #222);
        }
        :host([dark]) .sort-label,
        body.dark-mode .sort-label {
          color: var(--ddd-theme-default-limestoneGray, #f5f5f5);
        }
        .sort-select {
          width: 100%;
          padding: var(--ddd-spacing-1, 4px) var(--ddd-spacing-2, 8px);
          font-size: var(--ddd-font-size-3xs, 11px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-slateGray, #666);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-coalyGray, #222);
          font-family: var(--ddd-font-primary, sans-serif);
          box-sizing: border-box;
        }
        :host([dark]) .sort-select,
        body.dark-mode .sort-select {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        .sort-select:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        .reset-button {
          margin-top: var(--ddd-spacing-1, 4px);
          background: var(--ddd-theme-default-original87Pink, #e4007c);
          border: var(--ddd-border-xs, 1px solid) transparent;
          color: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-sm, 4px);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--ddd-spacing-1, 4px);
          box-shadow: var(--ddd-boxShadow-sm);
          cursor: pointer;
          transition: all 0.2s ease;
          min-height: var(--ddd-spacing-7, 28px);
        }
        .reset-button:hover,
        .reset-button:focus {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
          transform: translateY(-1px);
        }
        .reset-button:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        :host([dark]) .reset-button,
        body.dark-mode .reset-button {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
        }
        :host([dark]) .reset-button:hover,
        :host([dark]) .reset-button:focus,
        body.dark-mode .reset-button:hover,
        body.dark-mode .reset-button:focus {
          background: var(--ddd-theme-default-original87Pink, #e4007c);
        }
        .collapseFilter {
          display: none;
        }

        /* Visually hidden content for screen readers */
        .visually-hidden {
          position: absolute !important;
          width: 1px !important;
          height: 1px !important;
          padding: 0 !important;
          margin: -1px !important;
          overflow: hidden !important;
          clip: rect(0, 0, 0, 0) !important;
          white-space: nowrap !important;
          border: 0 !important;
        }

        /* Loading and fallback messages */
        .loading-message,
        .no-results {
          grid-column: 1 / -1;
          text-align: center;
          padding: var(--ddd-spacing-8, 32px);
          font-size: var(--ddd-font-size-s, 16px);
          color: var(--ddd-theme-default-slateGray, #666);
          font-family: var(--ddd-font-primary, sans-serif);
        }
        :host([dark]) .loading-message,
        :host([dark]) .no-results,
        body.dark-mode .loading-message,
        body.dark-mode .no-results {
          color: var(--ddd-theme-default-limestoneGray, #ccc);
        }
        .fallback-message {
          grid-column: 1 / -1;
          display: flex;
          flex-direction: column;
          align-items: center;
          gap: var(--ddd-spacing-4, 16px);
        }
        .fallback-message .no-results {
          padding: var(--ddd-spacing-4, 16px) 0;
        }
        .fallback-message app-hax-use-case {
          max-width: 300px;
        }

        /* 780px - 601px */
        @media (max-width: 780px) {
          :host {
            padding-left: var(--ddd-spacing-5, 20px);
            padding-right: var(--ddd-spacing-5, 20px);
          }
          :host([show-filter]) .filter {
            display: flex;
            width: 250px;
            max-width: 250px;
          }
          :host .collapseFilter {
            display: flex;
          }
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-m, 20px);
          }

        }

        /* 600px - 481px*/
        @media (max-width: 600px) {
          :host([show-filter]) .filter {
            width: 200px;
          }
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-s, 18px);
          }
        }

        /* 480px - 0px*/
        @media (max-width: 480px) {
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-s, 16px);
            margin-bottom: var(--ddd-spacing-3, 12px);
          }
          #returnToSection app-hax-search-results {
            min-width: 100%;
          }
        }
        
        .no-results {
          font-size: var(--ddd-font-size-s, 16px);
          color: light-dark(
            var(--ddd-theme-default-coalyGray, #222),
            var(--ddd-theme-default-white, white)
          );
        }
      `]}updated(e){super.updated&&super.updated(e),e.forEach((e,t)=>{["basePath"].includes(t)&&this[t]&&(d.AppHaxAPI[t]=this[t]),"responsiveSize"===t&&this._syncResponsiveStoreState(),["token"].includes(t)&&this[t]&&(d[t]=this[t])})}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),this._syncResponsiveStoreState(),d.AppHaxAPI&&this.basePath&&(d.AppHaxAPI.basePath=this.basePath),import("./lib/v2/app-hax-steps.js"),import("./lib/v2/app-hax-site-button.js"),import("../../wired-elements/lib/wired-button.js"),import("./lib/v2/app-hax-toast.js"),import("./lib/v2/app-hax-wired-toggle.js"),import("./lib/v2/app-hax-search-results.js"),import("./lib/v2/app-hax-user-menu.js"),import("./lib/v2/app-hax-user-menu-button.js"),this.dispatchEvent(new CustomEvent("app-hax-loaded",{composed:!0,bubbles:!0,cancelable:!1,detail:!0})),d.appReady=!0,o(()=>{i(d.appReady)?globalThis.document.body.classList.add("app-loaded"):globalThis.document.body.classList.remove("app-loaded")}),d.appEl=this,o(()=>{d.activeItem&&(this.activeItem=i(d.activeItem))}),o(()=>{this.soundIcon=i(d.soundStatus)?new URL("./lib/assets/images/FullVolume.svg",import.meta.url).href:new URL("./lib/assets/images/Silence.svg",import.meta.url).href,i(d.soundStatus)?d.toast("You can hear me now? Good.",2e3,{hat:"random"}):d.toast("Sound off.. hey! HELLO!?!",2e3,{fire:!0})}),o(async()=>{if(i(d.appReady)&&d.AppHaxAPI&&i(d.isLoggedIn)&&i(d.appSettings)&&i(d.refreshSiteList)){const e=await h.makeCall("getSitesList");d.manifest=e.data}else i(d.appReady)&&!i(d.isLoggedIn)&&this.login()}),this._syncHeaderPreSlot(),this._applyInitialAdminRoutePath()}toggleMenu(){this.userMenuOpen=!this.userMenuOpen,d.appEl.playSound("click")}closeMenu(){this.userMenuOpen=!1}_isHAXiamDeployment(e=null){return!(!e||"object"!=typeof e)&&(!0===e.iam||!!e.haxiamAddUserAccess)}_allowSystemSettings(e=null){const t=e&&"object"==typeof e?e:i(d.appSettings);return!(!t||"object"!=typeof t)&&!this._isHAXiamDeployment(t)}_syncSystemSettingsMerlinContext(e=null){this._allowSystemSettings(e)?p.appendContext("system-settings"):p.removeContext("system-settings")}_normalizeSystemSettingsPanel(e="dashboard"){if("string"!=typeof e)return"dashboard";const t=e.trim().toLowerCase();return Object.prototype.hasOwnProperty.call(u,t)?u[t]:Object.prototype.hasOwnProperty.call(c,t)?t:"dashboard"}_systemSettingsPanelMeta(e="dashboard"){const t=this._normalizeSystemSettingsPanel(e);return c[t]?c[t]:c.dashboard}_systemSettingsPanelFromRouteSegment(e=""){if("string"!=typeof e)return null;const t=e.trim().toLowerCase().replace(/^\/+/,"").replace(/\/+$/,"");return""===t?"dashboard":Object.prototype.hasOwnProperty.call(u,t)?u[t]:null}_systemSettingsRouteSegment(e="dashboard"){const t=this._systemSettingsPanelMeta(e);return t&&t.route?t.route:""}_systemSettingsRoutePath(e="dashboard"){const t=this._systemSettingsRouteSegment(e);return t?`admin/${t}`:"admin"}_systemSettingsPanelFromAdminRoutePath(e="admin"){const t=this._normalizeAdminRoutePath(e);return t?"admin"===t?"dashboard":0===t.indexOf("admin/")?this._systemSettingsPanelFromRouteSegment(t.replace("admin/","")):null:null}_systemSettingsModalRouteState(e="dashboard"){const t=this._normalizeSystemSettingsPanel(e),a=this._systemSettingsPanelMeta(t);return"dashboard"===t?{title:this.t.settings,titleIcon:"settings",breadcrumbs:[]}:{title:`${this.t.settings} > ${a.label}`,titleIcon:a.icon,breadcrumbs:[{label:this.t.settings,icon:"settings",action:"system-settings-dashboard",clickable:!0},{label:a.label,icon:a.icon,clickable:!1}]}}_systemSettingsModalStyles(){return{"--simple-modal-titlebar-background":"black","--simple-modal-titlebar-color":"var(--ddd-theme-default-white)","--simple-modal-background":"light-dark(var(--ddd-theme-default-white), var(--ddd-theme-default-coalyGray))","--simple-modal-z-index":"100000000","--simple-modal-titlebar-height":"80px","--simple-modal-width":"80vw","--simple-modal-max-width":"80vw","--simple-modal-height":"80vh","--simple-modal-max-height":"80vh","--simple-modal-border-radius":"var(--ddd-radius-md)"}}_currentSystemSettingsDialog(){if(this.__systemSettingsDialog&&this.__systemSettingsDialog.isConnected)return this.__systemSettingsDialog;if(globalThis.document&&globalThis.document.querySelector){const e=globalThis.document.querySelector("haxcms-system-settings");if(e)return this.__systemSettingsDialog=e,this.__systemSettingsDialog}return null}_setSystemSettingsDialogPanel(e="dashboard",t=null){const a=t||this._currentSystemSettingsDialog();if(!a)return;const i=this._normalizeSystemSettingsPanel(e);"function"==typeof a.setActivePanelFromRoute?a.setActivePanelFromRoute(i):a.activePanel=i}_showSystemSettingsModal(e=null,t="dashboard"){const a=e||this._currentSystemSettingsDialog();if(!a)return!1;const i=this._systemSettingsModalRouteState(t);return globalThis.dispatchEvent(new CustomEvent("simple-modal-show",{bubbles:!0,composed:!0,cancelable:!1,detail:{title:i.title,titleIcon:i.titleIcon,breadcrumbs:i.breadcrumbs,styles:this._systemSettingsModalStyles(),elements:{content:a},invokedBy:this._getSystemSettingsInvokedBy(),clone:!1,modal:!0,showClose:!0}})),!0}_normalizeAdminRoutePath(e=""){if("string"!=typeof e)return null;let t=e.trim().toLowerCase();if(!t)return null;if(t=t.replace(/^\/+/,"").replace(/\/+$/,""),"admin"===t)return"admin";if(0===t.indexOf("admin/")){const e=this._systemSettingsPanelFromRouteSegment(t.replace("admin/",""));return e?this._systemSettingsRoutePath(e):null}return null}_getAdminRoutePathFromLocation(){const e=new URLSearchParams(globalThis.location.search);return this._normalizeAdminRoutePath(e.get("admin"))}_setAdminRoutePathOnLocation(e=null,t="push"){const a=new URLSearchParams(globalThis.location.search),i=e?this._normalizeAdminRoutePath(e):null;i?a.set("admin",i):a.delete("admin");const o=a.toString(),s=`${globalThis.location.pathname}${o?`?${o}`:""}${globalThis.location.hash||""}`;"replace"===t?globalThis.history.replaceState(globalThis.history.state,"",s):globalThis.history.pushState(globalThis.history.state,"",s)}_clearAdminRoutePathFromLocation(e="replace"){this.__currentAdminRoutePath=null,this.__systemSettingsModalOpen=!1,this.__systemSettingsActivePanel="dashboard",this.__systemSettingsDialog=null,this._setAdminRoutePathOnLocation(null,e)}_canApplyAdminRoutePath(){return d.appReady&&i(d.isLoggedIn)&&this._allowSystemSettings()}_applyAdminRoutePath(e,t=0){const a=this._normalizeAdminRoutePath(e);if(!a)return;if(!this._canApplyAdminRoutePath())return void(t<120&&setTimeout(()=>{this._applyAdminRoutePath(a,t+1)},100));const i=this._systemSettingsPanelFromAdminRoutePath(a)||"dashboard";this.openSystemSettings(!1,{skipUrlUpdate:!0,silent:!0,panel:i})||(this._clearAdminRoutePathFromLocation("replace"),globalThis.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,cancelable:!0,detail:{}})))}_applyInitialAdminRoutePath(){const e=this._getAdminRoutePathFromLocation();e&&this._applyAdminRoutePath(e)}_adminRoutePopState(){const e=this._getAdminRoutePathFromLocation();e?this._applyAdminRoutePath(e):("string"==typeof this.__currentAdminRoutePath&&0===this.__currentAdminRoutePath.indexOf("admin")||this.__systemSettingsModalOpen)&&(this.__currentAdminRoutePath=null,this.__systemSettingsModalOpen=!1,this.__systemSettingsActivePanel="dashboard",globalThis.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,cancelable:!0,detail:{}})))}_simpleModalClosed(){("string"==typeof this.__currentAdminRoutePath&&0===this.__currentAdminRoutePath.indexOf("admin")||this.__systemSettingsModalOpen)&&setTimeout(()=>{let e=!1;if(globalThis.SimpleModal&&globalThis.SimpleModal.requestAvailability&&(e=globalThis.SimpleModal.requestAvailability().opened),e)return;const t=this._getAdminRoutePathFromLocation();this.__systemSettingsModalOpen=!1,this.__systemSettingsDialog=null,this.__systemSettingsActivePanel="dashboard",t&&0===t.indexOf("admin")?this._clearAdminRoutePathFromLocation("replace"):this.__currentAdminRoutePath=null},40)}_systemSettingsPanelChanged(e){if(!this.__systemSettingsModalOpen)return;const t=e&&e.detail?e.detail:{},a=this._normalizeSystemSettingsPanel("string"==typeof t.panel?t.panel:"dashboard"),i=t&&"replace"===t.historyMode?"replace":"push",o=this._systemSettingsRoutePath(a);this.__systemSettingsActivePanel=a,this.__currentAdminRoutePath=o,this._getAdminRoutePathFromLocation()!==o&&this._setAdminRoutePathOnLocation(o,i),this._showSystemSettingsModal(null,a)}_simpleModalBreadcrumbClick(e){if(!e||!e.detail||!e.detail.breadcrumb)return;if(!this.__systemSettingsModalOpen)return;"system-settings-dashboard"===e.detail.breadcrumb.action&&this.openSystemSettings(!1,{panel:"dashboard",historyMode:"push"})}_openSystemSettingsMenuTap(e){e&&(e.preventDefault(),e.stopPropagation()),this.openSystemSettings(!1)}openSystemSettingsFromMerlin(){this.openSystemSettings(!0)}_getSystemSettingsInvokedBy(){if(this.shadowRoot){const e=this.shadowRoot.querySelector("#systemsettingsbutton");if(e)return e;const t=this.shadowRoot.querySelector("#tbchar");if(t)return t}return null}openSystemSettings(e=!1,t={}){const a=t&&"object"==typeof t?t:{};if(!this._allowSystemSettings()){const t="System settings are not available in HAXiam deployments.";return a.silent||(e&&p&&"function"==typeof p.merlinSpeak?p.merlinSpeak(t):d.toast(t,3e3,{hat:"random"})),!1}const i=this._normalizeSystemSettingsPanel("string"==typeof a.panel?a.panel:"dashboard"),o=this._systemSettingsRoutePath(i);if(!a.skipUrlUpdate){this._getAdminRoutePathFromLocation()!==o&&this._setAdminRoutePathOnLocation(o,a.historyMode||"push")}return this.__currentAdminRoutePath=o,this.__systemSettingsActivePanel=i,this.__systemSettingsModalOpen?(this._setSystemSettingsDialogPanel(i),this._showSystemSettingsModal(null,i),!0):(this.__systemSettingsModalOpen=!0,this.closeMenu(),import("../haxcms-elements/lib/core/haxcms-system-settings.js").then(()=>{import("../simple-modal/simple-modal.js").then(()=>{const e=globalThis.document.createElement("haxcms-system-settings");this.__systemSettingsDialog=e,this._setSystemSettingsDialogPanel(i,e),this._showSystemSettingsModal(e,i)}).catch(()=>{this.__systemSettingsModalOpen=!1,this.__systemSettingsDialog=null,a.silent||d.toast("Unable to load system settings UI",3e3,{hat:"random"})})}).catch(()=>{this.__systemSettingsModalOpen=!1,this.__systemSettingsDialog=null,a.silent||d.toast("Unable to load system settings UI",3e3,{hat:"random"})}),!0)}_syncHeaderPreSlot(e){let t=!1;e&&e.target&&e.target.assignedElements?t=e.target.assignedElements({flatten:!0}).length>0:this.querySelector('[slot="app-header-pre"]')&&(t=!0),this.hasHeaderPre=t}_isDashboardModalOpen(){const e=globalThis.document&&globalThis.document.querySelector?globalThis.document.querySelector("simple-modal"):null;if(e&&e.opened)return!0;const t=this.shadowRoot&&this.shadowRoot.querySelector?this.shadowRoot.querySelector("app-hax-use-case-filter"):null;if(t&&t.shadowRoot&&t.shadowRoot.querySelector){const e=t.shadowRoot.querySelector("app-hax-site-creation-modal");if(e&&e.open)return!0}return!1}_handleSuperDaemonClose(e){requestAnimationFrame(()=>{if(!this._isDashboardModalOpen())try{globalThis.document&&globalThis.document.body&&globalThis.document.body.style&&"hidden"===globalThis.document.body.style.overflow&&(globalThis.document.body.style.overflow=""),globalThis.document&&globalThis.document.documentElement&&globalThis.document.documentElement.style&&"hidden"===globalThis.document.documentElement.style.overflow&&(globalThis.document.documentElement.style.overflow="")}catch(e){}})}openMerlin(){d.appEl.playSound("click"),p.open()}render(){return t`<app-hax-router></app-hax-router>
      <header>
        <app-hax-top-bar>
          <div slot="left" class="left-group">
            <slot
              name="app-header-pre"
              @slotchange="${this._syncHeaderPreSlot}"
            ></slot>
            ${this.hasHeaderPre?"":t`<a
                    id="home"
                    title="${this.t.home}"
                    href="home"
                  >
                    <simple-icon-lite
                      id="hlogo"
                      icon="hax:hax2022"
                      class="haxLogo"
                      title="${this.t.home}"
                    ></simple-icon-lite>
                  </a>
                  <simple-tooltip for="hlogo" position="right">
                    ${this.t.home}
                  </simple-tooltip>`}
          </div>

          <div slot="center">HAXcms Site Dashboard</div>
          
          <div slot="right" class="right-group">
            <app-hax-user-menu
              slot="right"
              id="user-menu"
              ?is-open="${this.userMenuOpen}"
            >
                <button
                  @click="${this.toggleMenu}"
                  @mouseover="${this.rpgStartWalk}"
                  @focusin="${this.rpgStartWalk}"
                  @mouseout="${this.rpgStopWalk}"
                  @focusout="${this.rpgStopWalk}"
                  class="topbar-character"
                  slot="menuButton"
                  id="tbchar"
                  title="System menu"
                  aria-label="User menu for ${this.userName}"
                  aria-expanded="${this.userMenuOpen}"
                  aria-haspopup="menu"
                >
                <rpg-character
                  seed="${this.userName}"
                  width="68"
                  height="68"
                  hat="${this.userMenuOpen?"edit":"none"}"
                  alt="Avatar for ${this.userName}"
                  ?walking="${this.rpgWalk}"
                  role="img"
                ></rpg-character>
                </button>
                <simple-tooltip
                  for="tbchar"
                  slot="menuButton"
                  position="bottom"
                >
                  ${this.userName}
                </simple-tooltip>
                <div slot="pre-menu" class="ops-panel">
                  <wired-button
                    elevation="1"
                    class="soundToggle"
                    @click="${soundToggle}"
                    aria-label="Toggle sound effects ${this.soundIcon&&-1!==this.soundIcon.indexOf("Full")?"off":"on"}"
                    aria-pressed="${this.soundIcon&&-1!==this.soundIcon.indexOf("Full")?"true":"false"}"
                  >
                    <simple-icon-lite
                      src="${this.soundIcon}"
                      loading="lazy"
                      decoding="async"
                      aria-hidden="true"
                    ></simple-icon-lite>
                  </wired-button>
                  <app-hax-wired-toggle id="wt"></app-hax-wired-toggle>
                </div>
                ${this._allowSystemSettings()?t`
                      <app-hax-user-menu-button
                        id="systemsettingsbutton"
                        slot="main-menu"
                        icon="settings"
                        label="${this.t.settings}"
                        @click="${this._openSystemSettingsMenuTap}"
                      ></app-hax-user-menu-button>
                    `:""}
                <app-hax-user-menu-button
                  slot="post-menu"
                  class="logout"
                  label="${this.t.logOut}"
                  @click=${this.logout}
                ></app-hax-user-menu-button>
            </app-hax-user-menu>
          </div>
        </app-hax-top-bar>
      </header>
      <main @click="${this.closeMenu}">
        <confetti-container id="confetti">
          <section class="content">
            <div class="start-journey">
              <app-hax-use-case-filter></app-hax-use-case-filter>
            </div>
          </section>
        </confetti-container>
      </main>`}rpgStartWalk(){globalThis.matchMedia("(prefers-reduced-motion: reduce)").matches||(this.rpgWalk=!0)}rpgStopWalk(){this.rpgWalk=!1}siteReadyToGo(e){import("../multiple-choice/lib/confetti-container.js").then(e=>{setTimeout(()=>{this.shadowRoot.querySelector("#confetti").setAttribute("popped","")},0)}),e.detail&&(d.siteReady=!0)}}customElements.define(AppHax.tag,AppHax),globalThis.AppHax=globalThis.AppHax||{},globalThis.AppHax.requestAvailability=()=>(!globalThis.AppHax.instance&&globalThis.document&&(globalThis.AppHax.instance=globalThis.document.querySelector(AppHax.tag)),globalThis.AppHax.instance);
