// @lrnwebcomponents/bootstrap-theme/bootstrap-theme.js
define(["exports", "meta", "require", "../../lit-element/lit-element.js", "../haxcms-elements/lib/core/HAXCMSLitElementTheme.js", "../haxcms-elements/lib/core/utils/HAXCMSMobileMenu.js", "./lib/BootstrapUserStylesMenuMixin.js", "../haxcms-elements/lib/core/utils/HAXCMSUserStylesMenu.js", "../haxcms-elements/lib/core/utils/HAXCMSThemeParts.js", "../simple-icon/lib/simple-icon-lite.js", "../simple-icon/lib/simple-icons.js", "./lib/BootstrapBreadcrumb.js", "./lib/BootstrapFooter.js", "./lib/BootstrapMenu.js", "./lib/BootstrapSearch.js", "../haxcms-elements/lib/core/haxcms-site-store.js", "../../mobx/dist/mobx.esm.js"], function (_exports, meta, _require, _litElement, _HAXCMSLitElementTheme, _HAXCMSMobileMenu, _BootstrapUserStylesMenuMixin, _HAXCMSUserStylesMenu, _HAXCMSThemeParts2, _simpleIconLite, _simpleIcons, _BootstrapBreadcrumb, _BootstrapFooter, _BootstrapMenu, _BootstrapSearch, _haxcmsSiteStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapTheme = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_20702a108e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <div class=\"site container-fluid\">\n        <div class=\"menu-outline\">\n          <div id=\"site-search-input\" role=\"search\">\n            <!-- <input\n              type=\"text\"\n              aria-label=\"Search site content\"\n              placeholder=\"Type to search\"\n              .value=\"", "\"\n              id=\"search\"\n              @input=\"", "\"\n            /> -->\n            <bootstrap-search\n              color-theme=\"", "\"\n            ></bootstrap-search>\n            <bootstrap-menu color-theme=\"", "\"> </bootstrap-menu>\n          </div>\n        </div>\n        <div id=\"body\" class=\"site-body\">\n          <div id=\"top\"></div>\n          <div class=\"site-inner\">\n            <header\n              class=\"site-header\"\n              role=\"navigation\"\n              .part=\"", "\"\n            >\n              <div class=\"btn-container\">\n                ", "\n                ", "\n                <div class=\"pull-right link-actions\">\n                  <site-print-button\n                    class=\"btn js-toolbar-action\"\n                  ></site-print-button>\n                  <site-rss-button\n                    type=\"rss\"\n                    class=\"btn js-toolbar-action\"\n                  ></site-rss-button>\n                  <site-git-corner size=\"small\"></site-git-corner>\n                </div>\n              </div>\n              <site-active-title></site-active-title>\n            </header>\n            <main class=\"page-wrapper\" role=\"main\">\n              <bootstrap-breadcrumb color-theme=\"", "\">\n              </bootstrap-breadcrumb>\n              <article class=\"main-content page-inner\">\n                <div class=\"normal main-section\">\n                  <site-search\n                    hide-input\n                    search=\"", "\"\n                    ?hidden=\"", "\"\n                  ></site-search>\n                  <section\n                    class=\"card p-4\"\n                    id=\"contentcontainer\"\n                    ?hidden=\"", "\"\n                  >\n                    <div id=\"slot\">\n                      <slot id=\"main-content\"></slot>\n                    </div>\n                  </section>\n                </div>\n              </article>\n            </main>\n            <footer>\n              <bootstrap-footer color-theme=\"", "\">\n              </bootstrap-footer>\n            </footer>\n          </div>\n        </div>\n      </div>\n    "]);

    _templateObject2_20702a108e7d11eba1fe15f944a3ea4d = function _templateObject2_20702a108e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_20702a108e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          background-color: rgb(242, 244, 244);\n          width: 100%;\n          display: flex;\n          padding: 0;\n          margin: 0;\n          min-height: 100vh;\n          flex-direction: column;\n          -webkit-box-orient: vertical;\n          -webkit-box-direction: normal;\n          font-family: system-ui, -apple-system, BlinkMacSystemFont, \"Segoe UI\",\n            Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", \"Liberation Sans\",\n            sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\",\n            \"Noto Color Emoji\";\n          --map-menu-item-a-color: var(--bootstrap-menu-item-a-color, #0d6efd);\n          --bootstrap-light-theme-color: #000000;\n          --bootstrap-light-theme-background-color: #ffffff;\n          --bootstrap-dark-theme-background-color: #000000;\n          --bootstrap-dark-theme-secondary-background-color: #343a40;\n          --bootstrap-dark-theme-color: #ffffff;\n          --bootstrap-link-color: #0d6efd;\n        }\n        :host([hidden]) {\n          display: none;\n        }\n        [hidden] {\n          display: none;\n        }\n        :host([menu-open]) .menu-outline {\n          left: 0;\n        }\n        :host([is-logged-in][menu-open]) .menu-outline {\n          left: 48px;\n        }\n\n        :host([responsive-size=\"xs\"]) .main-content,\n        :host([responsive-size=\"sm\"]) .main-content {\n          overflow-x: hidden;\n        }\n        :host([responsive-size=\"xs\"]) .page-inner {\n          overflow-x: auto;\n        }\n\n        /* main content */\n        .menu-outline {\n          /* font-family: \"Helvetica Neue\", Helvetica, Arial, sans-serif; */\n          position: absolute;\n          top: 0;\n          left: -300px;\n          bottom: 0;\n          z-index: 1;\n          overflow-y: auto;\n          width: 300px;\n          color: #364149;\n          background: #fafafa;\n          border-right: 1px solid rgba(0, 0, 0, 0.07);\n          transition: left 250ms ease;\n        }\n        .site {\n          background-color: rgb(242, 244, 244);\n        }\n        .site-body {\n          position: absolute;\n          top: 0;\n          right: 0;\n          left: 0;\n          bottom: 0;\n          min-width: 400px;\n          overflow-y: auto;\n          transition: left 250ms ease;\n        }\n        .site-body .site-inner {\n          position: relative;\n          top: 0;\n          right: 0;\n          left: 0;\n          bottom: 0;\n          overflow-y: auto;\n        }\n        :host([menu-open]) .site-body {\n          left: 300px;\n        }\n        :host([is-logged-in][menu-open]) .site-body {\n          left: 348px;\n        }\n        :host([is-logged-in]) .site-body {\n          left: 48px;\n        }\n        :host([responsive-size=\"xs\"]) .site-body,\n        :host([responsive-size=\"sm\"]) .site-body {\n          overflow-x: hidden;\n          position: fixed;\n        }\n        :host([responsive-size=\"xs\"]) .main-content,\n        :host([responsive-size=\"sm\"]) .main-content {\n          overflow-x: hidden;\n        }\n        :host([responsive-size=\"xs\"]) .site-inner {\n          max-width: 100vw;\n        }\n        :host([responsive-size=\"xs\"]) .page-inner {\n          overflow-x: auto;\n        }\n        .page-wrapper {\n          position: relative;\n          outline: 0;\n        }\n        .main-content > :first-child {\n          margin-top: 0 !important;\n        }\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content h6 {\n          margin-top: 1.275em;\n          margin-bottom: 0.85em;\n          font-weight: 700;\n        }\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5 {\n          page-break-after: avoid;\n        }\n        .main-content *,\n        .main-content ::slotted(*) {\n          box-sizing: border-box;\n          -webkit-box-sizing: border-box;\n          font-size: inherit;\n        }\n\n        .main-content h2 {\n          font-size: 1.75em;\n          font-family: Roboto;\n        }\n\n        .main-content h3 {\n          font-size: 1.5em;\n        }\n        .main-content h4 {\n          font-size: 1.25em;\n        }\n        .main-content h5 {\n          font-size: 1em;\n        }\n        .main-content h6 {\n          font-size: 1em;\n          color: #777;\n        }\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content h6 {\n          margin-top: 1.275em;\n          margin-bottom: 0.85em;\n          font-weight: 700;\n        }\n\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5 {\n          page-break-after: avoid;\n        }\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content p {\n          orphans: 3;\n          widows: 3;\n        }\n        .main-content blockquote,\n        .main-content dl,\n        .main-content ol,\n        .main-content p,\n        .main-content table,\n        .main-content ul {\n          margin-top: 0;\n          margin-bottom: 0.85em;\n        }\n        .main-content ol,\n        .main-content ul {\n          padding: 0;\n          margin: 0;\n          margin-bottom: 0.85em;\n          padding-left: 2em;\n        }\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content p {\n          orphans: 3;\n          widows: 3;\n        }\n        .main-content * {\n          box-sizing: border-box;\n          -webkit-box-sizing: border-box;\n          font-size: inherit;\n        }\n        .page-inner {\n          position: relative;\n          max-width: 1000px;\n          margin: 0 auto;\n          min-height: 90vh;\n          padding: 20px 15px 40px 15px;\n        }\n        .main-section {\n          display: block;\n          word-wrap: break-word;\n          color: var(--haxcms-user-styles-color-theme-color-color);\n          line-height: 1.7;\n          text-size-adjust: 100%;\n          -ms-text-size-adjust: 100%;\n          -webkit-text-size-adjust: 100%;\n          -moz-text-size-adjust: 100%;\n        }\n        article,\n        aside,\n        details,\n        figcaption,\n        figure,\n        footer,\n        header,\n        hgroup,\n        main,\n        nav,\n        section,\n        summary {\n          display: block;\n        }\n\n        /* site menu navigation buttons */\n        site-menu-button:not(:defined) {\n          display: none;\n        }\n        site-menu-button {\n          --site-menu-button-icon-fill-color: var(\n            --haxcms-user-styles-color-theme-color-1\n          );\n          --haxcms-tooltip-color: var(--haxcms-user-styles-color-theme-color-2);\n          --haxcms-tooltip-background-color: var(\n            --haxcms-user-styles-color-theme-color-1\n          );\n        }\n        site-menu-button {\n          --site-menu-button-icon-width: 64px;\n          --site-menu-button-icon-height: 64px;\n        }\n        :host([is-logged-in][menu-open]) site-menu-button[type=\"prev\"] {\n          left: 348px;\n        }\n        :host([is-logged-in]) site-menu-button[type=\"prev\"] {\n          left: 48px;\n        }\n        :host([menu-open]) site-menu-button[type=\"prev\"] {\n          left: 300px;\n        }\n        site-menu-button[type=\"prev\"] {\n          left: 0;\n        }\n        site-menu-button[type=\"next\"] {\n          right: 0;\n        }\n        simple-icon-button,\n        simple-icon-button-lite,\n        site-rss-button,\n        site-print-button,\n        site-git-corner {\n          color: var(--site-print-button-color, black);\n          --site-git-corner-background: var(--bootstrap-light-theme-color);\n          --site-git-corner-color: var(\n            --bootstrap-light-theme-background-color\n          );\n          --simple-icon-fill-color: var(\n            --haxcms-user-styles-color-theme-color-1\n          );\n          --haxcms-tooltip-color: var(--haxcms-user-styles-color-theme-color-2);\n          --haxcms-tooltip-background-color: var(\n            --haxcms-user-styles-color-theme-color-1\n          );\n        }\n        /* header */\n        .pull-right {\n          float: right;\n        }\n        .btn-container {\n          z-index: 2;\n          height: 50px;\n          padding: 6px;\n        }\n        .navigation {\n          position: fixed;\n          top: 50px;\n          bottom: 0px;\n          margin: 0px 20px;\n          max-width: 150px;\n          min-width: 90px;\n          display: flex;\n          place-content: center;\n          flex-direction: column;\n          font-size: 40px;\n          color: rgb(204, 204, 204);\n          text-align: center;\n          transition: all 0.35s ease 0s;\n        }\n        .site-header {\n          overflow: visible;\n          height: 50px;\n          padding: 0 8px;\n          z-index: 2;\n          font-size: 0.85em;\n          color: #7e888b;\n          background: 0 0;\n        }\n        .site-header site-active-title h1 {\n          color: inherit;\n          text-decoration: none;\n          position: absolute;\n          z-index: 1;\n          top: 0;\n          left: 0;\n          right: 0;\n          margin: 10px 200px;\n          font-size: 20px;\n          font-weight: 200;\n          text-align: center;\n          line-height: 50px;\n          opacity: 0;\n          transition: opacity 0.2s ease;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          white-space: nowrap;\n        }\n        .site-header site-active-title h1:hover {\n          opacity: 1;\n        }\n        /* site search stuff */\n        #site-search-input {\n          padding: 6px;\n          background: 0 0;\n          transition: top 0.5s ease;\n          background: #fff;\n          border-bottom: 1px solid rgba(0, 0, 0, 0.07);\n          border-top: 1px solid rgba(0, 0, 0, 0.07);\n          margin-bottom: 10px;\n          margin-top: -1px;\n        }\n        #site-search-input input,\n        #site-search-input input:focus,\n        #site-search-input input:hover {\n          width: 100%;\n          background: 0 0;\n          border: 1px solid transparent;\n          box-shadow: none;\n          outline: 0;\n          line-height: 22px;\n          padding: 7px 7px;\n          color: inherit;\n        }\n        site-search {\n          height: auto;\n          width: auto;\n          font-size: inherit;\n        }\n        /* Light Theme */\n        :host([color-theme=\"0\"]) site-search {\n          color: #252737;\n          --site-search-result-background-color: transparent;\n          --site-search-result-background-color-hover: #f5f5f5;\n          --site-search-link-color-hover: #252737;\n          --site-search-link-text-color: #252737;\n          --site-search-link-color: #252737;\n          --site-search-result-color: #252737;\n        }\n        /* Dark Theme */\n        :host([color-theme=\"1\"]) {\n          background-color: black;\n          --simple-fields-background-color: transparent;\n          --github-corner-background: white;\n          --map-menu-item-a-color: var(--bootstrap-dark-theme-color);\n          --haxcms-user-styles-color-theme-color-color: var(\n            --bootstrap-dark-theme-color\n          );\n          --hax-base-styles-a-color: var(--bootstrap-link-color);\n          --haxcms-tooltip-background-color: var(--bootstrap-dark-theme-color);\n          --haxcms-tooltip-color: var(--bootstrap-dark-theme-background-color);\n          --site-menu-button-icon-fill-color: white;\n        }\n\n        :host([color-theme=\"1\"]) simple-fields-field {\n        }\n\n        :host([color-theme=\"1\"]) site-menu-button {\n          --haxcms-tooltip-color: var(---bootstrap-dark-theme-color);\n          --haxcms-tooltip-background-color: var(--bootstrap-dark-theme-color);\n        }\n\n        :host([color-theme=\"1\"]) .main-section {\n          color: var(--bootstrap-dark-theme-color);\n        }\n\n        :host([color-theme=\"1\"]) #site-search-input {\n          background-color: var(\n            --bootstrap-dark-theme-secondary-background-color\n          );\n        }\n\n        :host([color-theme=\"1\"]) site-git-corner {\n          --site-git-corner-background: var(--bootstrap-dark-theme-color);\n          --site-git-corner-color: var(--bootstrap-dark-theme-background-color);\n        }\n\n        :host([color-theme=\"1\"]) .card {\n          background-color: var(\n            --bootstrap-dark-theme-secondary-background-color\n          );\n        }\n        :host([color-theme=\"1\"]) .card h1 h2 h3 h4 h5 h6 p {\n          color: #fff;\n        }\n        :host([color-theme=\"1\"]) .site-header .btn {\n          color: white;\n        }\n        :host([color-theme=\"1\"]) .site-header .btn:hover,\n        :host([color-theme=\"1\"]) .site-header .btn:focus,\n        :host([color-theme=\"1\"]) .site-header .btn:active {\n          color: #fffff5;\n          background: none;\n        }\n\n        :host([color-theme=\"1\"]) .site-header site-active-title {\n          color: #fff;\n        }\n\n        :host([color-theme=\"1\"]) site-search {\n          color: var(--bootstrap-dark-theme-color);\n          --site-search-result-background-color: transparent;\n          --site-search-result-background-color-hover: transparent;\n          --site-search-link-color-hover: var(--bootstrap-dark-theme-color);\n          --site-search-link-text-color: var(--bootstrap-dark-theme-color);\n          --site-search-link-color: var(--bootstrap-dark-theme-color);\n          --site-search-result-color: var(--bootstrap-dark-theme-color);\n        }\n        :host([color-theme=\"2\"]) site-search {\n          color: var(--simple-colors-default-theme-light-blue-1, #cfd4e3);\n          --site-search-result-background-color: transparent;\n          --site-search-result-background-color-hover: transparent;\n          --site-search-link-color-hover: var(\n            --simple-colors-default-theme-light-blue-1,\n            #cfd4e3\n          );\n          --site-search-link-text-color: var(\n            --simple-colors-default-theme-light-blue-1,\n            #cfd4e3\n          );\n          --site-search-link-color: var(\n            --simple-colors-default-theme-light-blue-1,\n            #cfd4e3\n          );\n          --site-search-result-color: var(\n            --simple-colors-default-theme-light-blue-1,\n            #cfd4e3\n          );\n        }\n      "]);

    _templateObject_20702a108e7d11eba1fe15f944a3ea4d = function _templateObject_20702a108e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `bootstrap-theme`
   * `Hax bootstrap themeing`
   * @demo demo/index.html
   * @element bootstrap-theme
   */
  var BootstrapTheme = /*#__PURE__*/function (_HAXCMSThemeParts) {
    babelHelpers.inherits(BootstrapTheme, _HAXCMSThemeParts);

    var _super = _createSuper(BootstrapTheme);

    babelHelpers.createClass(BootstrapTheme, null, [{
      key: "tag",

      /**
       * Convention we use
       */
      get: function get() {
        return "bootstrap-theme";
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme), "styles", this)), [(0, _litElement.css)(_templateObject_20702a108e7d11eba1fe15f944a3ea4d())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme), "properties", this)), {}, {
          searchTerm: {
            type: String
          },
          menuOpen: {
            type: Boolean,
            attribute: "menu-open",
            reflect: true
          },
          colorTheme: {
            type: String,
            attribute: "color-theme",
            reflect: true
          }
        });
      }
      /**
       * HTMLElement
       */

    }]);

    function BootstrapTheme() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapTheme);
      _this = _super.call(this);
      _this.HAXCMSThemeSettings.autoScroll = true;
      _this.menuOpen = true;
      _this.headerTags = ["H2", "H3", "H4", "H5", "H6"];

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this._themeElements = [];
      _this.colorTheme = "0";
      _this.searchTerm = ""; // event listener for search changed

      _this.addEventListener("searchChanged", function (evt) {
        if (evt.detail.searchText) {
          new Promise(function (res, rej) {
            return _require.default(["../haxcms-elements/lib/ui-components/site/site-search.js"], res, rej);
          }).then(function () {
            _this.searchTerm = evt.detail.searchText;
          });
        }
      }); // prettier-ignore


      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/active-item/site-active-title.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/navigation/site-menu-button.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/site/site-print-button.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/site/site-rss-button.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/active-item/site-git-corner.js"], res, rej);
      });
      _this.__disposer = _this.__disposer ? _this.__disposer : [];
      (0, _mobxEsm.autorun)(function (reaction) {
        _this.activeManifestIndex = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeManifestIndex);
        _this.searchTerm = "";

        _this.__disposer.push(reaction);
      });
      return _this;
    }

    babelHelpers.createClass(BootstrapTheme, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_20702a108e7d11eba1fe15f944a3ea4d(), this._bootstrapPath, this.searchTerm, this.searchChanged, this.colorTheme, this.colorTheme, this.editMode ? "edit-mode-active" : "", this.HAXCMSMobileMenuButton(), this.BootstrapUserStylesMenu(), this.colorTheme, this.searchTerm, this.searchTerm != "" ? false : true, this.searchTerm != "" ? true : false, this.colorTheme);
      }
    }, {
      key: "_wrapSections",
      value: function _wrapSections() {
        if (this._themeElements) {
          this._applyClasses(this._themeElements);
        }
      }
    }, {
      key: "_applyClasses",
      value: function _applyClasses(nodes) {
        var _this2 = this;

        nodes.forEach(function (node) {
          if (_this2.headerTags.includes(node.nodeName)) {
            node.classList.add("display-6");
          }
        });
      }
    }, {
      key: "_applyMutationObserver",
      value: function _applyMutationObserver() {
        var _this3 = this;

        this._observer = new MutationObserver(function (mutationList, observer) {
          mutationList.forEach(function (mutationEntry) {
            if (mutationEntry.addedNodes) {
              mutationEntry.addedNodes.forEach(function (item) {
                _this3._themeElements.push(item);
              });

              _this3._wrapSections();
            } else if (mutationEntry.removedNodes) {
              mutationEntry.removedNodes.forEach(function (item) {
                _this3._themeElements.splice(_this3._themeElements.indexOf(item), 1);
              });
            }
          });
        });

        this._observer.observe(this, {
          childList: true
        });
      }
    }, {
      key: "_generateBootstrapLink",
      value: function _generateBootstrapLink() {
        if (this._bootstrapLink) {
          document.head.removeChild(this._bootstrapLink);
        }

        var basePath = this.getBasePath(decodeURIComponent(meta.url));
        console.log(basePath);
        var link = document.createElement("link");
        link.setAttribute("rel", "stylesheet");
        link.setAttribute("href", basePath + "bootstrap/dist/css/bootstrap.min.css");
        document.head.appendChild(link);
        return link;
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        if (this._bootstrapLink) {
          document.head.removeChild(this._bootstrapLink);
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * LitElement ready
       */

    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "firstUpdated", this).call(this, changedProperties);
        }

        this._loadScripts();

        this._bootstrapLink = this._generateBootstrapLink();

        this._applyMutationObserver();

        this._wrapSections();
      }
      /*
       * Loads jquery first because bootstrap requires jquery to be present first
       * Jquery callback function then loads bootstrap
       */

    }, {
      key: "_loadScripts",
      value: function _loadScripts() {
        var basePath = this.getBasePath(decodeURIComponent(meta.url));
        var jqueryPath = "jquery/dist/jquery.min.js";
        window.ESGlobalBridge.requestAvailability();
        window.ESGlobalBridge.instance.load("jquery", basePath + jqueryPath);
        window.addEventListener("es-bridge-jquery-loaded", this._jqueryLoaded.bind(this));
      }
    }, {
      key: "_bootstrapLoaded",
      value: function _bootstrapLoaded(e) {
        this._bootstrap = true;
      }
    }, {
      key: "_loadBootstrap",
      value: function _loadBootstrap() {
        var basePath = this.getBasePath(decodeURIComponent(meta.url));
        var bootstrapPath = "bootstrap/dist/js/bootstrap.bundle.min.js";
        window.ESGlobalBridge.instance.load("bootstrap", basePath + bootstrapPath);
        window.addEventListener("es-bridge-bootstrap-loaded", this._bootstrapLoaded.bind(this));
      }
    }, {
      key: "_jqueryLoaded",
      value: function _jqueryLoaded(e) {
        this._jquery = true;

        this._loadBootstrap();
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }, {
      key: "prevPage",
      value: function prevPage(e) {
        babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "prevPage", this).call(this, e);
      }
    }, {
      key: "nextPage",
      value: function nextPage(e) {
        babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "nextPage", this).call(this, e);
      }
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {});
      }
    }]);
    return BootstrapTheme;
  }((0, _HAXCMSThemeParts2.HAXCMSThemeParts)((0, _BootstrapUserStylesMenuMixin.BootstrapUserStylesMenuMixin)((0, _HAXCMSMobileMenu.HAXCMSMobileMenuMixin)(_HAXCMSLitElementTheme.HAXCMSLitElementTheme))));

  _exports.BootstrapTheme = BootstrapTheme;
  customElements.define(BootstrapTheme.tag, BootstrapTheme);
});
