// @lrnwebcomponents/bootstrap-theme/bootstrap-theme.js
define(["exports", "meta", "require", "../../lit-element/lit-element.js", "../haxcms-elements/lib/core/HAXCMSLitElementTheme.js", "../haxcms-elements/lib/core/utils/HAXCMSMobileMenu.js", "../haxcms-elements/lib/core/utils/HAXCMSUserStylesMenu.js", "../haxcms-elements/lib/core/utils/HAXCMSThemeParts.js", "../haxcms-elements/lib/core/haxcms-site-store.js", "../../mobx/dist/mobx.esm.js", "../es-global-bridge/es-global-bridge.js"], function (_exports, meta, _require, _litElement, _HAXCMSLitElementTheme, _HAXCMSMobileMenu, _HAXCMSUserStylesMenu, _HAXCMSThemeParts2, _haxcmsSiteStore, _mobxEsm, _esGlobalBridge) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapTheme = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);
  _require = babelHelpers.interopRequireWildcard(_require);

  function _templateObject2_b38140308b3211ebb67059cf696085af() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <div class=\"site container-fluid\">\n        <div class=\"menu-outline\">\n          <div id=\"site-search-input\" role=\"search\">\n            <input\n              type=\"text\"\n              aria-label=\"Search site content\"\n              placeholder=\"Type to search\"\n              .value=\"", "\"\n              id=\"search\"\n              @input=\"", "\"\n            />\n            ", "\n          </div>\n        </div>\n        <div id=\"body\" class=\"site-body\">\n          <div id=\"top\"></div>\n          <div class=\"site-inner\">\n            <header\n              class=\"site-header\"\n              role=\"navigation\"\n              .part=\"", "\"\n            >\n              <div class=\"btn-container\">\n                ", " ", "\n                <div class=\"pull-right link-actions\">\n                  <site-print-button\n                    class=\"btn js-toolbar-action\"\n                  ></site-print-button>\n                  <site-rss-button\n                    type=\"rss\"\n                    class=\"btn js-toolbar-action\"\n                  ></site-rss-button>\n                  <site-git-corner size=\"small\"></site-git-corner>\n                </div>\n              </div>\n              <site-active-title></site-active-title>\n            </header>\n            <main class=\"page-wrapper\" role=\"main\">\n              <article class=\"main-content page-inner\">\n                <div class=\"normal main-section\">\n                  <site-search\n                    hide-input\n                    search=\"", "\"\n                    ?hidden=\"", "\"\n                  ></site-search>\n                  <section\n                    class=\"card bg-white p-4\"\n                    id=\"contentcontainer\"\n                    ?hidden=\"", "\"\n                  >\n                    <div id=\"slot\">\n                      <slot id=\"main-content\"></slot>\n                    </div>\n                  </section>\n                </div>\n              </article>\n            </main>\n          </div>\n        </div>\n      </div>\n    "]);

    _templateObject2_b38140308b3211ebb67059cf696085af = function _templateObject2_b38140308b3211ebb67059cf696085af() {
      return data;
    };

    return data;
  }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject_b38140308b3211ebb67059cf696085af() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          --haxcms-color: white !important;\n          --hax-base-styles-p-font-size: 12px;\n        }\n        :host([menu-open]) .menu-outline {\n          left: 0;\n        }\n        :host([is-logged-in][menu-open]) .menu-outline {\n          left: 48px;\n        }\n\n        :host([responsive-size=\"xs\"]) .main-content,\n        :host([responsive-size=\"sm\"]) .main-content {\n          overflow-x: hidden;\n        }\n        :host([responsive-size=\"xs\"]) .page-inner {\n          overflow-x: auto;\n        }\n\n        .menu-outline {\n          font-family: \"Helvetica Neue\", Helvetica, Arial, sans-serif;\n          position: absolute;\n          top: 0;\n          left: -300px;\n          bottom: 0;\n          z-index: 1;\n          overflow-y: auto;\n          width: 300px;\n          color: #364149;\n          background: #fafafa;\n          border-right: 1px solid rgba(0, 0, 0, 0.07);\n          transition: left 250ms ease;\n        }\n        .page-wrapper {\n          position: relative;\n          outline: 0;\n        }\n        .main-content > :first-child {\n          margin-top: 0 !important;\n        }\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content h6 {\n          margin-top: 1.275em;\n          margin-bottom: 0.85em;\n          font-weight: 700;\n        }\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5 {\n          page-break-after: avoid;\n        }\n        .main-content *,\n        .main-content ::slotted(*) {\n          box-sizing: border-box;\n          -webkit-box-sizing: border-box;\n          font-size: inherit;\n        }\n\n        main-content h2 {\n          font-size: 1.75em;\n        }\n        .main-content h3 {\n          font-size: 1.5em;\n        }\n        .main-content h4 {\n          font-size: 1.25em;\n        }\n        .main-content h5 {\n          font-size: 1em;\n        }\n        .main-content h6 {\n          font-size: 1em;\n          color: #777;\n        }\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content h6 {\n          margin-top: 1.275em;\n          margin-bottom: 0.85em;\n          font-weight: 700;\n        }\n\n        .main-content h1,\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5 {\n          page-break-after: avoid;\n        }\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content p {\n          orphans: 3;\n          widows: 3;\n        }\n        .main-content blockquote,\n        .main-content dl,\n        .main-content ol,\n        .main-content p,\n        .main-content table,\n        .main-content ul {\n          margin-top: 0;\n          margin-bottom: 0.85em;\n        }\n        .main-content ol,\n        .main-content ul {\n          padding: 0;\n          margin: 0;\n          margin-bottom: 0.85em;\n          padding-left: 2em;\n        }\n        .main-content h2,\n        .main-content h3,\n        .main-content h4,\n        .main-content h5,\n        .main-content p {\n          orphans: 3;\n          widows: 3;\n        }\n        .main-content * {\n          box-sizing: border-box;\n          -webkit-box-sizing: border-box;\n          font-size: inherit;\n        }\n        .page-inner {\n          position: relative;\n          max-width: 800px;\n          margin: 0 auto;\n          min-height: 90vh;\n          padding: 20px 15px 40px 15px;\n        }\n        .main-section {\n          display: block;\n          word-wrap: break-word;\n          color: var(--haxcms-user-styles-color-theme-color-color);\n          line-height: 1.7;\n          text-size-adjust: 100%;\n          -ms-text-size-adjust: 100%;\n          -webkit-text-size-adjust: 100%;\n          -moz-text-size-adjust: 100%;\n        }\n      "]);

    _templateObject_b38140308b3211ebb67059cf696085af = function _templateObject_b38140308b3211ebb67059cf696085af() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `bootstrap-theme`
   * `Hax bootstrap themeing`
   * @demo demo/index.html
   * @element bootstrap-theme
   */
  var BootstrapTheme = /*#__PURE__*/function (_HAXCMSThemeParts) {
    babelHelpers.inherits(BootstrapTheme, _HAXCMSThemeParts);

    var _super = _createSuper(BootstrapTheme);

    babelHelpers.createClass(BootstrapTheme, null, [{
      key: "tag",

      /**
       * Convention we use
       */
      get: function get() {
        return "bootstrap-theme";
      }
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme), "styles", this)), [(0, _litElement.css)(_templateObject_b38140308b3211ebb67059cf696085af())]);
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme), "properties", this)), {}, {
          searchTerm: {
            type: String
          },
          menuOpen: {
            type: Boolean,
            attribute: "menu-open",
            reflect: true
          }
        });
      }
      /**
       * HTMLElement
       */

    }]);

    function BootstrapTheme() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapTheme);
      _this = _super.call(this);
      _this.HAXCMSThemeSettings.autoScroll = true;
      _this.menuOpen = true;
      _this.headerTags = ["H2", "H3", "H4", "H5", "H6"];

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css"; // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/active-item/site-active-title.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/navigation/site-menu-button.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/site/site-print-button.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/site/site-rss-button.js"], res, rej);
      }); // prettier-ignore

      new Promise(function (res, rej) {
        return _require.default(["../haxcms-elements/lib/ui-components/active-item/site-git-corner.js"], res, rej);
      });
      _this.__disposer = _this.__disposer ? _this.__disposer : [];
      (0, _mobxEsm.autorun)(function (reaction) {
        _this.activeManifestIndex = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeManifestIndex);
        _this.searchTerm = "";

        _this.__disposer.push(reaction);
      });
      return _this;
    }

    babelHelpers.createClass(BootstrapTheme, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_b38140308b3211ebb67059cf696085af(), this._bootstrapPath, this.searchTerm, this.searchChanged, this.HAXCMSMobileMenu(), this.editMode ? "edit-mode-active" : "", this.HAXCMSMobileMenuButton(), this.HAXCMSUserStylesMenu(), this.searchTerm, this.searchTerm != "" ? false : true, this.searchTerm != "" ? true : false);
      }
    }, {
      key: "_wrapSections",
      value: function _wrapSections() {}
    }, {
      key: "_applyClasses",
      value: function _applyClasses(nodes) {
        var _this2 = this;

        nodes.forEach(function (node) {
          if (_this2.headerTags.includes(node.nodeName)) {
            node.classList.add("display-6");
          }
        });
      }
    }, {
      key: "_applyMutationObserver",
      value: function _applyMutationObserver() {
        var _this3 = this;

        this._observer = new MutationObserver(function (mutationList, observer) {
          mutationList.forEach(function (mutationEntry) {
            _this3._applyClasses(mutationEntry.addedNodes);
          });
        });

        this._observer.observe(this, {
          childList: true
        });
      }
    }, {
      key: "_generateBootstrapLink",
      value: function _generateBootstrapLink() {
        if (this._bootstrapLink) {
          document.head.removeChild(this._bootstrapLink);
        }

        var basePath = this.getBasePath(decodeURIComponent(meta.url));
        console.log(basePath);
        var link = document.createElement("link");
        link.setAttribute("rel", "stylesheet");
        link.setAttribute("href", basePath + "bootstrap/dist/css/bootstrap.min.css");
        document.head.appendChild(link);
        return link;
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        if (this._bootstrapLink) {
          document.head.removeChild(this._bootstrapLink);
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * LitElement ready
       */

    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "firstUpdated", this).call(this, changedProperties);
        }

        this._loadScripts();

        this._bootstrapLink = this._generateBootstrapLink();

        this._applyMutationObserver();
      }
      /*
       * Loads jquery first because bootstrap requires jquery to be present first
       * Jquery callback function then loads bootstrap
       */

    }, {
      key: "_loadScripts",
      value: function _loadScripts() {
        var basePath = this.getBasePath(decodeURIComponent(meta.url));
        var jqueryPath = "jquery/dist/jquery.min.js";
        window.ESGlobalBridge.requestAvailability();
        window.ESGlobalBridge.instance.load("jquery", basePath + jqueryPath);
        window.addEventListener("es-bridge-jquery-loaded", this._jqueryLoaded.bind(this));
      }
    }, {
      key: "_bootstrapLoaded",
      value: function _bootstrapLoaded(e) {
        this._bootstrap = true;
      }
    }, {
      key: "_loadBootstrap",
      value: function _loadBootstrap() {
        var basePath = this.getBasePath(decodeURIComponent(meta.url));
        var bootstrapPath = "bootstrap/dist/js/bootstrap.bundle.min.js";
        window.ESGlobalBridge.instance.load("bootstrap", basePath + bootstrapPath);
        window.addEventListener("es-bridge-bootstrap-loaded", this._bootstrapLoaded.bind(this));
      }
    }, {
      key: "_jqueryLoaded",
      value: function _jqueryLoaded(e) {
        this._jquery = true;

        this._loadBootstrap();
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }, {
      key: "prevPage",
      value: function prevPage(e) {
        babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "prevPage", this).call(this, e);
      }
    }, {
      key: "nextPage",
      value: function nextPage(e) {
        babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "nextPage", this).call(this, e);
      }
      /**
       * LitElement life cycle - property changed
       */

    }, {
      key: "updated",
      value: function updated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(BootstrapTheme.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {});
      }
    }]);
    return BootstrapTheme;
  }((0, _HAXCMSThemeParts2.HAXCMSThemeParts)((0, _HAXCMSUserStylesMenu.HAXCMSUserStylesMenuMixin)((0, _HAXCMSMobileMenu.HAXCMSMobileMenuMixin)(_HAXCMSLitElementTheme.HAXCMSLitElementTheme))));

  _exports.BootstrapTheme = BootstrapTheme;
  customElements.define(BootstrapTheme.tag, BootstrapTheme);
});
