// @lrnwebcomponents/bootstrap-theme/lib/BootstrapMenu.js
define(["exports", "meta", "../../../lit-element/lit-element.js", "../../../mobx/dist/mobx.esm.js", "../../haxcms-elements/lib/core/haxcms-site-store.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js"], function (_exports, meta, _litElement, _mobxEsm, _haxcmsSiteStore, _simpleIconLite, _simpleIcons) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapMenu = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  function _templateObject7_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          --simple-icon-color: #007bff;\n        }\n\n        .bullet {\n          --simple-icon-height: 10px;\n          --simple-icon-width: 10px;\n        }\n\n        .indent-0 {\n          padding-left: 0px;\n        }\n        .indent-1,\n        .indent-2 {\n          padding-left: 16px;\n        }\n        .indent-3,\n        .indent-4,\n        .indent-5,\n        .indent-6 {\n          padding-left: 32px;\n        }\n\n        simple-icon-lite {\n          transform: rotate(270deg);\n        }\n        /* dark theme */\n        :host([color-theme=\"1\"]) {\n          --simple-icon-color: #999;\n        }\n\n        :host([color-theme=\"1\"]) .link {\n          color: #999;\n        }\n      "]);

    _templateObject7_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject7_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _templateObject6_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"indent-", "\">\n              <a class=\"link\" href=\"", "\"> ", " </a>\n            </div>\n          "]);

    _templateObject6_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject6_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _templateObject5_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      ", "\n    "]);

    _templateObject5_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject5_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _templateObject4_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["<simple-icon-lite\n                    class=\"bullet\"\n                    icon=\"radio-button-unchecked\"\n                  ></simple-icon-lite>"]);

    _templateObject4_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject4_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _templateObject3_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["<simple-icon-lite\n                    icon=\"arrow-drop-down\"\n                  ></simple-icon-lite>"]);

    _templateObject3_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject3_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _templateObject2_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["\n            <div class=\"indent-", "\">\n              ", "\n              <a class=\"link\" href=\"", "\"> ", " </a>\n            </div>\n            ", "\n          "]);

    _templateObject2_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject2_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _templateObject_85c4e69090bb11eb9010f5aa583cef45() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n\n      ", "\n    "]);

    _templateObject_85c4e69090bb11eb9010f5aa583cef45 = function _templateObject_85c4e69090bb11eb9010f5aa583cef45() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var BootstrapMenu = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(BootstrapMenu, _LitElement);

    var _super = _createSuper(BootstrapMenu);

    function BootstrapMenu() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapMenu);
      _this = _super.call(this);

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this._manifestItems = {};
      (0, _mobxEsm.autorun)(function () {
        var storeManifestItems = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest.items);

        if (_this._manifestItems !== storeManifestItems) {
          _this._manifestItems = storeManifestItems;
        }
      });
      return _this;
    }

    babelHelpers.createClass(BootstrapMenu, [{
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject_85c4e69090bb11eb9010f5aa583cef45(), this._bootstrapPath, this._manifestItems.map(function (item) {
          if (item.parent === null) {
            return (0, _litElement.html)(_templateObject2_85c4e69090bb11eb9010f5aa583cef45(), item.indent, item.children ? (0, _litElement.html)(_templateObject3_85c4e69090bb11eb9010f5aa583cef45()) : (0, _litElement.html)(_templateObject4_85c4e69090bb11eb9010f5aa583cef45()), item.slug, item.title, item.children ? _this2.renderChildren(item) : "");
          }
        }));
      }
    }, {
      key: "renderChildren",
      value: function renderChildren(item) {
        var _this3 = this;

        return (0, _litElement.html)(_templateObject5_85c4e69090bb11eb9010f5aa583cef45(), item.children.map(function (item) {
          if (item.children) {
            _this3.renderChildren(item);
          } else {
            return (0, _litElement.html)(_templateObject6_85c4e69090bb11eb9010f5aa583cef45(), item.indent, item.slug, item.title);
          }
        }));
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "bootstrap-menu";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          colorTheme: {
            type: String,
            reflect: true,
            attribute: "color-theme"
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject7_85c4e69090bb11eb9010f5aa583cef45())];
      }
    }]);
    return BootstrapMenu;
  }(_litElement.LitElement);

  _exports.BootstrapMenu = BootstrapMenu;
  customElements.define(BootstrapMenu.tag, BootstrapMenu);
});
