// @lrnwebcomponents/bootstrap-theme/lib/BootstrapFooter.js
define(["exports", "meta", "../../../lit-element/lit-element.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js", "../../haxcms-elements/lib/core/haxcms-site-store.js", "../../../mobx/dist/mobx.esm.js"], function (_exports, meta, _litElement, _simpleIconLite, _simpleIcons, _haxcmsSiteStore, _mobxEsm) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapFooter = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  function _templateObject2_82dd410091ad11eb950c05786044fa93() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <div class=\"container\">\n        <a class=\"backward\" href=\"", "\">\n          <button class=\"btn btn-outline-primary\">\n            <simple-icon-lite icon=\"av:fast-rewind\"></simple-icon-lite>\n            ", "\n          </button>\n        </a>\n        <a class=\"forward\" href=\"", "\">\n          <button class=\"btn btn-outline-primary forward\">\n            ", "\n            <simple-icon-lite icon=\"av:fast-forward\"></simple-icon-lite>\n          </button>\n        </a>\n      </div>\n    "]);

    _templateObject2_82dd410091ad11eb950c05786044fa93 = function _templateObject2_82dd410091ad11eb950c05786044fa93() {
      return data;
    };

    return data;
  }

  function _templateObject_82dd410091ad11eb950c05786044fa93() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          --bootstrap-dark-theme-secondary-background-color: #343a40;\n          margin-bottom: 10px;\n          --simple-icon-height: 18px;\n          --simple-icon-width: 18px;\n          --simple-icon-color: #007bff;\n        }\n\n        button:hover {\n          --simple-icon-color: white;\n        }\n\n        .container {\n          display: flex;\n          background-color: #e9ecef;\n          border-radius: 5px;\n          padding: 10px;\n        }\n\n        .forward {\n          margin-left: auto;\n        }\n\n        .backward {\n          margin-right: auto;\n        }\n\n        simple-icon-lite {\n          margin-bottom: 3px;\n        }\n\n        a {\n          text-decoration: none;\n        }\n        /* dark theme */\n        :host([color-theme=\"1\"]) {\n          --simple-icon-color: #fff;\n        }\n\n        :host([color-theme=\"1\"]) .btn {\n          color: #fff;\n          background-color: transparent;\n          background-image: none;\n          border-color: #6c757d;\n        }\n\n        :host([color-theme=\"1\"]) .btn:hover {\n          color: #fff;\n          background-color: #6c757d;\n          border-color: #6c757d;\n        }\n\n        :host([color-theme=\"1\"]) .container {\n          background-color: var(\n            --bootstrap-dark-theme-secondary-background-color\n          );\n        }\n      "]);

    _templateObject_82dd410091ad11eb950c05786044fa93 = function _templateObject_82dd410091ad11eb950c05786044fa93() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var BootstrapFooter = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(BootstrapFooter, _LitElement);

    var _super = _createSuper(BootstrapFooter);

    babelHelpers.createClass(BootstrapFooter, null, [{
      key: "tag",
      get: function get() {
        return "bootstrap-footer";
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject_82dd410091ad11eb950c05786044fa93())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          _backwardItem: {
            type: Object,
            reflect: true
          },
          _forwardItem: {
            type: Object,
            reflect: true
          },
          colorTheme: {
            type: String,
            reflect: true,
            attribute: "color-theme"
          }
        };
      }
    }]);

    function BootstrapFooter() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapFooter);
      _this = _super.call(this);

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this._forwardItem = {};
      _this._backwardItem = {};
      _this._activeItemIndex = -1;
      _this._routerManifest = {};
      _this.__disposer = _this.__disposer ? _this.__disposer : [];
      (0, _mobxEsm.autorun)(function (reaction) {
        var storeActiveItemIndex = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeManifestIndex);
        var storeRouterManifest = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.routerManifest);

        if (_this._activeItem !== storeActiveItemIndex || _this._routerManifest !== storeRouterManifest) {
          _this._activeItem = storeActiveItemIndex;
          _this._routerManifest = storeRouterManifest;

          if (storeRouterManifest.items[storeActiveItemIndex - 1]) {
            _this._backwardItem = storeRouterManifest.items[storeActiveItemIndex - 1];
          }

          if (storeRouterManifest.items[storeActiveItemIndex + 1]) {
            _this._forwardItem = storeRouterManifest.items[storeActiveItemIndex + 1];
          }
        }

        _this.__disposer.push(reaction);
      });
      return _this;
    }

    babelHelpers.createClass(BootstrapFooter, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_82dd410091ad11eb950c05786044fa93(), this._bootstrapPath, this._backwardItem.slug, this._backwardItem.title, this._forwardItem.slug, this._forwardItem.title);
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }]);
    return BootstrapFooter;
  }(_litElement.LitElement);

  _exports.BootstrapFooter = BootstrapFooter;
  customElements.define(BootstrapFooter.tag, BootstrapFooter);
});
