// @lrnwebcomponents/bootstrap-theme/lib/BootstrapSearch.js
define(["exports", "meta", "../../../lit-element/lit-element.js", "../../../mobx/dist/mobx.esm.js", "../../haxcms-elements/lib/core/haxcms-site-store.js"], function (_exports, meta, _litElement, _mobxEsm, _haxcmsSiteStore) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.BootstrapSearch = void 0;
  meta = babelHelpers.interopRequireWildcard(meta);

  function _templateObject2_3792b5f08e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject2_3792b5f08e7d11eba1fe15f944a3ea4d = function _templateObject2_3792b5f08e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function _templateObject_3792b5f08e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <link rel=\"stylesheet\" href=\"", "\" />\n      <form class=\"d-flex\">\n        <input\n          class=\"form-control me-2\"\n          type=\"search\"\n          placeholder=\"Type to search\"\n          aria-label=\"Search\"\n          @input=", "\n        />\n      </form>\n    "]);

    _templateObject_3792b5f08e7d11eba1fe15f944a3ea4d = function _templateObject_3792b5f08e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var BootstrapSearch = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(BootstrapSearch, _LitElement);

    var _super = _createSuper(BootstrapSearch);

    function BootstrapSearch() {
      var _this;

      babelHelpers.classCallCheck(this, BootstrapSearch);
      _this = _super.call(this);

      var basePath = _this.getBasePath(decodeURIComponent(meta.url));

      _this._bootstrapPath = basePath + "bootstrap/dist/css/bootstrap.min.css";
      _this.searchText = "";
      return _this;
    }

    babelHelpers.createClass(BootstrapSearch, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_3792b5f08e7d11eba1fe15f944a3ea4d(), this._bootstrapPath, this.inputChanged);
      }
    }, {
      key: "inputChanged",
      value: function inputChanged(evt) {
        if (evt.target.value) {
          this.searchText = evt.target.value;
          this.dispatchEvent(new CustomEvent("searchChanged", {
            bubbles: true,
            composed: true,
            detail: this
          }));
        }
      }
    }, {
      key: "getBasePath",
      value: function getBasePath(url) {
        return url.substring(0, url.lastIndexOf("/@lrnwebcomponents/") + 1);
      }
    }], [{
      key: "tag",
      get: function get() {
        return "bootstrap-search";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          colorTheme: {
            type: String,
            reflect: true,
            attribute: "color-theme"
          },
          searchText: {
            type: String
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_3792b5f08e7d11eba1fe15f944a3ea4d())];
      }
    }]);
    return BootstrapSearch;
  }(_litElement.LitElement);

  _exports.BootstrapSearch = BootstrapSearch;
  customElements.define(BootstrapSearch.tag, BootstrapSearch);
});
