// @haxtheweb/link-card-theme/link-card-theme.js
/**
 * Copyright 2025 haxtheweb
 * @license Apache-2.0, see LICENSE for full text.
 */
import{css as e,html as t}from"../../lit/index.js";import{ifDefined as i}from"../../lit/directives/if-defined.js";import{HAXCMSLitElementTheme as a}from"../haxcms-elements/lib/core/HAXCMSLitElementTheme.js";import{HAXCMSThemeParts as r}from"../haxcms-elements/lib/core/utils/HAXCMSThemeParts.js";import{store as d}from"../haxcms-elements/lib/core/haxcms-site-store.js";import{autorun as o,toJS as l}from"../../mobx/dist/mobx.esm.js";import{DDDSuper as n}from"../d-d-d/d-d-d.js";import{DDDVariables as s,DDDPaletteStyles as c}from"../d-d-d/lib/DDDStyles.js";import{UserScaffoldInstance as p}from"../user-scaffold/user-scaffold.js";import"../simple-icon/lib/simple-icons.js";import"../simple-icon/lib/simple-icon-lite.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../hax-iconset/lib/simple-hax-iconset.js";export class LinkCardTheme extends(r(n(a))){static get tag(){return"link-card-theme"}constructor(){super(),this.manifest={},this.linkItems=[],this.socialItems=[],this.profileImage="",this.profileName="My profile",this.profileDescription="";const e=p.readMemory("HAXCMSSitePalette");this.dataPalette=null===e||""===e?"0":String(e),this.__disposer=this.__disposer||[],this.__disposer.push(o(()=>{const e=l(d.manifest);Promise.resolve().then(()=>{this.manifest=e||{},this._syncManifestData()})}))}static get properties(){return{...super.properties,manifest:{type:Object},linkItems:{type:Array},socialItems:{type:Array},profileImage:{type:String,attribute:"profile-image"},profileName:{type:String,attribute:"profile-name"},profileDescription:{type:String,attribute:"profile-description"},dataPalette:{type:String,reflect:!0,attribute:"data-palette"}}}static get styles(){return[s,c,super.styles,e`
        :host {
          display: block;
          min-height: 100vh;
          --link-card-shell-bg: var(--ddd-palette-color-2);
          --link-card-shell-text: #000000;
          --link-card-card-bg: var(--ddd-theme-default-white);
          --link-card-card-text: #000000;
          --link-card-button-bg: var(--ddd-palette-color-1);
          --link-card-button-hover-bg: var(--ddd-palette-color-3);
          --link-card-button-text: light-dark(black, white);
          --link-card-button-hover-text: var(--link-card-button-text);
          --link-card-outline: light-dark(
            var(--ddd-theme-default-limestoneGray),
            var(--ddd-palette-color-5)
          );
        }
        @media (prefers-color-scheme: dark) {
          :host {
            --link-card-shell-bg: var(--ddd-palette-color-6);
            --link-card-shell-text: #ffffff;
            --link-card-card-bg: var(--ddd-theme-default-coalyGray);
            --link-card-card-text: #ffffff;
            --link-card-button-bg: var(--ddd-palette-color-7);
            --link-card-button-hover-bg: var(--ddd-palette-color-5);
            --link-card-button-hover-text: var(--link-card-button-text);
          }
        }
        :host([dark-mode]) {
          --link-card-shell-bg: var(--ddd-palette-color-6);
          --link-card-shell-text: #ffffff;
          --link-card-card-bg: var(--ddd-theme-default-coalyGray);
          --link-card-card-text: #ffffff;
          --link-card-button-bg: var(--ddd-palette-color-7);
          --link-card-button-hover-bg: var(--ddd-palette-color-5);
          --link-card-button-hover-text: var(--link-card-button-text);
        }
        :host([data-palette="0"]),
        :host([data-palette="1"]),
        :host([data-palette="2"]),
        :host([data-palette="3"]),
        :host([data-palette="5"]),
        :host([data-palette="8"]),
        :host([data-palette="11"]),
        :host([data-palette="15"]) {
          --link-card-button-hover-text: light-dark(white, black);
        }
        :host([data-palette="10"]) {
          --link-card-button-hover-text: white;
        }

        .theme-shell {
          min-height: 100vh;
          padding: var(--ddd-spacing-8) var(--ddd-spacing-4);
          box-sizing: border-box;
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: center;
          gap: var(--ddd-spacing-6);
          background-color: var(--link-card-shell-bg);
          color: var(--link-card-shell-text);
        }
        .card {
          width: min(100%, 620px);
          border: var(--ddd-border-sm) solid var(--link-card-outline);
          border-radius: var(--ddd-radius-xl);
          box-shadow: var(--ddd-boxShadow-sm);
          padding: var(--ddd-spacing-8) var(--ddd-spacing-6);
          box-sizing: border-box;
          background-color: var(--link-card-card-bg);
          color: var(--link-card-card-text);
          text-align: center;
        }
        .profile-image {
          width: var(--ddd-spacing-30);
          height: var(--ddd-spacing-30);
          object-fit: cover;
          object-position: center;
          border-radius: var(--ddd-radius-circle);
          border: var(--ddd-border-sm) solid var(--link-card-outline);
          box-shadow: var(--ddd-boxShadow-xs);
        }
        .profile-image-fallback {
          width: var(--ddd-spacing-30);
          height: var(--ddd-spacing-30);
          margin: 0 auto;
          border-radius: var(--ddd-radius-circle);
          border: var(--ddd-border-sm) solid var(--link-card-outline);
          display: flex;
          align-items: center;
          justify-content: center;
          color: var(--link-card-card-text);
        }
        .profile-image-fallback simple-icon-lite {
          --simple-icon-width: var(--ddd-icon-3xl);
          --simple-icon-height: var(--ddd-icon-3xl);
        }
        h1 {
          margin: var(--ddd-spacing-4) 0 var(--ddd-spacing-2);
          font-size: clamp(
            var(--ddd-font-size-l),
            5vw,
            var(--ddd-font-size-xl)
          );
          font-family: var(--ddd-font-navigation, sans-serif);
          font-weight: var(--ddd-font-weight-bold);
          line-height: 1.2;
        }
        .description {
          margin: 0 0 var(--ddd-spacing-5);
          font-size: var(--ddd-font-size-s);
          line-height: 1.5;
          color: var(--link-card-card-text);
        }
        .primary-links {
          display: grid;
          gap: var(--ddd-spacing-3);
        }
        .link-button {
          display: inline-flex;
          align-items: center;
          justify-content: center;
          min-height: 48px;
          padding: var(--ddd-spacing-3) var(--ddd-spacing-4);
          border-radius: var(--ddd-radius-pill);
          border: var(--ddd-border-sm) solid var(--link-card-outline);
          font-size: var(--ddd-font-size-s);
          font-family: var(--ddd-font-navigation, sans-serif);
          font-weight: var(--ddd-font-weight-bold);
          text-decoration: none;
          color: var(--link-card-button-text);
          background-color: var(--link-card-button-bg);
          box-shadow: var(--ddd-boxShadow-xs);
          transition:
            transform 0.2s ease,
            background-color 0.2s ease;
        }
        .link-button:hover,
        .link-button:focus-visible {
          transform: translateY(-1px);
          background-color: var(--link-card-button-hover-bg);
          color: var(--link-card-button-hover-text);
          text-decoration: none;
        }
        .link-button:focus-visible,
        .social-link:focus-visible,
        .palette-picker:focus-visible {
          outline: var(--ddd-border-sm) solid
            light-dark(
              var(--ddd-theme-default-nittanyNavy),
              var(--ddd-theme-default-keystoneYellow)
            );
          outline-offset: var(--ddd-spacing-1);
        }
        .empty-state {
          margin: var(--ddd-spacing-2) 0 0;
          font-size: var(--ddd-font-size-xs);
          line-height: 1.5;
        }
        .social-links {
          list-style: none;
          margin: var(--ddd-spacing-6) 0 0;
          padding: 0;
          display: flex;
          justify-content: center;
          gap: var(--ddd-spacing-3);
        }
        .social-link {
          width: var(--ddd-spacing-9);
          height: var(--ddd-spacing-9);
          border-radius: var(--ddd-radius-circle);
          border: var(--ddd-border-sm) solid var(--link-card-outline);
          display: inline-flex;
          align-items: center;
          justify-content: center;
          text-decoration: none;
          color: var(--link-card-button-text);
          background-color: var(--link-card-button-bg);
          box-shadow: var(--ddd-boxShadow-xs);
          transition: background-color 0.2s ease;
        }
        .social-link:hover,
        .social-link:focus-visible {
          background-color: var(--link-card-button-hover-bg);
          color: var(--link-card-button-hover-text);
        }
        .social-link simple-icon-lite {
          --simple-icon-width: var(--ddd-icon-sm);
          --simple-icon-height: var(--ddd-icon-sm);
        }
        .palette-picker {
          --simple-icon-button-border-radius: var(--ddd-radius-circle);
          --simple-icon-button-border: var(--ddd-border-sm) solid
            var(--link-card-outline);
          --simple-icon-button-color: var(--link-card-button-text);
          --simple-icon-button-background: var(--link-card-button-bg);
        }
        #contentcontainer {
          width: min(100%, 620px);
          box-sizing: border-box;
          border: var(--ddd-border-sm) dashed var(--link-card-outline);
          border-radius: var(--ddd-radius-md);
          padding: var(--ddd-spacing-4);
          background-color: var(--link-card-card-bg);
          color: var(--link-card-card-text);
        }
        .contrast-color-resolver {
          position: absolute;
          width: 0;
          height: 0;
          overflow: hidden;
          opacity: 0;
          pointer-events: none;
        }
        :host(:not([edit-mode])) #contentcontainer {
          display: none;
        }
      `]}updated(e){super.updated&&super.updated(e),e.has("dataPalette")&&p.writeMemory("HAXCMSSitePalette",this.dataPalette,"long")}_itemOrder(e){if(!e||void 0===e.order||null===e.order)return 0;const t=Number(e.order);return Number.isNaN(t)?0:t}_itemHidden(e){return!(!e||!e.metadata)&&(!0===e.metadata.hideInMenu||!1===e.metadata.published)}_isExternalHref(e){return!!e&&/^https?:\/\//i.test(e)}_resolveItemHref(e){if(!e)return"";let t="";return e.metadata&&e.metadata.linkUrl&&(t=`${e.metadata.linkUrl}`.trim()),""===t&&e.slug&&(t=`${e.slug}`.trim()),""===t&&e.location&&(t=`${e.location}`.trim()),t}_resolveTarget(e,t){return e&&e.metadata&&e.metadata.linkTarget&&["_self","_blank","_parent","_top"].includes(e.metadata.linkTarget)?e.metadata.linkTarget:this._isExternalHref(t)?"_blank":"_self"}_detectSocialType(e){if(!this._isExternalHref(e))return"";const t=e.toLowerCase();return-1!==t.indexOf("linkedin.com")?"linkedin":-1!==t.indexOf("x.com")||-1!==t.indexOf("twitter.com")?"x":""}_socialIcon(e){return"linkedin"===e?"mdi-social:linkedin":"mdi-social:twitter"}_socialLabel(e){return"linkedin"===e?"LinkedIn":"X"}_socialAriaLabel(e){const t=this._socialLabel(e.socialType);return e.title?`${e.title} on ${t}`:`Open ${t}`}_syncManifestData(){const e=this.manifest&&"object"==typeof this.manifest?this.manifest:{},t=e.metadata&&"object"==typeof e.metadata?e.metadata:{},i=t.theme&&"object"==typeof t.theme?t.theme:{},a=i.variables&&"object"==typeof i.variables?i.variables:{},r=t.author&&"object"==typeof t.author?t.author:{},d=t.site&&"object"==typeof t.site?t.site:{},o=Array.isArray(e.items)?e.items:[];void 0!==a.palette&&null!==a.palette&&""!==`${a.palette}`.trim()&&(this.dataPalette=`${a.palette}`.trim()),this.profileName=e.title?e.title:r.name?r.name:d.name?d.name:"My profile",this.profileDescription=e.description?e.description:"",this.profileImage=a.image?a.image:r.image?r.image:d.logo?d.logo:"";const l=o.filter(e=>e&&null===e.parent&&!this._itemHidden(e)).sort((e,t)=>this._itemOrder(e)-this._itemOrder(t)),n=[],s=[];l.forEach(e=>{const t=this._resolveItemHref(e);if(!t)return;const i=this._resolveTarget(e,t),a=this._detectSocialType(t),r={title:e.title?e.title:t,href:t,target:i,rel:"_blank"===i?"noopener noreferrer":null,socialType:a};a?s.push(r):n.push(r)}),this.linkItems=n,this.socialItems=s}togglePalette(){let e=Number(this.dataPalette);Number.isNaN(e)&&(e=0),e+=1,e>11&&(e=0),this.dataPalette=`${e}`}testEditMode(e){this.editMode?(e.preventDefault(),e.stopPropagation(),e.stopImmediatePropagation()):e.currentTarget.blur()}render(){const e=this.profileName?`${this.profileName} profile image`:"Profile image",a=this.profileName?this.profileName:"Link card profile";return t`
      <div class="theme-shell">
        <span class="contrast-color-resolver" aria-hidden="true"></span>
        <simple-icon-button-lite
          class="palette-picker"
          icon="image:style"
          label="Change palette"
          title="Change palette"
          @click="${this.togglePalette}"
        ></simple-icon-button-lite>
        <main class="card" role="main" aria-label="${a}">
          ${this.profileImage?t`<img
                class="profile-image"
                src="${this.profileImage}"
                alt="${e}"
                loading="lazy"
                decoding="async"
                fetchpriority="low"
              />`:t`<div class="profile-image-fallback" aria-hidden="true">
                <simple-icon-lite icon="account-circle"></simple-icon-lite>
              </div>`}
          <h1>${this.profileName}</h1>
          ${this.profileDescription?t`<p class="description">${this.profileDescription}</p>`:""}
          <nav class="primary-links" aria-label="Primary links">
            ${this.linkItems.length>0?this.linkItems.map(e=>t`<a
                    class="link-button"
                    href="${e.href}"
                    target="${e.target}"
                    rel="${i(e.rel)}"
                    @click="${this.testEditMode}"
                    >${e.title}</a
                  >`):t`<p class="empty-state">
                  Add top-level menu items in site.json to render link buttons.
                </p>`}
          </nav>
          ${this.socialItems.length>0?t`<ul class="social-links" aria-label="Social links">
                ${this.socialItems.map(e=>t`<li>
                    <a
                      class="social-link"
                      href="${e.href}"
                      target="${e.target}"
                      rel="${i(e.rel)}"
                      @click="${this.testEditMode}"
                      aria-label="${this._socialAriaLabel(e)}"
                    >
                      <simple-icon-lite
                        icon="${this._socialIcon(e.socialType)}"
                      ></simple-icon-lite>
                    </a>
                  </li>`)}
              </ul>`:""}
        </main>
        <div id="contentcontainer">
          <div id="slot"><slot></slot></div>
        </div>
      </div>
    `}disconnectedCallback(){this.__disposer&&(this.__disposer.forEach(e=>{"function"==typeof e&&e()}),this.__disposer=[]),this.__darkModeMediaQuery&&(this.__darkModeMediaQuery.removeEventListener?this.__darkModeMediaQuery.removeEventListener("change",this.__onColorSchemeChange):this.__darkModeMediaQuery.removeListener&&this.__darkModeMediaQuery.removeListener(this.__onColorSchemeChange)),super.disconnectedCallback()}}globalThis.customElements.define(LinkCardTheme.tag,LinkCardTheme);
