// @lrnwebcomponents/simple-fields/lib/simple-tags.js
define(["exports", "meta", "../../../lit/index.js", "./simple-tag.js"], function (_exports, meta, _index, _simpleTag) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleTags = void 0;
  meta = _interopRequireWildcard(meta);
  var _templateObject_164ca200e8e111ed8789699b2ad5335a, _templateObject2_164ca200e8e111ed8789699b2ad5335a, _templateObject3_164ca200e8e111ed8789699b2ad5335a;
  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  var SimpleTags = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SimpleTags, _LitElement);
    var _super = _createSuper(SimpleTags);
    function SimpleTags() {
      var _this;
      babelHelpers.classCallCheck(this, SimpleTags);
      _this = _super.call(this);
      _this.tags = "";
      _this.autoAccentColor = false;
      _this.accentColor = null;
      return _this;
    }
    babelHelpers.createClass(SimpleTags, [{
      key: "render",
      value: function render() {
        var _this2 = this;
        return (0, _index.html)(_templateObject_164ca200e8e111ed8789699b2ad5335a || (_templateObject_164ca200e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral([" <div class=\"tag-container\">\n      ", "\n    </div>"])), this.tags ? this.tags.split(",").map(function (tag) {
          return (0, _index.html)(_templateObject2_164ca200e8e111ed8789699b2ad5335a || (_templateObject2_164ca200e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n          <simple-tag\n            ?auto-accent-color=\"", "\"\n            value=\"", "\"\n            accent-color=\"", "\"\n          ></simple-tag>\n        "])), _this2.autoAccentColor, tag.trim(), _this2.accentColor);
        }) : "");
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          tags: {
            type: String
          },
          accentColor: {
            type: String,
            attribute: "accent-color"
          },
          autoAccentColor: {
            type: Boolean,
            attribute: "auto-accent-color"
          }
        };
      }
    }, {
      key: "tag",
      get: function get() {
        return "simple-tags";
      }

      /**
       * haxProperties integration via file reference
       */
    }, {
      key: "haxProperties",
      get: function get() {
        return new URL("./".concat(this.tag, ".haxProperties.json"), meta.url).href;
      }
    }, {
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject3_164ca200e8e111ed8789699b2ad5335a || (_templateObject3_164ca200e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        .tag-container {\n          display: flex;\n          flex-wrap: wrap;\n          justify-content: center;\n        }\n        simple-tag {\n          margin: 2px 2px;\n        }\n      "])))];
      }
    }]);
    return SimpleTags;
  }(_index.LitElement);
  _exports.SimpleTags = SimpleTags;
  customElements.define(SimpleTags.tag, SimpleTags);
});
