// @lrnwebcomponents/flash-card/flash-card.js
define(["exports", "meta", "require", "../../lit/index.js", "../simple-colors/simple-colors.js", "./lib/flash-card-answer-box.js", "./lib/flash-card-prompt-img.js"], function (_exports, meta, _require, _index, _simpleColors, _flashCardAnswerBox, _flashCardPromptImg) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.FlashCard = void 0;
  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);
  var _templateObject_f2b85b40e8e011ed8789699b2ad5335a, _templateObject2_f2b85b40e8e011ed8789699b2ad5335a, _templateObject3_f2b85b40e8e011ed8789699b2ad5335a;
  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  var FlashCard = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(FlashCard, _SimpleColors);
    var _super = _createSuper(FlashCard);
    // HTMLElement life-cycle, built in; use this for setting defaults
    function FlashCard() {
      var _this;
      babelHelpers.classCallCheck(this, FlashCard);
      _this = _super.call(this);
      _this.imgKeyword = "";
      _this.imgSource = "";
      return _this;
    }

    // properties that you wish to use as data in HTML, CSS, and the updated life-cycle
    babelHelpers.createClass(FlashCard, [{
      key: "statusChanged",
      value: function statusChanged(e) {
        var _this2 = this;
        this.status = e.detail;
        if (this.status === "correct") {
          new Promise(function (res, rej) {
            return _require.default(["../multiple-choice/lib/confetti-container.js"], res, rej);
          }).then(function (module) {
            setTimeout(function () {
              _this2.shadowRoot.querySelector("#confetti").setAttribute("popped", "");
            }, 0);
          });
        }
      }

      // HTML - specific to Lit
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_f2b85b40e8e011ed8789699b2ad5335a || (_templateObject_f2b85b40e8e011ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n      <confetti-container id=\"confetti\">\n        ", "\n        <flash-card-answer-box\n          @flash-card-status-change=\"", "\"\n        >\n          <div slot=\"front\">\n            <slot slot=\"front\" name=\"front\"></slot>\n          </div>\n          <div slot=\"back\">\n            <slot slot=\"back\" name=\"back\"></slot>\n          </div>\n        </flash-card-answer-box>\n      </confetti-container>\n    "])), !this.imgSource && !this.imgKeyword ? "" : (0, _index.html)(_templateObject2_f2b85b40e8e011ed8789699b2ad5335a || (_templateObject2_f2b85b40e8e011ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n              <flash-card-image-prompt\n                img-src=\"", "\"\n                img-keyword=\"", "\"\n                status=\"", "\"\n              ></flash-card-image-prompt>\n            "])), this.imgSource, this.imgKeyword, this.status), this.statusChanged);
      }

      /**
       * haxProperties integration via file reference
       */
    }], [{
      key: "tag",
      get: function get() {
        return "flash-card";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(FlashCard), "properties", this)), {}, {
          inverted: {
            type: Boolean
          },
          imgSource: {
            type: String,
            attribute: "img-source",
            reflect: true
          },
          imgKeyword: {
            type: String,
            attribute: "img-keyword"
          },
          status: {
            type: String,
            reflect: true
          }
        });
      }

      // CSS - specific to Lit
    }, {
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(FlashCard), "styles", this)), [(0, _index.css)(_templateObject3_f2b85b40e8e011ed8789699b2ad5335a || (_templateObject3_f2b85b40e8e011ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          border: 1px solid var(--simple-colors-default-theme-accent-6);\n          min-width: 320px;\n          min-height: 155px;\n          border-radius: 20px;\n          padding: 20px;\n          width: 5em;\n          background-color: var(--simple-colors-default-theme-accent-2);\n          box-shadow: 0 0 5px var(--simple-colors-default-theme-accent-7);\n          margin: 10px;\n        }\n        p {\n          color: var(--simple-colors-default-theme-accent-10);\n        }\n      "])))]);
      }
    }, {
      key: "haxProperties",
      get: function get() {
        return new URL("./lib/".concat(this.tag, ".haxProperties.json"), meta.url).href;
      }
    }]);
    return FlashCard;
  }(_simpleColors.SimpleColors);
  _exports.FlashCard = FlashCard;
  customElements.define(FlashCard.tag, FlashCard);
});
