// @haxtheweb/app-hax/lib/v2/app-hax-site-bar.js
import{html as e,css as t,unsafeCSS as i}from"../../../../lit/index.js";import"../../../simple-icon/lib/simple-icons.js";import"../../../simple-icon/lib/simple-icon-button-lite.js";import{SimpleColors as a}from"../../../simple-colors/simple-colors.js";import"../../../simple-toolbar/lib/simple-toolbar-button.js";import"../../../simple-fields/lib/simple-context-menu.js";import{toJS as o}from"../../../../mobx/dist/mobx.esm.js";import{store as s}from"./AppHaxStore.js";import"./app-hax-user-access-modal.js";import"./app-hax-confirmation-modal.js";let l=0;export class AppHaxSiteBars extends a{static get tag(){return"app-hax-site-bar"}constructor(){super(),this.showOptions=!1,this.inprogress=!1,this.textInfo={},this.siteId="",this.description="",this.lastUpdatedTime=0,this.moreOptionsId=`moreOptions-${l}`,l+=1}static get properties(){return{...super.properties,showOptions:{type:Boolean},inprogress:{type:Boolean,reflect:!0},textInfo:{type:Object},siteId:{type:String,reflect:!0,attribute:"site-id"},title:{type:String},slug:{type:String},description:{type:String},siteUrl:{type:String,attribute:"site-url"},image:{type:String},lastUpdatedTime:{type:Number,attribute:"last-updated-time"},moreOptionsId:{type:String}}}toggleOptionsMenu(){const e=this.shadowRoot.querySelector("simple-context-menu"),t=this.shadowRoot.querySelector(`#${this.moreOptionsId}`);e&&t&&e.toggle(t)}closeOptionsMenu(){const e=this.shadowRoot.querySelector("simple-context-menu");e&&e.close()}handleKeydown(e,t){"Enter"!==e.key&&" "!==e.key||(e.preventDefault(),t.call(this))}getSiteMachineName(){if(!this.slug)return"";const e=this.slug.split("/").filter(e=>e.length>0);return e.length>0?e[e.length-1]:this.slug}copySite(){this.closeOptionsMenu(),this.siteOperation("copySite","Copy","icons:content-copy")}downloadSite(){this.closeOptionsMenu(),this.siteOperation("downloadSite","Download","file-download")}archiveSite(){this.closeOptionsMenu(),this.siteOperation("archiveSite","Archive","icons:archive")}createTemplate(){this.closeOptionsMenu();const e=this.getSiteRecord();e&&this.openCreateTemplateDialog(e)}getSiteRecord(){return s&&s.manifest&&Array.isArray(s.manifest.items)?o(s.manifest.items.filter(e=>e.id===this.siteId).pop()):null}openCreateTemplateDialog(e){if(!e||!e.metadata||!e.metadata.site)return;const t=e.metadata.site.name||this.title||"site",i=document.createElement("app-hax-confirmation-modal");i.title=`Create template from ${t}?`,i.message="Choose how to turn this site into a reusable template: download the skeleton file or save it to your templates.",i.confirmText="Save to templates",i.cancelText="Download skeleton",i.modal=!0,i.cancelIsNeutral=!0,i.cancelActionOnPassiveClose=!1,i.confirmAction=this.saveSiteAsTemplate.bind(this,e),i.cancelAction=this.downloadSiteSkeleton.bind(this,e),document.body.appendChild(i),i.openModal(),i.addEventListener("close",()=>{document.body.contains(i)&&document.body.removeChild(i)},{once:!0})}triggerJsonDownload(e,t){const i=t&&"string"==typeof t&&""!==t.trim()?t.trim():"site-template.json",a=JSON.stringify(e,null,2),o=new Blob([a],{type:"application/json"}),s=globalThis.URL.createObjectURL(o),l=globalThis.document.createElement("a");l.href=s,l.setAttribute("download",i),l.style.display="none",globalThis.document.body.appendChild(l),l.click(),globalThis.document.body.removeChild(l),globalThis.URL.revokeObjectURL(s)}refreshTemplateResults(){let e=globalThis.document.querySelector("app-hax-use-case-filter");!e&&s&&s.appEl&&s.appEl.shadowRoot&&(e=s.appEl.shadowRoot.querySelector("app-hax-use-case-filter")),e&&"function"==typeof e.updateSkeletonResults&&e.updateSkeletonResults()}async downloadSiteSkeleton(e){if(!(s&&s.appSettings&&s.appSettings.downloadSiteSkeleton&&s.AppHaxAPI&&s.AppHaxAPI.makeCall))return void s.toast("Download skeleton endpoint is not configured.",3e3,{hat:"random"});const t=e&&e.metadata&&e.metadata.site&&"string"==typeof e.metadata.site.name?e.metadata.site.name:"";if(t)try{const i=await s.AppHaxAPI.makeCall("downloadSiteSkeleton",{site:{name:t,id:e.id}},!0),a=i&&i.data?i.data:null,o=a&&a.skeleton?a.skeleton:a;if(!o||"object"!=typeof o)throw new Error("Skeleton payload missing from response");const l=a&&a.filename&&"string"==typeof a.filename?a.filename:`${this.getSiteMachineName()||"site-template"}.json`;this.triggerJsonDownload(o,l),s.toast(`${t} skeleton downloaded!`,3e3,{hat:"random"})}catch(e){s.appEl&&s.appEl.playSound&&s.appEl.playSound("error"),console.error("downloadSiteSkeleton failed:",e),s.toast("Unable to download site skeleton.",4e3,{hat:"random"})}}async saveSiteAsTemplate(e){if(!(s&&s.appSettings&&s.appSettings.saveSiteAsTemplate&&s.AppHaxAPI&&s.AppHaxAPI.makeCall))return void s.toast("Save template endpoint is not configured.",3e3,{hat:"random"});const t=e&&e.metadata&&e.metadata.site&&"string"==typeof e.metadata.site.name?e.metadata.site.name:"";if(t)try{const i=await s.AppHaxAPI.makeCall("saveSiteAsTemplate",{site:{name:t,id:e.id}},!0);if(!i||200!==i.status)throw new Error("saveSiteAsTemplate did not return status 200");this.refreshTemplateResults(),s.toast(`${t} saved to templates!`,3e3,{hat:"random"})}catch(e){s.appEl&&s.appEl.playSound&&s.appEl.playSound("error"),console.error("saveSiteAsTemplate failed:",e),s.toast("Unable to save site as template.",4e3,{hat:"random"})}}openUserAccess(){this.closeOptionsMenu(),s.activeSiteId=this.siteId,import("../../../simple-modal/simple-modal.js").then(()=>{const e=document.createElement("app-hax-user-access-modal");this.title&&(e.siteTitle=this.title);const t=new CustomEvent("simple-modal-show",{bubbles:!0,composed:!0,cancelable:!0,detail:{title:"User Access",elements:{content:e},invokedBy:this,styles:{"--simple-modal-titlebar-background":"var(--ddd-theme-default-nittanyNavy, #001e44)","--simple-modal-titlebar-color":"var(--ddd-theme-default-white, white)","--simple-modal-width":"80vw","--simple-modal-max-width":"var(--ddd-spacing-32, 480px)","--simple-modal-min-width":"300px","--simple-modal-z-index":"1000","--simple-modal-height":"auto","--simple-modal-min-height":"300px","--simple-modal-max-height":"80vh","--simple-modal-titlebar-height":"80px","--simple-modal-border-radius":"var(--ddd-radius-md, 8px)","--simple-modal-background":"light-dark(var(--ddd-theme-default-white, white), var(--ddd-theme-default-coalyGray, #1c1c1c))","--simple-modal-content-container-background":"light-dark(var(--ddd-theme-default-white, white), var(--ddd-theme-default-coalyGray, #1c1c1c))","--simple-modal-content-container-color":"light-dark(var(--ddd-theme-default-coalyGray, #1c1c1c), var(--ddd-theme-default-white, white))","--simple-modal-buttons-background":"light-dark(var(--ddd-theme-default-white, white), var(--ddd-theme-default-coalyGray, #1c1c1c))","--simple-modal-box-shadow":"var(--ddd-boxShadow-xl)"}}});this.dispatchEvent(t)})}siteOperation(e,t,i){s.appEl&&s.appEl.playSound&&s.appEl.playSound("click"),s.activeSiteOp=e,s.activeSiteId=this.siteId;const a=o(s.manifest.items.filter(e=>e.id===this.siteId).pop());if(!a)return void console.error("Site not found for ID:",this.siteId);const l=document.createElement("app-hax-confirmation-modal");l.title=`${t} ${a.metadata.site.name}?`,l.message=`Are you sure you want to ${e.replace("Site","")} ${a.metadata.site.name}?`,l.confirmText="Confirm",l.cancelText="Cancel",l.dangerous="archiveSite"===e,l.confirmAction=this.confirmOperation.bind(this),l.cancelAction=this.cancelOperation.bind(this),document.body.appendChild(l),l.openModal(),l.addEventListener("close",()=>{document.body.removeChild(l)},{once:!0})}cancelOperation(){s.activeSiteOp="",s.activeSiteId=null,globalThis.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0}))}async confirmOperation(){const e=o(s.activeSiteOp),t=o(s.activeSite);t?(await s.AppHaxAPI.makeCall(e,{site:{name:t.metadata.site.name,id:t.id}},!0,()=>{if("downloadSite"===o(s.activeSiteOp)){const e=s.AppHaxAPI.lastResponse.downloadSite;if(e&&e.data&&e.data.link){const t=e.data.link,i=e.data.name||"",a=globalThis.document.createElement("a");a.href=t,i&&a.setAttribute("download",i),a.style.display="none",globalThis.document.body.appendChild(a),a.click(),globalThis.document.body.removeChild(a)}else console.error("downloadSite response missing data.link:",e)}else s.refreshSiteListing()}),globalThis.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0})),s.toast(`${t.metadata.site.name} ${e.replace("Site","")} successful!`,3e3,{hat:"random"})):console.error("No active site found for operation:",e)}static get styles(){return[super.styles,t`
        :host {
          text-align: left;
          --app-hax-site-card-width: 180px;
          --app-hax-site-card-min-height: 260px;
          --app-hax-site-image-height: 180px;
          --app-hax-site-title-font-size: var(--ddd-font-size-xxs, 13px);
          --app-hax-site-date-font-size: var(--ddd-font-size-4xs, 12px);
          --app-hax-site-card-content-padding: var(--ddd-spacing-2, 8px);
          width: var(--app-hax-site-card-width);
          max-width: var(--app-hax-site-card-width);
          font-family: var(--ddd-font-primary);
          background-color: white;
          border: var(--ddd-border-sm);
          min-height: var(--app-hax-site-card-min-height);
          box-shadow: light-dark(
            2px 2px 12px #1c1c1c,
            2px 2px 12px rgba(0, 0, 0, 0.3)
          );
          transition: transform 0.2s ease, box-shadow 0.2s ease, border 0.2s ease;
          overflow: visible;
          
        }

        :host(:hover),
        :host(:focus),
        :host(:focus-within) {
          border: var(--ddd-border-md);
          border-color:var(--ddd-theme-default-beaverBlue);
          box-shadow: light-dark(
            4px 8px 24px rgba(28, 28, 28, 0.15),
            4px 8px 24px rgba(0, 0, 0, 0.5)
          );
          transform: scale(1.03);
          transform-origin: center top;
          z-index: 2;
        }

        :host([dark]:hover),
        :host([dark]:focus),
        :host([dark]:focus-within) {
          border-color: var(--ddd-theme-default-skyBlue);
        }


        #mainCard {
          display: flex;
          flex-direction: column;
        }

        .cardContent {
          padding: var(--app-hax-site-card-content-padding);
        }

        .cardImage {
          width: 100%;
          object-fit: cover;
          border-top-left-radius: var(--ddd-radius-md, 8px);
          border-top-right-radius: var(--ddd-radius-md, 8px);
          border-bottom: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-limestoneGray);
          background-color: var(--ddd-theme-default-skyLight);
        }

        .titleBar {
          display: flex;
          justify-content: space-between;
          align-items: center;
          position: relative;
          overflow: visible;
          margin-bottom: var(--ddd-spacing-2, 8px);
        }

        .more-options {
          display: flex;
          align-items: center;
        }

        .more-options simple-icon-button-lite {
          --simple-icon-color: var(--ddd-theme-default-nittanyNavy);
        }

        .more-options simple-icon-button-lite:hover,
        .more-options simple-icon-button-lite:focus,
        .more-options simple-icon-button-lite::part(button):hover,
        .more-options simple-icon-button-lite::part(button):focus {
          color: var(--ddd-theme-default-accent);
          --simple-icon-color: var(--ddd-theme-default-accent);
        }

        ::slotted([slot="heading"]) {
          font-size: var(--app-hax-site-title-font-size);
          font-weight: var(--ddd-font-weight-bold, 700);
          color: var(--ddd-theme-default-nittanyNavy);
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        ::slotted(a[slot="heading"]) {
          text-decoration: none;
        }

        .date {
          display: flex;
          gap: var(--ddd-spacing-1, 4px);
          margin-top: var(--ddd-spacing-2, 8px);
          font-size: var(--app-hax-site-date-font-size);
          color: var(--ddd-theme-default-nittanyNavy);
          align-items: center;
          line-height: 1;
        }

        .date simple-icon {
          --simple-icon-color: var(--ddd-theme-default-nittanyNavy);
        }

        img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          display: block;
        }

        .imageLink {
          display: block;
          overflow: hidden;
          height: var(--app-hax-site-image-height);
        }

        @media (max-width: 640px) {
          :host {
            --app-hax-site-card-width: clamp(120px, 30vw, 148px);
            --app-hax-site-card-min-height: 210px;
            --app-hax-site-image-height: clamp(96px, 27vw, 120px);
            --app-hax-site-title-font-size: var(--ddd-font-size-5xs, 12px);
            --app-hax-site-date-font-size: var(--ddd-font-size-5xs, 11px);
            --app-hax-site-card-content-padding: var(--ddd-spacing-1, 4px);
          }

          .titleBar {
            margin-bottom: var(--ddd-spacing-1, 4px);
          }
        }
      `]}__clickButton(){this.opened=!this.opened}render(){return e`
      <div id="mainCard">
        <div class="cardImage">
        ${this.image?e`
              <a
                class="imageLink"
                href="${this.siteUrl}"
                aria-label="Open ${this.title||"site"}"
              >
                <img
                  src="${this.image}"
                  alt="Screenshot of ${this.title||"site"} theme"
                  loading="lazy"
                  decoding="async"
                  fetchpriority="low"
                />
              </a>
            `:""}

        </div>
        <div class="cardContent">
          <div class="titleBar">
            <slot name="heading"></slot>

            <div class="more-options">
              <simple-icon-button-lite
                id="${this.moreOptionsId}"
                icon="lrn:more-vert"
                @click="${this.toggleOptionsMenu}"
                aria-label="Open options"
                aria-haspopup="menu"
                title="Options"
              ></simple-icon-button-lite>
              <simple-context-menu title="Options">
                <simple-toolbar-button
                  icon="content-copy"
                  icon-position="left"
                  align-horizontal="left"
                  show-text-label
                  label="Copy"
                  @click=${this.copySite}
                ></simple-toolbar-button>
                <simple-toolbar-button
                  icon="file-download"
                  icon-position="left"
                  align-horizontal="left"
                  show-text-label
                  label="Download"
                  @click=${this.downloadSite}
                ></simple-toolbar-button>
                <simple-toolbar-button
                  icon="icons:add-circle"
                  icon-position="left"
                  align-horizontal="left"
                  show-text-label
                  label="Create Template"
                  @click=${this.createTemplate}
                ></simple-toolbar-button>
                <simple-toolbar-button
                  icon="archive"
                  icon-position="left"
                  align-horizontal="left"
                  show-text-label
                  label="Archive"
                  @click=${this.archiveSite}
                ></simple-toolbar-button>
                ${s.appSettings&&s.appSettings.haxiamAddUserAccess?e`
                      <simple-toolbar-button
                        icon="account-circle"
                        icon-position="left"
                        align-horizontal="left"
                        show-text-label
                        label="User Access"
                        @click=${this.openUserAccess}
                      ></simple-toolbar-button>
                    `:""}
              </simple-context-menu>
            </div>
          </div>

          <div class="date">
            <simple-icon icon="hax:calendar" title="Last updated"></simple-icon>
            <simple-datetime
                title="Last updated"
                format="m/j/y"
                .timestamp="${this.lastUpdatedTime}"
                unix
            ></simple-datetime>
          </div>
        </div>
      </div>
    `}}customElements.define(AppHaxSiteBars.tag,AppHaxSiteBars);
