// @haxtheweb/app-hax/lib/v2/app-hax-use-case-filter.js
import{LitElement as e,html as t,css as i}from"../../../../lit/index.js";import"../../../simple-tooltip/simple-tooltip.js";import"../../../simple-icon/lib/simple-icons.js";import"../../../simple-icon/lib/simple-icon-lite.js";import{store as r}from"./AppHaxStore.js";import"./app-hax-use-case.js";import"./app-hax-search-results.js";import"./app-hax-filter-tag.js";import"./app-hax-scroll-button.js";import"./app-hax-site-creation-modal.js";export class AppHaxUseCaseFilter extends e{static get tag(){return"app-hax-use-case-filter"}constructor(){super(),this.searchTerm="",this.disabled=!1,this.showSearch=!1,this.items=[],this.filteredItems=[],this.filteredSites=[],this.activeFilters=[],this.filters=[],this.searchQuery="",this.demoLink="",this.errorMessage="",this.loading=!1,this.selectedCardIndex=null,this.returningSites=[],this.allFilters=new Set,this.dark=!1,void 0!==r&&import("../../../../mobx/dist/mobx.esm.js").then((({autorun:e,toJS:t})=>{e((()=>{this.dark=t(r.darkMode)})),e((()=>{const e=t(r.manifest);e&&e.items&&e.items.length>0&&this.updateSiteResults()}))}))}static get properties(){return{searchTerm:{type:String},showSearch:{type:Boolean,reflect:!0,attribute:"show-search"},showFilter:{type:Boolean,reflect:!0,attribute:"show-filter"},disabled:{type:Boolean,reflect:!0},items:{type:Array},filteredItems:{type:Array},filteredSites:{type:Array},activeFilters:{type:Array},filters:{type:Array},searchQuery:{type:String},demoLink:{type:String},errorMessage:{type:String},loading:{type:Boolean},selectedCardIndex:{type:Number},returningSites:{type:Array},allFilters:{attribute:!1},dark:{type:Boolean,reflect:!0}}}static get styles(){return[i`
        :host {
          overflow: hidden;
          display: block;
          max-width: 100%;
          font-family: var(--ddd-font-primary, sans-serif);
        }
        .contentSection {
          display: flex;
          align-items: flex-start;
          justify-content: flex-start;
          gap: var(--ddd-spacing-4, 16px);
          width: 100%;
          margin: 0 var(--ddd-spacing-4, 16px);
          padding-right: var(--ddd-spacing-4, 16px);
          box-sizing: border-box;
        }
        .leftSection,
        .rightSection {
          display: flex;
          flex-direction: column;
          flex: 1 1 0;
        }
        .leftSection {
          width: 340px;
          min-width: 260px;
          max-width: 380px;
          margin-left: 0;
          margin-right: var(--ddd-spacing-1, 4px);
          padding-top: 0;
          box-sizing: border-box;
          position: sticky;
        }
        .rightSection {
          flex: 1;
          width: calc(100vw - 420px);
          box-sizing: border-box;
          display: flex;
          flex-direction: column;
        }
        .template-results {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(240px, 1fr));
          width: 100%;
          min-height: 330px;
          box-sizing: border-box;
          gap: var(--ddd-spacing-2, 8px);
        }
        #returnToSection {
          width: 100%;
        }
        #returnToSection app-hax-search-results {
          display: flex;
          gap: var(--ddd-spacing-6, 24px);
          min-width: calc(3 * 264px + 2 * var(--ddd-spacing-6, 24px));
          min-height: 120px;
          box-sizing: border-box;
          justify-content: flex-start;
          align-items: stretch;
          flex-direction: row;
          flex-wrap: nowrap;
          scroll-behavior: smooth;
          overflow: hidden;
          height: 300px;
        }
        :host(:not([show-filter])) app-hax-search-results {
          width: 100%;
        }

        h4,
        .returnTo h4,
        .startNew h4 {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-l, 24px);
          color: var(--app-hax-accent-color, var(--accent-color));
          margin: 0 0 var(--ddd-spacing-4, 16px) 0;
        }
        .startNew,
        .returnTo {
          padding-top: var(--ddd-spacing-2, 8px);
          position: relative;
          display: flex;
          flex-direction: column;
          justify-content: flex-start;
          align-items: flex-start;
          margin-left: 0;
          margin-right: 0;
        }
        .upper-filter {
          margin-bottom: var(--ddd-spacing-4, 16px);
          position: relative;
          display: inline-block;
          width: 100%;
        }
        input[type="text"] {
          width: 100%;
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-2, 8px)
            var(--ddd-spacing-2, 8px) var(--ddd-spacing-8, 32px);
          font-size: var(--ddd-font-size-xs, 12px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid);
          border-color: var(--ddd-theme-default-slateGray, #666);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-coalyGray, #222);
          transition: all 0.2s ease;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          margin: 0;
          min-height: var(--ddd-spacing-8, 32px);
        }
        :host([dark]) input[type="text"],
        body.dark-mode input[type="text"] {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        input[type="text"]:focus {
          border: var(--ddd-border-md, 2px solid);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          background: var(--ddd-theme-default-white, white);
          outline: none;
        }
        :host([dark]) input[type="text"]:focus,
        body.dark-mode input[type="text"]:focus {
          background: var(--ddd-theme-default-coalyGray, #333);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
        }
        .search-icon {
          position: absolute;
          left: var(--ddd-spacing-2, 8px);
          top: 50%;
          transform: translateY(-50%);
          font-size: var(--ddd-font-size-xs, 14px);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          pointer-events: none;
          z-index: 1;
          --simple-icon-width: var(--ddd-icon-3xs, 20px);
          --simple-icon-height: var(--ddd-icon-3xs, 20px);
        }
        :host([dark]) .search-icon,
        body.dark-mode .search-icon {
          color: var(--ddd-theme-default-white, white);
        }
        .filter {
          position: sticky;
          top: 0;
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-5, 20px);
          background: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-lg, 12px);
          box-shadow: var(--ddd-boxShadow-lg);
          border: var(--ddd-border-xs, 1px solid);
          border-color: var(--ddd-theme-default-slateGray, #666);
          padding: var(--ddd-spacing-6, 24px);
          margin-top: 0;
          margin-bottom: 0;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          transition: box-shadow 0.2s ease;
        }
        :host([dark]) .filter,
        body.dark-mode .filter {
          background: var(--ddd-theme-default-coalyGray, #222);
          border-color: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
        }
        .filter:hover {
          box-shadow: var(--ddd-boxShadow-xl);
        }
        .filterButtons {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-3, 12px);
          margin-top: 0;
          width: 100%;
        }
        .filter-btn {
          display: flex;
          align-items: center;
          justify-content: flex-start;
          gap: var(--ddd-spacing-1, 4px);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid) transparent;
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          cursor: pointer;
          box-shadow: var(--ddd-boxShadow-sm);
          transition: all 0.2s ease;
          outline: none;
          min-height: var(--ddd-spacing-7, 28px);
          text-align: left;
        }
        :host([dark]) .filter-btn,
        body.dark-mode .filter-btn {
          background: var(--ddd-theme-default-slateGray, #444);
          color: var(--ddd-theme-default-white, white);
        }
        .filter-btn.active,
        .filter-btn:active {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-keystoneYellow, #ffd100);
          box-shadow: var(--ddd-boxShadow-md);
        }
        :host([dark]) .filter-btn.active,
        :host([dark]) .filter-btn:active,
        body.dark-mode .filter-btn.active,
        body.dark-mode .filter-btn:active {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          border-color: var(--ddd-theme-default-white, white);
        }
        .filter-btn:hover {
          background: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
          transform: translateY(-1px);
        }
        :host([dark]) .filter-btn:hover,
        body.dark-mode .filter-btn:hover {
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }
        .filter-btn .icon {
          font-size: var(--ddd-font-size-3xs, 12px);
          color: inherit;
          display: flex;
          align-items: center;
          flex-shrink: 0;
          width: var(--ddd-icon-3xs, 20px);
          height: var(--ddd-icon-3xs, 20px);
        }
        .filter-btn .icon simple-icon-lite {
          color: inherit;
          --simple-icon-width: var(--ddd-icon-3xs, 20px);
          --simple-icon-height: var(--ddd-icon-3xs, 20px);
        }
        .filter-btn.active .icon,
        .filter-btn.active .icon simple-icon-lite {
          color: inherit;
        }
        :host([dark]) .filter-btn.active .icon,
        :host([dark]) .filter-btn.active .icon simple-icon-lite,
        body.dark-mode .filter-btn.active .icon,
        body.dark-mode .filter-btn.active .icon simple-icon-lite {
          color: inherit;
        }
        .filter-btn:hover .icon simple-icon-lite,
        .filter-btn:focus .icon simple-icon-lite {
          color: inherit;
        }
        input[type="checkbox"] {
          display: none;
        }
        .reset-button {
          margin-top: var(--ddd-spacing-1, 4px);
          background: var(--ddd-theme-default-original87Pink, #e4007c);
          border: var(--ddd-border-xs, 1px solid) transparent;
          color: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-sm, 4px);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--ddd-spacing-1, 4px);
          box-shadow: var(--ddd-boxShadow-sm);
          cursor: pointer;
          transition: all 0.2s ease;
          min-height: var(--ddd-spacing-7, 28px);
        }
        .reset-button:hover {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
          transform: translateY(-1px);
        }
        :host([dark]) .reset-button,
        body.dark-mode .reset-button {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
        }
        :host([dark]) .reset-button:hover,
        body.dark-mode .reset-button:hover {
          background: var(--ddd-theme-default-original87Pink, #e4007c);
        }
        .collapseFilter {
          display: none;
        }

        @media (max-width: 780px) {
          .contentSection {
            display: block;
          }
          .leftSection {
            width: 100%;
            max-width: 100%;
            margin-bottom: var(--ddd-spacing-4, 16px);
            position: relative;
          }
          .rightSection {
            width: 100%;
          }
          :host([show-filter]) .filter {
            display: flex;
            width: 250px;
            max-width: 20vw;
          }
          :host .collapseFilter {
            display: flex;
          }
          h4,
          .returnTo h4,
          .startNew h4 {
            font-size: var(--ddd-font-size-m, 20px);
          }
          .template-results {
            grid-template-columns: repeat(auto-fit, minmax(280px, 1fr));
            gap: var(--ddd-spacing-3, 12px);
          }
        }

        @media (max-width: 600px) {
          .contentSection {
            display: block;
            margin: 0 var(--ddd-spacing-2, 8px);
            padding-right: var(--ddd-spacing-2, 8px);
          }
          .leftSection {
            width: 100%;
            max-width: 100%;
            margin-bottom: var(--ddd-spacing-3, 12px);
            position: relative;
          }
          .rightSection {
            width: 100%;
          }
          :host([show-filter]) .filter {
            display: flex;
            width: 200px;
            max-width: 20vw;
          }
          :host .collapseFilter {
            display: flex;
          }
          h4,
          .returnTo h4,
          .startNew h4 {
            font-size: var(--ddd-font-size-s, 18px);
          }
          .template-results {
            grid-template-columns: 1fr;
            gap: var(--ddd-spacing-2, 8px);
          }
        }

        @media (max-width: 480px) {
          .contentSection {
            margin: 0 var(--ddd-spacing-1, 4px);
            padding-right: var(--ddd-spacing-1, 4px);
          }
          h4,
          .returnTo h4,
          .startNew h4 {
            font-size: var(--ddd-font-size-s, 16px);
            margin: 0 0 var(--ddd-spacing-3, 12px) 0;
          }
          .template-results {
            grid-template-columns: 1fr;
            gap: var(--ddd-spacing-2, 8px);
          }
          #returnToSection app-hax-search-results {
            min-width: 100%;
          }
        }
      `]}testKeydown(e){"Escape"!==e.key&&"Enter"!==e.key||this.toggleSearch()}toggleFilterVisibility(){this.showFilter=!this.showFilter}render(){return t`
      <div class="contentSection">
        <div class="leftSection">
          <div class="filter">
            <!-- Search bar -->
            <div class="upper-filter">
              <slot>
                <simple-icon-lite
                  class="search-icon"
                  icon="icons:search"
                ></simple-icon-lite>
              </slot>
              <input
                id="searchField"
                @input="${this.handleSearch}"
                @keydown="${this.testKeydown}"
                type="text"
                placeholder="Search Sites & Templates"
              />
            </div>
            <!-- Filter Buttons -->
            <div class="filterButtons">
              ${this.filters.map(((e,i)=>t`
                  <input
                    type="checkbox"
                    id="filter-${i}"
                    class="filter-checkbox"
                    .checked=${this.activeFilters.includes(e)}
                    @change=${()=>this.toggleFilterByButton(e)}
                  />
                  <label
                    for="filter-${i}"
                    class="filter-btn ${this.activeFilters.includes(e)?"active":""}"
                    aria-pressed=${this.activeFilters.includes(e)}
                  >
                    <span class="icon">
                      <simple-icon-lite
                        icon="${this.iconForFilter(e)}"
                      ></simple-icon-lite>
                    </span>
                    ${e}
                  </label>
                `))}
            </div>
            <button class="reset-button" @click="${this.resetFilters}">
              Reset
            </button>
          </div>
        </div>
        <!-- Content Section -->
        <div class="rightSection">
          <!-- Returning Sites -->
          <div id="returnToSection" class="returnTo">
            <h4>Return to...</h4>
            <app-hax-search-results
              .displayItems=${this.filteredSites}
              .searchTerm=${this.searchTerm}
              ?dark="${this.dark}"
            >
            </app-hax-search-results>
          </div>

          <!-- Templates -->
          <div id="startJourneySection" class="startNew">
            <h4>Create New Site</h4>
            <div class="template-results">
              ${this.filteredItems.length>0?this.filteredItems.map(((e,i)=>t`
                      <div>
                        <a
                          href="${e.demoLink}"
                          target="_blank"
                          class="${i===this.activeUseCase?"active-card":""}"
                        ></a>
                        <app-hax-use-case
                          .source=${e.useCaseImage||""}
                          .title=${e.useCaseTitle||""}
                          .description=${e.useCaseDescription||""}
                          .demoLink=${e.demoLink||""}
                          .iconImage=${e.useCaseIcon||[]}
                          .isSelected=${e.isSelected||!1}
                          .showContinue=${e.showContinue||!1}
                          ?dark="${this.dark}"
                          @toggle-display=${e=>this.toggleDisplay(i,e)}
                          @continue-action=${()=>this.continueAction(i)}
                        ></app-hax-use-case>
                      </div>
                    `)):t`<p>No templates match the filters specified.</p>`}
            </div>
          </div>
        </div>
      </div>

      <!-- Site Creation Modal -->
      <app-hax-site-creation-modal
        id="siteCreationModal"
        @modal-closed=${this.handleModalClosed}
      ></app-hax-site-creation-modal>
    `}iconForFilter(e){switch(e.toLowerCase()){case"blog":return"lrn:write";case"brochure":return"icons:description";case"course":return"hax:lesson";case"portfolio":return"icons:perm-identity";default:return"icons:label"}}toggleFilterByButton(e){this.activeFilters.includes(e)?this.activeFilters=this.activeFilters.filter((t=>t!==e)):this.activeFilters=[...this.activeFilters,e],this.applyFilters(),this.requestUpdate()}firstUpdated(){super.firstUpdated(),this.updateRecipeResults(),this.updateSiteResults()}updated(e){(e.has("searchQuery")||e.has("activeFilters")||e.has("items"))&&this.applyFilters()}toggleSearch(){this.disabled||(this.shadowRoot.querySelector("#searchField").value="",r.appEl.playSound("click"),this.showSearch=!this.showSearch,setTimeout((()=>{this.shadowRoot.querySelector("#searchField").focus()}),300))}toggleSelection(e){this.activeUseCase===e?this.activeUseCase=!1:this.activeUseCase=e,this.requestUpdate()}handleSearch(e){const t=e.target.value.toLowerCase();this.searchTerm=t,r.searchTerm=t,this.filteredItems=[...this.items.filter((e=>"recipe"===e.dataType&&(e.useCaseTitle.toLowerCase().includes(t)||e.useCaseTag&&e.useCaseTag.some((e=>e.toLowerCase().includes(t))))))],this.filteredSites=[...this.items.filter((e=>"site"===e.dataType&&(e.originalData.title&&e.originalData.title.toLowerCase().includes(t)||e.useCaseTag&&e.useCaseTag.some((e=>e.toLowerCase().includes(t))))))],this.requestUpdate()}toggleFilter(e){const t=e.target.value;this.activeFilters.includes(t)?this.activeFilters=[...this.activeFilters.filter((e=>e!==t))]:this.activeFilters=[...this.activeFilters,t],this.applyFilters()}applyFilters(){const e=this.searchTerm.toLowerCase();this.filteredItems=[...this.items.filter((t=>{if("recipe"!==t.dataType)return!1;const i=""===e||t.useCaseTitle.toLowerCase().includes(e)||t.useCaseTag&&t.useCaseTag.some((t=>t.toLowerCase().includes(e))),r=0===this.activeFilters.length||t.useCaseTag&&this.activeFilters.some((e=>t.useCaseTag.includes(e)));return i&&r}))],this.filteredSites=[...this.returningSites.filter((t=>{if("site"!==t.dataType)return!1;const i=t.originalData.metadata?.site?.category||[],r=""===e||t.originalData.category&&t.originalData.category&&t.originalData.category.includes(e)||t.useCaseTag&&t.useCaseTag.some((t=>t.toLowerCase().includes(e))),a=0===this.activeFilters.length||this.activeFilters.some((e=>i.includes(e)));return r&&a}))]}removeFilter(e){const t=e.detail;this.activeFilters=this.activeFilters.filter((e=>e!==t)),this.applyFilters(),this.requestUpdate()}resetFilters(){this.searchTerm="",r.searchTerm="",this.activeFilters=[],this.filteredItems=[...this.items.filter((e=>"recipe"===e.dataType))],this.filteredSites=[...this.returningSites],this.shadowRoot.querySelector("#searchField").value="",this.shadowRoot.querySelectorAll('input[type="checkbox"]').forEach((e=>e.checked=!1)),this.requestUpdate()}updateRecipeResults(){this.loading=!0,this.errorMessage="";const e=new URL("./app-hax-recipes.json",import.meta.url).href;fetch(e).then((e=>{if(!e.ok)throw new Error(`Failed Recipes (${e.status})`);return e.json()})).then((e=>{const t=Array.isArray(e.item)?e.item.map((e=>{let t=[];Array.isArray(e.category)?t=e.category.filter((e=>"string"==typeof e&&""!==e.trim())):"string"==typeof e.category&&""!==e.category.trim()&&(t=[e.category.trim()]),0===t.length&&(t=["Empty"]),t.forEach((e=>this.allFilters.add(e)));const i=Array.isArray(e.attributes)?e.attributes.map((e=>({icon:e.icon||"",tooltip:e.tooltip||""}))):[];return{dataType:"recipe",useCaseTitle:e.title||"Untitled Template",useCaseImage:e.image||"",useCaseDescription:e.description||"",useCaseIcon:i,useCaseTag:t,demoLink:e["demo-url"]||"#",originalData:e}})):[];this.items=t,this.filters=Array.from(this.allFilters).sort(),0!==this.items.length||this.errorMessage||(this.errorMessage="No Recipes Found"),this.resetFilters()})).catch((e=>{this.errorMessage=`Failed to load data: ${e.message}`,this.items=[],this.filters=[]})).finally((()=>{this.loading=!1}))}updateSiteResults(){this.loading=!0,this.errorMessage="";try{const e=r.manifest;if(!e||!e.items)throw new Error("No manifest data available");const t=Array.isArray(e.items)?e.items.map((e=>{let t=e.metadata&&e.metadata.site?e.metadata.site.category:null,i=[];return Array.isArray(t)?i=t.filter((e=>"string"==typeof e&&""!==e.trim())):"string"==typeof t&&""!==t.trim()&&(i=[t.trim()]),0===i.length&&(i=["Empty"]),i.forEach((e=>this.allFilters.add(e))),{dataType:"site",useCaseTag:i,originalData:e,...e}})):[];this.returningSites=[...t],this.filters=Array.from(this.allFilters).sort(),this.filteredSites=[...t],0!==t.length||this.errorMessage||(this.errorMessage="No Sites Found"),this.requestUpdate(),this.loading=!1}catch(e){this.errorMessage=`Failed to load data: ${e.message}`,this.returningSites=[],this.filteredSites=[],this.filters=[],this.loading=!1}}toggleDisplay(e,t){const i=t.detail.isSelected;null!==this.selectedCardIndex&&this.selectedCardIndex!==e&&(this.filteredItems[this.selectedCardIndex].isSelected=!1,this.filteredItems[this.selectedCardIndex].showContinue=!1),this.selectedCardIndex=i?e:null,this.filteredItems[e].isSelected=i,this.filteredItems[e].showContinue=i,this.requestUpdate()}continueAction(e){const t=this.filteredItems[e],i=this.shadowRoot.querySelector("#siteCreationModal");i&&t&&(i.title=t.useCaseTitle,i.description=t.useCaseDescription,i.source=t.useCaseImage,i.template=t.useCaseTitle,i.openModal())}handleModalClosed(e){e.detail&&e.detail.cancelled&&null!==this.selectedCardIndex&&this.filteredItems[this.selectedCardIndex]&&(this.filteredItems[this.selectedCardIndex].isSelected=!1,this.filteredItems[this.selectedCardIndex].showContinue=!1,this.selectedCardIndex=null,this.requestUpdate()),console.log("Site creation modal closed",e.detail)}}customElements.define("app-hax-use-case-filter",AppHaxUseCaseFilter);
