// @haxtheweb/app-hax/lib/v2/app-hax-site-bar.js
import{html as e,css as t,unsafeCSS as i}from"../../../../lit/index.js";import"../../../simple-icon/lib/simple-icons.js";import"../../../simple-icon/lib/simple-icon-button-lite.js";import{SimpleColors as o}from"../../../simple-colors/simple-colors.js";import"../../../simple-tooltip/simple-tooltip.js";import{toJS as d}from"../../../../mobx/dist/mobx.esm.js";import{store as a}from"./AppHaxStore.js";import"./app-hax-user-access-modal.js";import"./app-hax-confirmation-modal.js";new URL("../assets/images/DropDownBorder.svg",import.meta.url);export class AppHaxSiteBars extends o{static get tag(){return"app-hax-site-bar"}constructor(){super(),this.showOptions=!1,this.inprogress=!1,this.textInfo={},this.siteId="",this.description="",this._boundKeydownHandler=this._handleMenuKeydown.bind(this)}disconnectedCallback(){super.disconnectedCallback(),document.removeEventListener("keydown",this._boundKeydownHandler)}static get properties(){return{...super.properties,showOptions:{type:Boolean},inprogress:{type:Boolean,reflect:!0},textInfo:{type:Object},siteId:{type:String,reflect:!0,attribute:"site-id"},title:{type:String},description:{type:String},siteUrl:{type:String,attribute:"site-url"},image:{type:String}}}updated(e){e.has("showOptions")&&(this.showOptions?(document.addEventListener("keydown",this._boundKeydownHandler),this._focusFirstMenuItem()):document.removeEventListener("keydown",this._boundKeydownHandler))}toggleOptionsMenu(){const e=this.showOptions;this.showOptions=!this.showOptions,e&&!this.showOptions&&this._returnFocusToSettingsButton()}closeOptionsMenu(){this.showOptions=!1,this._returnFocusToSettingsButton()}handleKeydown(e,t){"Enter"!==e.key&&" "!==e.key||(e.preventDefault(),t.call(this))}_focusFirstMenuItem(){setTimeout((()=>{const e=this.shadowRoot.querySelector(".close-menu-btn");if(e)e.focus();else{const e=this.shadowRoot.querySelector(".menu-item");e&&e.focus()}}),0)}_returnFocusToSettingsButton(){setTimeout((()=>{const e=this.shadowRoot.querySelector("#settingsIcon");e&&e.focus()}),0)}_handleMenuKeydown(e){if(!this.showOptions)return;const t=Array.from(this.shadowRoot.querySelectorAll(".close-menu-btn, .menu-item"));switch(e.key){case"Escape":e.preventDefault(),this.closeOptionsMenu();break;case"ArrowDown":case"ArrowUp":const i=t.indexOf(e.target);if(-1!==i)if(e.preventDefault(),"ArrowDown"===e.key){t[i<t.length-1?i+1:0].focus()}else{t[i>0?i-1:t.length-1].focus()}break;case"Tab":e.preventDefault();let o=-1;for(let e=0;e<t.length;e++)if(t[e]===document.activeElement||t[e]===this.shadowRoot.activeElement){o=e;break}if(-1===o){t[0].focus();break}if(e.shiftKey){t[o>0?o-1:t.length-1].focus()}else{t[o<t.length-1?o+1:0].focus()}}}copySite(){this.showOptions=!1,this.siteOperation("copySite","Copy","icons:content-copy")}downloadSite(){this.showOptions=!1,this.siteOperation("downloadSite","Download","file-download")}archiveSite(){this.showOptions=!1,this.siteOperation("archiveSite","Archive","icons:archive")}openUserAccess(){console.log("User Access clicked"),this.showOptions=!1,a.activeSiteId=this.siteId,import("../../../simple-modal/simple-modal.js").then((()=>{const e=document.createElement("app-hax-user-access-modal");this.title&&(e.siteTitle=this.title);const t=new CustomEvent("simple-modal-show",{bubbles:!0,composed:!0,cancelable:!0,detail:{title:"User Access",elements:{content:e},invokedBy:this,styles:{"--simple-modal-titlebar-background":"var(--ddd-theme-default-nittanyNavy, #001e44)","--simple-modal-titlebar-color":"light-dark(black, white)","--simple-modal-width":"90vw","--simple-modal-max-width":"var(--ddd-spacing-32, 480px)","--simple-modal-min-width":"300px","--simple-modal-z-index":"1000","--simple-modal-height":"auto","--simple-modal-min-height":"300px","--simple-modal-max-height":"80vh","--simple-modal-titlebar-height":"80px","--simple-modal-border-radius":"var(--ddd-radius-md, 8px)","--simple-modal-background":"var(--ddd-theme-default-white, white)","--simple-modal-box-shadow":"var(--ddd-boxShadow-xl)"}}});this.dispatchEvent(t)}))}siteOperation(e,t,i){a.appEl&&a.appEl.playSound&&a.appEl.playSound("click"),a.activeSiteOp=e,a.activeSiteId=this.siteId;const o=d(a.manifest.items.filter((e=>e.id===this.siteId)).pop());if(!o)return void console.error("Site not found for ID:",this.siteId);const s=document.createElement("app-hax-confirmation-modal");s.title=`${t} ${o.metadata.site.name}?`,s.message=`Are you sure you want to ${e.replace("Site","")} ${o.metadata.site.name}?`,s.confirmText="Confirm",s.cancelText="Cancel",s.dangerous="archiveSite"===e,s.confirmAction=this.confirmOperation.bind(this),s.cancelAction=this.cancelOperation.bind(this),document.body.appendChild(s),s.openModal(),s.addEventListener("close",(()=>{document.body.removeChild(s)}),{once:!0})}cancelOperation(){a.activeSiteOp="",a.activeSiteId=null,globalThis.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0})),a.appEl&&a.appEl.playSound&&a.appEl.playSound("error")}async confirmOperation(){const e=d(a.activeSiteOp),t=d(a.activeSite);t?(await a.AppHaxAPI.makeCall(e,{site:{name:t.metadata.site.name,id:t.id}},!0,(()=>{"downloadSite"===d(a.activeSiteOp)?globalThis.open(a.AppHaxAPI.lastResponse.downloadSite.data.link,"_blank"):a.refreshSiteListing()})),globalThis.dispatchEvent(new CustomEvent("simple-modal-hide",{bubbles:!0,composed:!0})),a.appEl&&a.appEl.playSound&&a.appEl.playSound("success"),a.toast(`${t.metadata.site.name} ${e.replace("Site","")} successful!`,3e3,{hat:"random"})):console.error("No active site found for operation:",e)}static get styles(){return[super.styles,t`
        :host {
          text-align: left;
          width: 264px;
          max-width: 264px;
          font-family: var(--ddd-font-primary);
          color: light-dark(
            var(--ddd-theme-default-nittanyNavy),
            var(--ddd-theme-default-white)
          );
          background-color: light-dark(
            white,
            var(--ddd-theme-default-coalyGray, #222)
          );
          border: var(--ddd-border-sm);
          border-color: light-dark(
            var(--ddd-theme-default-slateGray, #c4c4c4),
            var(--ddd-theme-default-slateGray, #666)
          );
          min-height: 260px;
          box-shadow: light-dark(
            2px 2px 12px #1c1c1c,
            2px 2px 12px rgba(0, 0, 0, 0.3)
          );
          border-radius: var(--ddd-radius-sm, 4px);
          transition: all 0.2s ease;
        }
        :host(:hover),
        :host(:focus-within) {
          transform: translateY(-2px);
          border-color: light-dark(
            var(--ddd-theme-default-keystoneYellow, #ffd100),
            var(--ddd-theme-default-keystoneYellow, #ffd100)
          );
          box-shadow: light-dark(
            4px 8px 24px rgba(28, 28, 28, 0.15),
            4px 8px 24px rgba(0, 0, 0, 0.5)
          );
        }
        #mainCard {
          display: flex;
          flex-direction: column;
        }
        .cardContent {
          padding: 12px 16px 20px;
        }
        .imageLink img {
          width: 100%;
          height: 125px;
          object-fit: cover;
          border-top-right-radius: 8px;
          border-top-left-radius: 8px;
          border-bottom: solid var(--ddd-theme-default-nittanyNavy) 8px;
          overflow: clip;
          justify-self: center;
        }
        .imageLink {
          position: relative;
          display: block;
          width: 100%;
        }
        .settings-button {
          position: relative;
          display: inline-block;
          align-self: center;
        }

        .options-menu {
          position: absolute;
          top: -125px;
          right: 0;
          background: var(--ddd-theme-default-white, white);
          border: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-slateGray, #666);
          box-shadow: var(--ddd-boxShadow-lg);
          border-radius: var(--ddd-radius-md, 8px);
          padding: var(--ddd-spacing-2, 8px);
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-1, 4px);
          z-index: 1000;
          overflow: visible;
          min-width: 140px;
        }
        :host([dark]) .options-menu,
        body.dark-mode .options-menu {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        .close-menu-btn {
          z-index: 10;
          position: absolute;
          top: var(--ddd-spacing-1, 4px);
          right: var(--ddd-spacing-1, 4px);
          background: transparent;
          border: none;
          color: var(--ddd-theme-default-slateGray, #666);
          cursor: pointer;
          font-size: var(--ddd-font-size-s, 16px);
          font-weight: var(--ddd-font-weight-bold, 700);
          padding: var(--ddd-spacing-1, 4px);
          border-radius: var(--ddd-radius-xs, 2px);
          line-height: 1;
          min-height: auto;
          box-shadow: none;
          width: var(--ddd-spacing-5, 20px);
          height: var(--ddd-spacing-5, 20px);
          display: flex;
          align-items: center;
          justify-content: center;
          transition: all 0.2s ease;
        }
        :host([dark]) .close-menu-btn,
        body.dark-mode .close-menu-btn {
          color: var(--ddd-theme-default-white, white);
        }
        .close-menu-btn:hover,
        .close-menu-btn:focus {
          background: var(--ddd-theme-default-slateGray, #666);
          color: var(--ddd-theme-default-white, white);
          transform: none;
          box-shadow: none;
          outline: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
        }
        :host([dark]) .close-menu-btn:hover,
        :host([dark]) .close-menu-btn:focus,
        body.dark-mode .close-menu-btn:hover,
        body.dark-mode .close-menu-btn:focus {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          outline-color: var(--ddd-theme-default-white, white);
        }
        .menu-item {
          display: flex;
          align-items: center;
          padding: var(--ddd-spacing-1, 4px) var(--ddd-spacing-2, 8px);
          border-radius: var(--ddd-radius-xs, 2px);
          cursor: pointer;
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          transition: all 0.2s ease;
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          text-decoration: none;
          min-height: var(--ddd-spacing-5, 20px);
          background: transparent;
          border: none;
          width: 100%;
          text-align: left;
          gap: var(--ddd-spacing-1, 4px);
        }
        :host([dark]) .menu-item,
        body.dark-mode .menu-item {
          color: var(--ddd-theme-default-white, white);
        }
        .menu-item:hover,
        .menu-item:focus {
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          outline: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
        }
        :host([dark]) .menu-item:hover,
        :host([dark]) .menu-item:focus,
        body.dark-mode .menu-item:hover,
        body.dark-mode .menu-item:focus {
          background: var(--ddd-theme-default-slateGray, #555);
          color: var(--ddd-theme-default-white, white);
        }
        .menu-item simple-icon-lite {
          flex-shrink: 0;
          width: var(--ddd-icon-3xs, 16px);
          height: var(--ddd-icon-3xs, 16px);
        }

        .titleBar {
          display: flex;
          justify-content: space-between;
          align-items: center;
          font-size: var(--ddd-font-size-xs, 14px);
          color: light-dark(
            var(--ddd-theme-default-nittanyNavy),
            var(--ddd-theme-default-white)
          );
        }
        p {
          font-size: var(--ddd-font-size-4xs, 12px);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-2, 8px)
            var(--ddd-spacing-1, 6px) var(--ddd-spacing-2, 10px);
          margin: 0;
          line-height: 1.4;
        }
        ::slotted([slot="heading"]) {
          font-size: var(--ddd-font-size-xs, 14px);
          font-weight: var(--ddd-font-weight-bold, 700);
          color: light-dark(
            var(--ddd-theme-default-nittanyNavy),
            var(--ddd-theme-default-white)
          );
          text-decoration: none;
          display: block;
          margin: 0;
          line-height: 1.2;
        }
        button {
          display: flex;
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
          border: var(--ddd-border-xs, 1px solid) transparent;
          border-radius: var(--ddd-radius-sm, 4px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-weight: var(--ddd-font-weight-medium, 500);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          min-height: var(--ddd-spacing-7, 28px);
          align-items: center;
          justify-content: start;
          cursor: pointer;
          transition: all 0.2s ease;
          box-shadow: var(--ddd-boxShadow-sm);
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }
        button:hover {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          transform: translateY(-1px);
          box-shadow: var(--ddd-boxShadow-md);
        }
        .cardBottom {
          display: flex;
          justify-content: space-between;
          align-items: center;
          margin-top: 6px;
          padding: 0px 12px 16px 12px;
          gap: 4px;
        }
        .cardBottom button {
          flex: 1;
          margin: 0 2px;
          min-width: 0;
        }
        ::slotted(a[slot="heading"]),
        ::slotted(span[slot="subHeading"]),
        ::slotted(div[slot="pre"]) {
          display: block;
        }
      `]}__clickButton(){this.opened=!this.opened}render(){return e`
      <div id="mainCard">
        ${this.image?e`
              <div class="imageLink">
                <img
                  src="${this.image}"
                  alt="Screenshot of ${this.title||"site"} theme"
                  loading="lazy"
                />
              </div>
            `:""}

        <div class="cardContent">
          <div class="titleBar">
            <slot name="heading"></slot>
            <div class="settings-button">
              <simple-tooltip
                for="settingsIcon"
                position="top"
                animation-delay="0"
              >
                Options
              </simple-tooltip>
              <simple-icon-button-lite
                id="settingsIcon"
                icon="hax:settings"
                @click="${this.toggleOptionsMenu}"
                aria-label="Open options"
                aria-expanded="${this.showOptions}"
                aria-haspopup="menu"
              ></simple-icon-button-lite>

              ${this.showOptions?e`
                    <div
                      id="options-menu-${this.siteId}"
                      class="options-menu"
                      role="menu"
                      aria-labelledby="settingsIcon"
                    >
                      <button
                        class="close-menu-btn"
                        @click="${this.toggleOptionsMenu}"
                        aria-label="Close menu"
                      >
                        ×
                      </button>
                      <button
                        class="menu-item"
                        @click="${this.copySite}"
                        role="menuitem"
                        aria-label="Copy site"
                      >
                        <simple-icon-lite
                          icon="content-copy"
                        ></simple-icon-lite>
                        Copy
                      </button>
                      <button
                        class="menu-item"
                        @click="${this.downloadSite}"
                        role="menuitem"
                        aria-label="Download site"
                      >
                        <simple-icon-lite
                          icon="file-download"
                        ></simple-icon-lite>
                        Download
                      </button>
                      <button
                        class="menu-item"
                        @click="${this.archiveSite}"
                        role="menuitem"
                        aria-label="Archive site"
                      >
                        <simple-icon-lite icon="archive"></simple-icon-lite>
                        Archive
                      </button>
                      ${a.appSettings&&a.appSettings.haxiamAddUserAccess?e`
                            <button
                              class="menu-item"
                              @click="${this.openUserAccess}"
                              role="menuitem"
                              aria-label="Manage user access"
                            >
                              <simple-icon-lite
                                icon="account-circle"
                              ></simple-icon-lite>
                              User Access
                            </button>
                          `:""}
                    </div>
                  `:""}
            </div>
          </div>

          <slot name="pre"></slot>

          <div class="cardBottom">
            <button
              class="open"
              @click=${()=>window.open(this.siteUrl,"_blank")}
            >
              Open
            </button>
          </div>
        </div>
      </div>
    `}}customElements.define(AppHaxSiteBars.tag,AppHaxSiteBars);
