// @haxtheweb/app-hax/lib/v2/app-hax-user-menu-button.js
import{LitElement as t,html as e,css as i}from"../../../../lit/index.js";import{DDDSuper as n}from"../../../d-d-d/d-d-d.js";import"../../../simple-icon/lib/simple-icon-lite.js";export class AppHaxUserMenuButton extends(n(t)){static get tag(){return"app-hax-user-menu-button"}constructor(){super(),this.icon="account-circle",this.label="Menu Item"}handleClick(t){const e=this.parentElement;e&&"a"===e.tagName.toLowerCase()&&(t.stopPropagation(),e.click())}static get properties(){return{icon:{type:String},label:{type:String}}}static get styles(){return[super.styles,i`
        :host {
          font-family: var(--ddd-font-primary, sans-serif);
          text-align: center;
          width: 100%;
          display: block;
        }

        .menu-button {
          display: flex;
          text-decoration: none;
          align-items: center;
          width: 100%;
          border: none;
          margin: 0;
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          font-size: var(--ddd-font-size-6xs, 12px);
          text-align: left;
          color: light-dark(var(--ddd-theme-default-coalyGray, #222), var(--ddd-theme-default-white, white));
          background: transparent;
          cursor: pointer;
          font-family: var(--ddd-font-primary, sans-serif);
          transition: all 0.2s ease;
          min-height: var(--ddd-spacing-8, 32px);
          box-sizing: border-box;
        }

        .icon {
          padding-right: var(--ddd-spacing-2, 8px);
          flex-shrink: 0;
          display: flex;
          align-items: center;
          --simple-icon-width: var(--ddd-icon-xs, 24px);
          --simple-icon-height: var(--ddd-icon-xs, 24px);
        }

        .label {
          flex: 1;
          text-align: left;
        }
      `]}render(){return e`
      <button
        class="menu-button"
        part="menu-button"
        @click="${this.handleClick}"
      >
        ${this.icon?e`<simple-icon-lite
              class="icon"
              icon="${this.icon}"
            ></simple-icon-lite>`:""}
        <span class="label">${this.label}</span>
      </button>
    `}}customElements.define(AppHaxUserMenuButton.tag,AppHaxUserMenuButton);
