// @haxtheweb/app-hax/lib/v2/app-hax-search-results.js
import{SimpleColors as t}from"../../../simple-colors/simple-colors.js";import{html as e,css as r}from"../../../../lit/index.js";import{autorun as s,toJS as a}from"../../../../mobx/dist/mobx.esm.js";import{varGet as o}from"../../../utils/utils.js";import{store as i}from"./AppHaxStore.js";import"./app-hax-site-bar.js";import"./app-hax-site-details.js";import"../../../simple-tooltip/simple-tooltip.js";export class AppHaxSearchResults extends t{static get tag(){return"app-hax-search-results"}constructor(){super(),this.searchItems=[],this.displayItems=[],this.searchTerm="",this.dark=!1,this.isPointerDown=!1,this.startX=0,this.scrollLeftVal=0,this.isDragging=!1,this.currentIndex=1,this.totalItems=0,this.sortOption="newest",s((()=>{this.searchTerm=a(i.searchTerm)})),s((()=>{this.dark=a(i.darkMode)})),s((()=>{const t=a(i.manifest);t&&t.items&&(this.searchItems=t.items,this.displayItems=[...this.searchItems],i.loadThemesData())}))}static get properties(){return{...super.properties,searchTerm:{type:String,reflect:!0},searchItems:{type:Array},displayItems:{type:Array},siteUrl:{type:String,attribute:"site-url"},isPointerDown:{type:Boolean},startX:{type:Number},scrollLeftVal:{type:Number},isDragging:{type:Boolean},currentIndex:{type:Number},totalItems:{type:Number},sortOption:{type:String,attribute:"sort-option"}}}static get styles(){return[super.styles,r`
        :host {
          display: block;
          width: 100%;
          overflow: visible;
        }

        .carousel-container {
          display: flex;
          align-items: center;
          justify-content: flex-start;
          gap: var(--ddd-spacing-6, 24px);
          width: 100%;
          max-width: 1350px;
          margin: 0 auto;
          position: relative;
        }
        .pager-container {
          display: flex;
          flex-direction: column;
          justify-content: center;
          align-items: center;
          gap: var(--ddd-spacing-2, 8px);
          margin-right: var(--ddd-spacing-4, 16px);
          padding: var(--ddd-spacing-1, 4px);
          flex-shrink: 0;
          scrollbar-width: none;
          -ms-overflow-style: none;
        }

        .pager-container::-webkit-scrollbar {
          display: none;
        }

        .scroll-left,
        .scroll-right {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--ddd-theme-default-white, white);
          border: var(--ddd-border-sm, 2px solid)
            white;
          border-radius: var(--ddd-radius-circle);
          padding: var(--ddd-spacing-4, 16px);
          cursor: pointer;
          height: var(--ddd-spacing-12);
          width: var(--ddd-spacing-12);
          display: flex;
          align-items: center;
          justify-content: center;
          font-size: var(--ddd-font-size-s, 20px);
          transition: all 0.2s ease;
          flex-shrink: 0;
          z-index: 10;
          position: relative;
        }

        :host([dark]) .scroll-left,
        :host([dark]) .scroll-right,
        body.dark-mode .scroll-left,
        body.dark-mode .scroll-right {
          background: var(--ddd-theme-default-accent);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        .scroll-left:hover:not(:disabled),
        .scroll-right:hover:not(:disabled) {
          background: var(--ddd-theme-default-accent);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          transform: translateY(-2px);
          box-shadow: var(--ddd-boxShadow-md);
        }

        .scroll-left:disabled,
        .scroll-right:disabled {
          opacity: 0.3;
          cursor: not-allowed;
          background: var(--ddd-theme-default-limestoneGray, #a2aaad);
          color: var(--ddd-theme-default-slateGray, #666);
          border-color: var(--ddd-theme-default-limestoneGray, #a2aaad);
          transform: none;
          box-shadow: none;
        }

        :host([dark]) .scroll-left:hover:not(:disabled),
        :host([dark]) .scroll-right:hover:not(:disabled),
        body.dark-mode .scroll-left:hover:not(:disabled),
        body.dark-mode .scroll-right:hover:not(:disabled) {
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        :host([dark]) .scroll-left:disabled,
        :host([dark]) .scroll-right:disabled,
        body.dark-mode .scroll-left:disabled,
        body.dark-mode .scroll-right:disabled {
          opacity: 0.3;
          cursor: not-allowed;
          background: var(--ddd-theme-default-limestoneGray, #a2aaad);
          color: var(--ddd-theme-default-coalyGray, #444);
        }
        
        #results {
          display: flex;
          flex-wrap: nowrap;
          overflow-x: auto;
          scroll-snap-type: x mandatory;
          gap: var(--ddd-spacing-6, 24px);
            padding: var(--ddd-spacing-2, 8px) 0 0 var(--ddd-spacing-2, 8px);
          flex: 1;
          min-width: 0;
          cursor: grab;
          user-select: none;
          /* Keep scrollbar visible for multiple interaction methods */
          scrollbar-width: thin;
          scrollbar-color: var(--ddd-theme-default-slateGray, #666)
            var(--ddd-theme-default-limestoneLight, #e4e5e7);
        }

        :host([dark]) #results,
        body.dark-mode #results {
          scrollbar-color: var(--ddd-theme-default-limestoneGray, #a2aaad)
            var(--ddd-theme-default-coalyGray, #444);
        }

        #results.dragging {
          cursor: grabbing;
          scroll-behavior: auto;
        }

        #results::-webkit-scrollbar {
          height: 8px;
        }

        #results::-webkit-scrollbar-track {
          background: var(--ddd-theme-default-limestoneLight, #e4e5e7);
          border-radius: var(--ddd-radius-xs, 2px);
        }

        #results::-webkit-scrollbar-thumb {
          background: var(--ddd-theme-default-slateGray, #666);
          border-radius: var(--ddd-radius-xs, 2px);
          transition: background 0.2s ease;
        }

        #results::-webkit-scrollbar-thumb:hover {
          background: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        :host([dark]) #results::-webkit-scrollbar-track,
        body.dark-mode #results::-webkit-scrollbar-track {
          background: var(--ddd-theme-default-coalyGray, #444);
        }

        :host([dark]) #results::-webkit-scrollbar-thumb,
        body.dark-mode #results::-webkit-scrollbar-thumb {
          background: var(--ddd-theme-default-limestoneGray, #a2aaad);
        }

        :host([dark]) #results::-webkit-scrollbar-thumb:hover,
        body.dark-mode #results::-webkit-scrollbar-thumb:hover {
          background: var(--ddd-theme-default-keystoneYellow, #ffd100);
        }

        li {
          flex: 0 0 auto;
          scroll-snap-align: center;
          width: 180px;
          min-width: 180px;
          height: 300px;
          display: flex;
          flex-direction: column;
          align-items: center;
          justify-content: flex-start; 
        }


        app-hax-site-bar {
          margin: 0 var(--ddd-spacing-3, 12px);
          width: 100%;
        }
        .description {
          max-height: 64px;
          overflow: hidden;
          max-width: 80%;
          text-overflow: ellipsis;
          word-break: break-word;
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-xs, 14px);
          line-height: var(--ddd-lh-150, 1.5);
        }

        @media (max-width: 1200px) {
          :host {
            min-width: calc(
              2 * 180px + var(--ddd-spacing-6, 24px) + 2 *
                var(--ddd-spacing-12, 56px)
            );
          }
        }

        @media (max-width: 800px) {
          :host {
            min-width: calc(180px + 2 * var(--ddd-spacing-12, 56px));
          }

          app-hax-site-bar {
            --main-banner-width: 60vw;
          }

          .description {
            max-height: 24px;
            font-size: var(--ddd-font-size-3xs, 8px);
            font-family: var(--ddd-font-primary, sans-serif);
          }

          .scroll-left,
          .scroll-right {
            min-width: var(--ddd-spacing-10, 40px);
            padding: var(--ddd-spacing-2, 8px);
            font-size: var(--ddd-font-size-s, 16px);
          }
        }
        @media (max-width: 640px) {
          app-hax-site-bar a {
            font-size: var(--ddd-font-size-xs, 14px);
          }
          app-hax-site-bar {
            --main-banner-width: 70vw;
          }

          li {
            width: 180px;
            min-width: 180px;
          }

          /* Mobile: Show only 1 item, hide arrows */
          .scroll-left,
          .scroll-right {
            display: none;
          }

          #results {
            /* Single item takes full width on mobile */
            gap: 0;
            scroll-snap-type: x mandatory;
            overflow-x: auto;
            scrollbar-width: none;
            -ms-overflow-style: none;
          }
          #results::-webkit-scrollbar {
            display: none;               
          }
        }
        span[slot="band"] {
          height: var(--ddd-spacing-12, 48px);
          overflow: hidden;
          text-overflow: ellipsis;
          margin-bottom: var(--ddd-spacing-2, 8px);
        }
        :host([dark]) #noResult,
        body.dark-mode #noResult {
          color: var(--ddd-theme-default-limestoneGray, #f5f5f5);
        }

        #noResult {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-s, 16px);
          color: var(--ddd-theme-default-coalyGray, #444);
          text-align: center;
          padding: var(--ddd-spacing-6, 24px);
          margin: var(--ddd-spacing-4, 16px);
        }

        /* Screen reader only text */
        .sr-only {
          position: absolute;
          width: 1px;
          height: 1px;
          padding: 0;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          white-space: nowrap;
          border: 0;
        }
      `]}getSortedItems(){const t=Array.isArray(this.displayItems)?[...this.displayItems]:[],e=this.sortOption||"az",getTitle=t=>(t.title||"").toString().toLowerCase(),getDate=(t,e)=>{const r=Number(o(t,`metadata.site.${e}`,o(t,"metadata.site.created",0)));return Number.isNaN(r)?0:r},getThemeName=t=>{const e=o(t,"metadata.theme.element","");return e&&i.themesData&&i.themesData[e]&&i.themesData[e].name?i.themesData[e].name.toString().toLowerCase():e.toString().toLowerCase()};return t.sort(((t,r)=>{switch(e){case"za":return getTitle(r).localeCompare(getTitle(t));case"newest":return getDate(r,"updated")-getDate(t,"updated");case"oldest":return getDate(t,"updated")-getDate(r,"updated");case"theme":return getThemeName(t).localeCompare(getThemeName(r));default:return getTitle(t).localeCompare(getTitle(r))}})),t}render(){const t=this.getSortedItems();return this.totalItems=t.length,e`
      <div
        class="carousel-container"
        role="region"
        aria-label="Site results carousel"
        aria-live="polite"
      >
        <simple-tooltip for="scroll-left-btn" position="top"
          >Previous</simple-tooltip
        >
        <button
          id="scroll-left-btn"
          class="scroll-left"
          @click="${this.scrollLeft}"
          ?disabled="${this.isAtStart||this.totalItems<=1}"
          aria-label="Previous sites"
          aria-describedby="scroll-left-desc"
        >
          ◀
          <span id="scroll-left-desc" class="sr-only"
            >View previous sites in the carousel</span
          >
        </button>
        <ul
          id="results"
          class="${this.isDragging?"dragging":""}"
          @mousedown="${this.handlePointerStart}"
          @touchstart="${this.handlePointerStart}"
          @mousemove="${this.handlePointerMove}"
          @touchmove="${this.handlePointerMove}"
          @mouseup="${this.handlePointerEnd}"
          @touchend="${this.handlePointerEnd}"
          @mouseleave="${this.handlePointerEnd}"
          @scroll="${this.handleScroll}"
>\n          ${t.length>0?t.map((t=>e` <li>
                    <app-hax-site-bar
                      ?dark="${this.dark}"
                      site-id="${t.id}"
                      .siteUrl="${t.slug}"
                      .slug="${t.slug}"
                      .image="${this.getThemeImage(t)}"
                      last-updated-time="${o(t,"metadata.site.updated",0)}"
                      accent-color="${o(t,"metadata.theme.variables.cssVariable","").replace("--simple-colors-default-theme-","").replace("-7","")}"
                    >
                      <a
                        slot="heading"
                        href="${t.slug}"
                        target="_blank"
                        tabindex="-1"
                        rel="noopener noreferrer"
                        title="${t.title}"
                      >
                        ${t.title}
                      </a>
                      <span slot="subHeading">${t.author}</span>
                      <app-hax-site-details
                        slot="band"
                        .details="${this.getItemDetails(t)}"
                        site-id="${t.id}"
                      >
                        <div class="description" slot="pre">
                          ${t.description}
                        </div>
                      </app-hax-site-details>
                    </app-hax-site-bar>
                  </li>`)):e`<div id="noResult">
                No
                results${""!==this.searchTerm?e`<strong>"${this.searchTerm}"</strong>`:", Create a new site!"}
              </div>`}
        </ul>
        <simple-tooltip for="scroll-right-btn" position="top"
          >Next</simple-tooltip
        >
        <button
          id="scroll-right-btn"
          class="scroll-right"
          @click="${this.scrollRight}"
          ?disabled="${this.isAtEnd||this.totalItems<=1}"
          this.totalItems <= 1}"
          aria-label="Next sites"
          aria-describedby="scroll-right-desc"
        >
          ▶
          <span id="scroll-right-desc" class="sr-only"
            >View next sites in the carousel</span
          >
        </button>
      </div>
    `}get isAtStart(){const t=this.shadowRoot?.querySelector("#results");return!t||t.scrollLeft<=2}get isAtEnd(){const t=this.shadowRoot?.querySelector("#results");return!t||t.scrollLeft+t.clientWidth>=t.scrollWidth-2}scrollLeft(){const t=this.shadowRoot.querySelector("#results");if(!t||this.totalItems<=1)return;t.scrollLeft<=204?t.scrollTo({left:0,behavior:"smooth"}):t.scrollBy({left:-204,behavior:"smooth"})}scrollRight(){const t=this.shadowRoot.querySelector("#results");if(!t||this.totalItems<=1)return;t.scrollBy({left:204,behavior:"smooth"})}handlePointerStart(t){const e=this.shadowRoot.querySelector("#results");this.isPointerDown=!0,this.isDragging=!1,this.startX=("mousedown"===t.type?t.clientX:t.touches[0].clientX)-e.offsetLeft,this.scrollLeftVal=e.scrollLeft,t.preventDefault()}handlePointerMove(t){if(!this.isPointerDown)return;const e=this.shadowRoot.querySelector("#results"),r=2*(("mousemove"===t.type?t.clientX:t.touches[0].clientX)-e.offsetLeft-this.startX);Math.abs(r)>5&&(this.isDragging=!0,e.style.scrollBehavior="auto"),this.isDragging&&(e.scrollLeft=this.scrollLeftVal-r,t.preventDefault())}handlePointerEnd(t){this.isPointerDown=!1;const e=this.shadowRoot.querySelector("#results");e&&(e.style.scrollBehavior="smooth"),setTimeout((()=>{this.isDragging=!1}),100)}handleScroll(t){if(this.totalItems<=1)return;const e=t.target,r=e.scrollLeft,s=e.scrollWidth;if(r+e.clientWidth>=s-50)this.currentIndex=this.totalItems;else{const t=r/204;this.currentIndex=Math.min(Math.max(1,Math.floor(t)+1),this.totalItems)}this.requestUpdate()}goToPage(t){const e=204*(t-1);this.shadowRoot.querySelector("#results").scrollTo({left:e,behavior:"smooth"}),this.currentIndex=t}getThemeImage(t){const e=o(t,"metadata.theme.element","");if(e&&i.themesData&&i.themesData[e]){let t=i.themesData[e].thumbnail||"";if(t&&t.startsWith("@haxtheweb/")){t=new URL("../../../../"+t,import.meta.url).href}return t}return""}getItemDetails(t){return{created:o(t,"metadata.site.created",new Date/1e3),updated:o(t,"metadata.site.updated",new Date/1e3),pages:o(t,"metadata.pageCount",0),url:t.slug}}openedChanged(t){i.appEl.playSound("click"),t.detail.value?this.shadowRoot.querySelector("app-hax-site-details").removeAttribute("tabindex"):this.shadowRoot.querySelector("app-hax-site-details").setAttribute("tabindex","-1")}}customElements.define(AppHaxSearchResults.tag,AppHaxSearchResults);
