// @haxtheweb/app-hax/lib/v2/app-hax-use-case-filter.js
import{LitElement as e,html as t,css as a}from"../../../../lit/index.js";import"../../../simple-tooltip/simple-tooltip.js";import"../../../simple-icon/lib/simple-icons.js";import"../../../simple-icon/lib/simple-icon-lite.js";import"../../../hax-iconset/lib/simple-hax-iconset.js";import{store as i}from"./AppHaxStore.js";import"./app-hax-use-case.js";import"./app-hax-search-results.js";import"./app-hax-filter-tag.js";import"./app-hax-scroll-button.js";import"./app-hax-site-creation-modal.js";export class AppHaxUseCaseFilter extends e{static get tag(){return"app-hax-use-case-filter"}constructor(){super(),this.windowControllers=new AbortController,this.searchTerm="",this.disabled=!1,this.showSearch=!1,this.items=[],this.filteredItems=[],this.filteredSites=[],this.activeFilters=[],this.filters=[],this.searchQuery="",this.demoLink="",this.errorMessage="",this.loading=!1,this.selectedCardIndex=null,this.returningSites=[],this.allFilters=new Set,this.dark=!1,this.isMobile=!1,this.isLoggedIn=!1,this.sortMenuOpen=!1,this.sortOption="newest",this.showTerrible=!1,this.showHidden=!1,this.__useCaseParamHandled=!1,this.__useCaseParamHandling=!1,void 0!==i&&import("../../../../mobx/dist/mobx.esm.js").then(({autorun:e,toJS:t})=>{e(()=>{this.dark=t(i.darkMode)}),e(()=>{this.isMobile=t(i.isMobile)});let a=!1;e(()=>{const e=t(i.appReady),s=t(i.isLoggedIn);this.isLoggedIn=s,this.showTerrible=t(i.showTerribleTemplates),this.showHidden=t(i.showHiddenTemplates),e&&s&&!a&&(a=!0,this.updateSkeletonResults(),this.updateSiteResults())});let s=null,r=null;e(()=>{const e=t(i.appReady),a=t(i.isLoggedIn),o=t(i.showTerribleTemplates),n=t(i.showHiddenTemplates);if(null===s&&null===r)return s=o,void(r=n);e&&a&&(o!==s||n!==r)&&(s=o,r=n,this.updateSkeletonResults(),this.updateSiteResults())}),e(()=>{const e=t(i.manifest);e&&e.items&&e.items.length>0&&this.updateSiteResults()})})}static get properties(){return{searchTerm:{type:String},showSearch:{type:Boolean,reflect:!0,attribute:"show-search"},showFilter:{type:Boolean,reflect:!0,attribute:"show-filter"},disabled:{type:Boolean,reflect:!0},items:{type:Array},filteredItems:{type:Array},filteredSites:{type:Array},activeFilters:{type:Array},filters:{type:Array},searchQuery:{type:String},demoLink:{type:String},errorMessage:{type:String},loading:{type:Boolean},selectedCardIndex:{type:Number},returningSites:{type:Array},allFilters:{attribute:!1},dark:{type:Boolean,reflect:!0},isMobile:{type:Boolean,attribute:!1},isLoggedIn:{type:Boolean},sortOption:{type:String,attribute:"sort-option"},sortMenuOpen:{type:Boolean},showTerrible:{type:Boolean,attribute:!1},showHidden:{type:Boolean,attribute:!1}}}static get styles(){return[a`
        :root {
          --page-column-width: 800px;
        }
        :host {
          overflow: hidden;
          display: block;
          max-width: 100%;
          font-family: var(--ddd-font-primary, sans-serif);
          margin-top: 2px;
        }
        :host([dark]) {
          --accent-color: var(--ddd-theme-default-white, white);
        }
        :host([light]) {
          --accent-color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }
        .contentSection {
          width: 100%;
          margin: 0;
          padding: 0;
          box-sizing: border-box;
        }

        .quick-create {
          padding: var(--ddd-spacing-5, 20px) var(--ddd-spacing-5, 20px)
            var(--ddd-spacing-4, 16px);
          background-color: var(--ddd-theme-default-limestoneLight);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
        }

        :host([dark]) .quick-create,
        body.dark-mode .quick-create {
          background-color: var(--ddd-theme-default-nittanyNavy, #001e44);
          color: var(--accent-color);
        }

        .quick-create h2 {
          color: inherit;
          margin: 0 0 var(--ddd-spacing-4, 16px) 0;
        }

        .quick-create-grid {
          display: flex;
          justify-content: center;
          gap: var(--ddd-spacing-4, 16px);
          overflow-x: auto;
          overflow-y: hidden;
          padding-bottom: var(--ddd-spacing-2, 8px);
          scrollbar-width: thin;
          scroll-snap-type: x mandatory;
        }

        .quick-create-grid > div {
          flex: 0 0 auto;
          scroll-snap-align: start;
        }

        .quick-create-grid app-hax-use-case {
          flex: 0 0 auto;
        }

        .template-group {
          width: 100%;
          box-sizing: border-box;
        }
        
        .template-group-heading {
          font-family: var(--ddd-font-primary, sans-serif);
          color: var(--accent-color);
          margin: 0 0 var(--ddd-spacing-4, 16px) var(--ddd-spacing-7, 28px);
          text-decoration: underline;
          text-decoration-thickness: var(--ddd-border-size-xs, 1px);
          text-underline-offset: var(--ddd-spacing-1, 4px);
        }

        .template-results {
          display: grid;
          grid-template-columns: repeat(auto-fit, minmax(200px, 0fr));
          gap: var(--ddd-spacing-4, 16px);
          justify-content: flex-start;
          justify-items: center;
          padding: 0 var(--ddd-spacing-5, 20px);
          box-sizing: border-box;
        }

        #returnToSection {
          padding-left: var(--ddd-spacing-6, 24px);
          padding-right: var(--ddd-spacing-6, 24px);
          
        }
        #returnToSection app-hax-search-results {
          
          box-sizing: border-box;
        }

        #startJourneySection {
           padding-left: var(--ddd-spacing-6, 24px);
          padding-right: var(--ddd-spacing-6, 24px);
        }

        .decorative-line {
          border-bottom: 4px solid var(--ddd-theme-default-skyBlue);
          box-sizing: border-box;
          width: 100%;
          margin: 0 0 var(--ddd-spacing-3, 12px) 0;
        }

        #from-template-heading {
          margin-top: var(--ddd-spacing-5, 20px);
        }

        :host(:not([show-filter])) app-hax-search-results {
          width: 100%;
        }

        h2,
        .returnTo h2,
        .startNew h2 {
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-m);
          color: var(--accent-color);
          margin: 0 0 var(--ddd-spacing-4, 16px) 0;
        }
        .startNew,
        .returnTo {
          position: relative;
          display: flex;
          flex-direction: column;
          justify-content: flex-start;
          align-items: stretch;
        }
        .upper-filter {
          position: relative;
          display: inline-block;
          flex: 1;
          min-width: 200px;
          width: 60%;
        }
        input[type="text"] {
          width: 100%;
          padding: var(--ddd-spacing-3, 12px) var(--ddd-spacing-3, 12px)
            var(--ddd-spacing-3, 12px) var(--ddd-spacing-3, 12px);
          font-size: var(--ddd-font-size-xs, 12px);
          border-radius: 12px;
          border: var(--ddd-border-s, 2px solid);
          border-color: var(--ddd-theme-default-limestoneGray);
          background: var(--ddd-theme-default-limestoneLight);
          color: var(--ddd-theme-default-coalyGray, #222);
          transition: all 0.2s ease;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          margin: 0;
          min-height: var(--ddd-spacing-12, 48px);
        }
        :host([dark]) input[type="text"],
        body.dark-mode input[type="text"] {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        input[type="text"]:focus {
          border: var(--ddd-border-md, 2px solid);
          border-color: var(--ddd-theme-default-accent);
          outline: none;
        }
        :host([dark]) input[type="text"]:focus,
        body.dark-mode input[type="text"]:focus {
          background: var(--ddd-theme-default-coalyGray, #333);
          border-color: var(--ddd-theme-default-accent);
        }
        .search-icon {
          position: absolute;
          right: var(--ddd-spacing-3, 12px);
          top: 50%;
          transform: translateY(-50%);
          font-size: var(--ddd-font-size-xs, 14px);
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          pointer-events: none;
          z-index: 1;
          --simple-icon-width: 25px;
          --simple-icon-height: 25px;
        }
        :host([dark]) .search-icon,
        body.dark-mode .search-icon {
          color: var(--ddd-theme-default-white, white);
        }
        .filter {
          margin: var(--ddd-spacing-4, 16px) auto;
          width: 100%;
          max-width: 700px;
          box-sizing: border-box;
          font-family: var(--ddd-font-primary, sans-serif);
          transition: box-shadow 0.2s ease;
          padding-left: var(--ddd-spacing-5, 20px);
          padding-right: var(--ddd-spacing-5, 20px);
          display: flex;
          flex-wrap: wrap;
          gap: var(--ddd-spacing-3, 12px);
          align-items: center;
        }

        .contentSection > div:has(.filter) {
          display: flex;
          justify-content: center;
        }
      
        input[type="checkbox"] {
          display: none;
        }
        .sort-control {
          margin-top: var(--ddd-spacing-2, 8px);
          margin-bottom: var(--ddd-spacing-3, 12px);
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-1, 4px);
          position: relative;
        }
        .sort-button {
          color: var(--ddd-theme-default-nittanyNavy, #001e44);
          background: transparent;
          border: none;
          box-shadow: none;
          padding: var(--ddd-spacing-2, 8px);
          cursor: pointer;
          display: flex;
          align-items: center;
          justify-content: center;
          transition: all 0.2s ease;
          outline: none;
          min-height: var(--ddd-spacing-7, 28px);
          min-width: var(--ddd-spacing-7, 28px);
          --simple-icon-width: 40px;
          --simple-icon-height: 40px;
        }

        :host([dark]) .sort-button, body.dark-mode .sort-button {
          color: var(--ddd-theme-default-white, white);
        }

        :host([dark]) .sort-button:hover, body.dark-mode .sort-button:hover {
          color: var(--ddd-theme-default-limestoneGray);
        }

        .sort-button:hover, body.dark-mode .sort-button:hover{
          color: var(--ddd-theme-default-accent);
        }
        
        .sort-menu {
          position: absolute;
          top: 100%;
          left: 0;
          margin-top: var(--ddd-spacing-1, 4px);
          background: var(--ddd-theme-default-limestoneLight);
          border: var(--ddd-border-xs, 1px solid) var(--ddd-theme-default-limestoneLight);
          border-radius: var(--ddd-radius-sm, 4px);
          box-shadow: var(--ddd-boxShadow-md);
          min-width: 200px;
          z-index: 1000;
        }
        .sort-menu-item {
          display: block;
          width: 100%;
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          background: none;
          border: none;
          text-align: left;
          cursor: pointer;
          font-family: var(--ddd-font-primary, sans-serif);
          font-size: var(--ddd-font-size-3xs, 11px);
          color: var(--ddd-theme-default-coalyGray, #222);
          transition: all 0.2s ease;
        }
        .sort-menu-item:hover {
          background: var(--ddd-theme-default-limestoneGray, #f5f5f5);
        }
        .sort-menu-item:focus {
          outline: var(--ddd-border-md, 2px solid) var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: -2px;
        }
        :host([dark]) .sort-menu-item,
        body.dark-mode .sort-menu-item {
          color: var(--ddd-theme-default-black);
        }
        :host([dark]) .sort-menu-item:hover,
        body.dark-mode .sort-menu-item:hover {
          background: var(--ddd-theme-default-limestoneGray);
        }
        .sort-select {
          width: 100%;
          padding: var(--ddd-spacing-1, 4px) var(--ddd-spacing-2, 8px);
          font-size: var(--ddd-font-size-3xs, 11px);
          border-radius: var(--ddd-radius-sm, 4px);
          border: var(--ddd-border-xs, 1px solid)
            var(--ddd-theme-default-slateGray, #666);
          background: var(--ddd-theme-default-white, white);
          color: var(--ddd-theme-default-coalyGray, #222);
          font-family: var(--ddd-font-primary, sans-serif);
          box-sizing: border-box;
        }
        :host([dark]) .sort-select,
        body.dark-mode .sort-select {
          background: var(--ddd-theme-default-coalyGray, #333);
          color: var(--ddd-theme-default-white, white);
          border-color: var(--ddd-theme-default-slateGray, #666);
        }
        .sort-select:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        .reset-button {
          margin-top: var(--ddd-spacing-1, 4px);
          background: var(--ddd-theme-default-original87Pink, #e4007c);
          border: var(--ddd-border-xs, 1px solid) transparent;
          color: var(--ddd-theme-default-white, white);
          border-radius: var(--ddd-radius-sm, 4px);
          font-size: var(--ddd-font-size-3xs, 11px);
          font-family: var(--ddd-font-primary, sans-serif);
          font-weight: var(--ddd-font-weight-medium, 500);
          padding: var(--ddd-spacing-2, 8px) var(--ddd-spacing-3, 12px);
          display: flex;
          align-items: center;
          justify-content: center;
          gap: var(--ddd-spacing-1, 4px);
          box-shadow: var(--ddd-boxShadow-sm);
          cursor: pointer;
          transition: all 0.2s ease;
          min-height: var(--ddd-spacing-7, 28px);
        }
        .reset-button:hover,
        .reset-button:focus {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
          transform: translateY(-1px);
        }
        .reset-button:focus {
          outline: var(--ddd-border-md, 2px solid)
            var(--ddd-theme-default-keystoneYellow, #ffd100);
          outline-offset: var(--ddd-spacing-1, 2px);
        }
        :host([dark]) .reset-button,
        body.dark-mode .reset-button {
          background: var(--ddd-theme-default-beaver70, #c85c2c);
        }
        :host([dark]) .reset-button:hover,
        :host([dark]) .reset-button:focus,
        body.dark-mode .reset-button:hover,
        body.dark-mode .reset-button:focus {
          background: var(--ddd-theme-default-original87Pink, #e4007c);
        }
        .collapseFilter {
          display: none;
        }

        /* Visually hidden content for screen readers */
        .visually-hidden {
          position: absolute !important;
          width: 1px !important;
          height: 1px !important;
          padding: 0 !important;
          margin: -1px !important;
          overflow: hidden !important;
          clip: rect(0, 0, 0, 0) !important;
          white-space: nowrap !important;
          border: 0 !important;
        }

        /* Loading and fallback messages */
        .loading-message,
        .no-results {
          grid-column: 1 / -1;
          text-align: center;
          padding: var(--ddd-spacing-8, 32px);
          font-size: var(--ddd-font-size-s, 16px);
          color: var(--ddd-theme-default-slateGray, #666);
          font-family: var(--ddd-font-primary, sans-serif);
        }
        :host([dark]) .loading-message,
        :host([dark]) .no-results,
        body.dark-mode .loading-message,
        body.dark-mode .no-results {
          color: var(--ddd-theme-default-limestoneGray, #ccc);
        }
        .fallback-message {
          grid-column: 1 / -1;
          display: flex;
          flex-direction: column;
          align-items: center;
          gap: var(--ddd-spacing-4, 16px);
        }
        .fallback-message .no-results {
          padding: var(--ddd-spacing-4, 16px) 0;
        }
        .fallback-message app-hax-use-case {
          max-width: 300px;
        }

        @media (max-width: 780px) {
          .contentSection {
            display: block;
          }
          :host([show-filter]) .filter {
            display: flex;
            width: 250px;
            max-width: 20vw;
          }
          :host .collapseFilter {
            display: flex;
          }
          h4,
          .returnTo h4,
          .startNew h4 {
            font-size: var(--ddd-font-size-m, 20px);
          }
        }

        @media (max-width: 600px) {
          .contentSection {
            display: block;
          }
          :host .collapseFilter {
            display: flex;
          }
          h4,
          .returnTo h4,
          .startNew h4 {
            font-size: var(--ddd-font-size-s, 18px);
          }

          .template-group-heading {
            margin-left: var(--ddd-spacing-3, 12px);
          }

          .template-results {
            grid-template-columns: repeat(
              auto-fit,
              minmax(clamp(120px, 30vw, 144px), 1fr)
            );
            gap: var(--ddd-spacing-2, 8px);
            padding: 0 var(--ddd-spacing-3, 12px);
          }
        }

        @media (max-width: 480px) {
          h4, .returnTo h4, .startNew h4 {
            font-size: var(--ddd-font-size-s, 16px);
            margin: 0 0 var(--ddd-spacing-3, 12px) 0;
          }
          .template-results {
            grid-template-columns: repeat(2, minmax(0, 1fr));
          }
          #returnToSection app-hax-search-results {
            min-width: 100%;
          }
        }
        .no-results {
          font-size: var(--ddd-font-size-s, 16px);
          color: light-dark(
            var(--ddd-theme-default-coalyGray, #222),
            var(--ddd-theme-default-white, white)
          );
        }
      `]}_getUseCaseParamFromUrl(){try{const e=new URL(globalThis.location.href),t=e.searchParams.get("use-case");if(t)return t;const a=e.searchParams.get("use_case");if(a)return a;const i=e.searchParams.get("useCase");if(i)return i}catch(e){}return""}_updateUrlQueryParam(e,t){if(e&&"string"==typeof e)try{const a=new URL(globalThis.location.href);null==t||""===t?a.searchParams.delete(e):a.searchParams.set(e,t),globalThis.history.replaceState(globalThis.history.state,"",`${a.pathname}${a.search}${a.hash}`)}catch(e){}}_getUseCaseMachineNameForTemplate(e){if(!e)return"";const t=this._getTemplateMachineNameCandidates(e);return t&&t.length>0?t[0]:""}_normalizeUseCaseMachineName(e){if(!e||"string"!=typeof e)return"";let t=e.trim().toLowerCase();return-1!==t.indexOf("/")&&(t=t.split("/").pop()),-1!==t.indexOf("?")&&(t=t.split("?").shift()),-1!==t.indexOf("#")&&(t=t.split("#").shift()),t=t.replace(/\.[^/.]+$/,""),t=t.replace(/[\s_]+/g,"-"),t=t.replace(/[^a-z0-9-]/g,"-"),t=t.replace(/-+/g,"-"),t=t.replace(/^-+/,"").replace(/-+$/,""),t}_looseNormalizeUseCaseMachineName(e){let t=this._normalizeUseCaseMachineName(e);return t?(t=t.replace(/-site-/g,"-"),-1!==t.indexOf("-site")&&t.lastIndexOf("-site")===t.length-5&&(t=t.replace(/-site$/,"")),t=t.replace(/-+/g,"-"),t=t.replace(/^-+/,"").replace(/-+$/,""),t):""}_deriveTemplateMachineName(e,t=""){const a=this._normalizeUseCaseMachineName(e);return a||this._normalizeUseCaseMachineName(t)}_getTemplateMachineNameCandidates(e){const t=[];if(!e)return t;e.machineName&&t.push(e.machineName),e.skeletonUrl&&t.push(e.skeletonUrl),e.importType&&(t.push("import-"+e.importType),t.push(e.importType)),e.themeElement&&t.push(e.themeElement),e.originalData&&e.originalData.element&&t.push(e.originalData.element),e.useCaseTitle&&t.push(e.useCaseTitle);const a=t.map(e=>this._normalizeUseCaseMachineName(e)).filter(e=>e);return[...new Set(a)]}_matchesUseCaseParam(e,t){const a=this._normalizeUseCaseMachineName(t),i=this._looseNormalizeUseCaseMachineName(t);if(!a)return!1;const s=this._getTemplateMachineNameCandidates(e);for(let e=0;e<s.length;e++){const t=s[e];if(t===a)return!0;const r=this._looseNormalizeUseCaseMachineName(t);if(r&&r===i)return!0}return!1}_scrollUseCaseIndexIntoView(e){if("number"!=typeof e||e<0)return;const t=this.shadowRoot&&this.shadowRoot.querySelector(`app-hax-use-case[data-item-index="${e}"]`);if(t&&"function"==typeof t.scrollIntoView)try{t.scrollIntoView({behavior:"smooth",block:"center"})}catch(e){t.scrollIntoView()}}_findTemplateByUseCaseParam(e){const t=Array.isArray(this.__allItems)&&this.__allItems.length>0?this.__allItems:this.items;if(!e||!Array.isArray(t)||0===t.length)return null;const a=this._normalizeUseCaseMachineName(e),i=this._looseNormalizeUseCaseMachineName(e);for(let e=0;e<t.length;e++){const i=t[e];if(-1!==this._getTemplateMachineNameCandidates(i).indexOf(a))return i}for(let e=0;e<t.length;e++){const a=t[e],s=this._getTemplateMachineNameCandidates(a);for(let e=0;e<s.length;e++){const t=this._looseNormalizeUseCaseMachineName(s[e]);if(t&&t===i)return a}}return null}async _maybeAutoOpenUseCaseFromUrl(){if(this.__useCaseParamHandled||this.__useCaseParamHandling)return;const e=this._getUseCaseParamFromUrl();if(e){if(Array.isArray(this.items)&&0!==this.items.length||Array.isArray(this.__allItems)&&0!==this.__allItems.length){this.__useCaseParamHandling=!0;try{await this.updateComplete;const t=this._findTemplateByUseCaseParam(e);if(t){if(Array.isArray(this.filteredItems)&&this.filteredItems.length>0){const a=this.filteredItems.findIndex(t=>this._matchesUseCaseParam(t,e));-1!==a?(this.selectedCardIndex=a,this.filteredItems[a]&&(this.filteredItems[a].isSelected=!0,this.filteredItems[a].showContinue=!0),this._scrollUseCaseIndexIntoView(a)):t&&t.dataType&&("skeleton"===t.dataType&&this.scrollToGroup("from-template-heading","skeleton"),"blank"===t.dataType&&this.scrollToGroup("from-scratch-heading","blank"),"import"===t.dataType&&this.scrollToGroup("from-existing-heading","import"))}await this.openTemplateModal(t)}else i&&i.toast&&i.toast(`Could not find use-case: ${e}`,4e3);this.__useCaseParamHandled=!0}catch(e){this.__useCaseParamHandled=!0}this.__useCaseParamHandling=!1}}else this.__useCaseParamHandled=!0}testKeydown(e){"Escape"!==e.key&&"Enter"!==e.key||this.toggleSearch()}toggleFilterVisibility(){this.showFilter=!this.showFilter}handleFilterKeydown(e,t){" "!==e.key&&"Enter"!==e.key||(e.preventDefault(),this.toggleFilterByButton(t))}scrollToGroup(e,t){const a=this.shadowRoot&&e?this.shadowRoot.querySelector("#"+e):null;a&&(a.scrollIntoView({behavior:"smooth",block:"start"}),setTimeout(()=>{this.focusFirstOption(t)},350))}focusFirstOption(e){if(!e||!Array.isArray(this.filteredItems))return;const t=this.filteredItems.findIndex(t=>t&&t.dataType===e);if(-1===t)return;const a=this.shadowRoot&&this.shadowRoot.querySelector(`app-hax-use-case[data-item-index="${t}"]`);if(a&&a.shadowRoot){const e=a.shadowRoot.querySelector('button.card, button.select, button, a[href], input, select, textarea, [tabindex]:not([tabindex="-1"])');if(e&&"function"==typeof e.focus){try{e.focus({preventScroll:!0})}catch(t){e.focus()}return}}a&&"function"==typeof a.focus&&(a.hasAttribute("tabindex")||a.setAttribute("tabindex","-1"),a.focus())}render(){const e=this.isMobile,a=Array.isArray(this.items)?this.items.filter(e=>e&&"skeleton"===e.dataType):[],i=e?[]:a.slice(0,3);return t`
      <div class="contentSection">
        <section class="quick-create" aria-labelledby="quick-create-heading">
          <h2 id="quick-create-heading">Create New Site...</h2>
          <div
            class="quick-create-grid"
            role="list"
            aria-label="Quick create options"
          >
            ${i.map(e=>t`
                <div role="listitem">
                  <app-hax-use-case
                    .source=${e.useCaseImage||""}
                    .title=${e.useCaseTitle||""}
                    .description=${e.useCaseDescription||""}
                    .demoLink=${e.demoLink||""}
                    .iconImage=${e.useCaseIcon||[]}
                    ?dark="${this.dark}"
                    aria-label="Create from template: ${e.useCaseTitle}"
                    @click=${()=>this.openTemplateModal(e)}
                  ></app-hax-use-case>
                </div>
              `)}

            <div role="listitem">
              <app-hax-use-case
                .source=${""}
                .title=${e?"Templates":"More Templates"}
                .description=${"Browse all template starters"}
                .iconImage=${[{icon:"icons:add",tooltip:"More templates"}]}
                ?dark="${this.dark}"
                aria-label=${e?"Templates":"More templates"}
                @click=${()=>this.scrollToGroup("from-template-heading","skeleton")}
              ></app-hax-use-case>
            </div>
            ${e?"":t`
                  <div role="listitem">
                    <app-hax-use-case
                      .source=${""}
                      .title=${"From Scratch"}
                      .description=${"Start from a blank site using a theme"}
                      .iconImage=${[{icon:"editor:insert-drive-file",tooltip:"From scratch"}]}
                      ?dark="${this.dark}"
                      aria-label="From scratch"
                      @click=${()=>this.scrollToGroup("from-scratch-heading","blank")}
                    ></app-hax-use-case>
                  </div>
                `}

            ${e?"":t`
                  <div role="listitem">
                    <app-hax-use-case
                      .source=${""}
                      .title=${"Import"}
                      .description=${"Import content from an existing source"}
                      .iconImage=${[{icon:"icons:cloud-download",tooltip:"Import content"}]}
                      ?dark="${this.dark}"
                      aria-label="Import"
                      @click=${()=>this.scrollToGroup("from-existing-heading","import")}
                    ></app-hax-use-case>
                  </div>
                `}
          </div>
        </section>

        <div>
          <div
            class="filter"
            role="search"
            aria-label="Filter and search site templates"
          >
            <!-- Search bar -->
            <div class="upper-filter">
              <label for="searchField" class="visually-hidden"
                >Search</label
              >
              <slot>
                <simple-icon-lite
                  class="search-icon"
                  icon="icons:search"
                  aria-hidden="true"
                ></simple-icon-lite>
              </slot>
              <input
                id="searchField"
                @input="${this.handleSearch}"
                @keydown="${this.testKeydown}"
                type="text"
                placeholder="Search"
                aria-label="Search"
                aria-describedby="search-help"
              />
              <div id="search-help" class="visually-hidden">
                Type to search for site templates and existing sites. Press
                Escape to clear.
              </div>
            </div>
            <!-- Sort options for returning sites -->
            <div class="sort-control">
              <button
                class="sort-button"
                @click=${e=>{e.stopPropagation(),this.sortMenuOpen=!this.sortMenuOpen}}
                aria-label="Sort options"
                aria-expanded=${this.sortMenuOpen}
                aria-has-popup="menu"
              >
                <simple-icon-lite 
                  icon="av:sort-by-alpha"
                  aria-hidden="true"
                ></simple-icon-lite>
              </button>
              ${this.sortMenuOpen?t`
                <div class="sort-menu" role="menu">
                  <button 
                    role="menuitem" 
                    class="sort-menu-item" 
                    @click=${()=>this.handleSortChange("newest")}
                  >
                    Recently updated
                  </button>
                  <button 
                    role="menuitem" 
                    class="sort-menu-item" 
                    @click=${()=>this.handleSortChange("az")}
                  >
                    Title A–Z
                  </button>
                  <button 
                    role="menuitem" 
                    class="sort-menu-item" 
                    @click=${()=>this.handleSortChange("za")}
                  >
                    Title Z–A
                  </button>
                  <button 
                    role="menuitem" 
                    class="sort-menu-item" 
                    @click=${()=>this.handleSortChange("oldest")}
                  >
                    Least recently updated
                  </button>
                  <button 
                    role="menuitem" 
                    class="sort-menu-item" 
                    @click=${()=>this.handleSortChange("theme")}
                  >
                    Theme name
                  </button>
                </div>
              `:""}
            </div>

            <div id="reset-help" class="visually-hidden">
              Clear all active filters and search terms
            </div>
          </div>
        </div>
        <!-- Content Section -->
        <div>
          <!-- Returning Sites -->
          <section
            id="returnToSection"
            class="returnTo"
            aria-labelledby="return-to-heading"
          >
            <h2 id="return-to-heading">Return to...</h2>
            <div class="decorative-line"></div>
            <div
              role="region"
              aria-label="Previously created sites"
              aria-live="polite"
            >
              <app-hax-search-results
                .displayItems=${this.filteredSites}
                .searchTerm=${this.searchTerm}
                .sortOption=${this.sortOption}
                ?dark="${this.dark}"
              >
              </app-hax-search-results>
            </div>
          </section>

          <!-- Templates -->
          <section
            id="startJourneySection"
            class="startNew"
            aria-labelledby="create-site-heading"
          >
            <h2 id="create-site-heading">Create New Site</h2>
            <div class="decorative-line"></div>
            <div id="template-count" class="visually-hidden">
              ${this.filteredItems.length} options available
            </div>

            ${this.filteredItems.length>0?(()=>{const a=[],i=[],s=[];this.filteredItems.forEach((e,t)=>{e&&"skeleton"===e.dataType&&a.push({item:e,index:t}),e&&"blank"===e.dataType&&i.push({item:e,index:t}),e&&"import"===e.dataType&&s.push({item:e,index:t})});const renderEntries=(e,a)=>e.map((e,i)=>t`
                        <div
                          role="gridcell"
                          aria-rowindex="${i+1}"
                          aria-colindex="1"
                        >
                          <app-hax-use-case
                            data-item-index="${e.index}"
                            data-data-type="${e.item.dataType}"
                            .source=${e.item.useCaseImage||""}
                            .title=${e.item.useCaseTitle||""}
                            .description=${e.item.useCaseDescription||""}
                            .demoLink=${e.item.demoLink||""}
                            .iconImage=${e.item.useCaseIcon||[]}
                            .isSelected=${e.item.isSelected||!1}
                            .showContinue=${e.item.showContinue||!1}
                            ?dark="${this.dark}"
                            aria-label="${a}: ${e.item.useCaseTitle}"
                            @toggle-display=${t=>this.toggleDisplay(e.index,t)}
                            @continue-action=${()=>this.continueAction(e.index)}
                          ></app-hax-use-case>
                        </div>
                      `);return t`
                    <div
                      class="template-group"
                      role="region"
                      aria-labelledby="from-template-heading"
                      aria-live="polite"
                    >
                      <h3
                        id="from-template-heading"
                        class="template-group-heading"
                      >
                        From Template
                      </h3>
                      ${a.length>0?t`<div
                            class="template-results"
                            role="grid"
                            aria-label="Template-based starters"
                            aria-describedby="template-count"
                          >
                            ${renderEntries(a,"Template")}
                          </div>`:t`<p
                            role="status"
                            class="no-results"
                            aria-live="polite"
                          >
                            No templates match your current filters.
                          </p>`}
                    </div>

                    <div
                      class="template-group"
                      role="region"
                      aria-labelledby="from-scratch-heading"
                      aria-live="polite"
                    >
                      <h3
                        id="from-scratch-heading"
                        class="template-group-heading"
                      >
                        From Scratch Using Theme
                      </h3>
                      ${i.length>0?t`<div
                            class="template-results"
                            role="grid"
                            aria-label="Theme-based starters"
                            aria-describedby="template-count"
                          >
                            ${renderEntries(i,"Theme")}
                          </div>`:t`<p
                            role="status"
                            class="no-results"
                            aria-live="polite"
                          >
                            No themes match your current filters.
                          </p>`}
                    </div>

                    ${e?"":t`
                          <div
                            class="template-group"
                            role="region"
                            aria-labelledby="from-existing-heading"
                            aria-live="polite"
                          >
                            <h3
                              id="from-existing-heading"
                              class="template-group-heading"
                            >
                              From Existing Site
                            </h3>
                            ${s.length>0?t`<div
                                  class="template-results"
                                  role="grid"
                                  aria-label="Import options"
                                  aria-describedby="template-count"
                                >
                                  ${renderEntries(s,"Import")}
                                </div>`:t`<p
                                  role="status"
                                  class="no-results"
                                  aria-live="polite"
                                >
                                  No import options match your current filters.
                                </p>`}
                          </div>
                        `}
                  `})():this.loading?t`<p
                    role="status"
                    class="loading-message"
                    aria-live="polite"
                  >
                    Loading templates...
                  </p>`:this.items&&this.items.length>0?t`<p
                      role="status"
                      class="no-results"
                      aria-live="polite"
                    >
                      No templates match your current filters. Reset filters to
                      see more options.
                    </p>`:t`
                      <div class="fallback-message">
                        <p
                          role="status"
                          class="no-results"
                          aria-live="polite"
                        >
                          ${this.errorMessage||"No templates available. You can still create a blank site."}
                        </p>
                        <app-hax-use-case
                          .source=""
                          .title="Blank Site"
                          .description="Create a blank site using the clean-one theme"
                          .iconImage=${[]}
                          .isSelected=${-1===this.selectedCardIndex}
                          .showContinue=${-1===this.selectedCardIndex}
                          ?dark="${this.dark}"
                          aria-label="Template: Blank Site"
                          @toggle-display=${e=>this.toggleDisplay(-1,e)}
                          @continue-action=${()=>this.continueAction(-1)}
                        ></app-hax-use-case>
                      </div>
                    `}
          </section>
        </div>
      </div>

      <!-- Site Creation Modal -->
      <app-hax-site-creation-modal
        id="siteCreationModal"
        @modal-closed=${this.handleModalClosed}
      ></app-hax-site-creation-modal>
    `}getImportItems(){return[{dataType:"import",importKind:"file",importType:"docx",callback:"@haxcms/docxToSite",fileType:"docx",useCaseTitle:"Word Doc",useCaseImage:"",useCaseDescription:"Import a .docx outline into a new HAX site",useCaseIcon:[{icon:"hax:file-docx",tooltip:"DOCX import"}],useCaseTag:["Import","Docx"],demoLink:"#"},{dataType:"import",importKind:"url",importType:"gitbook",callback:"@haxcms/gitbookToSite",prompt:"URL for the GitBook git repo",param:"md",useCaseTitle:"GitBook",useCaseImage:"",useCaseDescription:"Import a GitBook repo into a new HAX site",useCaseIcon:[{icon:"mdi-social:github-circle",tooltip:"Git-based import"}],useCaseTag:["Import","Git"],demoLink:"#"},{dataType:"import",importKind:"url",importType:"haxcms",callback:"@haxcms/haxcmsToSite",prompt:"URL for the HAXcms site",param:"repoUrl",useCaseTitle:"HAXcms",useCaseImage:"",useCaseDescription:"Import content from an existing HAXcms site",useCaseIcon:[{icon:"hax:hax2022",tooltip:"HAXcms import"}],useCaseTag:["Import","HAXcms"],demoLink:"#"},{dataType:"import",importKind:"url",importType:"notion",callback:"@haxcms/notionToSite",prompt:"URL for the Notion git repo",param:"repoUrl",useCaseTitle:"Notion",useCaseImage:"",useCaseDescription:"Import a Notion-exported git repo",useCaseIcon:[{icon:"book",tooltip:"Notion import"}],useCaseTag:["Import","Notion"],demoLink:"#"},{dataType:"import",importKind:"url",importType:"html",callback:"@haxcms/htmlToSite",prompt:"URL for the HTML content",param:"repoUrl",useCaseTitle:"HTML",useCaseImage:"",useCaseDescription:"Import a remote HTML site into HAX",useCaseIcon:[{icon:"icons:code",tooltip:"HTML import"}],useCaseTag:["Import","HTML"],demoLink:"#"},{dataType:"import",importKind:"file",importType:"pressbooks",callback:"@haxcms/pressbooksToSite",fileType:"html",useCaseTitle:"Pressbooks",useCaseImage:"",useCaseDescription:"Import a Pressbooks export (HTML) into HAX",useCaseIcon:[{icon:"hax:wordpress",tooltip:"Pressbooks import"}],useCaseTag:["Import","Pressbooks"],demoLink:"#"}]}iconForFilter(e){switch(e.toLowerCase()){case"blog":return"lrn:write";case"brochure":return"icons:description";case"course":return"hax:lesson";case"portfolio":return"icons:perm-identity";case"blank":return"hax:bricks";case"import":return"icons:cloud-download";default:return"icons:label"}}toggleFilterByButton(e){this.activeFilters.includes(e)?this.activeFilters=this.activeFilters.filter(t=>t!==e):this.activeFilters=[...this.activeFilters,e],this.applyFilters(),this.requestUpdate()}_handleOutsideSortClick(e){if(!this.sortMenuOpen)return;const t=this.shadowRoot.querySelector(".sort-control");t&&(t.contains(e.target)||(this.sortMenuOpen=!1,this.requestUpdate()))}connectedCallback(){super.connectedCallback(),globalThis.addEventListener("jwt-logged-in",this._jwtLoggedIn.bind(this),{signal:this.windowControllers.signal}),this._boundOutsideSortClick=this._handleOutsideSortClick.bind(this),document.addEventListener("click",this._boundOutsideSortClick)}disconnectedCallback(){document.removeEventListener("click",this._boundOutsideSortClick),this.windowControllers.abort(),super.disconnectedCallback()}_jwtLoggedIn(e){!0===e.detail?(this.isLoggedIn=!0,this.updateSkeletonResults(),this.updateSiteResults()):!1===e.detail&&(this.isLoggedIn=!1)}firstUpdated(){super.firstUpdated()}updated(e){(e.has("searchQuery")||e.has("activeFilters")||e.has("items"))&&this.applyFilters(),e.has("items")&&this._maybeAutoOpenUseCaseFromUrl()}toggleSearch(){this.disabled||(this.shadowRoot.querySelector("#searchField").value="",i.appEl.playSound("click"),this.showSearch=!this.showSearch,setTimeout(()=>{this.shadowRoot.querySelector("#searchField").focus()},300))}toggleSelection(e){this.activeUseCase===e?this.activeUseCase=!1:this.activeUseCase=e,this.requestUpdate()}handleSearch(e){const t=e.target.value.toLowerCase();this.searchTerm=t,i.searchTerm=t,this.applyFilters()}handleSortChange(e){const t="string"==typeof e?e:e&&e.target&&e.target.value?e.target.value:"newest";this.sortOption=t,this.sortMenuOpen=!1,this.requestUpdate()}toggleFilter(e){const t=e.target.value;this.activeFilters.includes(t)?this.activeFilters=[...this.activeFilters.filter(e=>e!==t)]:this.activeFilters=[...this.activeFilters,t],this.applyFilters()}applyFilters(){const e=this.searchTerm.toLowerCase();this.filteredItems=[...this.items.filter(t=>{if("skeleton"!==t.dataType&&"blank"!==t.dataType&&"import"!==t.dataType)return!1;const a=""===e||t.useCaseTitle.toLowerCase().includes(e)||t.useCaseTag&&t.useCaseTag.some(t=>t.toLowerCase().includes(e)),i=0===this.activeFilters.length||t.useCaseTag&&this.activeFilters.some(e=>t.useCaseTag.includes(e));return a&&i})],this.filteredSites=[...this.returningSites.filter(t=>{if("site"!==t.dataType)return!1;const a=t.originalData||{},i=(a.metadata||{}).site||{};let s=void 0!==i.category&&null!==i.category?i.category:t.useCaseTag||[];const r=(a.title||"").toLowerCase(),o=(a.description||"").toLowerCase(),n=(a.author||"").toLowerCase(),l=(a.slug||"").toLowerCase(),d=t.useCaseTag||[],c=""===e||-1!==r.indexOf(e)||-1!==o.indexOf(e)||-1!==n.indexOf(e)||-1!==l.indexOf(e)||d.some(t=>-1!==t.toLowerCase().indexOf(e))||"string"==typeof s&&-1!==s.toLowerCase().indexOf(e)||Array.isArray(s)&&s.some(t=>"string"==typeof t&&-1!==t.toLowerCase().indexOf(e)),p=0===this.activeFilters.length||this.activeFilters.some(e=>"string"==typeof s?s===e:!!Array.isArray(s)&&-1!==s.indexOf(e));return c&&p})]}removeFilter(e){const t=e.detail;this.activeFilters=this.activeFilters.filter(e=>e!==t),this.applyFilters(),this.requestUpdate()}resetFilters(){this.searchTerm="",i.searchTerm="",this.activeFilters=[],this.filteredItems=[...this.items.filter(e=>"skeleton"===e.dataType||"blank"===e.dataType||"import"===e.dataType)],this.filteredSites=[...this.returningSites],this.shadowRoot.querySelector("#searchField").value="",this.shadowRoot.querySelectorAll('input[type="checkbox"]').forEach(e=>e.checked=!1),this.requestUpdate()}updateSkeletonResults(){if(this.loading=!0,this.errorMessage="",!i.appSettings||!i.appSettings.skeletonsList)return this.errorMessage="Skeletons endpoint not configured",void(this.loading=!1);this.allFilters=new Set;const e=i.AppHaxAPI&&i.AppHaxAPI.makeCall?i.AppHaxAPI.makeCall("skeletonsList"):Promise.reject(new Error("API not available"));let t;if(i.appSettings&&i.appSettings.themes&&Object.keys(i.appSettings.themes).length>0)t=Promise.resolve(i.appSettings.themes);else{const e=new URL("../../../haxcms-elements/lib/themes.json",import.meta.url).href;t=fetch(e).then(e=>{if(!e.ok)throw new Error(`Failed Themes (${e.status})`);return e.json()}).catch(e=>(console.warn("Failed to load themes.json, using minimal fallback:",e),{"clean-one":{element:"clean-one",name:"Clean One",description:"A clean, simple theme",category:["Blank"],hidden:!1}}))}Promise.allSettled([e,t]).then(([e,t])=>{const a=e.value&&e.value.data&&Array.isArray(e.value.data)?e.value.data:[],_getUseCasePriority=e=>{if(!e)return 0;const t=e.originalData?e.originalData:{},a=void 0!==t.priority?t.priority:void 0!==e.priority?e.priority:0,i="number"==typeof a?a:Number(a);return Number.isFinite(i)?i:0},_sortUseCasesByPriority=(e,t)=>{const a=_getUseCasePriority(e),i=_getUseCasePriority(t);if(a!==i)return a-i;const s=e&&e.useCaseTitle?e.useCaseTitle:"",r=t&&t.useCaseTitle?t.useCaseTitle:"";return s.localeCompare(r)},i=a.map(e=>{let t=[];Array.isArray(e.category)?t=e.category.filter(e=>"string"==typeof e&&""!==e.trim()):"string"==typeof e.category&&""!==e.category.trim()&&(t=[e.category.trim()]),0===t.length&&(t=["Empty"]);const a=!(!e||!e.hidden),i=!(!e||!e.terrible);(!a||this.showHidden)&&(!i||this.showTerrible)&&t.forEach(e=>this.allFilters.add(e));const s=Array.isArray(e.attributes)?e.attributes.map(e=>({icon:e.icon||"",tooltip:e.tooltip||""})):[];let r=e.image||"";if(r&&r.startsWith("@haxtheweb/")){r=new URL("../../../../"+r,import.meta.url).href}const o=e["skeleton-url"]||"";return{dataType:"skeleton",machineName:this._deriveTemplateMachineName(e.machineName||e.name||e["machine-name"]||o||e.title||"",e.title||""),hidden:a,terrible:i,useCaseTitle:e.title||"Untitled Template",useCaseImage:r||"",useCaseDescription:e.description||"",useCaseIcon:s,useCaseTag:t,demoLink:e["demo-url"]||"#",skeletonUrl:o,originalData:e}})||[];i.sort(_sortUseCasesByPriority);const s=i.filter(e=>e&&(!e.hidden||this.showHidden)&&(!e.terrible||this.showTerrible)),r=t.value||{},o=Object.entries(r).map(([e,t])=>{let a=[];Array.isArray(t.category)?a=t.category.filter(e=>"string"==typeof e&&""!==e.trim()):"string"==typeof t.category&&""!==t.category.trim()&&(a=[t.category.trim()]),0===a.length&&(a=["Blank"]);const i=!(!t||!t.hidden);let s=!(!t||!t.terrible);s||"string"!=typeof e||0!==e.indexOf("terrible")||(s=!0);(!i||this.showHidden)&&(!s||this.showTerrible)&&a.forEach(e=>this.allFilters.add(e));let r=t.thumbnail||"";if(r&&r.startsWith("@haxtheweb/")){r=new URL("../../../../"+r,import.meta.url).href}return{dataType:"blank",machineName:e,themeElement:t&&t.element?t.element:e,hidden:i,terrible:s,useCaseTitle:t.name||"Untitled Theme",useCaseImage:r||"",useCaseDescription:t.description||"Start with a blank site using this theme",useCaseIcon:[{icon:"icons:build",tooltip:"Customizable"}],useCaseTag:a,demoLink:`https://playground.hax.cloud/site.html?theme=${e}`,originalData:t}});o.sort(_sortUseCasesByPriority);const n=o.filter(e=>e&&(!e.hidden||this.showHidden)&&(!e.terrible||this.showTerrible)),l=this.getImportItems();l.forEach(e=>{e&&Array.isArray(e.useCaseTag)&&e.useCaseTag.forEach(e=>this.allFilters.add(e))}),this.__allItems=[...i,...o,...l],this.items=[...s,...n,...l],this.filters=Array.from(this.allFilters).sort(),0!==this.items.length||this.errorMessage||(this.errorMessage="No Templates Found"),this.resetFilters()}).catch(e=>{this.errorMessage=`Failed to load data: ${e.message}`,this.items=[],this.filters=[]}).finally(()=>{this.loading=!1})}updateSiteResults(){this.loading=!0,this.errorMessage="";try{const e=i.manifest;if(!e||!e.items)throw new Error("No manifest data available");const t=Array.isArray(e.items)?e.items.map(e=>{let t=e.metadata&&e.metadata.site?e.metadata.site.category:null,a=[];Array.isArray(t)?a=t.filter(e=>"string"==typeof e&&""!==e.trim()):"string"==typeof t&&""!==t.trim()&&(a=[t.trim()]);let i=null;e.metadata&&e.metadata.site&&e.metadata.site.tags?i=e.metadata.site.tags:e.metadata&&e.metadata.tags&&(i=e.metadata.tags),Array.isArray(i)?i.filter(e=>"string"==typeof e&&""!==e.trim()).forEach(e=>a.push(e.trim())):"string"==typeof i&&""!==i.trim()&&i.split(",").map(e=>e.trim()).filter(e=>""!==e).forEach(e=>a.push(e));let s=null;return e.build&&e.build.type?s=e.build.type:e.metadata&&e.metadata.build&&e.metadata.build.type&&(s=e.metadata.build.type),"string"==typeof s&&""!==s.trim()&&a.push(s.trim()),a=[...new Set(a.filter(e=>"string"==typeof e&&""!==e.trim()))],0===a.length&&(a=["Website"]),a.forEach(e=>this.allFilters.add(e)),{dataType:"site",useCaseTag:a,originalData:e,...e}}):[];this.returningSites=[...t],this.filters=Array.from(this.allFilters).sort(),this.filteredSites=[...t],0!==t.length||this.errorMessage||(this.errorMessage="No Sites Found"),this.requestUpdate(),this.loading=!1}catch(e){this.errorMessage=`Failed to load data: ${e.message}`,this.returningSites=[],this.filteredSites=[],this.filters=[],this.loading=!1}}toggleDisplay(e,t){const a=t.detail.isSelected;null!==this.selectedCardIndex&&this.selectedCardIndex!==e&&-1!==this.selectedCardIndex&&this.filteredItems[this.selectedCardIndex]&&(this.filteredItems[this.selectedCardIndex].isSelected=!1,this.filteredItems[this.selectedCardIndex].showContinue=!1),this.selectedCardIndex=a?e:null,-1!==e&&this.filteredItems[e]&&(this.filteredItems[e].isSelected=a,this.filteredItems[e].showContinue=a),this.requestUpdate()}async continueAction(e){const t=this.shadowRoot.querySelector("#siteCreationModal");if(!t)return;if(-1===e)return this._updateUrlQueryParam("use-case","blank-site"),t.title="Blank Site",t.description="Create a blank site using the clean-one theme",t.source="",t.template="Blank Site",t.themeElement="clean-one",t.skeletonMachineName=null,t.skeletonData={meta:{name:"clean-one",type:"skeleton"},site:{name:"clean-one",theme:"clean-one"},build:{type:"skeleton",structure:"from-skeleton",items:[{id:"item-home-clean-one",title:"Home",slug:"home",order:0,parent:null,indent:0,content:"<p>Edit this page to get started on your HAX site!</p>",metadata:{published:!0,hideInMenu:!1,tags:[]}}],files:[]},theme:{}},t.siteName=t.title,void t.openModal();const a=Array.isArray(this.filteredItems)&&"number"==typeof e?this.filteredItems[e]:null;a?await this.openTemplateModal(a):console.warn("No template found at index:",e)}_resolveSameOriginPath(e){if(!e||"string"!=typeof e)return"";const t=e.trim();if(""===t||"#"===t)return"";if(-1!==t.indexOf("://")||0===t.indexOf("//"))return"";if(0===t.indexOf("/"))return t;try{const e=new URL(t,globalThis.location.href);return e.origin!==globalThis.location.origin?"":`${e.pathname}${e.search}${e.hash}`}catch(e){return""}}async openTemplateModal(e){const t=this.shadowRoot.querySelector("#siteCreationModal");if(!t||!e)return;const a=this._getUseCaseMachineNameForTemplate(e);if(a&&this._updateUrlQueryParam("use-case",a),t.title=e.useCaseTitle,t.description=e.useCaseDescription,t.source=e.useCaseImage,t.template=e.useCaseTitle,t.skeletonMachineName=null,"import"===e.dataType)return t.themeElement="clean-one",void await this.processImportTemplate(e,t);if("skeleton"===e.dataType&&!e.skeletonUrl)return i&&i.appEl&&i.appEl.playSound&&i.appEl.playSound("error"),void(i&&i.toast&&i.toast("This template isn't configured yet (missing skeleton URL).",4e3));if("skeleton"===e.dataType&&e.skeletonUrl)try{e.machineName&&"string"==typeof e.machineName&&(t.skeletonMachineName=e.machineName);const a=e.skeletonUrl||"",s=this._resolveSameOriginPath(a);if(!s)return i&&i.appEl&&i.appEl.playSound&&i.appEl.playSound("error"),i&&i.toast&&i.toast("This template isn't configured yet (invalid skeleton URL).",4e3),void console.warn("Refusing to load skeleton from URL:",a);const r=await fetch(s);if(!r.ok)return i&&i.appEl&&i.appEl.playSound&&i.appEl.playSound("error"),i&&i.toast&&i.toast("Failed to load template skeleton.",4e3),console.warn(`Failed to load skeleton from ${s}`),void(t.themeElement="clean-one");{const e=await r.json();t.skeletonData=e.data||e,!t.skeletonMachineName&&t.skeletonData.meta&&(t.skeletonData.meta.machineName&&"string"==typeof t.skeletonData.meta.machineName?t.skeletonMachineName=t.skeletonData.meta.machineName:t.skeletonData.meta.name&&"string"==typeof t.skeletonData.meta.name&&(t.skeletonMachineName=t.skeletonData.meta.name)),t.themeElement=t.skeletonData.site&&t.skeletonData.site.theme||"clean-one"}}catch(e){return i&&i.appEl&&i.appEl.playSound&&i.appEl.playSound("error"),i&&i.toast&&i.toast("Failed to load template skeleton.",4e3),console.warn("Error loading skeleton:",e),void(t.themeElement="clean-one")}else if("blank"===e.dataType){t.skeletonMachineName=null;const a=e.themeElement||(e.originalData&&e.originalData.element?e.originalData.element:e.machineName||"");a?(t.themeElement=a,t.skeletonData={meta:{name:a,type:"skeleton"},site:{name:a,theme:a},build:{type:"skeleton",structure:"from-skeleton",items:[{id:`item-home-${a}`,title:"Home",slug:"home",order:0,parent:null,indent:0,content:"<p>Edit this page to get started on your HAX site!</p>",metadata:{published:!0,hideInMenu:!1,tags:[]}}],files:[]},theme:{}}):t.themeElement="clean-one"}else t.skeletonMachineName=null,t.themeElement="clean-one";t.siteName=e.useCaseTitle||t.title||"New site",t.openModal()}_normalizeImportedName(e){if(!e||"string"!=typeof e)return"Imported site";let t=e.replace(/\.[^/.]+$/,"");return t=t.replace(/[_-]+/g," "),t=t.replace(/\s+/g," ").trim(),t||"Imported site"}async processImportTemplate(e,t){try{const[{MicroFrontendRegistry:a},{enableServices:s}]=await Promise.all([import("../../../micro-frontend-registry/micro-frontend-registry.js"),import("../../../micro-frontend-registry/lib/microServices.js")]);s(["haxcms"]);let r=null;if("file"===e.importKind){const t=e.fileType||"docx";await import("../../../file-system-broker/lib/docx-file-system-broker.js");const i=globalThis.FileSystemBroker.requestAvailability(),s=await i.loadFile(t);if(!s)return;const o=new FormData;o.append("method","site"),o.append("type","import"),o.append("upload",s),r=await a.call(e.callback,o)}else if("url"===e.importKind){const t=e.prompt||"Enter a URL",i=globalThis.prompt(t);if(!i)return;const s={};s[e.param||"repoUrl"]=i,r=await a.call(e.callback,s)}if(r&&200==r.status&&r.data&&""!=r.data.contents&&Array.isArray(r.data.items)){const a=r.data.items,i=r.data.files?r.data.files:[],s=this._normalizeImportedName(r.data.filename||e.useCaseTitle);t.themeElement=t.themeElement||"clean-one",t.skeletonMachineName=null,t.skeletonData={meta:{name:e.importType,type:"import"},build:{type:e.importType,structure:"import",items:a,files:i},site:{name:s,theme:t.themeElement},theme:{}},t.siteName=s,t.openModal()}else i&&i.appEl&&i.appEl.playSound&&i.appEl.playSound("error"),i&&i.toast&&i.toast("Import did not return a valid structure",3e3),console.warn("Import did not return a valid structure:",r)}catch(e){i&&i.appEl&&i.appEl.playSound&&i.appEl.playSound("error"),i&&i.toast&&i.toast("Import failed",3e3),console.warn("Import error:",e)}}handleModalClosed(e){e.detail&&e.detail.cancelled&&null!==this.selectedCardIndex&&this.filteredItems[this.selectedCardIndex]&&(this.filteredItems[this.selectedCardIndex].isSelected=!1,this.filteredItems[this.selectedCardIndex].showContinue=!1,this.selectedCardIndex=null,this.requestUpdate())}}customElements.define("app-hax-use-case-filter",AppHaxUseCaseFilter);
