// @haxtheweb/app-hax/lib/v2/app-hax-wired-toggle.js
import{autorun as e,toJS as d}from"../../../../mobx/dist/mobx.esm.js";import{html as t,css as r}from"../../../../lit/index.js";import{store as a}from"./AppHaxStore.js";import{WiredDarkmodeToggle as i}from"../../../haxcms-elements/lib/core/ui/wired-darkmode-toggle/wired-darkmode-toggle.js";export class AppHAXWiredToggle extends i{constructor(){super(),this.darkModeMediaQuery=globalThis.matchMedia("(prefers-color-scheme: dark)"),this._updateToggleState=()=>{this.checked=d(a.darkMode),this.darkModeMediaQuery.matches?this.disabled=!0:this.disabled=!1},e(this._updateToggleState),this.darkModeMediaQuery.addEventListener("change",this._updateToggleState)}static get tag(){return"app-hax-wired-toggle"}disconnectedCallback(){super.disconnectedCallback(),this.darkModeMediaQuery&&this._updateToggleState&&this.darkModeMediaQuery.removeEventListener("change",this._updateToggleState)}updated(e){super.updated&&super.updated(e),e.forEach((e,d)=>{"checked"===d&&void 0!==e&&(a.darkMode=this[d])})}static get styles(){return[super.styles,r`
        /* Screen reader only text */
        .sr-only {
          position: absolute;
          width: 1px;
          height: 1px;
          padding: 0;
          margin: -1px;
          overflow: hidden;
          clip: rect(0, 0, 0, 0);
          white-space: nowrap;
          border: 0;
        }
      `]}render(){return t`
      <div
        data-label="${this.label}"
        aria-describedby="dark-mode-desc"
      >
        ${super.render()}
        <div id="dark-mode-desc" class="sr-only">
          Toggle between light and dark mode themes
        </div>
      </div>
    `}}globalThis.customElements.define(AppHAXWiredToggle.tag,AppHAXWiredToggle);
