// @haxtheweb/app-hax/lib/v2/app-hax-scroll-button.js
import{LitElement as e,html as t,css as o}from"../../../../lit/index.js";import"../../../simple-tooltip/simple-tooltip.js";export class AppHaxScrollButton extends e{static get tag(){return"app-hax-scroll-button"}constructor(){super(),this.label="",this.targetId="",this.isDarkMode=document.body.classList.contains("dark-mode")}static get properties(){return{label:{type:String},targetId:{type:String},isDarkMode:{type:Boolean,reflect:!0}}}scrollToTarget(){if(this.targetId){let e=null,t=this.closest("app-hax")||document.querySelector("app-hax");if(t){let o=t.shadowRoot?t.shadowRoot.querySelector("app-hax-use-case-filter"):t.querySelector("app-hax-use-case-filter");o&&(e=o.shadowRoot?o.shadowRoot.getElementById(this.targetId):o.querySelector(`#${this.targetId}`))}e||(e=document.getElementById(this.targetId)),e?(console.log("Scrolling to:",e),e.scrollIntoView({behavior:"smooth",block:"start"})):console.warn(`Element with id '${this.targetId}' not found inside app-hax-use-case-filter.`)}}connectedCallback(){super.connectedCallback(),this._darkModeObserver=new MutationObserver((()=>{this.isDarkMode=document.body.classList.contains("dark-mode")})),this._darkModeObserver.observe(document.body,{attributes:!0,attributeFilter:["class"]})}disconnectedCallback(){super.disconnectedCallback(),this._darkModeObserver&&this._darkModeObserver.disconnect()}static get styles(){return[o`
        :host {
          display: flex;
          background-image: url("/elements/app-hax/lib/assets/images/scroll-button-LM.png");
          background-size: cover;
          background-position: center;
          width: 300px;
          height: 50px;
          color: var(--app-hax-accent-color, var(--accent-color));
          align-items: center;
          justify-content: center;
          text-align: center;
          margin: 8px;
        }
        :host([isDarkMode]) {
          background-image: url("/elements/app-hax/lib/assets/images/scroll-button-DM.png");
        }
        div {
          display: flex;
          align-items: center;
        }
      `]}render(){return t`
      <div @click="${this.scrollToTarget}" tabindex="0" role="button">
        <h5>${this.label}</h5>
      </div>
    `}}customElements.define(AppHaxScrollButton.tag,AppHaxScrollButton);
