// @lrnwebcomponents/haxcms-elements/lib/ui-components/active-item/site-active-tags.js
define(["exports", "../../../../../lit/index.js", "../../core/haxcms-site-store.js", "../../../../../mobx/dist/mobx.esm.js", "../../../../simple-fields/lib/simple-tags.js"], function (_exports, _index, _haxcmsSiteStore, _mobxEsm, _simpleTags) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SiteActiveTags = void 0;
  var _templateObject_2a96a080e51211edbf9f41986c80248c, _templateObject2_2a96a080e51211edbf9f41986c80248c;
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  /**
   * `site-active-tags`
   * `Tags of the active page`
   *
   * @demo demo/index.html
   */
  var SiteActiveTags = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SiteActiveTags, _LitElement);
    var _super = _createSuper(SiteActiveTags);
    /**
     * HTMLElement
     */
    function SiteActiveTags() {
      var _this;
      babelHelpers.classCallCheck(this, SiteActiveTags);
      _this = _super.call(this);
      _this.autoAccentColor = false;
      _this.accentColor = null;
      _this.tags = null;
      _this.__disposer = [];
      (0, _mobxEsm.autorun)(function (reaction) {
        _this.tags = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeTags);
        _this.__disposer.push(reaction);
      });
      return _this;
    }
    /**
     * HTMLElement
     */
    babelHelpers.createClass(SiteActiveTags, [{
      key: "render",
      value:
      /**
       * LitElement
       */
      function render() {
        var _this2 = this;
        return (0, _index.html)(_templateObject_2a96a080e51211edbf9f41986c80248c || (_templateObject_2a96a080e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["", ""])), this.tags && this.tags != "" ? this.tags.split(",").map(function (tag) {
          return (0, _index.html)(_templateObject2_2a96a080e51211edbf9f41986c80248c || (_templateObject2_2a96a080e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral([" <simple-tag\n              ?auto-accent-color=\"", "\"\n              value=\"", "\"\n              accent-color=\"", "\"\n            ></simple-tag>"])), _this2.autoAccentColor, tag.trim(), _this2.accentColor);
        }) : "");
      }
      /**
       * LitElement / popular convention
       */
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        // clean up state
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }
        babelHelpers.get(babelHelpers.getPrototypeOf(SiteActiveTags.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * Break shadowRoot
       */
    }, {
      key: "createRenderRoot",
      value: function createRenderRoot() {
        return this;
      }
    }], [{
      key: "tag",
      get:
      /**
       * Store the tag name to make it easier to obtain directly.
       */
      function get() {
        return "site-active-tags";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          tags: {
            type: String
          },
          autoAccentColor: {
            type: Boolean,
            attribute: "auto-accent-color"
          },
          accentColor: {
            type: String,
            attribute: "accent-color"
          }
        };
      }
    }]);
    return SiteActiveTags;
  }(_index.LitElement);
  _exports.SiteActiveTags = SiteActiveTags;
  customElements.define(SiteActiveTags.tag, SiteActiveTags);
});
