// @lrnwebcomponents/responsive-grid/lib/responsive-grid-clear.js
define(["exports", "../../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.ResponsiveGridClear = void 0;

  function _templateObject2_f5d10e502c1711ea878dbb11c9f11137() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: none;\n          clear: both;\n        }\n      "]);

    _templateObject2_f5d10e502c1711ea878dbb11c9f11137 = function _templateObject2_f5d10e502c1711ea878dbb11c9f11137() {
      return data;
    };

    return data;
  }

  function _templateObject_f5d10e502c1711ea878dbb11c9f11137() {
    var data = babelHelpers.taggedTemplateLiteral([""]);

    _templateObject_f5d10e502c1711ea878dbb11c9f11137 = function _templateObject_f5d10e502c1711ea878dbb11c9f11137() {
      return data;
    };

    return data;
  }

  /**  
  `responsive-grid-clear`
  A clearumn for the responsive grid
  @microcopy - the mental model for this element
    Example: 
    <responsive-grid-clear 
      xs    //clears the floated columns on an extra small-width screen
      sm    //clears the floated columns on a small-width screen
      md    //clears the floated columns on a medium-width screen
      lg/>  //clears the floated columns on a large-width screen
  */
  var ResponsiveGridClear =
  /*#__PURE__*/
  function (_LitElement) {
    babelHelpers.inherits(ResponsiveGridClear, _LitElement);
    babelHelpers.createClass(ResponsiveGridClear, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_f5d10e502c1711ea878dbb11c9f11137());
      }
    }], [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [(0, _litElement.css)(_templateObject2_f5d10e502c1711ea878dbb11c9f11137())];
      }
    }, {
      key: "tag",
      get: function get() {
        return "responsive-grid-clear";
      }
    }]);

    function ResponsiveGridClear() {
      var _this;

      babelHelpers.classCallCheck(this, ResponsiveGridClear);
      _this = babelHelpers.possibleConstructorReturn(this, babelHelpers.getPrototypeOf(ResponsiveGridClear).call(this));
      _this.xl = false;
      _this.lg = false;
      _this.md = false;
      _this.sm = false;
      _this.xs = false;
      return _this;
    }

    babelHelpers.createClass(ResponsiveGridClear, null, [{
      key: "properties",
      get: function get() {
        return {
          /**
           * clear the float and force a new worw on an extra-large screen?
           */
          xl: {
            type: Boolean
          },

          /**
           * clear the float and force a new worw on a large screen?
           */
          lg: {
            type: Boolean
          },

          /**
           * clear the float and force a new worw on a medium screen?
           */
          md: {
            type: Boolean
          },

          /**
           * clear the float and force a new worw on a small screen?
           */
          sm: {
            type: Boolean
          },

          /**
           * clear the float and force a new worw on an extra-small screen?
           */
          xs: {
            type: Boolean
          }
        };
      }
    }]);
    return ResponsiveGridClear;
  }(_litElement.LitElement);

  _exports.ResponsiveGridClear = ResponsiveGridClear;
  window.customElements.define(ResponsiveGridClear.tag, ResponsiveGridClear);
});
