// @lrnwebcomponents/word-count/word-count.js
define(["exports","../../@polymer/polymer/polymer-legacy.js","../../@polymer/polymer/lib/legacy/polymer.dom.js","../../@polymer/polymer/lib/utils/flattened-nodes-observer.js"],function(_exports,_polymerLegacy,_polymerDom,_flattenedNodesObserver){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.WordCount=void 0;/**
 * `word-count`
 * `Count the words on whatever this wraps`
 *
 * @demo demo/index.html
 */let WordCount=(0,_polymerLegacy.Polymer)({_template:_polymerLegacy.html`
    <style>
      :host {
        display: block;
        --word-count-color: #888888;
        --word-count-color-hover: #000000;
      }
      :host:after {
        content: attr(words-text);
        font-size: 10px;
        position: relative;
        transition: 0.3s font-size, color ease;
        display: flex;
        line-height: 16px;
        flex-direction: row-reverse;
        margin: 12px;
        color: var(--word-count-color);
        @apply --word-count-text;
      }
      :host(:hover):after {
        font-size: 12px;
        font-weight: bold;
        color: var(--word-count-color-hover);
        @apply --word-count-text-hover;
      }
      :host(:focus):after {
        font-size: 12px;
        font-weight: bold;
        color: var(--word-count-color-hover);
        @apply --word-count-text-hover;
      }
      :host(:active):after {
        font-size: 12px;
        font-weight: bold;
        color: var(--word-count-color-hover);
        @apply --word-count-text-hover;
      }
    </style>
    <slot></slot>
  `,is:"word-count",hostAttributes:{tabindex:"0"},properties:{/**
     * Words
     */words:{type:Number},/**
     * Prefix text.
     */wordsPrefix:{type:String,value:"Words:"},/**
     * Text to output
     */wordsText:{type:String,computed:"_computeWordsText(words, wordsPrefix)",reflectToAttribute:!0}},/**
   * Ready life cycle
   */ready:function(){// mutation observer that ensures state applied correctly
this._observer=new _flattenedNodesObserver.FlattenedNodesObserver(this,info=>{if(0<info.addedNodes.length||0<info.removedNodes.length){this._updateWords()}})},/**
   * Update words based on data in the slot.
   */_updateWords:function(){if(""!==(0,_polymerDom.dom)(this).textContent){this.words=parseInt((0,_polymerDom.dom)(this).textContent.split(/\s+/g).length-1)}else{this.words=0}},/**
   * Computer the text to output for words
   */_computeWordsText:function(words,prefix){return prefix+" "+words}});_exports.WordCount=WordCount});
