// @lrnwebcomponents/hax-body/lib/hax-app-picker.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-modal/lib/simple-modal-template.js", "./hax-picker.js", "./hax-ui-styles.js"], function (_exports, _litElement, _simpleModalTemplate, _haxPicker, _haxUiStyles) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxAppPicker = void 0;

  function _templateObject2_0c4ef8a0994411ebb1a5332c5e6bdde2() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <simple-modal-template\n        .title=\"", "\"\n        modal-id=\"hax-picker\"\n        mode=\"hax-ui\"\n      >\n        <hax-picker slot=\"content\"></hax-picker>\n      </simple-modal-template>\n    "]);

    _templateObject2_0c4ef8a0994411ebb1a5332c5e6bdde2 = function _templateObject2_0c4ef8a0994411ebb1a5332c5e6bdde2() {
      return data;
    };

    return data;
  }

  function _templateObject_0c4ef8a0994411ebb1a5332c5e6bdde2() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: none;\n        }\n      "]);

    _templateObject_0c4ef8a0994411ebb1a5332c5e6bdde2 = function _templateObject_0c4ef8a0994411ebb1a5332c5e6bdde2() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `hax-app-picker`
   * `app pop over + picker with options`
   */
  var HaxAppPicker = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(HaxAppPicker, _LitElement);

    var _super = _createSuper(HaxAppPicker);

    babelHelpers.createClass(HaxAppPicker, null, [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(_haxUiStyles.HaxComponentStyles), [(0, _litElement.css)(_templateObject_0c4ef8a0994411ebb1a5332c5e6bdde2())]);
      }
    }]);

    function HaxAppPicker() {
      var _this;

      babelHelpers.classCallCheck(this, HaxAppPicker);
      _this = _super.call(this);
      _this.title = "Select an option";
      return _this;
    }

    babelHelpers.createClass(HaxAppPicker, [{
      key: "connectedCallback",
      value: function connectedCallback() {
        babelHelpers.get(babelHelpers.getPrototypeOf(HaxAppPicker.prototype), "connectedCallback", this).call(this);
        window.addEventListener("simple-modal-show", this.modalToggle.bind(this));
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        window.removeEventListener("simple-modal-show", this.modalToggle.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(HaxAppPicker.prototype), "disconnectedCallback", this).call(this);
      }
      /**
       * a bit hacky but gets around the cloning element and events issue
       */

    }, {
      key: "modalToggle",
      value: function modalToggle(e) {
        if (e.detail.id == "hax-picker") {
          // present options AFTER we know the picker has opened :)
          e.detail.elements.content.children[0].buildOptions(this.elements, this.type, this.title, this.pickerType, this.target);
        }
      }
    }, {
      key: "presentOptions",

      /**
       * Bridge since everything calls here
       */
      value: function presentOptions(elements) {
        var _this2 = this;

        var type = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : "element";
        var title = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "Select an option";
        var pickerType = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : "gizmo";
        var target = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : window;
        this.elements = elements;
        this.type = type;
        this.title = title;
        this.pickerType = pickerType;
        this.target = target;
        setTimeout(function () {
          _this2.shadowRoot.querySelector('[modal-id="hax-picker"]').openModal(target);
        }, 0);
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject2_0c4ef8a0994411ebb1a5332c5e6bdde2(), this.title);
      }
    }, {
      key: "firstUpdated",

      /**
       * Attached to the DOM, now fire that we exist.
       */
      value: function firstUpdated() {
        // fire an event that this is a core piece of the system
        this.dispatchEvent(new CustomEvent("hax-register-core-piece", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {
            piece: "haxAppPicker",
            object: this
          }
        }));
      }
    }], [{
      key: "properties",
      get: function get() {
        return {
          /**
           * Title for the dialog
           */
          title: {
            type: String
          }
        };
      }
    }, {
      key: "tag",
      get: function get() {
        return "hax-app-picker";
      }
    }]);
    return HaxAppPicker;
  }(_litElement.LitElement);

  _exports.HaxAppPicker = HaxAppPicker;
  window.customElements.define(HaxAppPicker.tag, HaxAppPicker);
});
