// @lrnwebcomponents/hax-body/lib/hax-map.js
define(["exports", "../../../lit-element/lit-element.js", "../../simple-icon/lib/simple-icon-lite.js", "./hax-toolbar-item.js", "../../hax-iconset/lib/simple-hax-iconset.js", "./hax-store.js", "../../utils/utils.js", "./hax-ui-styles.js", "../../i18n-manager/lib/I18NMixin.js"], function (_exports, _litElement, _simpleIconLite, _haxToolbarItem, _simpleHaxIconset, _haxStore, _utils, _haxUiStyles, _I18NMixin2) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxMap = void 0;

  function _templateObject3_395aa4b08e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral(["\n              <li>\n                <hax-toolbar-item\n                  align-horizontal=\"left\"\n                  class=\"heading-level-", "\"\n                  @click=\"", "\"\n                  data-index=\"", "\"\n                  icon=\"", "\"\n                  label=\"", "\"\n                  show-text-label\n                >\n                </hax-toolbar-item>\n              </li>\n            "]);

    _templateObject3_395aa4b08e7d11eba1fe15f944a3ea4d = function _templateObject3_395aa4b08e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function _templateObject2_395aa4b08e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"container\">\n        <table>\n          <caption>\n            ", "\n          </caption>\n          <thead>\n            <tr>\n              <th scope=\"col\">", "</th>\n              <th scope=\"col\">", "</th>\n              <th scope=\"col\">", "</th>\n              <th scope=\"col\">", "</th>\n              <th scope=\"col\">", "</th>\n            </tr>\n          </thead>\n          <tbody>\n            <tr>\n              <td>", "</td>\n              <td>", "</td>\n              <td>", "</td>\n              <td>", "</td>\n              <td>", "</td>\n            </tr>\n          </tbody>\n        </table>\n        <h5>", "</h5>\n        <ul>\n          ", "\n        </ul>\n      </div>\n    "]);

    _templateObject2_395aa4b08e7d11eba1fe15f944a3ea4d = function _templateObject2_395aa4b08e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function _templateObject_395aa4b08e7d11eba1fe15f944a3ea4d() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        .container {\n          text-align: left;\n        }\n        table {\n          font-size: var(--hax-ui-font-size-sm);\n          border-collapse: collapse;\n          width: calc(100% - 2px);\n          max-width: calc(100% - 2px);\n        }\n        table,\n        th,\n        td {\n          text-align: center;\n          border: 1px solid var(--hax-ui-border-color);\n        }\n        th {\n          font-weight: normal;\n          font-size: var(--hax-ui-font-size-xs);\n        }\n        td {\n          font-weight: bold;\n          font-size: 150%;\n        }\n        ul {\n          list-style: none;\n          padding: 0;\n          margin: 0;\n        }\n        ul li {\n          margin: 0;\n          padding: 0;\n        }\n        li > hax-toolbar-item {\n          width: 100%;\n        }\n        li > hax-toolbar-item::part(button),\n        li > hax-toolbar-item[icon=\"hax:h2\"].heading-level-h2::part(button) {\n          width: 100%;\n          border: none;\n          margin-left: 0px;\n        }\n        li > hax-toolbar-item.heading-level-h2::part(button),\n        li > hax-toolbar-item[icon=\"hax:h3\"].heading-level-h3::part(button) {\n          width: calc(100% - 26px);\n          margin-left: 26px;\n        }\n        li > hax-toolbar-item.heading-level-h3::part(button),\n        li > hax-toolbar-item[icon=\"hax:h4\"].heading-level-h4::part(button) {\n          width: calc(100% - calc(2 * 26px));\n          margin-left: calc(2 * 26px);\n        }\n        li > hax-toolbar-item.heading-level-h4::part(button),\n        li > hax-toolbar-item[icon=\"hax:h5\"].heading-level-h5::part(button) {\n          width: calc(100% - calc(3 * 26px));\n          margin-left: calc(3 * 26px);\n        }\n        li > hax-toolbar-item.heading-level-h5::part(button),\n        li > hax-toolbar-item[icon=\"hax:h6\"].heading-level-h6::part(button) {\n          width: calc(100% - calc(4 * 26px));\n          margin-left: calc(4 * 26px);\n        }\n        li > hax-toolbar-item.heading-level-h6::part(button) {\n          width: calc(100% - calc(5 * 26px));\n          margin-left: calc(5 * 26px);\n        }\n      "]);

    _templateObject_395aa4b08e7d11eba1fe15f944a3ea4d = function _templateObject_395aa4b08e7d11eba1fe15f944a3ea4d() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `hax-map`
   * @element hax-map
   * `Export dialog with all export options and settings provided.`
   */
  var HaxMap = /*#__PURE__*/function (_I18NMixin) {
    babelHelpers.inherits(HaxMap, _I18NMixin);

    var _super = _createSuper(HaxMap);

    babelHelpers.createClass(HaxMap, null, [{
      key: "styles",

      /**
       * LitElement constructable styles enhancement
       */
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(_haxUiStyles.HaxTrayDetailHeadings), [(0, _litElement.css)(_templateObject_395aa4b08e7d11eba1fe15f944a3ea4d())]);
      }
    }]);

    function HaxMap() {
      var _this;

      babelHelpers.classCallCheck(this, HaxMap);
      _this = _super.call(this);
      _this.elementList = [];
      _this.t = {
        contentStatistics: "Content Statistics",
        words: "Words",
        headings: "Headings",
        paragraphs: "Paragraphs",
        widgets: "Widgets",
        characters: "Characters",
        listView: "List view"
      };

      _this.registerLocalization({
        context: babelHelpers.assertThisInitialized(_this),
        namespace: "hax"
      });

      return _this;
    }

    babelHelpers.createClass(HaxMap, [{
      key: "updateHAXMap",
      value: function () {
        var _updateHAXMap = babelHelpers.asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(e) {
          var list, elements, i, def;
          return regeneratorRuntime.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  _context.t0 = _haxStore.HAXStore;
                  _context.next = 3;
                  return _haxStore.HAXStore.activeHaxBody.haxToContent();

                case 3:
                  _context.t1 = _context.sent;
                  list = _context.t0.htmlToHaxElements.call(_context.t0, _context.t1);
                  this.calcStats(list);
                  elements = [];

                  for (i = 0; i < list.length; i++) {
                    def = _haxStore.HAXStore.haxSchemaFromTag(list[i].tag);

                    if (def.gizmo) {
                      elements.push({
                        icon: def.gizmo.icon,
                        name: def.gizmo.title
                      });
                    } else {
                      if (list[i].tag && list[i].tag.includes("-")) {
                        elements.push({
                          icon: "hax:templates",
                          name: "Widget"
                        });
                      } else {
                        elements.push({
                          icon: "hax:paragraph",
                          name: "HTML block"
                        });
                      }
                    }
                  }

                  this.elementList = [].concat(elements);

                case 9:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this);
        }));

        function updateHAXMap(_x) {
          return _updateHAXMap.apply(this, arguments);
        }

        return updateHAXMap;
      }()
      /**
       * Calculate statistics from the array of hax elements
       */

    }, {
      key: "calcStats",
      value: function calcStats(elements) {
        if (elements && _haxStore.HAXStore.activeHaxBody.innerText) {
          var counts = {
            c: _haxStore.HAXStore.activeHaxBody.innerText.length,
            w: parseInt(_haxStore.HAXStore.activeHaxBody.innerText.split(/\s+/g).length - 1),
            h: 0,
            p: 0,
            e: 0
          };
          elements.forEach(function (el) {
            switch (el.tag) {
              case "blockquote":
              case "div":
              case "span":
              case "p":
                counts.p++;
                break;

              case "h1":
              case "h2":
              case "h3":
              case "h4":
              case "h5":
              case "h6":
              case "relative-heading":
                // special support for our own heading tag
                counts.h++;
                break;

              default:
                counts.e++;
                break;
            }
          });

          for (var i in counts) {
            this["".concat(i, "Count")] = counts[i];
          }
        }
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;

        return (0, _litElement.html)(_templateObject2_395aa4b08e7d11eba1fe15f944a3ea4d(), this.t.contentStatistics, this.t.words, this.t.headings, this.t.paragraphs, this.t.widgets, this.t.characters, this.wCount, this.hCount, this.pCount, this.eCount, this.cCount, this.t.listView, this.indentedElements.map(function (element, index) {
          return (0, _litElement.html)(_templateObject3_395aa4b08e7d11eba1fe15f944a3ea4d(), element.parent || "h1", function (e) {
            return _this2.goToItem(index);
          }, index, element.icon, element.name);
        }));
      }
    }, {
      key: "goToItem",
      value: function goToItem(index) {
        if (index) {
          var activeChild = _haxStore.HAXStore.activeHaxBody.children[parseInt(index)];

          activeChild.classList.add("blinkfocus");

          if (typeof activeChild.scrollIntoViewIfNeeded === "function") {
            activeChild.scrollIntoViewIfNeeded(true);
          } else {
            activeChild.scrollIntoView({
              behavior: "smooth",
              inline: "center"
            });
          }

          setTimeout(function () {
            activeChild.classList.remove("blinkfocus");
          }, 500);
        }
      }
    }, {
      key: "scrollInMap",
      value: function scrollInMap() {
        var target = (0, _utils.normalizeEventPath)(e)[0];
        this.goToItem(target.getAttribute("data-index"));
      }
    }, {
      key: "indentedElements",
      get: function get() {
        var prev = "h1";
        return this.elementList.map(function (element) {
          var el = element;
          el.parent = prev;

          if (el.name == "Heading") {
            var h = el.icon.replace("hax:", "").trim();
            el.parent = h;
            prev = h;
          }

          return el;
        });
      }
    }], [{
      key: "tag",
      get: function get() {
        return "hax-map";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Title when open.
           */
          opened: {
            type: Boolean
          },
          elementList: {
            type: Array
          },
          cCount: {
            type: String
          },
          wCount: {
            type: String
          },
          hCount: {
            type: String
          },
          pCount: {
            type: String
          },
          eCount: {
            type: String
          }
        };
      }
    }]);
    return HaxMap;
  }((0, _I18NMixin2.I18NMixin)(_litElement.LitElement));

  _exports.HaxMap = HaxMap;
  window.customElements.define(HaxMap.tag, HaxMap);
});
