// @lrnwebcomponents/hax-body/lib/hax-app-browser.js
define(["exports", "../../../lit/index.js", "./hax-store.js", "../../../mobx/dist/mobx.esm.js", "../../simple-toolbar/lib/simple-button-grid.js", "../../a11y-collapse/a11y-collapse.js", "./hax-app-search.js", "./hax-tray-button.js"], function (_exports, _index, _haxStore, _mobxEsm, _simpleButtonGrid, _a11yCollapse, _haxAppSearch, _haxTrayButton) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HaxAppBrowser = void 0;
  var _templateObject_a6e7ed10b48911edaa972d5aad30d551, _templateObject2_a6e7ed10b48911edaa972d5aad30d551, _templateObject3_a6e7ed10b48911edaa972d5aad30d551, _templateObject4_a6e7ed10b48911edaa972d5aad30d551, _templateObject5_a6e7ed10b48911edaa972d5aad30d551;
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  /**
   * `hax-app-browser`
   * @element hax-app-browser
   *
   * `Browse a list of apps. This provides a listing of our gizmos that we've integrated with.`
   * @microcopy - the mental model for this element
   * - hax-app - expression of how to communicate and visualize a data source
   * - gizmo - silly name for the general public when talking about hax-app and what it provides in the end
   */
  var HaxAppBrowser = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(HaxAppBrowser, _LitElement);
    var _super = _createSuper(HaxAppBrowser);
    function HaxAppBrowser() {
      var _this;
      babelHelpers.classCallCheck(this, HaxAppBrowser);
      _this = _super.call(this);
      _this.searching = false;
      _this.categories = [];
      _this.appList = [];
      _this.activeApp = null;
      _this.hasActive = false;
      (0, _mobxEsm.autorun)(function () {
        _this.appList = (0, _mobxEsm.toJS)(_haxStore.HAXStore.appList);
      });
      (0, _mobxEsm.autorun)(function () {
        _this.activeApp = (0, _mobxEsm.toJS)(_haxStore.HAXStore.activeApp);
      });
      _this.addEventListener("hax-tray-button-click", function (e) {
        if (e.detail.eventName === "search-selected") {
          _this.searching = true;
          _haxStore.HAXStore.activeApp = (0, _mobxEsm.toJS)(_this.appList[e.detail.index]);
        } else if (e.detail.eventName === "cancel-search") {
          _this.searching = false;
        }
      });
      return _this;
    }
    babelHelpers.createClass(HaxAppBrowser, [{
      key: "ucfirst",
      value: function ucfirst(str) {
        return str.charAt(0).toUpperCase() + str.slice(1);
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;
        return (0, _index.html)(_templateObject_a6e7ed10b48911edaa972d5aad30d551 || (_templateObject_a6e7ed10b48911edaa972d5aad30d551 = babelHelpers.taggedTemplateLiteral(["", "\n      <hax-app-search\n        id=\"haxappsearch\"\n        class=\"", "\"\n      ></hax-app-search>\n      <slot></slot> "])), this.categories.map(function (tag) {
          return (0, _index.html)(_templateObject2_a6e7ed10b48911edaa972d5aad30d551 || (_templateObject2_a6e7ed10b48911edaa972d5aad30d551 = babelHelpers.taggedTemplateLiteral(["\n          <a11y-collapse heading=\"", "\" heading-button>\n            <simple-button-grid\n              class=\"", "\"\n              always-expanded\n              columns=\"2\"\n            >\n              ", "\n            </simple-button-grid>\n          </a11y-collapse>\n        "])), _this2.ucfirst(tag), _this2.searching ? "collapse-hide" : "", _this2.appList.map(function (app) {
            return (0, _index.html)(_templateObject3_a6e7ed10b48911edaa972d5aad30d551 || (_templateObject3_a6e7ed10b48911edaa972d5aad30d551 = babelHelpers.taggedTemplateLiteral([" ", ""])), app.details.tags.includes(tag) ? (0, _index.html)(_templateObject4_a6e7ed10b48911edaa972d5aad30d551 || (_templateObject4_a6e7ed10b48911edaa972d5aad30d551 = babelHelpers.taggedTemplateLiteral([" <hax-tray-button\n                        class=\"", "\"\n                        show-text-label\n                        icon-position=\"top\"\n                        index=\"", "\"\n                        label=\"", "\"\n                        icon=\"", "\"\n                        color=\"", "\"\n                        event-name=\"search-selected\"\n                        event-data=\"", "\"\n                      >\n                      </hax-tray-button>"])), _this2.searching ? "visibility-hidden" : "", app.index, app.details.title, app.details.icon, app.details.color, app.index) : "");
          }));
        }), !this.searching ? "visibility-hidden" : "");
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this3 = this;
        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "activeApp" && _this3[propName]) {
            _this3._activeAppChanged(_this3[propName], oldValue);
          }
          if (propName == "appList" && _this3[propName] && _this3.shadowRoot) {
            _this3.searching = false;
            _this3.categories = babelHelpers.toConsumableArray(_this3.updateCategories(_this3.appList));
          }
        });
      }
    }, {
      key: "updateCategories",
      value: function updateCategories(list) {
        var tags = [];
        list.forEach(function (app) {
          if (app.details.tags) {
            app.details.tags.forEach(function (tag) {
              if (!tags.includes(tag)) {
                tags.push(tag);
              }
            });
          }
        });
        tags.sort();
        return tags;
      }

      /**
       * Active app updated, so scroll it into view
       */
    }, {
      key: "_activeAppChanged",
      value: function _activeAppChanged(newValue, oldValue) {
        if (babelHelpers.typeof(oldValue) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined)) && newValue != null) {
          this.hasActive = true;
        } else {
          this.hasActive = false;
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        if (babelHelpers.get(babelHelpers.getPrototypeOf(HaxAppBrowser.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(HaxAppBrowser.prototype), "firstUpdated", this).call(this, changedProperties);
        }
        // app list registers early and has no imports so on
        // fast environments it's alreayd loaded
        this.searching = false;
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _index.css)(_templateObject5_a6e7ed10b48911edaa972d5aad30d551 || (_templateObject5_a6e7ed10b48911edaa972d5aad30d551 = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          position: relative;\n        }\n        simple-button-grid::part(buttons) {\n          overflow: auto;\n        }\n        simple-button-grid.collapse-hide {\n          max-height: 0 !important;\n          transition: all 0.5s;\n        }\n        hax-tray-button {\n          font-size: 11px;\n          --hax-ui-font-size-sm: 11px;\n        }\n        .visibility-hidden {\n          z-index: -1;\n          visibility: hidden;\n          opacity: 0;\n          height: 0;\n          transition: all 0.5s;\n        }\n        a11y-collapse {\n          margin: 0;\n          --a11y-collapse-margin: 0;\n        }\n        a11y-collapse::part(heading) {\n          margin: 4px;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "hax-app-browser";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          /**
           * Search term
           */
          search: {
            type: String
          },
          /**
           * Searching mode
           */
          searching: {
            type: Boolean,
            reflect: true
          },
          /**
           * Global activeApp object.
           */
          activeApp: {
            type: Object
          },
          /**
           * If we have an active, scale everything
           */
          hasActive: {
            reflect: true,
            type: Boolean,
            attribute: "has-active"
          },
          appList: {
            type: Array
          },
          categories: {
            type: Array
          }
        };
      }
    }]);
    return HaxAppBrowser;
  }(_index.LitElement);
  _exports.HaxAppBrowser = HaxAppBrowser;
  customElements.define(HaxAppBrowser.tag, HaxAppBrowser);
});
