// @lrnwebcomponents/editable-table/lib/editable-table-editor-rowcol.js
define(["exports", "../../../lit-element/lit-element.js", "./editable-table-behaviors.js", "../../a11y-menu-button/a11y-menu-button.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-icon/lib/simple-icons.js", "../../hax-iconset/lib/simple-hax-iconset.js"], function (_exports, _litElement, _editableTableBehaviors, _a11yMenuButton, _simpleIconLite, _simpleIcons, _simpleHaxIconset) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.EditableTableEditorRowcol = void 0;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  function _templateObject4_3c95187099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          --paper-item-min-height: 24px;\n          --a11y-menu-button-border: none;\n          --a11y-menu-button-list-border: 1px solid\n            var(--editable-table-border-color, #999);\n          --a11y-menu-button-vertical-padding: var(\n            --editable-table-cell-vertical-padding,\n            10px\n          );\n          --a11y-menu-button-horizontal-padding: var(\n            --editable-table-cell-horizontal-padding,\n            6px\n          );\n          --a11y-menu-button-item-focus-bg-color: var(\n            --editable-table-heading-bg-color,\n            #e8e8e8\n          );\n          --a11y-menu-button-list-bg-color: var(\n            --editable-table-bg-color,\n            #fff\n          );\n        }\n        ul,\n        absolute-position-behavior {\n          width: 100%;\n        }\n        absolute-position-behavior.row,\n        absolute-position-behavior.row ul {\n          width: 150px;\n        }\n        a11y-menu-button-item {\n          font-family: var(\n            --editable-table-secondary-font-family,\n            \"Roboto\",\n            \"Noto\",\n            sans-serif\n          );\n          color: var(--editable-table-color, #222);\n          font-size: var(--editable-table-secondary-font-size, 12px);\n        }\n      "]);

    _templateObject4_3c95187099b011eb851ed31c596cbbd8 = function _templateObject4_3c95187099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _templateObject3_3c95187099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["<span class=\"sr-only\">", "</span>"]);

    _templateObject3_3c95187099b011eb851ed31c596cbbd8 = function _templateObject3_3c95187099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _templateObject2_3c95187099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral([" <a11y-menu-button-item\n      @click=\"", "\"\n    >\n      ", "\n      ", "", "\n      ", "\n    </a11y-menu-button-item>"]);

    _templateObject2_3c95187099b011eb851ed31c596cbbd8 = function _templateObject2_3c95187099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _templateObject_3c95187099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <button\n        id=\"menubutton\"\n        aria-haspopup=\"true\"\n        aria-controls=\"menu\"\n        aria-expanded=\"", "\"\n      >\n        <span class=\"sr-only\">", "</span>\n        <span id=\"label\">", " </span>\n        <span class=\"sr-only\">Menu</span>\n        <simple-icon-lite icon=\"arrow-drop-down\"></simple-icon-lite>\n      </button>\n      <absolute-position-behavior\n        ?auto=\"", "\"\n        class=\"", "\"\n        for=\"menubutton\"\n        position=\"", "\"\n        position-align=\"", "\"\n        offset=\"-3\"\n      >\n        <ul\n          id=\"menu\"\n          role=\"menu\"\n          aria-labelledby=\"menubutton\"\n          ?hidden=\"", "\"\n          @mousover=\"", "\"\n          @mousout=\"", "\"\n        >\n          ", " ", "\n          ", "\n        </ul>\n      </absolute-position-behavior>\n    "]);

    _templateObject_3c95187099b011eb851ed31c596cbbd8 = function _templateObject_3c95187099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  /**
   * `editable-table-editor-rowcol`
   * A header label and menu for inserting and deleting a row or a column of editable-table interface (editable-table.html).
   *
   * @demo ./demo/editor.html
   * @customElement
   * @extends cellBehaviors
   * @extends A11yMenuButton
   */
  var EditableTableEditorRowcol = /*#__PURE__*/function (_cellBehaviors) {
    babelHelpers.inherits(EditableTableEditorRowcol, _cellBehaviors);

    var _super = _createSuper(EditableTableEditorRowcol);

    function EditableTableEditorRowcol() {
      babelHelpers.classCallCheck(this, EditableTableEditorRowcol);
      return _super.apply(this, arguments);
    }

    babelHelpers.createClass(EditableTableEditorRowcol, [{
      key: "render",
      value: function render() {
        var _this = this;

        return (0, _litElement.html)(_templateObject_3c95187099b011eb851ed31c596cbbd8(), this.expanded ? "true" : "false", this.type, this.label || "", this.expanded, this.row ? "row" : "column", this.row ? "right" : "bottom", this.row ? "start" : "center", !this.expanded, function (e) {
          return _this.hover = true;
        }, function (e) {
          return _this.hover = false;
        }, this._getItem(), this._getItem(false, true), this._getItem(true));
      }
    }, {
      key: "rowColAction",

      /**
       * Fires when  selection is made from menu button
       * @event delete-rowcol
       * @param {number} index index to perform action
       * @param {boolean} whether action is to insert
       */
      value: function rowColAction() {
        var index = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.index;
        var insert = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
        this.dispatchEvent(new CustomEvent("rowcol-action", {
          bubbles: true,
          cancelable: true,
          composed: true,
          detail: {
            insert: insert,
            row: this.row,
            index: index
          }
        }));
      }
    }, {
      key: "_getItem",
      value: function _getItem() {
        var deleteItem = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : false;
        var after = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
        return (0, _litElement.html)(_templateObject2_3c95187099b011eb851ed31c596cbbd8(), deleteItem ? this._onDelete : after ? this._onInsertAfter : this._onInsertBefore, deleteItem ? "Delete" : "Insert", this.type, deleteItem ? " " : after ? " After " : " Before ", this.labelInfo);
      }
      /**
       * Handles when Delete Row/Column is clicked
       * @param {event} e button event
       */

    }, {
      key: "_onDelete",
      value: function _onDelete(e) {
        this.rowColAction(this.index, false);
      }
      /**
       * Handles when Insert Row/Column is clicked
       * @param {event} e button event
       */

    }, {
      key: "_onInsertBefore",
      value: function _onInsertBefore(e) {
        this.rowColAction(this.row ? this.index - 1 : this.index);
      }
      /**
       * Handles when Insert Row/Column After is clicked
       * @param {event} e button event
       */

    }, {
      key: "_onInsertAfter",
      value: function _onInsertAfter(e) {
        this.rowColAction(this.row ? this.index : this.index + 1);
      }
    }, {
      key: "controls",

      /**
       *
       * Gets first cell that menu controls
       * @readonly
       * @memberof EditableTableEditorRowcol
       */
      get: function get() {
        return this.row ? "cell-0-".concat(this.index) : "cell-".concat(this.index, "-0");
      }
      /**
       *
       * Gets row or column label based on type
       * @readonly
       * @memberof EditableTableEditorRowcol
       */

    }, {
      key: "label",
      get: function get() {
        return this.row ? this._getLabel(this.index, true) : this._getLabel(this.index, false);
      }
      /**
       *
       * get cell label
       * @readonly
       * @memberof EditableTableEditorRowcol
       */

    }, {
      key: "labelInfo",
      get: function get() {
        return (0, _litElement.html)(_templateObject3_3c95187099b011eb851ed31c596cbbd8(), this.label);
      }
      /**
       *
       * Gets row or column type
       * @readonly
       * @memberof EditableTableEditorRowcol
       */

    }, {
      key: "type",
      get: function get() {
        return this.row ? "Row" : "Column";
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(EditableTableEditorRowcol), "styles", this) || []), babelHelpers.toConsumableArray(_editableTableBehaviors.editableTableCellStyles), [(0, _litElement.css)(_templateObject4_3c95187099b011eb851ed31c596cbbd8())]);
      }
    }, {
      key: "tag",
      get: function get() {
        return "editable-table-editor-rowcol";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(EditableTableEditorRowcol), "properties", this)), {}, {
          /**
           * Index of row or column
           */
          index: {
            type: Number
          },

          /**
           * Whether menu button controls a row
           */
          row: {
            type: Boolean
          }
        });
      }
    }]);
    return EditableTableEditorRowcol;
  }((0, _editableTableBehaviors.cellBehaviors)(_a11yMenuButton.A11yMenuButton));

  _exports.EditableTableEditorRowcol = EditableTableEditorRowcol;
  window.customElements.define(EditableTableEditorRowcol.tag, EditableTableEditorRowcol);
});
