// @haxtheweb/journey-theme/lib/journey-menu.js
import{LitElement as e,html as t,css as i,unsafeCSS as r}from"../../../lit/index.js";export class JourneyMenu extends e{static get tag(){return"journey-menu"}static get properties(){return{items:{type:Array},activeID:{type:String}}}static get styles(){return i`
      nav {
        position: fixed;
        top: 35px;
        bottom: 0;
        left: 0;
        width: 308px;
        text-align: left;
        overflow-y: auto;
        padding: var(--ddd-spacing-4);
        background: var(--ddd-theme-default-gradient-footer);
        background-size: cover;
        background-position: center;
      }

      ul {
        list-style: none;
        padding: 0;
        margin: 0;
        display: flex;
        flex-direction: column;
      }

      li a {
        display: block;
        text-decoration: none;
        color: white;
        font-size: var(--ddd-font-size-s);
        margin: 20px 0;
        cursor: pointer;
      }

      li a:hover,
      li a:focus {
        text-decoration: underline;
      }

      li a.active {
        font-weight: bold;
        color: yellow;
      }
    `}render(){let e=[];return Array.isArray(this.items)&&(e=this.items),t`
      <nav>
        <ul>
          ${e.map((e=>t`
              <li>
                <a
                  href="${e.slug}"
                  class="${e.id===this.activeID?"active":""}"
                >
                  ${e.title}
                </a>
              </li>
            `))}
        </ul>
      </nav>
    `}}customElements.define(JourneyMenu.tag,JourneyMenu);
