// @haxtheweb/citation-builder/lib/citation-item.js
/**
 * Copyright 2026 winstonwumbo
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as t,html as i,css as e}from"../../../lit/index.js";import{DDDSuper as o}from"../../d-d-d/d-d-d.js";import{I18NMixin as a}from"../../i18n-manager/lib/I18NMixin.js";import"../../simple-icon/lib/simple-icons.js";import"../../simple-icon/lib/simple-icon-button.js";import"../../hax-iconset/lib/simple-hax-iconset.js";export class CitationItem extends(o(a(t))){static get tag(){return"citation-item"}constructor(){super(),this.title="We Will Steal Ze Moon",this.authors=[{given:"John Andrew",surname:"Smith"},{given:"Jane",surname:"Doe"},{given:"Kevin",surname:"Brown"}],this.publicationDate={month:"June",day:"1",year:"2025"},this.citationType="web",this.source="HAX PSU",this.url="https://hax.psu.edu"}static get properties(){return{...super.properties,title:{type:String},citationType:{type:String},source:{type:String},authors:{type:Array},publicationDate:{type:Object},url:{type:String},volume:{type:String},issue:{type:String}}}static get styles(){return[super.styles,e`
      :host {
        display: block;
        color: var(--ddd-theme-primary);
        background-color: var(--ddd-theme-accent);
        font-family: var(--ddd-font-navigation);
      }
      .wrapper {
        margin: var(--ddd-spacing-2);
        padding: var(--ddd-spacing-4);
        border: grey 1px solid;
        display: flex;
        justify-content: space-between;
      }
      .italic-section {
        display: inline;
        font-style: italic;
      }
    `]}render(){return i`
    <div class="wrapper">
        <div class="citation">
          ${this._authorsToAPA()}, ${this._dateToAPA()}, ${this.title}, ${this._sourceToAPA()}
        </div>
        <div class="copy-button">
            <simple-icon-button
              part="icon"
              class="icon"
              title="Copy"
              icon="content-copy"
              @click=${this._copyToClipboard}
            ></simple-icon-button>
        </div>
    </div>`}updated(t){super.updated&&super.updated(t),t.forEach((t,i)=>{"authors"===i&&void 0!==t?this._formatAuthors():"publicationDate"===i&&void 0!==t&&(this._formattedDate=this._formatDate())})}_authorsToAPA(){if(!this.authors)return"";const t=this.authors.map(t=>`${t.surname.trim()}, ${t.given.trim().replace(/\b(\w)\w*/g,"$1.").toUpperCase()}`);if(1===this.authors.length)return t[0]+".";{const i=t.pop();return t.join(", ")+", & "+i}}_dateToAPA(){return this.publicationDate.year?"journal"===this.citationType?`(${this.publicationDate.year})`:"web"===this.citationType?this.publicationDate.month&&this.publicationDate.day?`(${this.publicationDate.year}, ${this.publicationDate.month} ${parseInt(this.publicationDate.day)})`:this.publicationDate.month?`(${this.publicationDate.year}, ${this.publicationDate.month})`:`(${this.publicationDate.year})`:void 0:"(n.d.)"}_sourceToAPA(){if("journal"===this.citationType){const t=this.url;return i`${this.source}, <div class="italic-section">${this.volume}</div>(${this.issue}), ${pageStart}-${pageEnd}. ${t}`}if("web"===this.citationType)return i`<div class="italic-section">${this.source}</div>, ${this.url}`}_copyToClipboard(){const t=this.shadowRoot.querySelector(".citation");navigator.clipboard.writeText(t.textContent)}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}haxHooks(){}}globalThis.customElements.define(CitationItem.tag,CitationItem);
