// @lrnwebcomponents/haxcms-elements/lib/core/haxcms-site-editor-ui.js
define(["exports", "meta", "require", "../../../../lit/index.js", "./haxcms-site-store.js", "../../../hax-body/lib/hax-store.js", "../../../../mobx/dist/mobx.esm.js", "../../../utils/utils.js", "../../../simple-icon/simple-icon.js", "../../../simple-icon/lib/simple-icons.js", "./utils/HAXCMSThemeParts.js", "./utils/HAXCMSI18NMixin.js", "../../../rpg-character/rpg-character.js", "../../../app-hax/lib/v1/app-hax-top-bar.js", "../../../app-hax/lib/v1/app-hax-user-menu.js", "../../../app-hax/lib/v1/app-hax-user-menu-button.js", "../../../../wired-elements/lib/wired-button.js", "../../../simple-colors/simple-colors.js", "../../../super-daemon/super-daemon.js", "../../../simple-modal/simple-modal.js", "./haxcms-site-insights.js", "../../../simple-fields/lib/simple-fields-form.js", "./micros/haxcms-button-add.js", "./haxcms-darkmode-toggle.js", "../ui-components/site/site-remote-content.js", "../../../page-flag/page-flag.js"], function (_exports, meta, _require, _index, _haxcmsSiteStore, _haxStore, _mobxEsm, _utils, _simpleIcon, _simpleIcons, _HAXCMSThemeParts2, _HAXCMSI18NMixin, _rpgCharacter, _appHaxTopBar, _appHaxUserMenu, _appHaxUserMenuButton, _wiredButton, _simpleColors, _superDaemon, _simpleModal, _haxcmsSiteInsights, _simpleFieldsForm, _haxcmsButtonAdd, _haxcmsDarkmodeToggle, _siteRemoteContent, _pageFlag) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSSiteEditorUI = void 0;
  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);
  var _templateObject_724e82a0ddf411ed96ad39cc5dd30a65, _templateObject2_724e82a0ddf411ed96ad39cc5dd30a65, _templateObject3_724e82a0ddf411ed96ad39cc5dd30a65;
  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
  function _regeneratorRuntime() { "use strict"; /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/facebook/regenerator/blob/main/LICENSE */ _regeneratorRuntime = function _regeneratorRuntime() { return exports; }; var exports = {}, Op = Object.prototype, hasOwn = Op.hasOwnProperty, defineProperty = Object.defineProperty || function (obj, key, desc) { obj[key] = desc.value; }, $Symbol = "function" == typeof Symbol ? Symbol : {}, iteratorSymbol = $Symbol.iterator || "@@iterator", asyncIteratorSymbol = $Symbol.asyncIterator || "@@asyncIterator", toStringTagSymbol = $Symbol.toStringTag || "@@toStringTag"; function define(obj, key, value) { return Object.defineProperty(obj, key, { value: value, enumerable: !0, configurable: !0, writable: !0 }), obj[key]; } try { define({}, ""); } catch (err) { define = function define(obj, key, value) { return obj[key] = value; }; } function wrap(innerFn, outerFn, self, tryLocsList) { var protoGenerator = outerFn && babelHelpers.instanceof(outerFn.prototype, Generator) ? outerFn : Generator, generator = Object.create(protoGenerator.prototype), context = new Context(tryLocsList || []); return defineProperty(generator, "_invoke", { value: makeInvokeMethod(innerFn, self, context) }), generator; } function tryCatch(fn, obj, arg) { try { return { type: "normal", arg: fn.call(obj, arg) }; } catch (err) { return { type: "throw", arg: err }; } } exports.wrap = wrap; var ContinueSentinel = {}; function Generator() {} function GeneratorFunction() {} function GeneratorFunctionPrototype() {} var IteratorPrototype = {}; define(IteratorPrototype, iteratorSymbol, function () { return this; }); var getProto = Object.getPrototypeOf, NativeIteratorPrototype = getProto && getProto(getProto(values([]))); NativeIteratorPrototype && NativeIteratorPrototype !== Op && hasOwn.call(NativeIteratorPrototype, iteratorSymbol) && (IteratorPrototype = NativeIteratorPrototype); var Gp = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(IteratorPrototype); function defineIteratorMethods(prototype) { ["next", "throw", "return"].forEach(function (method) { define(prototype, method, function (arg) { return this._invoke(method, arg); }); }); } function AsyncIterator(generator, PromiseImpl) { function invoke(method, arg, resolve, reject) { var record = tryCatch(generator[method], generator, arg); if ("throw" !== record.type) { var result = record.arg, value = result.value; return value && "object" == babelHelpers.typeof(value) && hasOwn.call(value, "__await") ? PromiseImpl.resolve(value.__await).then(function (value) { invoke("next", value, resolve, reject); }, function (err) { invoke("throw", err, resolve, reject); }) : PromiseImpl.resolve(value).then(function (unwrapped) { result.value = unwrapped, resolve(result); }, function (error) { return invoke("throw", error, resolve, reject); }); } reject(record.arg); } var previousPromise; defineProperty(this, "_invoke", { value: function value(method, arg) { function callInvokeWithMethodAndArg() { return new PromiseImpl(function (resolve, reject) { invoke(method, arg, resolve, reject); }); } return previousPromise = previousPromise ? previousPromise.then(callInvokeWithMethodAndArg, callInvokeWithMethodAndArg) : callInvokeWithMethodAndArg(); } }); } function makeInvokeMethod(innerFn, self, context) { var state = "suspendedStart"; return function (method, arg) { if ("executing" === state) throw new Error("Generator is already running"); if ("completed" === state) { if ("throw" === method) throw arg; return doneResult(); } for (context.method = method, context.arg = arg;;) { var delegate = context.delegate; if (delegate) { var delegateResult = maybeInvokeDelegate(delegate, context); if (delegateResult) { if (delegateResult === ContinueSentinel) continue; return delegateResult; } } if ("next" === context.method) context.sent = context._sent = context.arg;else if ("throw" === context.method) { if ("suspendedStart" === state) throw state = "completed", context.arg; context.dispatchException(context.arg); } else "return" === context.method && context.abrupt("return", context.arg); state = "executing"; var record = tryCatch(innerFn, self, context); if ("normal" === record.type) { if (state = context.done ? "completed" : "suspendedYield", record.arg === ContinueSentinel) continue; return { value: record.arg, done: context.done }; } "throw" === record.type && (state = "completed", context.method = "throw", context.arg = record.arg); } }; } function maybeInvokeDelegate(delegate, context) { var methodName = context.method, method = delegate.iterator[methodName]; if (undefined === method) return context.delegate = null, "throw" === methodName && delegate.iterator.return && (context.method = "return", context.arg = undefined, maybeInvokeDelegate(delegate, context), "throw" === context.method) || "return" !== methodName && (context.method = "throw", context.arg = new TypeError("The iterator does not provide a '" + methodName + "' method")), ContinueSentinel; var record = tryCatch(method, delegate.iterator, context.arg); if ("throw" === record.type) return context.method = "throw", context.arg = record.arg, context.delegate = null, ContinueSentinel; var info = record.arg; return info ? info.done ? (context[delegate.resultName] = info.value, context.next = delegate.nextLoc, "return" !== context.method && (context.method = "next", context.arg = undefined), context.delegate = null, ContinueSentinel) : info : (context.method = "throw", context.arg = new TypeError("iterator result is not an object"), context.delegate = null, ContinueSentinel); } function pushTryEntry(locs) { var entry = { tryLoc: locs[0] }; 1 in locs && (entry.catchLoc = locs[1]), 2 in locs && (entry.finallyLoc = locs[2], entry.afterLoc = locs[3]), this.tryEntries.push(entry); } function resetTryEntry(entry) { var record = entry.completion || {}; record.type = "normal", delete record.arg, entry.completion = record; } function Context(tryLocsList) { this.tryEntries = [{ tryLoc: "root" }], tryLocsList.forEach(pushTryEntry, this), this.reset(!0); } function values(iterable) { if (iterable) { var iteratorMethod = iterable[iteratorSymbol]; if (iteratorMethod) return iteratorMethod.call(iterable); if ("function" == typeof iterable.next) return iterable; if (!isNaN(iterable.length)) { var i = -1, next = function next() { for (; ++i < iterable.length;) if (hasOwn.call(iterable, i)) return next.value = iterable[i], next.done = !1, next; return next.value = undefined, next.done = !0, next; }; return next.next = next; } } return { next: doneResult }; } function doneResult() { return { value: undefined, done: !0 }; } return GeneratorFunction.prototype = GeneratorFunctionPrototype, defineProperty(Gp, "constructor", { value: GeneratorFunctionPrototype, configurable: !0 }), defineProperty(GeneratorFunctionPrototype, "constructor", { value: GeneratorFunction, configurable: !0 }), GeneratorFunction.displayName = define(GeneratorFunctionPrototype, toStringTagSymbol, "GeneratorFunction"), exports.isGeneratorFunction = function (genFun) { var ctor = "function" == typeof genFun && genFun.constructor; return !!ctor && (ctor === GeneratorFunction || "GeneratorFunction" === (ctor.displayName || ctor.name)); }, exports.mark = function (genFun) { return Object.setPrototypeOf ? Object.setPrototypeOf(genFun, GeneratorFunctionPrototype) : (genFun.__proto__ = GeneratorFunctionPrototype, define(genFun, toStringTagSymbol, "GeneratorFunction")), genFun.prototype = Object.create(Gp), genFun; }, exports.awrap = function (arg) { return { __await: arg }; }, defineIteratorMethods(AsyncIterator.prototype), define(AsyncIterator.prototype, asyncIteratorSymbol, function () { return this; }), exports.AsyncIterator = AsyncIterator, exports.async = function (innerFn, outerFn, self, tryLocsList, PromiseImpl) { void 0 === PromiseImpl && (PromiseImpl = Promise); var iter = new AsyncIterator(wrap(innerFn, outerFn, self, tryLocsList), PromiseImpl); return exports.isGeneratorFunction(outerFn) ? iter : iter.next().then(function (result) { return result.done ? result.value : iter.next(); }); }, defineIteratorMethods(Gp), define(Gp, toStringTagSymbol, "Generator"), define(Gp, iteratorSymbol, function () { return this; }), define(Gp, "toString", function () { return "[object Generator]"; }), exports.keys = function (val) { var object = Object(val), keys = []; for (var key in object) keys.push(key); return keys.reverse(), function next() { for (; keys.length;) { var key = keys.pop(); if (key in object) return next.value = key, next.done = !1, next; } return next.done = !0, next; }; }, exports.values = values, Context.prototype = { constructor: Context, reset: function reset(skipTempReset) { if (this.prev = 0, this.next = 0, this.sent = this._sent = undefined, this.done = !1, this.delegate = null, this.method = "next", this.arg = undefined, this.tryEntries.forEach(resetTryEntry), !skipTempReset) for (var name in this) "t" === name.charAt(0) && hasOwn.call(this, name) && !isNaN(+name.slice(1)) && (this[name] = undefined); }, stop: function stop() { this.done = !0; var rootRecord = this.tryEntries[0].completion; if ("throw" === rootRecord.type) throw rootRecord.arg; return this.rval; }, dispatchException: function dispatchException(exception) { if (this.done) throw exception; var context = this; function handle(loc, caught) { return record.type = "throw", record.arg = exception, context.next = loc, caught && (context.method = "next", context.arg = undefined), !!caught; } for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i], record = entry.completion; if ("root" === entry.tryLoc) return handle("end"); if (entry.tryLoc <= this.prev) { var hasCatch = hasOwn.call(entry, "catchLoc"), hasFinally = hasOwn.call(entry, "finallyLoc"); if (hasCatch && hasFinally) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } else if (hasCatch) { if (this.prev < entry.catchLoc) return handle(entry.catchLoc, !0); } else { if (!hasFinally) throw new Error("try statement without catch or finally"); if (this.prev < entry.finallyLoc) return handle(entry.finallyLoc); } } } }, abrupt: function abrupt(type, arg) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc <= this.prev && hasOwn.call(entry, "finallyLoc") && this.prev < entry.finallyLoc) { var finallyEntry = entry; break; } } finallyEntry && ("break" === type || "continue" === type) && finallyEntry.tryLoc <= arg && arg <= finallyEntry.finallyLoc && (finallyEntry = null); var record = finallyEntry ? finallyEntry.completion : {}; return record.type = type, record.arg = arg, finallyEntry ? (this.method = "next", this.next = finallyEntry.finallyLoc, ContinueSentinel) : this.complete(record); }, complete: function complete(record, afterLoc) { if ("throw" === record.type) throw record.arg; return "break" === record.type || "continue" === record.type ? this.next = record.arg : "return" === record.type ? (this.rval = this.arg = record.arg, this.method = "return", this.next = "end") : "normal" === record.type && afterLoc && (this.next = afterLoc), ContinueSentinel; }, finish: function finish(finallyLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.finallyLoc === finallyLoc) return this.complete(entry.completion, entry.afterLoc), resetTryEntry(entry), ContinueSentinel; } }, catch: function _catch(tryLoc) { for (var i = this.tryEntries.length - 1; i >= 0; --i) { var entry = this.tryEntries[i]; if (entry.tryLoc === tryLoc) { var record = entry.completion; if ("throw" === record.type) { var thrown = record.arg; resetTryEntry(entry); } return thrown; } } throw new Error("illegal catch attempt"); }, delegateYield: function delegateYield(iterable, resultName, nextLoc) { return this.delegate = { iterator: values(iterable), resultName: resultName, nextLoc: nextLoc }, "next" === this.method && (this.arg = undefined), ContinueSentinel; } }, exports; }
  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  var haxLogo = new URL("../../../app-hax/lib/assets/images/HAXLogo.svg", meta.url).href;
  var ButtonBGLight = new URL("../../../app-hax/lib/assets/images/ButtonBGLM.svg", meta.url).href;
  var ButtonBGDark = new URL("../../../app-hax/lib/assets/images/ButtonBGDM.svg", meta.url).href;
  var LogOut = new URL("../../../app-hax/lib/assets/images/Logout.svg", meta.url).href;
  /**
   * `haxcms-site-editor-ui`
   * `haxcms editor element buttons that you see`
   *
   * @demo demo/index.html
   * @microcopy - the mental model for this element
   */
  var HAXCMSSiteEditorUI = /*#__PURE__*/function (_HAXCMSThemeParts) {
    babelHelpers.inherits(HAXCMSSiteEditorUI, _HAXCMSThemeParts);
    var _super = _createSuper(HAXCMSSiteEditorUI);
    function HAXCMSSiteEditorUI() {
      var _this;
      babelHelpers.classCallCheck(this, HAXCMSSiteEditorUI);
      _this = _super.call(this);
      _this.rpgHat = "none";
      _this.darkMode = false;
      _this.__settingsText = "";
      _this.__editText = "";
      _this.userMenuOpen = false;
      _this.soundIcon = "";
      _this.__disposer = _this.__disposer || [];
      _this.t = _this.t || {};
      // allow commands to go through at any time
      // hax-store default is only when editor is open to avoid conflicts w/ other UX
      _superDaemon.SuperDaemonInstance.allowedCallback = function () {
        return true;
      };
      // nothing message so we can suggest a link to make a suggestion
      _superDaemon.SuperDaemonInstance.noResultsSlot = function () {
        return (0, _index.html)(_templateObject_724e82a0ddf411ed96ad39cc5dd30a65 || (_templateObject_724e82a0ddf411ed96ad39cc5dd30a65 = babelHelpers.taggedTemplateLiteral(["Expecting to see results?\n        <a\n          @click=\"", "\"\n          >Make a suggestion</a\n        >\n        for what you thought you would see here!"])), function (e) {
          _haxStore.HAXStore._haxStoreContribute("merlin", "POP,enhancement", _superDaemon.SuperDaemonInstance.commandContext + " " + _superDaemon.SuperDaemonInstance.like);
        });
      };
      _superDaemon.SuperDaemonInstance.appendContext("CMS");
      window.addEventListener("hax-store-ready", _this.haxStoreReady.bind(babelHelpers.assertThisInitialized(_this)));
      if (_haxStore.HAXStore.ready) {
        var s = document.createElement("site-remote-content");
        _haxStore.HAXStore.haxAutoloader.appendChild(s);
        // site-remote-content injects citation element so ensure it's in there too!
        var ce = document.createElement("citation-element");
        _haxStore.HAXStore.haxAutoloader.appendChild(ce);
        // page-flag for author notes
        var flag = document.createElement("page-flag");
        _haxStore.HAXStore.haxAutoloader.appendChild(flag);

        // links need to be given support for internal linkage updates on the form
        if (!_haxStore.HAXStore.primativeHooks.a) {
          _haxStore.HAXStore.primativeHooks.a = {};
        }
        _haxStore.HAXStore.primativeHooks.a.setupActiveElementForm = function (props) {
          var itemManifest = _haxcmsSiteStore.store.getManifestItems(true);
          // default to null parent as the whole site
          var items = [{
            text: "-- ".concat(_this.t.selectPage, " --"),
            value: null
          }];
          itemManifest.forEach(function (el) {
            if (el.id != _this.itemId) {
              // calculate -- depth so it looks like a tree
              var itemBuilder = el;
              // walk back through parent tree
              var distance = "- ";
              while (itemBuilder && itemBuilder.parent != null) {
                itemBuilder = itemManifest.find(function (i) {
                  return i.id == itemBuilder.parent;
                });
                // double check structure is sound
                if (itemBuilder) {
                  distance = "--" + distance;
                }
              }
              items.push({
                text: distance + el.title,
                value: el.id
              });
            }
          });
          props.settings.configure.splice(1, 0, {
            property: "data-uuid",
            title: "Internal content",
            description: "Link to content internal to this site",
            inputMethod: "select",
            itemsList: items
          });
          setTimeout(function () {
            _haxStore.HAXStore.haxTray.shadowRoot.querySelector("[name='settings.configure.data-uuid']").addEventListener("value-changed", _this._internalLinkChanged.bind(babelHelpers.assertThisInitialized(_this)));
          }, 100);
        };
      }
      _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
        selectPage: "Select page",
        backToSiteList: "Back to site list",
        cancelEditing: "Cancel editing",
        editDetails: "Page details",
        addPage: "Add page",
        addChildPage: "Add child page",
        duplicatePage: "Duplicate page",
        importDocxFile: "Import docx file",
        deletePage: "Delete page",
        shareSite: "Share site",
        closeSiteSettings: "Close site settings",
        editSiteSettings: "Site settings",
        savePageContent: "Save page",
        editPageContent: "Edit page",
        newJourney: "New Journey",
        accountInfo: "Account Info",
        outlineDesigner: "Outline designer",
        more: "More",
        insights: "Insights",
        logOut: "Log out",
        menu: "Menu",
        showMore: "More"
      });
      _this.backText = "Site list";
      _this.painting = true;
      _this.pageAllowed = false;
      _this.editMode = false;
      _this.__editIcon = "hax:page-edit";
      _this.icon = "hax:site-settings";
      _this.manifestEditMode = false;
      _this.backLink = "../../";
      if (window.appSettings && window.appSettings.backLink) {
        _this.backLink = window.appSettings.backLink;
      }
      (0, _mobxEsm.autorun)(function () {
        var badDevice = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.badDevice);
        // good device, we can inject font we use
        if (badDevice === false) {
          var link = document.createElement("link");
          link.setAttribute("href", "https://fonts.googleapis.com/css2?family=Press+Start+2P&display=swap");
          link.setAttribute("rel", "stylesheet");
          link.setAttribute("fetchpriority", "low");
          document.head.appendChild(link);
        } else if (badDevice === true) {
          document.body.classList.add("bad-device");
        }
      });
      (0, _mobxEsm.autorun)(function () {
        _this.darkMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.darkMode);
        _this.dark = _this.darkMode;
        if ((0, _mobxEsm.toJS)(_haxcmsSiteStore.store.darkMode)) {
          _haxStore.HAXStore.globalPreferences.haxUiTheme = "haxdark";
        } else {
          _haxStore.HAXStore.globalPreferences.haxUiTheme = "hax";
        }
      });
      (0, _mobxEsm.autorun)(function () {
        _this.soundIcon = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.soundStatus) ? new URL("../../../app-hax/lib/assets/images/FullVolume.svg", meta.url).href : new URL("../../../app-hax/lib/assets/images/Silence.svg", meta.url).href;
      });
      setTimeout(function () {
        // prettier-ignore
        new Promise(function (res, rej) {
          return _require.default(["./haxcms-outline-editor-dialog.js"], res, rej);
        });
        // prettier-ignore
        new Promise(function (res, rej) {
          return _require.default(["./haxcms-share-dialog.js"], res, rej);
        });
        // prettier-ignore
        new Promise(function (res, rej) {
          return _require.default(["./haxcms-site-dashboard.js"], res, rej);
        });
      }, 0);
      return _this;
    }
    babelHelpers.createClass(HAXCMSSiteEditorUI, [{
      key: "_internalLinkChanged",
      value:
      // ability to link to internal content inside of the 'a' tag
      function _internalLinkChanged(e) {
        // look up the UUID
        if (e.detail.value && _haxcmsSiteStore.store.findItem(e.detail.value)) {
          var item = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.findItem(e.detail.value));
          _haxStore.HAXStore.haxTray.shadowRoot.querySelector("[name='settings.configure.href']").value = item.slug;
        }
      }
    }, {
      key: "haxStoreReady",
      value: function haxStoreReady(e) {
        // it is safe to add in elements that we want the editor to have every haxcms instance
        if (e.detail) {
          var s = document.createElement("site-remote-content");
          _haxStore.HAXStore.haxAutoloader.appendChild(s);
        }
      }
    }, {
      key: "soundToggle",
      value: function soundToggle() {
        var status = !(0, _mobxEsm.toJS)(_haxcmsSiteStore.store.soundStatus);
        _haxcmsSiteStore.store.soundStatus = status;
        (0, _utils.localStorageSet)("app-hax-soundStatus", status);
        if (!status) {
          _haxcmsSiteStore.store.toast("Sound off.. hey.. HELLO!?!", 2000, {
            fire: true
          });
        } else {
          _haxcmsSiteStore.store.toast("Can you hear me now? Good.", 2000, {
            hat: "random"
          });
          _haxcmsSiteStore.store.playSound("click");
        }
      }
    }, {
      key: "toggleMenu",
      value: function toggleMenu() {
        this.userMenuOpen = !this.userMenuOpen;
        _haxcmsSiteStore.store.playSound("click");
      }
      // render function
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject2_724e82a0ddf411ed96ad39cc5dd30a65 || (_templateObject2_724e82a0ddf411ed96ad39cc5dd30a65 = babelHelpers.taggedTemplateLiteral(["\n      <app-hax-top-bar part=\"top-bar\">\n        <span slot=\"left\" class=\"home-btn\">\n          <a\n            href=\"", "\"\n            class=\"haxLogo\"\n            id=\"backtosites\"\n            part=\"hax-logo\"\n          >\n            <simple-icon-lite src=\"", "\"></simple-icon-lite>\n          </a>\n          <simple-tooltip for=\"backtosites\" position=\"right\"\n            >", "</simple-tooltip\n          >\n          <slot name=\"haxcms-site-editor-ui-prefix-avatar\"></slot>\n        </span>\n        <simple-toolbar\n          slot=\"center\"\n          ?dark=\"", "\"\n          icon-position=\"top\"\n          show-text-label\n          label=\"", "\"\n        >\n          <slot name=\"haxcms-site-editor-ui-prefix-buttons\"></slot>\n          <simple-toolbar-button\n            hidden\n            id=\"editbutton\"\n            icon=\"", "\"\n            icon-position=\"top\"\n            @click=\"", "\"\n            label=\"", "\"\n            show-text-label\n            voice-command=\"edit (this) page\"\n          ></simple-toolbar-button>\n          <simple-toolbar-menu\n            id=\"addmenubutton\"\n            ?disabled=\"", "\"\n            icon=\"hax:add-page\"\n            icon-position=\"top\"\n            label=\"", "\"\n            tabindex=\"", "\"\n            @dblclick=\"", "\"\n            show-text-label\n          >\n            <simple-toolbar-menu-item>\n              <haxcms-button-add\n                hidden\n                ?disabled=\"", "\"\n                show-text-label\n                id=\"addbutton\"\n                show-text-label\n                data-super-daemon-label=\"", "\"\n                data-super-daemon-icon=\"hax:add-page\"\n                data-super-daemon-path=\"CMS/action/create/page\"\n              >\n              </haxcms-button-add>\n            </simple-toolbar-menu-item>\n            <simple-toolbar-menu-item>\n              <haxcms-button-add\n                hidden\n                ?disabled=\"", "\"\n                id=\"addbuttonchild\"\n                type=\"child\"\n                show-text-label\n                data-super-daemon-path=\"CMS/action/create/child\"\n                data-super-daemon-label=\"", "\"\n                data-super-daemon-icon=\"hax:add-child-page\"\n              ></haxcms-button-add>\n            </simple-toolbar-menu-item>\n            <simple-toolbar-menu-item>\n              <haxcms-button-add\n                hidden\n                ?disabled=\"", "\"\n                type=\"duplicate\"\n                id=\"duplicatebutton\"\n                show-text-label\n                data-super-daemon-path=\"CMS/action/create/duplicate\"\n                data-super-daemon-label=\"", "\"\n                data-super-daemon-icon=\"hax:duplicate\"\n              ></haxcms-button-add>\n            </simple-toolbar-menu-item>\n            <simple-toolbar-menu-item>\n              <haxcms-button-add\n                hidden\n                ?disabled=\"", "\"\n                type=\"docximport\"\n                id=\"docximport\"\n                show-text-label\n                data-super-daemon-path=\"CMS/action/import/docx\"\n                data-super-daemon-label=\"", "\"\n                data-super-daemon-icon=\"hax:file-docx\"\n              ></haxcms-button-add>\n            </simple-toolbar-menu-item>\n          </simple-toolbar-menu>\n          <simple-toolbar-button\n            ?disabled=\"", "\"\n            tabindex=\"", "\"\n            id=\"deletebutton\"\n            hidden\n            icon-position=\"top\"\n            icon=\"icons:delete\"\n            @click=\"", "\"\n            label=\"", "\"\n            show-text-label\n            voice-command=\"delete page\"\n          ></simple-toolbar-button>\n          <simple-toolbar-button\n            ?disabled=\"", "\"\n            tabindex=\"", "\"\n            id=\"outlinebutton\"\n            @click=\"", "\"\n            icon-position=\"top\"\n            icon=\"hax:site-map\"\n            part=\"outlinebtn\"\n            show-text-label\n            label=\"", "\"\n          ></simple-toolbar-button>\n          <simple-toolbar-button\n            ?disabled=\"", "\"\n            tabindex=\"", "\"\n            id=\"insightsbutton\"\n            icon=\"hax:clipboard-pulse\"\n            part=\"insightsbtn\"\n            icon-position=\"top\"\n            @click=\"", "\"\n            label=\"", "\"\n            show-text-label\n            voice-command=\"insights\"\n          ></simple-toolbar-button>\n          <simple-toolbar-menu\n            show-text-label\n            ?disabled=\"", "\"\n            icon=\"more-vert\"\n            part=\"morebtn\"\n            icon-position=\"top\"\n            label=\"", "\"\n            tabindex=\"", "\"\n          >\n            <simple-toolbar-menu-item>\n              <simple-toolbar-button\n                id=\"editdetails\"\n                hidden\n                ?disabled=\"", "\"\n                icon=\"hax:page-details\"\n                icon-position=\"left\"\n                @click=\"", "\"\n                label=\"", "\"\n                show-text-label\n                voice-command=\"edit (page) details\"\n                part=\"detailsbtn\"\n                tabindex=\"", "\"\n              ></simple-toolbar-button>\n            </simple-toolbar-menu-item>\n            <simple-toolbar-menu-item>\n              <simple-toolbar-button\n                @click=\"", "\"\n                icon-position=\"left\"\n                icon=\"", "\"\n                part=\"manifestbtn\"\n                id=\"manifestbtn\"\n                show-text-label\n                tabindex=\"", "\"\n                label=\"", "\"\n              ></simple-toolbar-button>\n            </simple-toolbar-menu-item>\n          </simple-toolbar-menu>\n          <simple-toolbar-button\n            id=\"cancelbutton\"\n            icon=\"icons:cancel\"\n            icon-position=\"top\"\n            @click=\"", "\"\n            .hidden=\"", "\"\n            show-text-label\n            tabindex=\"", "\"\n            label=\"", "\"\n            voice-command=\"cancel (editing)\"\n          ></simple-toolbar-button>\n          <slot name=\"haxcms-site-editor-ui-suffix-buttons\"></slot>\n        </simple-toolbar>\n\n        <app-hax-user-menu slot=\"right\" id=\"user-menu\" part=\"app-hax-user-menu\">\n          <button\n            class=\"topbar-character\"\n            slot=\"menuButton\"\n            @click=\"", "\"\n          >\n            <rpg-character\n              seed=\"", "\"\n              width=\"68\"\n              height=\"68\"\n              part=\"rpgcharacter\"\n              aria-label=\"", "\"\n              hat=\"", "\"\n            ></rpg-character>\n            <span class=\"characterbtn-name\">", "</span>\n            <slot name=\"haxcms-site-editor-ui-topbar-character-button\"></slot>\n          </button>\n          <div slot=\"pre-menu\" class=\"ops-panel\">\n            <slot name=\"haxcms-site-editor-ui-pre-menu\"></slot>\n            <wired-button\n              elevation=\"1\"\n              class=\"soundToggle\"\n              @click=\"", "\"\n            >\n              <simple-icon-lite\n                src=\"", "\"\n                loading=\"lazy\"\n                decoding=\"async\"\n              ></simple-icon-lite>\n            </wired-button>\n            <haxcms-darkmode-toggle></haxcms-darkmode-toggle>\n          </div>\n          <!-- <app-hax-user-menu-button\n          slot=\"main-menu\"\n          icon=\"face\"\n          label=\"", "\"\n        ></app-hax-user-menu-button> -->\n          <slot slot=\"main-menu\" name=\"haxcms-site-editor-ui-main-menu\"></slot>\n\n          <app-hax-user-menu-button\n            id=\"sharebutton\"\n            @click=\"", "\"\n            slot=\"main-menu\"\n            icon=\"social:share\"\n            part=\"sharebtn\"\n            label=\"", "\"\n          ></app-hax-user-menu-button>\n\n          <app-hax-user-menu-button\n            slot=\"main-menu\"\n            icon=\"add\"\n            label=\"", "\"\n            part=\"newjourneybtn\"\n            @click=\"", "\"\n          ></app-hax-user-menu-button>\n          <slot slot=\"post-menu\" name=\"haxcms-site-editor-ui-post-menu\"></slot>\n          <app-hax-user-menu-button\n            slot=\"post-menu\"\n            part=\"logoutbtn\"\n            class=\"logout\"\n            label=\"", "\"\n            @click=", "\n          ></app-hax-user-menu-button>\n        </app-hax-user-menu>\n      </app-hax-top-bar>\n    "])), this.backLink, haxLogo, this.backText, !this.darkMode, this.t.showMore, this.__editIcon, this._editButtonTap, this.__editText, this.editMode, this.t.addPage, this.editMode ? "-1" : "0", this._addPageClick, this.editMode, this.t.addPage, this.editMode, this.t.addChildPage, this.editMode, this.t.duplicatePage, this.editMode, this.t.importDocxFile, this.editMode, this.editMode ? "-1" : "0", this._deleteButtonTap, this.t.deletePage, this.editMode, this.editMode ? "-1" : "0", this._outlineButtonTap, this.t.outlineDesigner, this.editMode, this.editMode ? "-1" : "0", this._insightsButtonTap, this.t.insights, this.editMode, this.t.more, this.editMode ? "-1" : "0", this.editMode, this._editDetailsButtonTap, this.t.editDetails, this.editMode ? "-1" : "0", this._manifestButtonTap, this.icon, this.editMode ? "0" : "-1", this.__settingsText, this._cancelButtonTap, !this.editMode, this.editMode ? "0" : "-1", this.t.cancelEditing, this.toggleMenu, this.userName, this.t.menu, this.rpgHat, this.userName, this.soundToggle, this.soundIcon, this.t.accountInfo, this._shareButtonTap, this.t.shareSite, this.t.newJourney, this._addButtonTap, this.t.logOut, this._logout);
      }
    }, {
      key: "_logout",
      value: function _logout() {
        window.dispatchEvent(new CustomEvent("jwt-login-logout", {
          composed: true,
          bubbles: true,
          cancelable: false,
          detail: true
        }));
        this.__logoutUserAction = true;
      }
      // only care about logouts
    }, {
      key: "_jwtLoggedIn",
      value: function _jwtLoggedIn(e) {
        if (e.detail === false && this.__logoutUserAction) {
          this.__logoutUserAction = false;
          setTimeout(function () {
            window.location.reload();
          }, 100);
        }
      }
    }, {
      key: "_addPageClick",
      value: function _addPageClick(e) {
        // ensure they see thes operations
        if (!this.editMode && this.shadowRoot.querySelector("simple-toolbar-menu-item haxcms-button-add:not([hidden])")) {
          this.shadowRoot.querySelector("simple-toolbar-menu-item haxcms-button-add:not([hidden])").HAXCMSButtonClick();
        }
      }

      /**
       * update buttons since these are triggered by a mix of
       * differnet backend types we can't leverage the store
       * since a CMS needs to just hardcode these at run time
       * for some environments
       */
    }, {
      key: "updateAvailableButtons",
      value: function updateAvailableButtons() {
        var _this2 = this;
        if (this.shadowRoot) {
          setTimeout(function () {
            // backText
            if (window.appSettings && window.appSettings.backText) {
              _this2.backText = window.appSettings.backText;
            }
            var ary = [{
              varPath: "getNodeFieldsPath",
              selector: "#editdetails"
            }, {
              varPath: "deleteNodePath",
              selector: "#deletebutton"
            }, {
              varPath: "saveNodePath",
              selector: "#editbutton"
            }, {
              varPath: "createNodePath",
              selector: "#addbutton"
            }, {
              varPath: "createNodePath",
              selector: "#addbuttonchild"
            }, {
              varPath: "createNodePath",
              selector: "#duplicatebutton"
            }, {
              varPath: "createNodePath",
              selector: "#docximport"
            }];
            // see which features should be enabled
            ary.forEach(function (pair) {
              if (window.appSettings && window.appSettings[pair.varPath] && window.appSettings[pair.varPath] != null && window.appSettings[pair.varPath] != "" && window.appSettings[pair.varPath] != "null") {
                if (pair.dep) {
                  if (window.appSettings[pair.dep] != null && window.appSettings[pair.dep] != "" && window.appSettings[pair.dep] != "null") {
                    _this2.shadowRoot.querySelector(pair.selector).removeAttribute("hidden");
                  } else {
                    // a dependency didn't meet the requirement
                  }
                } else {
                  _this2.shadowRoot.querySelector(pair.selector).removeAttribute("hidden");
                }
              }
            });
          }, 100);
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this3 = this;
        if (babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "firstUpdated", this).call(this, changedProperties);
        }
        this.shadowRoot.querySelectorAll("[data-super-daemon-path]").forEach(function (item) {
          _superDaemon.SuperDaemonInstance.defineOption({
            title: item.getAttribute("data-super-daemon-label"),
            icon: item.getAttribute("data-super-daemon-icon"),
            tags: ["page", "operation", "command", "add", "create"],
            value: {
              target: item
            },
            context: "CMS",
            eventName: "super-daemon-element-click",
            path: item.getAttribute("data-super-daemon-path")
          });
        });
        // load up commands for daemon
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.savePageContent,
          icon: "icons:save",
          tags: ["CMS", "save", "page", "operation", "command"],
          value: {
            target: this,
            method: "_editButtonTap"
          },
          context: "HAX",
          eventName: "super-daemon-element-method",
          path: "CMS/action/save"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.insights,
          icon: "hax:clipboard-pulse",
          tags: ["CMS", "insights", "data", "operation"],
          value: {
            target: this,
            method: "_insightsButtonTap"
          },
          context: "CMS",
          eventName: "super-daemon-element-method",
          path: "CMS/site/insights"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.editPageContent,
          icon: "hax:page-edit",
          tags: ["CMS", "edit", "page", "operation", "command"],
          value: {
            target: this,
            method: "_editButtonTap"
          },
          context: "CMS",
          eventName: "super-daemon-element-method",
          path: "CMS/action/edit"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.editSiteSettings,
          icon: "hax:site-settings",
          tags: ["CMS", "site", "settings", "operation", "command"],
          value: {
            target: this,
            method: "_manifestButtonTap",
            args: [{
              target: this.shadowRoot.querySelector("#manifestbtn")
            }]
          },
          context: "CMS",
          eventName: "super-daemon-element-method",
          path: "CMS/action/site/settings"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.shareSite,
          icon: "social:share",
          tags: ["CMS", "share"],
          value: {
            target: this,
            method: "_shareButtonTap"
          },
          context: ["logged-in", "CMS", "HAX"],
          eventName: "super-daemon-element-method",
          path: "CMS/action/share"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.newJourney,
          icon: "add",
          tags: ["CMS", "create", "new site"],
          value: {
            target: this,
            method: "_addButtonTap"
          },
          context: "CMS",
          eventName: "super-daemon-element-method",
          path: "CMS/action/create-site"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.logOut,
          icon: "add",
          tags: ["CMS", "user", "logout"],
          value: {
            target: this,
            method: "_logout"
          },
          context: ["logged-in", "CMS"],
          eventName: "super-daemon-element-method",
          path: "CMS/user/logout"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: "Dark mode toggle",
          icon: "device:brightness-medium",
          tags: ["CMS", "dark mode"],
          value: {
            target: this.shadowRoot.querySelector("haxcms-darkmode-toggle"),
            method: "toggle"
          },
          context: ["logged-in", "CMS", "HAX"],
          eventName: "super-daemon-element-method",
          path: "CMS/action/darkMode"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: "Sound toggle",
          icon: "av:volume-up",
          tags: ["CMS", "sound"],
          value: {
            target: this.shadowRoot.querySelector(".soundToggle")
          },
          context: ["logged-in", "CMS", "HAX"],
          eventName: "super-daemon-element-click",
          path: "CMS/action/sound"
        });
        _superDaemon.SuperDaemonInstance.defineOption({
          title: this.t.outlineDesigner,
          icon: "hax:site-map",
          tags: ["CMS", "outline", "designer", "site outline", "operation", "command"],
          value: {
            target: this.shadowRoot.querySelector("#outlinebutton")
          },
          context: ["CMS"],
          eventName: "super-daemon-element-click",
          path: "CMS/action/outline"
        });
        // change theme program
        _superDaemon.SuperDaemonInstance.defineOption({
          title: "DEV: Change theme",
          icon: "image:style",
          tags: ["Developer", "theme"],
          eventName: "super-daemon-run-program",
          path: ">settings/theme",
          context: [">"],
          value: {
            name: "DEV: Change theme",
            context: ">",
            program: function () {
              var _program = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee2(input, values) {
                var results;
                return _regeneratorRuntime().wrap(function _callee2$(_context2) {
                  while (1) switch (_context2.prev = _context2.next) {
                    case 0:
                      results = [];
                      ["clean-one", "clean-two", "learn-two-theme", "bootstrap-theme", "outline-player", "haxor-slevin"].forEach( /*#__PURE__*/function () {
                        var _ref = babelHelpers.asyncToGenerator( /*#__PURE__*/_regeneratorRuntime().mark(function _callee(name) {
                          return _regeneratorRuntime().wrap(function _callee$(_context) {
                            while (1) switch (_context.prev = _context.next) {
                              case 0:
                                if (input == "" || name.includes(input)) {
                                  results.push({
                                    title: name.replace("-theme", "").replace("-", " "),
                                    icon: "image:style",
                                    tags: ["theme"],
                                    value: {
                                      target: window.HAXCMS,
                                      method: "setTheme",
                                      args: [name]
                                    },
                                    eventName: "super-daemon-element-method",
                                    context: [">", ">settings/theme/" + name],
                                    path: ">settings/theme/" + name
                                  });
                                }
                              case 1:
                              case "end":
                                return _context.stop();
                            }
                          }, _callee);
                        }));
                        return function (_x3) {
                          return _ref.apply(this, arguments);
                        };
                      }());
                      return _context2.abrupt("return", results);
                    case 3:
                    case "end":
                      return _context2.stop();
                  }
                }, _callee2);
              }));
              function program(_x, _x2) {
                return _program.apply(this, arguments);
              }
              return program;
            }()
          }
        });
        this.updateAvailableButtons();
        // load user data
        this.dispatchEvent(new CustomEvent("haxcms-load-user-data", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: true
        }));
        this.shadowRoot.querySelectorAll("[voice-command]").forEach(function (el) {
          if (el.getAttribute("id") == "editbutton") {
            _this3.dispatchEvent(new CustomEvent("hax-add-voice-command", {
              bubbles: true,
              composed: true,
              cancelable: false,
              detail: {
                command: ":name: save (this) page",
                context: el,
                callback: "click"
              }
            }));
          }
          _this3.dispatchEvent(new CustomEvent("hax-add-voice-command", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: {
              command: ":name: " + el.getAttribute("voice-command"),
              context: el,
              callback: "click"
            }
          }));
        });
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this4 = this;
        if (babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "updated", this).call(this, changedProperties);
        }
        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "userMenuOpen" && oldValue !== undefined) {
            if (_this4.userMenuOpen) {
              _this4.rpgHat = "edit";
            } else {
              _this4.rpgHat = "none";
            }
          }
          if (propName == "editMode") {
            if (_this4.editMode) {
              _this4.rpgHat = "construction";
            } else {
              _this4.rpgHat = "none";
            }
            _superDaemon.SuperDaemonInstance.close();
            // observer
            _this4._editModeChanged(_this4[propName], oldValue);
            // notify
            _this4.dispatchEvent(new CustomEvent("edit-mode-changed", {
              detail: _this4[propName]
            }));
          }
          if (propName == "manifestEditMode") {
            // observer
            _this4._manifestEditModeChanged(_this4[propName], oldValue);
            // notify
            _this4.dispatchEvent(new CustomEvent("manifest-edit-mode-changed", {
              detail: _this4[propName]
            }));
          }
          if (propName == "dashboardOpened" || propName == "t") {
            // observer
            _this4._dashboardOpenedChanged(_this4.dashboardOpened);
          }
          // make sure edit matches but state doesnt shift
          if (propName === "t") {
            if (_this4.editMode) {
              _this4.__editText = _this4.t.savePageContent;
            } else {
              _this4.__editText = _this4.t.editPageContent;
            }
          }
        });
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var _this5 = this;
        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "connectedCallback", this).call(this);
        window.addEventListener("jwt-logged-in", this._jwtLoggedIn.bind(this));
        (0, _mobxEsm.autorun)(function (reaction) {
          if (_haxcmsSiteStore.store.userData) {
            _this5.userName = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.userData.userName);
            _this5.userPicture = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.userData.userPicture);
            // update buttons to match since we got a state response
            _this5.updateAvailableButtons();
          }
          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this5.editMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.editMode);
          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this5.manifest = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest);
          _this5.icon = "hax:site-settings";
          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this5.dashboardOpened = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.dashboardOpened);
          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          var activeItem = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeItem);
          // update buttons to match since we got a state response
          _this5.updateAvailableButtons();
          if (activeItem && activeItem.id) {
            _this5.activeTitle = activeItem.title;
            _this5.pageAllowed = true;
          } else {
            _this5.pageAllowed = false;
          }
          _this5.__disposer.push(reaction);
        });
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }
        window.removeEventListener("jwt-logged-in", this._jwtLoggedIn.bind(this));
        window.removeEventListener("hax-store-ready", this.haxStoreReady.bind(this));
        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "_dashboardOpenedChanged",
      value: function _dashboardOpenedChanged(newValue) {
        if (newValue) {
          this.__settingsText = this.t.closeSiteSettings;
          this.icon = "icons:cancel";
        } else if (!newValue) {
          this.__settingsText = this.t.editSiteSettings;
          this.icon = "hax:site-settings";
        }
      }
      /**
       * toggle state on button tap
       */
    }, {
      key: "_editButtonTap",
      value: function _editButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        this.editMode = !this.editMode;
        // save button shifted to edit
        if (!this.editMode) {
          this.dispatchEvent(new CustomEvent("haxcms-save-node", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: _haxcmsSiteStore.store.activeItem
          }));
        }
        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
      }
    }, {
      key: "_editDetailsButtonTap",
      value: function _editDetailsButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        var evt = new CustomEvent("haxcms-load-node-fields", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.target
        });
        window.dispatchEvent(evt);
      }
    }, {
      key: "_insightsButtonTap",
      value: function _insightsButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        var c = document.createElement("haxcms-site-insights");
        var evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: this.t.insights,
            styles: {
              "--simple-modal-titlebar-background": "orange",
              "--simple-modal-titlebar-color": "black",
              "--simple-modal-width": "94vw",
              "--simple-modal-min-width": "300px",
              "--simple-modal-z-index": "100000000",
              "--simple-modal-height": "94vh",
              "--simple-modal-min-height": "300px",
              "--simple-modal-titlebar-height": "80px"
            },
            elements: {
              content: c
            },
            invokedBy: this.shadowRoot.querySelector("#insightsbutton"),
            clone: false,
            modal: false
          }
        });
        window.dispatchEvent(evt);
      }
    }, {
      key: "_cancelButtonTap",
      value: function _cancelButtonTap(e) {
        this.editMode = false;
        _haxcmsSiteStore.store.playSound("error");
        this.dispatchEvent(new CustomEvent("hax-cancel", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.detail
        }));
        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
      }
      /**
       * create new item
       */
    }, {
      key: "_createNewItem",
      value: function _createNewItem(e) {
        var evt = new CustomEvent("haxcms-create-node", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            values: this.__newForm.value
          }
        });
        this.dispatchEvent(evt);
      }
      /**
       * Fire item
       */
    }, {
      key: "_updateItem",
      value: function _updateItem(e) {
        var local = e.target;
        var values;
        if (!local.__form) {
          values = local.parentNode.__form.value;
        } else {
          values = local.__form.value;
        }
        // fire event with details for saving
        window.dispatchEvent(new CustomEvent("haxcms-save-node-details", {
          bubbles: true,
          composed: true,
          cancelable: true,
          detail: values
        }));
        // fire event to close the modal
        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          composed: true,
          cancelable: true,
          detail: {}
        }));
      }
      /**
       * Delete button hit, confirm they want to do this
       */
    }, {
      key: "_deleteButtonTap",
      value: function _deleteButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        var c = document.createElement("span");
        c.innerHTML = "\"".concat(_haxcmsSiteStore.store.activeItem.title, "\" will be removed from the outline but its content stays on the file system.");
        var b1 = document.createElement("button");
        b1.appendChild(document.createTextNode("Confirm"));
        b1.classList.add("hax-modal-btn");
        b1.addEventListener("click", this._deleteActive.bind(this));
        var b2 = document.createElement("button");
        b2.appendChild(document.createTextNode("cancel"));
        b2.addEventListener("click", function () {
          return _haxcmsSiteStore.store.playSound("error");
        });
        b2.setAttribute("dialog-dismiss", "dialog-dismiss");
        b2.classList.add("hax-modal-btn");
        b2.classList.add("cancel");
        var b = document.createElement("span");
        b.appendChild(b1);
        b.appendChild(b2);
        var evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: "Are you sure you want to delete this page?",
            styles: {
              "--simple-modal-titlebar-background": "orange",
              "--simple-modal-titlebar-color": "black",
              "--simple-modal-width": "25vw",
              "--simple-modal-min-width": "300px",
              "--simple-modal-z-index": "100000000",
              "--simple-modal-height": "15vh",
              "--simple-modal-min-height": "300px",
              "--simple-modal-titlebar-height": "80px"
            },
            elements: {
              content: c,
              buttons: b
            },
            invokedBy: this.shadowRoot.querySelector("#deletebutton"),
            clone: false,
            modal: true
          }
        });
        window.dispatchEvent(evt);
      }
      /**
       * delete active item
       */
    }, {
      key: "_deleteActive",
      value: function _deleteActive(e) {
        _haxcmsSiteStore.store.playSound("click");
        var evt = new CustomEvent("haxcms-delete-node", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            item: _haxcmsSiteStore.store.activeItem
          }
        });
        this.dispatchEvent(evt);
      }
      /**
       * toggle share button
       */
    }, {
      key: "_shareButtonTap",
      value: function _shareButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        var evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: this.t.shareSite,
            styles: {
              "--simple-modal-titlebar-background": "orange",
              "--simple-modal-titlebar-color": "black",
              "--simple-modal-width": "55vw",
              "--simple-modal-min-width": "300px",
              "--simple-modal-z-index": "100000000",
              "--simple-modal-height": "50vh",
              "--simple-modal-min-height": "300px",
              "--simple-modal-titlebar-height": "80px"
            },
            elements: {
              content: document.createElement("haxcms-share-dialog")
            },
            invokedBy: this.shadowRoot.querySelector("#sharebutton"),
            clone: false,
            modal: false
          }
        });
        window.dispatchEvent(evt);
      }
      /**
       * toggle state on button tap
       */
    }, {
      key: "_outlineButtonTap",
      value: function _outlineButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        var evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: this.t.outlineDesigner,
            styles: {
              "--simple-modal-titlebar-background": "orange",
              "--simple-modal-titlebar-color": "black",
              "--simple-modal-z-index": "100000000",
              "--simple-modal-titlebar-height": "80px",
              "--simple-modal-width": "85vw",
              "--simple-modal-max-width": "85vw",
              "--simple-modal-height": "90vh",
              "--simple-modal-max-height": "90vh"
            },
            elements: {
              content: document.createElement("haxcms-outline-editor-dialog")
            },
            invokedBy: this.shadowRoot.querySelector("#outlinebutton"),
            clone: false,
            modal: true
          }
        });
        window.dispatchEvent(evt);
      }
    }, {
      key: "_addButtonTap",
      value: function _addButtonTap() {
        var _this6 = this;
        _haxcmsSiteStore.store.playSound("click");
        setTimeout(function () {
          window.location = _this6.backLink + "createSite-step-1";
        }, 100);
      }
      /**
       * toggle state on button tap
       */
    }, {
      key: "_manifestButtonTap",
      value: function _manifestButtonTap(e) {
        _haxcmsSiteStore.store.playSound("click");
        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
        window.dispatchEvent(new CustomEvent("haxcms-load-site-dashboard", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.target
        }));
      }
      /**
       * Edit state has changed.
       */
    }, {
      key: "_editModeChanged",
      value: function _editModeChanged(newValue, oldValue) {
        if (newValue) {
          this.__editIcon = "icons:save";
          this.__editText = this.t.savePageContent;
          _superDaemon.SuperDaemonInstance.appendContext("HAX");
          _superDaemon.SuperDaemonInstance.removeContext("CMS");
        } else {
          this.__editIcon = "hax:page-edit";
          this.__editText = this.t.editPageContent;
          _superDaemon.SuperDaemonInstance.appendContext("CMS");
          _superDaemon.SuperDaemonInstance.removeContext("HAX");
        }
        if (babelHelpers.typeof(oldValue) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          _haxcmsSiteStore.store.editMode = newValue;
          // force tray status to be the opposite of the editMode
          // to open when editing and close when not
          _haxStore.HAXStore.haxTray.collapsed = !newValue;
        }
      }
      /**
       * Note changes to the outline / structure of the page's items
       */
    }, {
      key: "_outlineEditModeChanged",
      value: function _outlineEditModeChanged(newValue, oldValue) {
        this.dispatchEvent(new CustomEvent("haxcms-outline-edit-mode-changed", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: newValue
        }));
      }
      /**
       * Note changes to the outline / structure of the page's items
       */
    }, {
      key: "_manifestEditModeChanged",
      value: function _manifestEditModeChanged(newValue, oldValue) {
        this.dispatchEvent(new CustomEvent("haxcms-manifest-edit-mode-changed", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: newValue
        }));
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI), "styles", this)), [(0, _index.css)(_templateObject3_724e82a0ddf411ed96ad39cc5dd30a65 || (_templateObject3_724e82a0ddf411ed96ad39cc5dd30a65 = babelHelpers.taggedTemplateLiteral(["\n        :host *:not(:defined) {\n          display: none;\n        }\n        :host {\n          display: block;\n          position: relative;\n          height: 48px;\n          left: 0;\n          top: 0;\n          right: 0;\n          background-color: var(--haxcms-system-bg, #37474f);\n          z-index: 10000;\n          visibility: visible;\n        }\n        :host([edit-mode]) {\n          z-index: 9999;\n        }\n        :host([dashboard-opened]) {\n          left: 50vw;\n        }\n        /**\n         * Dashboard open trumps all contextual settings\n         */\n        :host([dashboard-opened]) #editbutton,\n        :host([dashboard-opened]) #editdetails,\n        :host([dashboard-opened]) #deletebutton,\n        :host([dashboard-opened]) #addbutton,\n        :host([dashboard-opened]) #outlinebutton,\n        :host([dashboard-opened]) #insightsbutton,\n        :host([dashboard-opened]) #addmenubutton,\n        :host([dashboard-opened]) #addbuttonchild,\n        :host([dashboard-opened]) #duplicatebutton {\n          display: none !important;\n        }\n        :host *[hidden] {\n          display: none;\n        }\n        simple-tooltip:not(:defined) {\n          display: none !important;\n        }\n        :host([painting]) {\n          opacity: 0;\n          visibility: hidden;\n        }\n        #editbutton {\n          visibility: hidden;\n          opacity: 0;\n        }\n        :host([page-allowed]) #editbutton,\n        :host([page-allowed]) #editdetails,\n        :host([page-allowed]) #deletebutton {\n          visibility: visible;\n          opacity: 1;\n        }\n        #editbutton[icon=\"icons:save\"]:focus,\n        #editbutton[icon=\"icons:save\"]:active,\n        #editbutton[icon=\"icons:save\"]:hover {\n          background-color: var(--simple-colors-default-theme-green-2);\n        }\n        #editbutton[icon=\"icons:save\"] {\n          background-color: var(--simple-colors-default-theme-green-1);\n          color: var(--simple-colors-default-theme-green-8);\n        }\n        #cancelbutton {\n          color: var(--hax-ui-color-danger-secondary);\n          background-color: var(--simple-colors-default-theme-red-1);\n        }\n        #cancelbutton:hover,\n        #cancelbutton:active,\n        #cancelbutton:focus {\n          background-color: var(--simple-colors-default-theme-red-2);\n        }\n        simple-toolbar-menu:hover,\n        simple-toolbar-menu:active,\n        simple-toolbar-menu:focus,\n        simple-toolbar-button:hover,\n        simple-toolbar-button:active,\n        simple-toolbar-button:focus {\n          background-color: var(--hax-ui-background-color-accent);\n          color: var(--hax-ui-color);\n        }\n\n        :host(:hover),\n        :host(:active),\n        :host(:focus) {\n          opacity: 1;\n        }\n        simple-tooltip {\n          --simple-tooltip-background: #000000;\n          --simple-tooltip-opacity: 1;\n          --simple-tooltip-text-color: #ffffff;\n          --simple-tooltip-delay-in: 0;\n          --simple-tooltip-duration-in: 200ms;\n          --simple-tooltip-duration-out: 0;\n          --simple-tooltip-border-radius: 0;\n          --simple-tooltip-font-size: 14px;\n          font-family: \"Press Start 2P\", sans-serif;\n        }\n        app-hax-top-bar {\n          z-index: 1000;\n          right: 0;\n          left: 0;\n          position: fixed;\n        }\n        :host([dark-mode]) app-hax-top-bar {\n          --bg-color: #222;\n          --accent-color: #fff;\n        }\n        app-hax-top-bar::part(top-bar) {\n          grid-template-columns: 20% 60% 20%;\n          overflow: visible;\n        }\n        .haxLogo {\n          color: var(--simple-colors-default-theme-accent-12, black);\n        }\n        :host([dark-mode]) .haxLogo {\n          color: var(--simple-colors-default-theme-accent-1, white);\n        }\n        :host([dark-mode]) .haxLogo:hover,\n        :host([dark-mode]) .haxLogo:focus,\n        :host([dark-mode]) .haxLogo:active,\n        .haxLogo:hover,\n        .haxLogo:focus,\n        .haxLogo:active {\n          color: var(--haxcms-color);\n        }\n        .haxLogo simple-icon-lite {\n          --simple-icon-height: 40px;\n          --simple-icon-width: 40px;\n          margin: 4px;\n        }\n        .soundToggle {\n          position: relative;\n          display: inline-flex;\n          vertical-align: top;\n        }\n\n        .soundToggle img {\n          width: 24px;\n          height: 24px;\n        }\n\n        app-hax-search-bar {\n          vertical-align: middle;\n          display: inline-flex;\n        }\n        simple-toolbar {\n          align-items: stretch;\n          justify-content: center;\n          height: var(--top-bar-height);\n          --simple-toolbar-button-disabled-border-color: transparent;\n          --simple-toolbar-button-disabled-opacity: 0.3;\n          --simple-toolbar-button-padding: 3px 6px;\n        }\n        simple-toolbar[dark-mode] {\n          --simple-toolbar-button-color: #e0e0e0;\n          --simple-toolbar-button-hover-color: #fff;\n          --simple-toolbar-button-bg: #222;\n          --simple-toolbar-button-hover-bg: #000;\n        }\n        simple-toolbar::part(buttons) {\n          flex: 0 1 auto;\n        }\n        .ops-panel {\n          justify-content: space-around;\n          display: flex;\n          padding: 4px 0px;\n        }\n\n        .topbar-character {\n          cursor: pointer;\n          display: inline-block;\n          border: none;\n          border-radius: 0px;\n          padding: 0 8px;\n          margin: 0 0 0 16px;\n          background-color: transparent;\n          height: 48px;\n          max-width: 160px;\n          transition: all 0.5 ease-in-out;\n        }\n        .characterbtn-name {\n          font-family: \"Press Start 2P\", sans-serif;\n          margin-left: 8px;\n          font-size: 12px;\n          vertical-align: bottom;\n          line-height: 48px;\n          overflow: hidden;\n          text-overflow: ellipsis;\n          height: 48px;\n          word-break: break-all;\n        }\n        :host([dark-mode]) .topbar-character,\n        :host([dark-mode]) .topbar-character {\n          color: #e0e0e0;\n          background-color: #222;\n        }\n        .topbar-character rpg-character {\n          margin: -4px -14px 0px -10px;\n          height: 52px;\n          width: 64px;\n          display: inline-block;\n          transform: scale(0.7);\n          transition: all 0.5 ease-in-out;\n        }\n        .topbar-character:hover rpg-character,\n        .topbar-character:focus rpg-character {\n          transform: scale(0.9);\n        }\n\n        .logout::part(menu-button) {\n          background-image: url(\"", "\");\n          background-repeat: no-repeat;\n          background-position: center center;\n          text-align: center;\n          background-size: cover;\n          border-top: 0px;\n          border-bottom: 0px;\n          padding: 10px;\n          font-family: \"Press Start 2P\", sans-serif;\n        }\n        app-hax-user-menu app-hax-user-menu-button::part(menu-button) {\n          font-family: \"Press Start 2P\", sans-serif;\n          font-size: 12px;\n        }\n\n        @media screen and (max-width: 800px) {\n          :host([dashboard-opened]) {\n            left: 90vw;\n          }\n          :host([edit-mode]) {\n            bottom: unset;\n          }\n          .topbar-character {\n            padding: 0;\n            margin: 0;\n          }\n          .characterbtn-name {\n            padding: 0;\n            margin: 0;\n            width: 30px;\n            display: inline-flex;\n            overflow: hidden;\n            text-overflow: ellipsis;\n            letter-spacing: -2px;\n            margin-left: -6px;\n          }\n          simple-toolbar {\n            --simple-toolbar-button-padding: 3px 3px;\n          }\n          .characterbtn-name,\n          haxcms-button-add,\n          simple-toolbar-menu,\n          simple-toolbar-button {\n            font-size: 10px;\n          }\n          simple-toolbar-menu {\n            --icon-offset-right: 2px;\n            --icon-offset-left: 2px;\n          }\n          simple-toolbar-menu::part(dropdown-icon) {\n            right: 0;\n            top: 10px;\n          }\n          :host([edit-mode]) #deletebutton,\n          :host([edit-mode]) #addmenubutton,\n          :host([edit-mode]) #editdetails {\n            display: none;\n          }\n        }\n        @media (max-width: 600px) {\n          simple-toolbar {\n            --simple-toolbar-button-padding: 0px;\n          }\n          app-hax-top-bar::part(top-bar) {\n            grid-template-columns: 10% 80% 10%;\n          }\n          .haxLogo simple-icon-lite {\n            margin: 0px;\n          }\n          .characterbtn-name {\n            display: none;\n          }\n        }\n      "])), (0, _index.unsafeCSS)(LogOut))]);
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */
    }, {
      key: "tag",
      get: function get() {
        return "haxcms-site-editor-ui";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI), "properties", this)), {}, {
          userName: {
            type: String,
            attribute: "user-name"
          },
          rpgHat: {
            type: String
          },
          userPicture: {
            type: String,
            attribute: "user-picture"
          },
          userMenuOpen: {
            type: Boolean
          },
          soundIcon: {
            type: String
          },
          darkMode: {
            type: Boolean,
            reflect: true,
            attribute: "dark-mode"
          },
          backLink: {
            type: String
          },
          backText: {
            type: String
          },
          __editIcon: {
            type: String
          },
          __editText: {
            type: String
          },
          __settingsText: {
            type: String
          },
          /**
           * small visual lock that events break on initial paint
           */
          painting: {
            type: Boolean,
            reflect: true
          },
          /**
           * page allowed
           */
          pageAllowed: {
            type: Boolean,
            attribute: "page-allowed",
            reflect: true
          },
          /**
           * if the page is in an edit state or not
           */
          editMode: {
            type: Boolean,
            reflect: true,
            attribute: "edit-mode"
          },
          /**
           * Manifest editing state
           */
          manifestEditMode: {
            type: Boolean,
            attribute: "manifest-edit-mode",
            reflect: true
          },
          activeTitle: {
            type: String,
            attribute: "active-title"
          },
          manifest: {
            type: Object
          },
          icon: {
            type: String
          },
          dashboardOpened: {
            type: Boolean,
            reflect: true,
            attribute: "dashboard-opened"
          }
        });
      }
    }]);
    return HAXCMSSiteEditorUI;
  }((0, _HAXCMSThemeParts2.HAXCMSThemeParts)((0, _HAXCMSI18NMixin.HAXCMSI18NMixin)(_simpleColors.SimpleColors)));
  _exports.HAXCMSSiteEditorUI = HAXCMSSiteEditorUI;
  customElements.define(HAXCMSSiteEditorUI.tag, HAXCMSSiteEditorUI);
});
