// @lrnwebcomponents/haxcms-elements/lib/core/haxcms-site-editor-ui.js
define(["exports", "meta", "require", "../../../../lit/index.js", "./haxcms-site-store.js", "../../../hax-body/lib/hax-store.js", "../../../../mobx/dist/mobx.esm.js", "../../../utils/utils.js", "../../../simple-icon/simple-icon.js", "../../../simple-icon/lib/simple-icons.js", "../../../simple-icon/lib/simple-icon-button.js", "./utils/HAXCMSThemeParts.js", "./utils/HAXCMSI18NMixin.js", "./micros/haxcms-button-add.js", "../../../rpg-character/rpg-character.js", "../../../app-hax/lib/v1/app-hax-top-bar.js", "./haxcms-darkmode-toggle.js", "../../../../wired-elements/lib/wired-button.js", "../../../simple-colors/simple-colors.js"], function (_exports, meta, _require, _index, _haxcmsSiteStore, _haxStore, _mobxEsm, _utils, _simpleIcon, _simpleIcons, _simpleIconButton, _HAXCMSThemeParts2, _HAXCMSI18NMixin, _haxcmsButtonAdd, _rpgCharacter, _appHaxTopBar, _haxcmsDarkmodeToggle, _wiredButton, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSSiteEditorUI = void 0;
  meta = _interopRequireWildcard(meta);
  _require = _interopRequireWildcard(_require);

  var _templateObject_60cb7f60c76b11eca96c833ece275368, _templateObject2_60cb7f60c76b11eca96c833ece275368;

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || babelHelpers.typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  var haxLogo = new URL('../../../app-hax/lib/assets/images/HAXLogo.svg', meta.url).href;
  var ButtonBGLight = new URL('../../../app-hax/lib/assets/images/ButtonBGLM.svg', meta.url).href;
  var ButtonBGDark = new URL('../../../app-hax/lib/assets/images/ButtonBGDM.svg', meta.url).href;
  var LogOut = new URL('../../../app-hax/lib/assets/images/Logout.svg', meta.url).href;
  /**
   * `haxcms-site-editor-ui`
   * `haxcms editor element buttons that you see`
   *
   * @demo demo/index.html
   * @microcopy - the mental model for this element
   */

  var HAXCMSSiteEditorUI = /*#__PURE__*/function (_HAXCMSThemeParts) {
    babelHelpers.inherits(HAXCMSSiteEditorUI, _HAXCMSThemeParts);

    var _super = _createSuper(HAXCMSSiteEditorUI);

    function HAXCMSSiteEditorUI() {
      var _this;

      babelHelpers.classCallCheck(this, HAXCMSSiteEditorUI);
      _this = _super.call(this);
      _this.rpgHat = "none";
      _this.darkMode = false;
      _this.userMenuOpen = false;
      _this.soundIcon = '';
      _this.__disposer = _this.__disposer || [];
      _this.t = _this.t || {};
      _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
        backToSiteList: "Back to site list",
        cancelEditing: "Cancel editing",
        editDetails: "Edit details",
        addPage: "Add page",
        deletePage: "Delete page",
        editSiteOutline: "Edit site outline",
        closeSiteSettings: "Close site settings",
        editSiteSettings: "Edit site settings",
        savePageContent: "Save page content",
        editPageContent: "Edit page content",
        startNewJourney: "Start New Journey",
        accountInfo: "Account Info",
        logOut: "Log out",
        systemMenu: "System menu"
      });
      _this.backText = "Back to site list";
      _this.painting = true;
      _this.pageAllowed = false;
      _this.editMode = false;
      _this.__editIcon = "hax:page-edit";
      _this.icon = "hax:site-settings";
      _this.manifestEditMode = false;
      _this.backLink = '../../';

      if (window.appSettings && window.appSettings.backLink) {
        _this.backLink = window.appSettings.backLink;
      }

      (0, _mobxEsm.autorun)(function () {
        _this.darkMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.darkMode);
        _this.dark = _this.darkMode;

        if ((0, _mobxEsm.toJS)(_haxcmsSiteStore.store.darkMode)) {
          _haxStore.HAXStore.globalPreferences.haxUiTheme = 'haxdark';
        } else {
          _haxStore.HAXStore.globalPreferences.haxUiTheme = 'hax';
        }
      });
      (0, _mobxEsm.autorun)(function () {
        _this.soundIcon = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.soundStatus) ? new URL('../../../app-hax/lib/assets/images/FullVolume.svg', meta.url).href : new URL('../../../app-hax/lib/assets/images/Silence.svg', meta.url).href;
      });
      setTimeout(function () {
        // prettier-ignore
        new Promise(function (res, rej) {
          return _require.default(["./haxcms-outline-editor-dialog.js"], res, rej);
        }); // prettier-ignore

        new Promise(function (res, rej) {
          return _require.default(["./haxcms-site-dashboard.js"], res, rej);
        });
        new Promise(function (res, rej) {
          return _require.default(["../../../simple-modal/simple-modal.js"], res, rej);
        });
        new Promise(function (res, rej) {
          return _require.default(["../../../simple-fields/lib/simple-fields-form.js"], res, rej);
        });
      }, 0);
      return _this;
    }

    babelHelpers.createClass(HAXCMSSiteEditorUI, [{
      key: "soundToggle",
      value: function soundToggle() {
        var status = !(0, _mobxEsm.toJS)(_haxcmsSiteStore.store.soundStatus);
        _haxcmsSiteStore.store.soundStatus = status;
        (0, _utils.localStorageSet)('app-hax-soundStatus', status);

        if (!status) {
          _haxcmsSiteStore.store.toast("Sound off.. hey.. HELLO!?!", 2000, {
            fire: true
          });
        } else {
          _haxcmsSiteStore.store.toast("Can you hear me now? Good.", 2000, {
            hat: 'random'
          });

          _haxcmsSiteStore.store.playSound('click');
        }
      }
    }, {
      key: "toggleMenu",
      value: function toggleMenu() {
        this.userMenuOpen = !this.userMenuOpen;

        _haxcmsSiteStore.store.playSound('click');
      } // render function

    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_60cb7f60c76b11eca96c833ece275368 || (_templateObject_60cb7f60c76b11eca96c833ece275368 = babelHelpers.taggedTemplateLiteral(["\n      <app-hax-top-bar>\n        <span slot=\"left\">\n          <a href=\"", "\" class=\"haxLogo\" id=\"backtosites\">\n            <simple-icon-lite src=\"", "\"></simple-icon-lite>\n          </a>\n          <slot name=\"haxcms-site-editor-ui-prefix-avatar\"></slot>\n        </span>\n      <span slot=\"center\">\n        <slot name=\"haxcms-site-editor-ui-prefix-buttons\"></slot>\n        <simple-icon-button\n          hidden\n          ?dark=\"", "\"\n          id=\"editbutton\"\n          icon=\"", "\"\n          @click=\"", "\"\n          label=\"", "\"\n          voice-command=\"edit (this) page\"\n          class=\"topbar-button\"\n          accent-color=\"green\"\n        ></simple-icon-button>\n        <simple-icon-button\n          id=\"cancelbutton\"\n          ?dark=\"", "\"\n          icon=\"icons:cancel\"\n          @click=\"", "\"\n          .hidden=\"", "\"\n          label=\"", "\"\n          voice-command=\"cancel (editing)\"\n          class=\"topbar-button\"\n          accent-color=\"red\"\n        ></simple-icon-button>\n        <simple-icon-button\n          ?dark=\"", "\"\n          id=\"editdetails\"\n          icon=\"hax:page-details\"\n          @click=\"", "\"\n          label=\"", "\"\n          voice-command=\"edit (page) details\"\n          class=\"topbar-button\"\n          accent-color=\"blue\"\n          part=\"", "\"\n          tabindex=\"", "\"\n        ></simple-icon-button>\n        <simple-icon-button\n          part=\"", "\"\n          tabindex=\"", "\"\n          ?dark=\"", "\"\n          id=\"deletebutton\"\n          icon=\"icons:delete\"\n          @click=\"", "\"\n          label=\"", "\"\n          voice-command=\"delete page\"\n          class=\"topbar-button\"\n          accent-color=\"red\"\n        ></simple-icon-button>\n        <haxcms-button-add\n        hidden\n        ?dark-mode=\"", "\"\n        accent-color=\"purple\"\n        id=\"addbutton\"></haxcms-button-add>\n        <haxcms-button-add\n          hidden\n          ?dark-mode=\"", "\"\n          id=\"addbuttonchild\"\n          type=\"child\"\n          accent-color=\"purple\"\n        ></haxcms-button-add>\n        <haxcms-button-add\n          hidden\n          ?dark-mode=\"", "\"\n          type=\"duplicate\"\n          id=\"duplicatebutton\"\n          accent-color=\"purple\"\n        ></haxcms-button-add>\n        <simple-tooltip for=\"backtosites\" position=\"right\"\n          >", "</simple-tooltip\n        >\n        <simple-tooltip for=\"cancelbutton\" position=\"bottom\" offset=\"14\"\n          >", "</simple-tooltip\n        >\n        <simple-tooltip for=\"editbutton\" position=\"bottom\" offset=\"14\"\n          >", "</simple-tooltip\n        >\n        <simple-tooltip for=\"editdetails\" position=\"bottom\" offset=\"14\"\n          >", "</simple-tooltip\n        >\n        <simple-tooltip for=\"deletebutton\" position=\"bottom\" offset=\"14\"\n          >", "</simple-tooltip\n        >\n        <slot name=\"haxcms-site-editor-ui-suffix-buttons\"></slot>\n      </span>\n      <span slot=\"right\">\n        <button class=\"topbar-character\" @click=\"", "\">\n          <rpg-character\n            seed=\"", "\"\n            width=\"68\"\n            height=\"68\"\n            aria-label=\"", "\"\n            title=\"", "\"\n            hat=\"", "\"\n          ></rpg-character>\n        </button>\n        <div class=\"user-menu ", "\">\n          <div class=\"ops-panel\">\n            <wired-button\n              elevation=\"1\"\n              class=\"soundToggle\"\n              @click=\"", "\"\n            >\n              <simple-icon-lite src=\"", "\" loading=\"lazy\" decoding=\"async\"></simple-icon-lite>\n            </wired-button>\n            <haxcms-darkmode-toggle></haxcms-darkmode-toggle>\n          </div>\n          <button @click=\"", "\">\n            <simple-icon-lite icon=\"", "\"></simple-icon-lite>", "</button>\n          <button\n          id=\"outlinebutton\"\n          @click=\"", "\"\n          ><simple-icon-lite icon=\"hax:site-map\"></simple-icon-lite>", "</button>\n          <button>\n          <simple-icon-lite icon=\"add\"></simple-icon-lite>", "</button>\n          <button>\n          <simple-icon-lite icon=\"face\"></simple-icon-lite>", "</button>\n          <button class=\"logout\">", "</button>\n        </div>\n      </span>\n    </app-hax-top-bar>\n    "])), this.backLink, haxLogo, !this.darkMode, this.__editIcon, this._editButtonTap, this.__editText, !this.darkMode, this._cancelButtonTap, !this.editMode, this.t.cancelEditing, !this.darkMode, this._editDetailsButtonTap, this.t.editDetails, this.editMode ? "edit-mode-active" : "", this.editMode ? "-1" : "", this.editMode ? "edit-mode-active" : "", this.editMode ? "-1" : "", !this.darkMode, this._deleteButtonTap, this.t.deletePage, this.darkMode, this.darkMode, this.darkMode, this.backText, this.t.cancelEditing, this.__editText, this.t.editDetails, this.t.deletePage, this.toggleMenu, this.userName, this.t.systemMenu, this.t.systemMenu, this.rpgHat, this.userMenuOpen ? 'open' : '', this.soundToggle, this.soundIcon, this._manifestButtonTap, this.icon, this.__settingsText, this._outlineButtonTap, this.t.editSiteOutline, this.t.startNewJourney, this.t.accountInfo, this.t.logOut);
      }
      /**
       * update buttons since these are triggered by a mix of
       * differnet backend types we can't leverage the store
       * since a CMS needs to just hardcode these at run time
       * for some environments
       */

    }, {
      key: "updateAvailableButtons",
      value: function updateAvailableButtons() {
        var _this2 = this;

        if (this.shadowRoot) {
          setTimeout(function () {
            // backText
            if (window.appSettings && window.appSettings.backText) {
              _this2.backText = window.appSettings.backText;
            }

            var ary = [{
              varPath: "getNodeFieldsPath",
              selector: "#editdetails"
            }, {
              varPath: "deleteNodePath",
              selector: "#deletebutton"
            }, {
              varPath: "saveNodePath",
              selector: "#editbutton"
            }, {
              varPath: "createNodePath",
              selector: "#addbutton"
            }, {
              varPath: "createNodePath",
              selector: "#addbuttonchild"
            }, {
              varPath: "createNodePath",
              selector: "#duplicatebutton"
            }]; // see which features should be enabled

            ary.forEach(function (pair) {
              if (window.appSettings && window.appSettings[pair.varPath] && window.appSettings[pair.varPath] != null && window.appSettings[pair.varPath] != "" && window.appSettings[pair.varPath] != "null") {
                if (pair.dep) {
                  if (window.appSettings[pair.dep] != null && window.appSettings[pair.dep] != "" && window.appSettings[pair.dep] != "null") {
                    _this2.shadowRoot.querySelector(pair.selector).removeAttribute("hidden");
                  } else {// a dependency didn't meet the requirement
                  }
                } else {
                  _this2.shadowRoot.querySelector(pair.selector).removeAttribute("hidden");
                }
              }
            });
          }, 100);
        }
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        var _this3 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "firstUpdated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "firstUpdated", this).call(this, changedProperties);
        }

        this.updateAvailableButtons(); // load user data

        this.dispatchEvent(new CustomEvent("haxcms-load-user-data", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: true
        }));
        this.shadowRoot.querySelectorAll("[voice-command]").forEach(function (el) {
          if (el.getAttribute("id") == "editbutton") {
            _this3.dispatchEvent(new CustomEvent("hax-add-voice-command", {
              bubbles: true,
              composed: true,
              cancelable: false,
              detail: {
                command: ":name: save (this) page",
                context: el,
                callback: "click"
              }
            }));
          }

          _this3.dispatchEvent(new CustomEvent("hax-add-voice-command", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: {
              command: ":name: " + el.getAttribute("voice-command"),
              context: el,
              callback: "click"
            }
          }));
        });
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this4 = this;

        if (babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "updated", this)) {
          babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "updated", this).call(this, changedProperties);
        }

        changedProperties.forEach(function (oldValue, propName) {
          if (propName === 'userMenuOpen' && oldValue !== undefined) {
            if (_this4.userMenuOpen) {
              _this4.rpgHat = 'construction';
            } else {
              _this4.rpgHat = 'none';
            }
          }

          if (propName == "editMode") {
            if (_this4.editMode) {
              _this4.rpgHat = 'edit';
            } else {
              _this4.rpgHat = 'none';
            } // observer


            _this4._editModeChanged(_this4[propName], oldValue); // notify


            _this4.dispatchEvent(new CustomEvent("edit-mode-changed", {
              detail: _this4[propName]
            }));
          }

          if (propName == "manifestEditMode") {
            // observer
            _this4._manifestEditModeChanged(_this4[propName], oldValue); // notify


            _this4.dispatchEvent(new CustomEvent("manifest-edit-mode-changed", {
              detail: _this4[propName]
            }));
          }

          if (propName == "dashboardOpened" || propName == "t") {
            // observer
            _this4._dashboardOpenedChanged(_this4.dashboardOpened);
          } // make sure edit matches but state doesnt shift


          if (propName === "t") {
            if (_this4.editMode) {
              _this4.__editText = _this4.t.savePageContent;
            } else {
              _this4.__editText = _this4.t.editPageContent;
            }
          }
        });
      }
    }, {
      key: "connectedCallback",
      value: function connectedCallback() {
        var _this5 = this;

        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "connectedCallback", this).call(this);
        (0, _mobxEsm.autorun)(function (reaction) {
          if (_haxcmsSiteStore.store.userData) {
            _this5.userName = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.userData.userName);
            _this5.userPicture = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.userData.userPicture); // update buttons to match since we got a state response

            _this5.updateAvailableButtons();
          }

          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this5.editMode = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.editMode);

          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this5.manifest = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest);
          _this5.icon = "hax:site-settings";

          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          _this5.dashboardOpened = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.dashboardOpened);

          _this5.__disposer.push(reaction);
        });
        (0, _mobxEsm.autorun)(function (reaction) {
          var activeItem = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeItem); // update buttons to match since we got a state response

          _this5.updateAvailableButtons();

          if (activeItem && activeItem.id) {
            _this5.activeTitle = activeItem.title;
            _this5.pageAllowed = true;
          } else {
            _this5.pageAllowed = false;
          }

          _this5.__disposer.push(reaction);
        });
      }
    }, {
      key: "disconnectedCallback",
      value: function disconnectedCallback() {
        for (var i in this.__disposer) {
          this.__disposer[i].dispose();
        }

        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI.prototype), "disconnectedCallback", this).call(this);
      }
    }, {
      key: "_dashboardOpenedChanged",
      value: function _dashboardOpenedChanged(newValue) {
        if (newValue) {
          this.__settingsText = this.t.closeSiteSettings;
          this.icon = "icons:cancel";
        } else if (!newValue) {
          this.__settingsText = this.t.editSiteSettings;
          this.icon = "hax:site-settings";
        }
      }
      /**
       * toggle state on button tap
       */

    }, {
      key: "_editButtonTap",
      value: function _editButtonTap(e) {
        _haxcmsSiteStore.store.playSound('click');

        this.editMode = !this.editMode; // save button shifted to edit

        if (!this.editMode) {
          this.dispatchEvent(new CustomEvent("haxcms-save-node", {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: _haxcmsSiteStore.store.activeItem
          }));
        }

        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
      }
    }, {
      key: "_editDetailsButtonTap",
      value: function _editDetailsButtonTap(e) {
        _haxcmsSiteStore.store.playSound('click');

        var evt = new CustomEvent("haxcms-load-node-fields", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.target
        });
        window.dispatchEvent(evt);
      }
    }, {
      key: "_cancelButtonTap",
      value: function _cancelButtonTap(e) {
        this.editMode = false;

        _haxcmsSiteStore.store.playSound('error');

        this.dispatchEvent(new CustomEvent("hax-cancel", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.detail
        }));
        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
      }
      /**
       * create new item
       */

    }, {
      key: "_createNewItem",
      value: function _createNewItem(e) {
        var evt = new CustomEvent("haxcms-create-node", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            values: this.__newForm.value
          }
        });
        this.dispatchEvent(evt);
      }
      /**
       * Fire item
       */

    }, {
      key: "_updateItem",
      value: function _updateItem(e) {
        var local = e.target;
        var values;

        if (!local.__form) {
          values = local.parentNode.__form.value;
        } else {
          values = local.__form.value;
        } // fire event with details for saving


        window.dispatchEvent(new CustomEvent("haxcms-save-node-details", {
          bubbles: true,
          composed: true,
          cancelable: true,
          detail: values
        })); // fire event to close the modal

        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          composed: true,
          cancelable: true,
          detail: {}
        }));
      }
      /**
       * Delete button hit, confirm they want to do this
       */

    }, {
      key: "_deleteButtonTap",
      value: function _deleteButtonTap(e) {
        _haxcmsSiteStore.store.playSound('click');

        var c = document.createElement("span");
        c.innerHTML = "\"".concat(_haxcmsSiteStore.store.activeItem.title, "\" will be removed from the outline but its content stays on the file system.");
        var b1 = document.createElement("button");
        var icon = document.createElement("simple-icon");
        icon.icon = "icons:delete";
        b1.appendChild(icon);
        b1.appendChild(document.createTextNode("Confirm"));
        b1.style.color = "white";
        b1.style.backgroundColor = "#ee0000";
        b1.addEventListener("click", this._deleteActive.bind(this));
        var b2 = document.createElement("button");
        var icon2 = document.createElement("simple-icon");
        icon2.icon = "icons:cancel";
        b2.appendChild(icon2);
        b2.appendChild(document.createTextNode("cancel"));
        b2.addEventListener("click", function () {
          return _haxcmsSiteStore.store.playSound('error');
        });
        b2.setAttribute("dialog-dismiss", "dialog-dismiss");
        var b = document.createElement("span");
        b.appendChild(b1);
        b.appendChild(b2);
        var evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: "Are you sure you want to delete this page?",
            styles: {
              "--simple-modal-z-index": "100000000",
              "--simple-modal-min-width": "30vw",
              "--simple-modal-min-height": "30vh"
            },
            elements: {
              content: c,
              buttons: b
            },
            invokedBy: this.shadowRoot.querySelector("#deletebutton"),
            clone: false,
            modal: true
          }
        });
        window.dispatchEvent(evt);
      }
      /**
       * delete active item
       */

    }, {
      key: "_deleteActive",
      value: function _deleteActive(e) {
        _haxcmsSiteStore.store.playSound('click');

        var evt = new CustomEvent("haxcms-delete-node", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            item: _haxcmsSiteStore.store.activeItem
          }
        });
        this.dispatchEvent(evt);
      }
      /**
       * toggle state on button tap
       */

    }, {
      key: "_outlineButtonTap",
      value: function _outlineButtonTap(e) {
        _haxcmsSiteStore.store.playSound('click');

        var evt = new CustomEvent("simple-modal-show", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: {
            title: this.t.editSiteOutline,
            styles: {
              "--simple-modal-width": "70vw",
              "--simple-modal-max-width": "70vw",
              "--simple-modal-z-index": "100000000",
              "--simple-modal-height": "70vh",
              "--simple-modal-max-height": "70vh"
            },
            elements: {
              content: document.createElement("haxcms-outline-editor-dialog")
            },
            invokedBy: this.shadowRoot.querySelector("#outlinebutton"),
            clone: false,
            modal: true
          }
        });
        window.dispatchEvent(evt);
      }
      /**
       * toggle state on button tap
       */

    }, {
      key: "_manifestButtonTap",
      value: function _manifestButtonTap(e) {
        _haxcmsSiteStore.store.playSound('click');

        window.dispatchEvent(new CustomEvent("simple-modal-hide", {
          bubbles: true,
          cancelable: true,
          detail: {}
        }));
        window.dispatchEvent(new CustomEvent("haxcms-load-site-dashboard", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: e.target
        }));
      }
      /**
       * Edit state has changed.
       */

    }, {
      key: "_editModeChanged",
      value: function _editModeChanged(newValue, oldValue) {
        if (newValue) {
          // enable it some how
          this.__editIcon = "icons:save";
          this.__editText = this.t.savePageContent;
        } else {
          // disable it some how
          this.__editIcon = "hax:page-edit";
          this.__editText = this.t.editPageContent;
        }

        if (babelHelpers.typeof(oldValue) !== (typeof undefined === "undefined" ? "undefined" : babelHelpers.typeof(undefined))) {
          _haxcmsSiteStore.store.editMode = newValue; // force tray status to be the opposite of the editMode
          // to open when editing and close when not

          _haxStore.HAXStore.haxTray.collapsed = !newValue;
        }
      }
      /**
       * Note changes to the outline / structure of the page's items
       */

    }, {
      key: "_outlineEditModeChanged",
      value: function _outlineEditModeChanged(newValue, oldValue) {
        this.dispatchEvent(new CustomEvent("haxcms-outline-edit-mode-changed", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: newValue
        }));
      }
      /**
       * Note changes to the outline / structure of the page's items
       */

    }, {
      key: "_manifestEditModeChanged",
      value: function _manifestEditModeChanged(newValue, oldValue) {
        this.dispatchEvent(new CustomEvent("haxcms-manifest-edit-mode-changed", {
          bubbles: true,
          composed: true,
          cancelable: false,
          detail: newValue
        }));
      }
    }], [{
      key: "styles",
      get: function get() {
        return [].concat(babelHelpers.toConsumableArray(babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI), "styles", this)), [(0, _index.css)(_templateObject2_60cb7f60c76b11eca96c833ece275368 || (_templateObject2_60cb7f60c76b11eca96c833ece275368 = babelHelpers.taggedTemplateLiteral(["\n        :host *:not(:defined) {\n          display: none;\n        }\n        :host {\n          display: flex;\n          position: fixed;\n          left: 0;\n          top: 0;\n          right: 0;\n          background-color: var(--haxcms-system-bg, #37474f);\n          z-index: 10000;\n          visibility: visible;\n        }\n        :host([edit-mode]) {\n          z-index: 9999;\n        }\n        :host([dashboard-opened]) {\n          left: 50vw;\n        }\n        /**\n         * Dashboard open trumps all contextual settings\n         */\n        :host([dashboard-opened]) #editbutton,\n        :host([dashboard-opened]) #editdetails,\n        :host([dashboard-opened]) #deletebutton,\n        :host([dashboard-opened]) #addbutton,\n        :host([dashboard-opened]) #addbuttonchild,\n        :host([dashboard-opened]) #duplicatebutton {\n          display: none !important;\n        }\n        :host([edit-mode]) #editdetails {\n          margin-left: 64px;\n        }\n        :host *[hidden] {\n          display: none;\n        }\n        simple-tooltip:not(:defined) {\n          display: none !important;\n        }\n        :host([painting]) {\n          opacity: 0;\n          visibility: hidden;\n        }\n        simple-icon-button {\n          vertical-align: text-bottom;\n          display: inline-flex;\n          --simple-icon-width: 24px;\n          --simple-icon-height: 24px;\n          border-radius: 50%;\n          border: none;\n          background-color: black;\n          color: white;\n          text-align: center;\n          line-height: 40px;\n          min-width: unset;\n          padding: 0;\n          margin: 4px 2px;\n          width: 40px;\n          height: 40px;\n          transition: 0.2s all ease-in-out;\n          position: relative;\n          box-shadow: 0 8px 10px 1px rgba(0, 0, 0, 0.14),\n            0 3px 14px 2px rgba(0, 0, 0, 0.12),\n            0 5px 5px -3px rgba(0, 0, 0, 0.4);\n        }\n        haxcms-button-add {\n          display: inline-flex;\n          margin: 4px 2px;\n          vertical-align: text-bottom;\n        }\n        .topbar-button:hover,\n        .topbar-button:focus,\n        .topbar-button:active {\n          opacity: .8;\n        }\n        .topbar-button {\n          background-image: url('", "');\n          background-color: var(--simple-colors-default-theme-accent-5, blue);\n        }\n        :host([dark-mode]) .topbar-button {\n          background-image: url('", "');\n        }\n        #cancelbutton {\n          background-color: var(--haxcms-system-danger-color);\n        }\n        #editbutton {\n          visibility: hidden;\n          opacity: 0;\n        }\n        :host([page-allowed]) #editbutton,\n        :host([page-allowed]) #editdetails,\n        :host([page-allowed]) #deletebutton {\n          visibility: visible;\n          opacity: 1;\n        }\n\n        :host(:hover),\n        :host(:active),\n        :host(:focus) {\n          opacity: 1;\n        }\n        simple-tooltip {\n          --simple-tooltip-background: #000000;\n          --simple-tooltip-opacity: 1;\n          --simple-tooltip-text-color: #ffffff;\n          --simple-tooltip-delay-in: 0;\n          --simple-tooltip-duration-in: 200ms;\n          --simple-tooltip-duration-out: 0;\n          --simple-tooltip-border-radius: 0;\n          --simple-tooltip-font-size: 14px;\n        }\n        app-hax-top-bar {\n          top: 0;\n          z-index: 1000;\n          right: 0;\n          left: 0;\n          position: fixed;\n        }\n        app-hax-top-bar::part(top-bar) {\n          grid-template-columns: 20% 60% 20%;\n        }\n        .haxLogo {\n          color: var(--simple-colors-default-theme-accent-12, black);\n        }\n        .haxLogo:hover,\n        .haxLogo:focus,\n        .haxLogo:active {\n          color: var(--haxcms-color);\n        }\n        .haxLogo simple-icon-lite {\n          --simple-icon-height: 40px;\n          --simple-icon-width: 40px;\n          margin: 4px;\n        }\n\n        .space-hack {\n          display: inline-flex;\n          width: 64px;\n          height: 48px;\n        }\n        .soundToggle {\n          position: relative;\n          display: inline-flex;\n          vertical-align: top;\n        }\n\n        .soundToggle img {\n          width: 24px;\n          height: 24px;\n        }\n\n        app-hax-search-bar {\n          vertical-align: middle;\n          display: inline-flex;\n        }\n        .user-menu {\n          display:none;\n        }\n        .user-menu.open {\n          display: block;\n          top: 49px;\n          right: 0px;\n          position: absolute;\n          border: 3px solid black;\n          background-color: white;\n        }\n        .user-menu.open button {\n          display: block;\n          width: 100%;\n          margin: 0;\n          padding: 8px;\n          font-size: 20px;\n          font-family: 'Press Start 2P', sans-serif;\n          color: black;\n          background-color: white;\n        }\n        .user-menu.open button:hover,\n        .user-menu.open button:active,\n        .user-menu.open button:focus {\n          background-color: black;\n          color: white;\n        }\n        .user-menu .ops-panel {\n          justify-content: space-around;\n          display: flex;\n          padding: 8px 16px;\n        }\n\n        :host([dark-mode]) .user-menu.open {\n          background-color: black;\n          color: white;\n          border: 3px solid white;\n        }\n        :host([dark-mode]) .user-menu.open button {\n          background-color: black;\n          color: white;\n        }\n\n        :host([dark-mode]) .user-menu.open button:hover,\n        :host([dark-mode]) .user-menu.open button:active,\n        :host([dark-mode]) .user-menu.open button:focus {\n          background-color: white;\n          color: black;\n        }\n\n        .topbar-character {\n          cursor: pointer;\n          display: inline-block;\n          border: none;\n          border-radius: 0px;\n          background-color: transparent;\n        }\n        .topbar-character:hover,\n        .topbar-character:focus {\n          background-color: var(--simple-colors-default-theme-light-blue-4);\n          outline: var(--haxcms-color) solid 3px;\n          outline-offset: -3px;\n          height: 48px;\n        }\n        .topbar-character rpg-character {\n          margin: -4px -14px 0px -10px;\n          height: 52px;\n          width: 64px;      \n          display: inline-block;\n        }\n\n        .user-menu.open > .logout {\n          background-image: url('", "');\n          background-repeat: no-repeat;\n          background-position: center;\n          text-align: center;\n        }\n\n        @media screen and (max-width: 800px) {\n          :host([dashboard-opened]) {\n            left: 90vw;\n          }\n          :host([edit-mode]) {\n            bottom: unset;\n          }\n        }\n      "])), (0, _index.unsafeCSS)(ButtonBGLight), (0, _index.unsafeCSS)(ButtonBGDark), (0, _index.unsafeCSS)(LogOut))]);
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "tag",
      get: function get() {
        return "haxcms-site-editor-ui";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSSiteEditorUI), "properties", this)), {}, {
          userName: {
            type: String,
            attribute: "user-name"
          },
          rpgHat: {
            type: String
          },
          userPicture: {
            type: String,
            attribute: "user-picture"
          },
          userMenuOpen: {
            type: Boolean
          },
          soundIcon: {
            type: String
          },
          darkMode: {
            type: Boolean,
            reflect: true,
            attribute: 'dark-mode'
          },
          backLink: {
            type: String
          },
          backText: {
            type: String
          },
          __editIcon: {
            type: String
          },
          __editText: {
            type: String
          },
          __settingsText: {
            type: String
          },

          /**
           * small visual lock that events break on initial paint
           */
          painting: {
            type: Boolean,
            reflect: true
          },

          /**
           * page allowed
           */
          pageAllowed: {
            type: Boolean,
            attribute: "page-allowed",
            reflect: true
          },

          /**
           * if the page is in an edit state or not
           */
          editMode: {
            type: Boolean,
            reflect: true,
            attribute: "edit-mode"
          },

          /**
           * Manifest editing state
           */
          manifestEditMode: {
            type: Boolean,
            attribute: "manifest-edit-mode",
            reflect: true
          },
          activeTitle: {
            type: String,
            attribute: "active-title"
          },
          manifest: {
            type: Object
          },
          icon: {
            type: String
          },
          dashboardOpened: {
            type: Boolean,
            reflect: true,
            attribute: "dashboard-opened"
          }
        });
      }
    }]);
    return HAXCMSSiteEditorUI;
  }((0, _HAXCMSThemeParts2.HAXCMSThemeParts)((0, _HAXCMSI18NMixin.HAXCMSI18NMixin)(_simpleColors.SimpleColors)));

  _exports.HAXCMSSiteEditorUI = HAXCMSSiteEditorUI;
  window.customElements.define(HAXCMSSiteEditorUI.tag, HAXCMSSiteEditorUI);
});
