// @lrnwebcomponents/haxcms-elements/lib/core/haxcms-site-insights.js
define(["exports", "./haxcms-site-store.js", "../../../../mobx/dist/mobx.esm.js", "../../../../lit/index.js", "./utils/HAXCMSI18NMixin.js", "../../../micro-frontend-registry/micro-frontend-registry.js", "../../../micro-frontend-registry/lib/microServices.js", "../../../simple-tooltip/simple-tooltip.js", "../../../a11y-tabs/a11y-tabs.js", "../../../../@github/time-elements/dist/relative-time-element.js", "../../../iframe-loader/lib/loading-indicator.js"], function (_exports, _haxcmsSiteStore, _mobxEsm, _index, _HAXCMSI18NMixin2, _microFrontendRegistry, _microServices, _simpleTooltip, _a11yTabs, _relativeTimeElement, _loadingIndicator) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.HAXCMSShareDialog = void 0;

  var _templateObject_f6d7bec039ed11ed89d1c353cb18487d, _templateObject2_f6d7bec039ed11ed89d1c353cb18487d, _templateObject3_f6d7bec039ed11ed89d1c353cb18487d, _templateObject4_f6d7bec039ed11ed89d1c353cb18487d, _templateObject5_f6d7bec039ed11ed89d1c353cb18487d, _templateObject6_f6d7bec039ed11ed89d1c353cb18487d, _templateObject7_f6d7bec039ed11ed89d1c353cb18487d, _templateObject8_f6d7bec039ed11ed89d1c353cb18487d, _templateObject9_f6d7bec039ed11ed89d1c353cb18487d, _templateObject10_f6d7bec039ed11ed89d1c353cb18487d, _templateObject11_f6d7bec039ed11ed89d1c353cb18487d, _templateObject12_f6d7bec039ed11ed89d1c353cb18487d, _templateObject13_f6d7bec039ed11ed89d1c353cb18487d, _templateObject14_f6d7bec039ed11ed89d1c353cb18487d, _templateObject15_f6d7bec039ed11ed89d1c353cb18487d, _templateObject16_f6d7bec039ed11ed89d1c353cb18487d, _templateObject17_f6d7bec039ed11ed89d1c353cb18487d, _templateObject18_f6d7bec039ed11ed89d1c353cb18487d, _templateObject19_f6d7bec039ed11ed89d1c353cb18487d, _templateObject20_f6d7bec039ed11ed89d1c353cb18487d;

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

  (0, _microServices.enableServices)(["haxcms"]);
  /**
   * `haxcms-outline-editor-dialog`
   * `Dialog for presenting an editable outline`
   *
   * @demo demo/index.html
   *
   * @microcopy - the mental model for this element
   */

  var HAXCMSShareDialog = /*#__PURE__*/function (_HAXCMSI18NMixin) {
    babelHelpers.inherits(HAXCMSShareDialog, _HAXCMSI18NMixin);

    var _super = _createSuper(HAXCMSShareDialog);

    function HAXCMSShareDialog() {
      var _this;

      babelHelpers.classCallCheck(this, HAXCMSShareDialog);
      _this = _super.call(this);
      _this.t = _this.t || {};
      _this.t = _objectSpread(_objectSpread({}, _this.t), {}, {
        refresh: "Refresh",
        loading: "Loading",
        course: "Course",
        lesson: "Lesson",
        page: "Page",
        hour: "hour",
        hours: "hours",
        minute: "minute",
        minutes: "minutes"
      });
      _this.data = {
        site: {
          readability: {},
          updatedItems: []
        },
        branch: {
          readability: {}
        },
        page: {
          readability: {}
        }
      };
      _this.loading = false;
      return _this;
    }

    babelHelpers.createClass(HAXCMSShareDialog, [{
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSShareDialog.prototype), "firstUpdated", this).call(this, changedProperties);
        this.refreshData();
      }
    }, {
      key: "_insightResponse",
      value: function _insightResponse(data) {
        this.loading = false;
        this.data = data.data;
      }
    }, {
      key: "refreshData",
      value: function refreshData() {
        var base = "";

        if (document.querySelector("base")) {
          base = document.querySelector("base").href;
        }

        var site = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.manifest);
        var ancestor = (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.ancestorItem);
        var params = {
          type: "site",
          site: {
            file: base + "site.json",
            id: site.id,
            title: site.title,
            author: site.author,
            description: site.description,
            license: site.license,
            metadata: site.metadata,
            items: site.items
          },
          activeId: (0, _mobxEsm.toJS)(_haxcmsSiteStore.store.activeId),
          ancestorId: ancestor ? ancestor.id : null,
          link: base
        };
        this.loading = true;

        _microFrontendRegistry.MicroFrontendRegistry.call("@haxcms/insights", params, this._insightResponse.bind(this), this);
      }
      /**
       * Store the tag name to make it easier to obtain directly.
       */

    }, {
      key: "getReadingTime",
      value: function getReadingTime(value) {
        var readVal = [];
        var hours = Math.floor(value / 60);
        var minutes = Math.floor(value - hours * 60); // handle hours of reading

        if (hours === 1) {
          readVal.push("".concat(hours, " ").concat(this.t.hour));
        } else if (hours === 0) {// do nothing for 0
        } else {
          readVal.push("".concat(hours, " ").concat(this.t.hours));
        } // minutes


        if (minutes === 1) {
          readVal.push("".concat(minutes, " ").concat(this.t.minute));
        } else if (minutes === 0) {// do nothing for 0
        } else {
          readVal.push("".concat(minutes, " ").concat(this.t.minutes));
        }

        return readVal.join(", ");
      }
    }, {
      key: "closeModal",
      value: function closeModal() {
        window.dispatchEvent(new CustomEvent('simple-modal-hide'));
      } // render function

    }, {
      key: "render",
      value: function render() {
        var _this2 = this;

        var site = this.data.site;
        var branch = this.data.branch;
        var page = this.data.page;
        return (0, _index.html)(_templateObject_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["\n    <simple-icon-button-lite @click=\"", "\" icon=\"refresh\" ?disabled=\"", "\">", "</simple-icon-button-lite>\n    <a11y-tabs id=\"tabs\" full-width>\n      <a11y-tab id=\"course\" icon=\"communication:import-contacts\" label=\"", "\">\n      <loading-indicator full ?loading=\"", "\"></loading-indicator>\n      ", "\n      </a11y-tab>\n      <a11y-tab id=\"lesson\" icon=\"editor:format-list-bulleted\" label=\"", "\">\n        <loading-indicator full ?loading=\"", "\"></loading-indicator>\n        <div>\n        ", "\n        </div>\n      </a11y-tab>\n      <a11y-tab id=\"page\" icon=\"icons:subject\" label=\"", "\">\n        <div>\n        <loading-indicator full ?loading=\"", "\"></loading-indicator>\n        ", "\n          </div>\n        </a11y-tab>\n      </a11y-tabs>\n    "])), this.refreshData, this.loading, this.t.refresh, this.t.course, this.loading, this.loading ? (0, _index.html)(_templateObject2_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject2_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<p>", "...</p>"])), this.t.loading) : (0, _index.html)(_templateObject3_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject3_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["\n        <div>\n          <h2>Course insights</h2>\n          <ul>\n            <li><strong>", "</strong> ", "</li>\n            <li><strong>", "</strong> special elements</li>\n            <li><strong>", "</strong> headings</li>\n            <li><strong>", "</strong> links</li>\n            ", "\n            <li><strong>", "</strong> Pages</li>\n            <li><strong>", "</strong> of reading</li>\n            <li><strong>", "</strong> grade reading level</li>\n            <li><strong>", "</strong> words</li>\n            <li><strong>", "</strong> difficult words</li>\n            <li>Created:\n              <relative-time datetime=\"", "\">\n              </relative-time></li>\n            <li>Last updated:\n              <relative-time datetime=\"", "\">\n              </relative-time></li>\n            <li>Recent updates:\n              <ol>\n                ", "\n              </ol>\n            </li>\n          </ul>\n        </div>\n      "])), site.objectives, site.objectives == 0 ? (0, _index.html)(_templateObject4_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject4_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["objectives, why not add some?"]))) : (0, _index.html)(_templateObject5_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject5_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["objectives in this course"]))), site.specialTags, site.headings, site.links, site.video == 0 ? "" : (0, _index.html)(_templateObject6_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject6_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<li><strong>", "</strong> videos ", "</li>"])), site.video, site.videoLength != 0 ? (0, _index.html)(_templateObject7_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject7_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["(<strong>", "</strong>)"])), toHHMMSS(site.videoLength)) : ""), site.pages, this.getReadingTime(site.readTime), site.readability.gradeLevel, site.readability.lexiconCount, site.readability.difficultWords, site.created, site.updated, site.updatedItems.map(function (item) {
          return (0, _index.html)(_templateObject8_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject8_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["\n                <li>\n                  <a @click=\"", "\" href=\"", "\">", " <relative-time datetime=\"", "\"></relative-time></a>\n                </li>"])), _this2.closeModal, item.slug, item.title, item.metadata.updated);
        })), this.t.lesson, this.loading, this.loading ? (0, _index.html)(_templateObject9_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject9_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<p>", "...</p>"])), this.t.loading) : (0, _index.html)(_templateObject10_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject10_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["\n        <h2>Lesson insights</h2>\n          <ul>\n            ", "\n            <li><strong>", "</strong> ", "</li>\n            <li><strong>", "</strong> special elements</li>\n            <li><strong>", "</strong> headings</li>\n            <li><strong>", "</strong> links</li>\n                <li><strong>", "</strong> Pages</li>\n                <li><strong>", "</strong> of reading</li>\n                <li><strong>", "</strong> grade reading level</li>\n                <li><strong>", "</strong> words</li>\n                <li><strong>", "</strong> difficult words</li>\n            <li>Created:\n              <relative-time datetime=\"", "\">\n              </relative-time>\n            </li>\n            <li>Last updated:\n              <relative-time datetime=\"", "\">\n              </relative-time>\n            </li>\n          </ul>\n          "])), branch.video == 0 ? "" : (0, _index.html)(_templateObject11_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject11_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<li><strong>", "</strong> videos ", "</li>"])), branch.video, branch.videoLength != 0 ? (0, _index.html)(_templateObject12_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject12_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["(<strong>", "</strong>)"])), toHHMMSS(branch.videoLength)) : ""), branch.objectives, branch.objectives == 0 ? (0, _index.html)(_templateObject13_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject13_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["objectives, why not add some?"]))) : (0, _index.html)(_templateObject14_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject14_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["objectives in this course"]))), branch.specialTags, branch.headings, branch.links, branch.pages, this.getReadingTime(branch.readTime), branch.readability.gradeLevel, branch.readability.lexiconCount, branch.readability.difficultWords, branch.created, branch.updated), this.t.page, this.loading, this.loading ? (0, _index.html)(_templateObject15_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject15_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<p>", "...</p>"])), this.t.loading) : (0, _index.html)(_templateObject16_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject16_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["\n            <h2>Page insights</h2>\n            <ul>\n              ", "\n              ", "\n              <li><strong>", "</strong> special elements</li>\n              <li><strong>", "</strong> headings</li>\n              <li><strong>", "</strong> links</li>\n              <li><strong>", "</strong> Pages</li>\n              <li><strong>", "</strong> of reading</li>\n              <li><strong>", "</strong> grade reading level</li>\n              <li><strong>", "</strong> words</li>\n              <li><strong>", "</strong> difficult words</li>\n              <li>Created:\n                <relative-time datetime=\"", "\">\n                </relative-time></li>\n              <li>Last updated:\n                <relative-time datetime=\"", "\">\n                </relative-time></li>\n            </ul>\n          "])), page.objectives == 0 ? "" : (0, _index.html)(_templateObject17_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject17_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<li><strong>", "</strong> objectives</li>"])), page.objectives), page.video == 0 ? "" : (0, _index.html)(_templateObject18_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject18_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["<li><strong>", "</strong> videos ", "</li>"])), page.video, page.videoLength != 0 ? (0, _index.html)(_templateObject19_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject19_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["(<strong>", "</strong>)"])), toHHMMSS(page.videoLength)) : ""), page.specialTags, page.headings, page.links, page.pages, this.getReadingTime(page.readTime), page.readability.gradeLevel, page.readability.lexiconCount, page.readability.difficultWords, page.created, page.updated));
      }
    }], [{
      key: "styles",
      get:
      /**
       * LitElement constructable styles enhancement
       */
      function get() {
        return [(0, _index.css)(_templateObject20_f6d7bec039ed11ed89d1c353cb18487d || (_templateObject20_f6d7bec039ed11ed89d1c353cb18487d = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          overflow: auto;\n        }\n      "])))];
      }
    }, {
      key: "tag",
      get: function get() {
        return "haxcms-site-insights";
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(HAXCMSShareDialog), "properties", this)), {}, {
          data: {
            type: Object
          },
          loading: {
            type: Boolean,
            reflect: true
          }
        });
      }
    }]);
    return HAXCMSShareDialog;
  }((0, _HAXCMSI18NMixin2.HAXCMSI18NMixin)(_index.LitElement));

  _exports.HAXCMSShareDialog = HAXCMSShareDialog;
  window.customElements.define(HAXCMSShareDialog.tag, HAXCMSShareDialog);

  // convert seconds back into full time stamp
  function toHHMMSS(seconds) {
    var out = "";
    var snum = parseInt(seconds, 10);
    var hours = Math.floor(snum / 3600);
    var minutes = Math.floor((snum - hours * 3600) / 60);

    if (hours !== 0) {
      out += "".concat(hours, " hour");

      if (hours !== 1) {
        out += "s";
      }

      out += ", ";
    }

    if (minutes !== 0) {
      out += "".concat(minutes, " minute");

      if (hours !== 1) {
        out += "s";
      }
    }

    return out;
  }
});
