// @haxtheweb/citation-builder/citation-builder.js
/**
 * Copyright 2026 winstonwumbo
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as t,html as i,css as e,render as o}from"../../lit/index.js";import{DDDSuper as n}from"../d-d-d/d-d-d.js";import{I18NMixin as a}from"../i18n-manager/lib/I18NMixin.js";import"../simple-modal/simple-modal.js";import"../simple-icon/lib/simple-icons.js";import"../simple-icon/lib/simple-icon-button-lite.js";import"../hax-iconset/lib/simple-hax-iconset.js";import"./lib/citation-item.js";export class CitationBuilder extends(n(a(t))){static get tag(){return"citation-builder"}constructor(){super(),this.title="",this.t=this.t||{},this.t={...this.t,title:"References"},this.registerLocalization({context:this,localesPath:new URL("./locales/citation-builder.ar.json",import.meta.url).href+"/../"}),this.exportMode="",this.citationArr=["Select a citation format below"],this.addEventListener("add-citation",this._addCitationHandler.bind(this)),this.addEventListener("remove-citation",this._removeCitationHandler.bind(this))}static get properties(){return{...super.properties,title:{type:String},exportMode:{type:String,attribute:"export-mode"},citationArr:{type:Array,attribute:"citation-array"}}}static get styles(){return[super.styles,e`
      :host {
        display: block;
        color: var(--ddd-theme-primary);
        background-color: var(--ddd-theme-accent);
        font-family: var(--ddd-font-navigation);
      }
      .wrapper {
        padding: var(--ddd-spacing-4);
        margin: var(--ddd-spacing-2);
        border: var(--ddd-border-md);
      }
      h3 {
        margin-top: 16px;
      }
      h3 span {
        font-size: var(--citation-builder-label-font-size, var(--ddd-font-size-s));
      }
      .title-card {
        display: flex;
        justify-content: space-between;
      }

      #export-button {
        align-self: center;
        background-color: white;
      }

      .citation-list {
        border: black solid 1px;
      }
    `]}render(){return i`
    <div class="wrapper">
      <div class="title-card">
        <h3><span>${this.t.title}:</span> ${this.title}</h3>
        <button id="export-button" @click=${this._showExportModal}>Export</button>
      </div>      

      <slot>
        <citation-item title="Make your own citation with the HAX Editor" publication-date="${new Date}"></citation-item>
      </slot>
    </div>`}updated(t){super.updated&&super.updated(t),console.log(this.editMode),t.forEach((t,i)=>{"citationArr"===i&&void 0!==t&&this.renderExport()})}_showExportModal(t){this._modalContent=this._modalContent||globalThis.document.createElement("div"),this.renderExport(),this.dispatchEvent(new CustomEvent("simple-modal-show",{bubbles:!0,composed:!0,cancelable:!0,detail:{title:"Export Citations",elements:{content:this._modalContent},invokedBy:t.target}}))}renderExport(){o(i`
      <style>
        .citation-list {
          max-height: 40vh;
          overflow-y: auto;
          border: black solid 1px;
          padding: 8px;
          margin: 8px;
        }
        .control-bar {
          --simple-icon-button-border: 1px black solid;
          --simple-icon-button-border-radius: 4px;
          --simple-icon-button-focus-border: 2px black solid;
          --simple-icon-button-padding: 6px;
          --simple-icon-height: 32px;
          --simple-icon-width: 32px;
          display: flex;
          justify-content: space-between;
          margin-top: 16px;
        }
        simple-icon-button-lite {
          align-self: center;
          width: fit-content;
        }
        simple-icon-button-lite::part(button) {
          font-size: 20px;
        }
      </style>
      <div class="citation-list">
        ${this.citationArr.map(t=>i`<div class="citation-entry">${t}</div>`)}
      </div>
      <div class="control-bar">
        <div class="copy-button">
          <simple-icon-button-lite
            part="icon"
            class="icon"
            title="Copy"
            icon="content-copy"
            @click=${this._copyToClipboard}
          >Copy</simple-icon-button-lite>
        </div>
        <div class="export-drop">
          <select id="export-dropdown" @change=${this._exportHandler}>
            <option value="Export">Export</option>
            <option value="APA">APA</option>
            <option value="BibTeX">BibTeX</option>
          </select>
        </div>
      </div>
      `,this._modalContent,{host:this})}_exportHandler(t){this.exportMode=t.target.value;const i=this.shadowRoot.querySelector("slot").assignedElements();"APA"===this.exportMode?this.citationArr=i.map(t=>t.exportAPA()):"BibTeX"===this.exportMode?this.citationArr=i.map(t=>t.exportBibtex()):this.citationArr=["Select a citation format below"]}_copyToClipboard(){const t=this._modalContent.querySelector(".citation-list"),i=t.innerHTML,e=t.textContent,o=new Blob([i],{type:"text/html"}),n=new Blob([e],{type:"text/plain"}),a=[new ClipboardItem({"text/html":o,"text/plain":n})];navigator.clipboard.write(a)}_addCitationHandler(t){const i=globalThis.document.createElement("citation-item"),e={"data-hax-layout":!0,"data-hax-ray":"citation-item"};for(const t in e)i.setAttribute(t,e[t]);"above"===t.detail.direction?t.detail.node.before(i):"below"===t.detail.direction?t.detail.node.after(i):this.appendChild(i)}_removeCitationHandler(t){t.detail.node.remove()}haxHooks(){return{activeElementChanged:"haxactiveElementChanged",inlineContextMenu:"haxinlineContextMenu"}}haxinlineContextMenu(t){t.ceButtons=[{icon:"icons:add",callback:"_addCitationHandler",label:"Add citation"}]}haxactiveElementChanged(t,i){this.editMode=i,console.log("el",t),console.log("val",i);const e=this.shadowRoot.querySelector("slot").assignedElements(),o={"data-hax-layout":!0,"data-hax-ray":"citation-item"};for(const t of e)for(const i in o)console.log(t.textContent),t.setAttribute(i,o[i])}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}}globalThis.customElements.define(CitationBuilder.tag,CitationBuilder);
