// @lrnwebcomponents/grade-book/lib/letter-grade.js
define(["exports", "../../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.LetterGrade = void 0;

  function _templateObject4_3d36d6b099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        :host([letter=\"F\"]) {\n          --letter-grade-default-letter-background-color: #ff4444;\n        }\n        :host([letter=\"D\"]) {\n          --letter-grade-default-letter-background-color: #ff7744;\n        }\n        :host([letter=\"C\"]),\n        :host([letter=\"C+\"]) {\n          --letter-grade-default-letter-background-color: #ffff00;\n        }\n        :host([letter=\"B\"]),\n        :host([letter=\"B-\"]) {\n          --letter-grade-default-letter-background-color: #ffff66;\n        }\n        :host([letter=\"B+\"]),\n        :host([letter=\"A-\"]) {\n          --letter-grade-default-letter-background-color: #aaffaa;\n        }\n        :host([letter=\"A\"]) {\n          --letter-grade-default-letter-background-color: #55ff55;\n        }\n        :host([show-scale]) .wrap {\n          height: var(--letter-grade-wrap-height, 64px);\n          width: var(--letter-grade-wrap-width, 64px);\n        }\n        .wrap {\n          height: var(--letter-grade-wrap-height, 40px);\n          width: var(--letter-grade-wrap-width, 40px);\n          padding: var(--letter-grade-wrap-padding, 16px);\n          border: var(--letter-grade-wrap-border, 4px solid black);\n          border-radius: 50%;\n          box-shadow: rgb(0 0 0 / 20%) 0px 5px 5px 0px;\n          text-align: center;\n          justify-content: center;\n          vertical-align: middle;\n          background-color: var(\n            --letter-grade-wrap-background-color,\n            var(--letter-grade-default-letter-background-color, transparent)\n          );\n        }\n        .letter {\n          font-weight: var(--letter-grade-letter-font-weight, bold);\n          font-size: var(--letter-grade-letter-font-size, 40px);\n          line-height: var(\n            --letter-grade-letter-line-height,\n            var(--letter-grade-letter-font-size, 40px)\n          );\n          letter-spacing: -2px;\n        }\n        .score,\n        .range {\n          font-weight: var(--letter-grade-range-font-weight, normal);\n          font-size: var(--letter-grade-range-font-size, 12px);\n        }\n      "]);

    _templateObject4_3d36d6b099b011eb851ed31c596cbbd8 = function _templateObject4_3d36d6b099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _templateObject3_3d36d6b099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["<div class=\"range\">\n              ", "% -\n              ", "%\n            </div>"]);

    _templateObject3_3d36d6b099b011eb851ed31c596cbbd8 = function _templateObject3_3d36d6b099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _templateObject2_3d36d6b099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["<div class=\"score\">\n              (", "/", " ", ")\n            </div>"]);

    _templateObject2_3d36d6b099b011eb851ed31c596cbbd8 = function _templateObject2_3d36d6b099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _templateObject_3d36d6b099b011eb851ed31c596cbbd8() {
    var data = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"wrap\">\n        <div class=\"letter\">", "</div>\n        ", "\n        ", "\n      </div>\n    "]);

    _templateObject_3d36d6b099b011eb851ed31c596cbbd8 = function _templateObject_3d36d6b099b011eb851ed31c596cbbd8() {
      return data;
    };

    return data;
  }

  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }

  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

  var LetterGrade = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(LetterGrade, _LitElement);

    var _super = _createSuper(LetterGrade);

    function LetterGrade() {
      var _this;

      babelHelpers.classCallCheck(this, LetterGrade);
      _this = _super.call(this);
      _this.pointsSystem = "pts";
      _this.score = null;
      _this.total = null;
      _this.gradeScale = [];
      _this._letterIndex = null;
      _this.letter = "";
      _this.showScale = false;
      return _this;
    }

    babelHelpers.createClass(LetterGrade, [{
      key: "render",
      value: function render() {
        return (0, _litElement.html)(_templateObject_3d36d6b099b011eb851ed31c596cbbd8(), this.letter, this.showScale && this._letterIndex !== null ? (0, _litElement.html)(_templateObject2_3d36d6b099b011eb851ed31c596cbbd8(), this.score, this.total, this.pointsSystem) : "", this.showScale && this._letterIndex !== null ? (0, _litElement.html)(_templateObject3_3d36d6b099b011eb851ed31c596cbbd8(), this.gradeScale[this._letterIndex].lowRange, this.gradeScale[this._letterIndex].highRange) : "");
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;

        changedProperties.forEach(function (oldValue, propName) {
          if (["total", "score", "gradeScale"].includes(propName) && _this2[propName]) {
            if (_this2.gradeScale.length > 0) {
              clearTimeout(_this2.__debounce);
              _this2.__debounce = setTimeout(function () {
                _this2.letter = _this2.calculateLetterGrade();
              }, 0);
            }
          }
        });
      }
      /**
       * return the letter grade based on score and gradeScale existing
       */

    }, {
      key: "calculateLetterGrade",
      value: function calculateLetterGrade() {
        var percent = this.score / this.total * 100; // account for us having MORE points than max

        if (this.gradeScale[0] && percent >= this.gradeScale[0].highRange) {
          this._letterIndex = 0;
          return this.gradeScale[0].letter;
        } else {
          for (var i in this.gradeScale) {
            var val = this.gradeScale[i];

            if (percent <= val.highRange && percent >= val.lowRange) {
              this._letterIndex = i;
              return this.gradeScale[i].letter;
            }
          }

          this._letterIndex = this.gradeScale.length - 1;
          return this.gradeScale[this.gradeScale.length - 1].letter;
        }
      }
    }], [{
      key: "styles",
      get: function get() {
        return [(0, _litElement.css)(_templateObject4_3d36d6b099b011eb851ed31c596cbbd8())];
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          score: {
            type: Number
          },
          total: {
            type: Number
          },
          pointsSystem: {
            type: String,
            attribute: "points-system"
          },
          gradeScale: {
            type: Array,
            attribute: "grade-scale"
          },
          letter: {
            type: String,
            reflect: true
          },
          showScale: {
            type: Boolean,
            reflect: true,
            attribute: "show-scale"
          }
        };
      }
    }, {
      key: "tag",
      get: function get() {
        return "letter-grade";
      }
    }]);
    return LetterGrade;
  }(_litElement.LitElement);

  _exports.LetterGrade = LetterGrade;
  customElements.define(LetterGrade.tag, LetterGrade);
});
