// @lrnwebcomponents/super-daemon/super-daemon.js
/**
 * Copyright 2023 The Pennsylvania State University
 * @license Apache-2.0, see License.md for full text.
 */
import{LitElement as e,html as o,css as t}from"../../lit/index.js";import"../../web-dialog/index.js";import"./lib/super-daemon-ui.js";class SuperDaemon extends e{static get properties(){return{opened:{type:Boolean,reflect:!0},key1:{type:String},key2:{type:String},icon:{type:String},items:{type:Array}}}constructor(){super(),this.icon="hardware:keyboard-return",this.opened=!1,this.items=[];const e=void 0!==window.safari;this.key1=e?"Ctrl":"Alt",this.key2="Shift"}connectedCallback(){super.connectedCallback(),window.addEventListener("keydown",this.keyHandler.bind(this)),window.addEventListener("super-daemon-define-option",this.defineOptionEvent.bind(this))}disconnectedCallback(){super.connectedCallback(),window.removeEventListener("keydown",this.keyHandler.bind(this)),window.removeEventListener("super-daemon-define-option",this.defineOptionEvent.bind(this))}defineOptionEvent(e){this.defineOption(e.detail)}defineOption(e){e&&e.value&&e.title&&e.eventName&&(e.index=e.tags.join(" ")+" "+e.title+" "+e.key+" "+e.path,this.items.push(e))}keyHandler(e){if(this.allowedCallback()&&("Shift"==this.key2&&e.shiftKey&&("Ctrl"==this.key1&&e.ctrlKey||"Alt"==this.key1&&e.altKey)&&(this.opened=!this.opened),"Escape"==e.key&&this.opened&&(this.opened=!1),this.opened&&this.shadowRoot.querySelector("super-daemon-ui").filtered.length>0))switch(e.key){case"Enter":this.shadowRoot.querySelector("super-daemon-ui").shadowRoot.querySelector("super-daemon-row").selected();break;case"ArrowUp":case"ArrowDown":this.shadowRoot.querySelector("super-daemon-ui").shadowRoot.querySelector("super-daemon-row").shadowRoot.querySelector("button").focus()}}static get styles(){let e=[];return super.styles&&(e=super.styles),[...e,t`
        :host {
          display: none;
        }
        :host([opened]) {
          display: block;
        }
        web-dialog {
          --dialog-border-radius: var(--simple-modal-border-radius, 2px);
          z-index: var(--simple-modal-z-index, 1) !important;
          padding: 0;
        }
        web-dialog::part(dialog) {
          border: 1px solid var(--simple-modal-border-color, #222);
          min-height: var(--simple-modal-min-height, unset);
          min-width: var(--simple-modal-min-width, unset);
          z-index: var(--simple-modal-z-index, 1000);
          resize: var(--simple-modal-resize, unset);
          padding: 0;
          --dialog-height: var(--simple-modal-height, auto);
          --dialog-width: var(--simple-modal-width, 75vw);
          --dialog-max-width: var(--simple-modal-max-width, 100vw);
          --dialog-max-height: var(--simple-modal-max-height, 100vh);
        }
        web-dialog.style-scope.simple-modal {
          display: none !important;
        }
        web-dialog[open].style-scope.simple-modal {
          display: flex !important;
          position: fixed !important;
          margin: auto;
        }
        :host([resize="none"]) web-dialog[open].style-scope.simple-modal,
        :host([resize="horizontal"]) web-dialog[open].style-scope.simple-modal {
          top: calc(50% - var(--simple-modal-height, auto) / 2);
        }
        :host([resize="none"]) web-dialog[open].style-scope.simple-modal,
        :host([resize="vertical"]) web-dialog[open].style-scope.simple-modal {
          left: calc(50% - var(--simple-modal-width, 75vw) / 2);
        }
      `]}close(){this.opened=!1,window.ShadyCSS&&!window.ShadyCSS.nativeShadow&&(this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.position="relative")}open(){this.items=[...this.items],this.opened=!0;const e=this.shadowRoot.querySelector("web-dialog");this.modal?(e.$backdrop.removeEventListener("click",e.onBackdropClick),e.removeEventListener("keydown",e.onKeyDown,{capture:!0})):(e.$backdrop.addEventListener("click",e.onBackdropClick),e.addEventListener("keydown",e.onKeyDown,{capture:!0,passive:!0})),window.ShadyCSS&&!window.ShadyCSS.nativeShadow&&(this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.position="fixed",this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.top=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.bottom=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.left=0,this.shadowRoot.querySelector("web-dialog").shadowRoot.querySelector("#backdrop").style.right=0),this.shadowRoot.querySelector("super-daemon-ui").shadowRoot.querySelector("simple-fields-field").focus()}render(){return o`
      <web-dialog
        id="dialog"
        center
        role="dialog"
        part="dialog"
        aria-label="Super Daemon"
        aria-modal="true"
        ?open="${this.opened}"
        @open="${this.open}"
        @close="${this.close}"
      >
        <super-daemon-ui
          ?open="${this.opened}"
          icon="${this.icon}"
          .items="${this.items}"
          @super-daemon-close="${this.close}"
        ></super-daemon-ui>
      </web-dialog>
    `}allowedCallback(){return!0}static get tag(){return"super-daemon"}}customElements.define(SuperDaemon.tag,SuperDaemon);export{SuperDaemon};window.SuperDaemonManager=window.SuperDaemonManager||{},window.SuperDaemonManager.requestAvailability=()=>(window.SuperDaemonManager.instance||(window.SuperDaemonManager.instance=document.createElement("super-daemon"),document.body.appendChild(window.SuperDaemonManager.instance)),window.SuperDaemonManager.instance);export const SuperDaemonInstance=window.SuperDaemonManager.requestAvailability();
