// @haxtheweb/hax-body/lib/hax-element-demo.js
import{LitElement as e,html as i,css as t}from"../../../lit/index.js";import{HAXStore as r}from"./hax-store.js";import{haxElementToNode as d}from"../../utils/utils.js";import{IntersectionObserverMixin as o}from"../../intersection-element/lib/IntersectionObserverMixin.js";import"../../simple-icon/lib/simple-icon-lite.js";export class HaxElementDemo extends(o(e)){static get tag(){return"hax-element-demo"}constructor(){super(),this.renderTag=null,this.activePickerSchema=-1,this.gizmoTitle="",this.gizmoDescription="",this.gizmoIcon=""}static get properties(){return{...super.properties,renderTag:{type:String,attribute:"render-tag"},activePickerSchema:{type:Number,attribute:"active-picker-schema"},gizmoTitle:{type:String,attribute:"gizmo-title"},gizmoDescription:{type:String,attribute:"gizmo-description"},gizmoIcon:{type:String,attribute:"gizmo-icon"}}}static get styles(){return[t`
        :host {
          display: flex;
          flex-direction: column;
          overflow: hidden;
          width: 300px;
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-white)
          );
          background-color: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
        }
        .preview-wrap {
          height: 200px;
          width: 300px;
          overflow: hidden;
          padding: var(--ddd-spacing-2);
        }
        .preview-wrap ::slotted(*) {
          transform: scale(0.5);
          transform-origin: top left;
          width: 575px;
          max-height: 300px;
          pointer-events: none;
        }
        .info {
          padding: var(--ddd-spacing-2);
          border-top: var(--ddd-border-xs) solid
            light-dark(
              var(--ddd-theme-default-limestoneGray),
              var(--ddd-primary-5)
            );
          background-color: light-dark(
            var(--ddd-theme-default-white),
            var(--ddd-theme-default-coalyGray)
          );
          max-height: var(--ddd-spacing-16);
          overflow: hidden;
        }
        .title {
          display: flex;
          align-items: center;
          gap: var(--ddd-spacing-1);
          font-weight: var(--ddd-font-weight-bold);
          font-size: var(--ddd-font-size-4xs);
          margin-bottom: var(--ddd-spacing-1);
          color: light-dark(
            var(--ddd-theme-default-coalyGray),
            var(--ddd-theme-default-white)
          );
        }
        .title simple-icon {
          --simple-icon-height: var(--ddd-spacing-4);
          --simple-icon-width: var(--ddd-spacing-4);
          flex-shrink: 0;
        }
        .description {
          font-size: var(--ddd-font-size-4xs);
          line-height: var(--ddd-lh-140);
          color: light-dark(
            var(--ddd-theme-default-slateGray),
            var(--ddd-theme-default-limestoneGray)
          );
          overflow: hidden;
          text-overflow: ellipsis;
          display: -webkit-box;
          -webkit-line-clamp: 4;
          -webkit-box-orient: vertical;
        }
      `]}render(){const e=this.gizmoDescription&&this.gizmoDescription.length>200?this.gizmoDescription.substring(0,200)+"...":this.gizmoDescription;return i`
      <div class="preview-wrap">
        <slot></slot>
      </div>
      ${this.gizmoTitle||e?i`
        <div class="info">
          ${this.gizmoTitle?i`
            <div class="title">
              ${this.gizmoIcon?i`<simple-icon-lite icon="${this.gizmoIcon}"></simple-icon-lite>`:""}
              <span>${this.gizmoTitle}</span>
            </div>
          `:""}
          ${e?i`<div class="description">${e}</div>`:""}
        </div>
      `:""}
    `}updated(e){super.updated&&super.updated(e),e.forEach((e,i)=>{if("renderTag"===i&&this[i]){this.innerHTML="";let e=r.haxSchemaFromTag(this[i]);var t;t=-1!==this.activePickerSchema&&globalThis.document.querySelector("hax-picker")&&globalThis.document.querySelector("hax-picker")._elements&&globalThis.document.querySelector("hax-picker")._elements.length>0?d(globalThis.document.querySelector("hax-picker")._elements[this.activePickerSchema]):e&&e.gizmo&&e.gizmo.tag&&e.demoSchema&&e.demoSchema[0]?d(e.demoSchema[0]):globalThis.document.createElement(this[i]),this.appendChild(t)}})}}globalThis.customElements.define(HaxElementDemo.tag,HaxElementDemo);
