// @lrnwebcomponents/page-flag/lib/page-flag-comment.js
define(["exports", "../../../lit/index.js", "../../simple-icon/lib/simple-icon-button.js", "../../simple-icon/lib/simple-icons.js", "../../rpg-character/rpg-character.js", "../../simple-datetime/simple-datetime.js", "../../simple-fields/lib/simple-fields-field.js", "../../simple-picker/lib/simple-emoji-picker.js"], function (_exports, _index, _simpleIconButton, _simpleIcons, _rpgCharacter, _simpleDatetime, _simpleFieldsField, _simpleEmojiPicker) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.PageFlagComment = void 0;
  var _templateObject_0e1a9ab0e51211edbf9f41986c80248c, _templateObject2_0e1a9ab0e51211edbf9f41986c80248c, _templateObject3_0e1a9ab0e51211edbf9f41986c80248c, _templateObject4_0e1a9ab0e51211edbf9f41986c80248c, _templateObject5_0e1a9ab0e51211edbf9f41986c80248c, _templateObject6_0e1a9ab0e51211edbf9f41986c80248c, _templateObject7_0e1a9ab0e51211edbf9f41986c80248c;
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  var PageFlagComment = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(PageFlagComment, _LitElement);
    var _super = _createSuper(PageFlagComment);
    function PageFlagComment() {
      var _this;
      babelHelpers.classCallCheck(this, PageFlagComment);
      _this = _super.call(this);
      _this.mood = null;
      _this.seed = "abc123";
      _this.timestamp = Date.now() / 1000;
      _this.reply = 0;
      _this.editMode = false;
      _this.canEdit = false;
      _this.readOnly = true;
      _this.haxUIElement = true;
      return _this;
    }
    babelHelpers.createClass(PageFlagComment, [{
      key: "testCanUpdate",
      value: function testCanUpdate(user) {
        if (this.seed === user) {
          this.canEdit = true;
        } else {
          this.canEdit = false;
        }
      }
    }, {
      key: "deleteOp",
      value: function deleteOp() {
        this.dispatchEvent(new CustomEvent("page-flag-comment-delete", {
          bubbles: true,
          cancelable: true,
          detail: this
        }));
      }
    }, {
      key: "replyOp",
      value: function replyOp() {
        this.dispatchEvent(new CustomEvent("page-flag-comment-reply", {
          bubbles: true,
          cancelable: true,
          detail: this
        }));
      }
    }, {
      key: "editOp",
      value: function editOp() {
        this.dispatchEvent(new CustomEvent("page-flag-comment-edit", {
          bubbles: true,
          cancelable: true,
          detail: this
        }));
      }
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["\n      <div class=\"comment\">\n        ", "\n        <div class=\"comment__header\">\n          <div class=\"comment__header__avatar\">\n            <rpg-character\n              seed=\"", "\"\n              width=\"40\"\n              height=\"40\"\n              hat=\"", "\"\n            ></rpg-character>\n          </div>\n          <div class=\"comment__header__info\">\n            <div class=\"comment__header__info__name\">", "</div>\n            <div class=\"comment__header__info__date\">\n              <simple-datetime\n                format=\"m/j/y h:i\"\n                .timestamp=\"", "\"\n                unix\n              ></simple-datetime>\n            </div>\n          </div>\n        </div>\n        <div class=\"comment__body\">\n          <simple-fields-field type=\"textarea\"></simple-fields-field>\n          <slot></slot>\n        </div>\n        <div class=\"comment__footer\">\n          <div class=\"comment__footer__actions\">\n            ", "\n          </div>\n        </div>\n      </div>\n    "])), this.editMode ? (0, _index.html)(_templateObject2_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject2_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<simple-emoji-picker\n              align-right\n              label=\"Reaction\"\n              value=\"", "\"\n              @value-changed=\"", "\"\n            ></simple-emoji-picker>"])), this.mood, this.emojiChanged) : (0, _index.html)(_templateObject3_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject3_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<div class=\"emoji\" .innerHTML=\"", "\"></div>"])), this.mood), this.seed, this.editMode ? "edit" : "none", this.seed, this.timestamp, !this.readOnly ? (0, _index.html)(_templateObject4_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject4_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["\n                  ", "\n                  <simple-icon-button\n                    icon=\"reply\"\n                    title=\"Reply\"\n                    @click=\"", "\"\n                  ></simple-icon-button>\n                  ", "\n                "])), this.canEdit ? (0, _index.html)(_templateObject5_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject5_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<simple-icon-button\n                        icon=\"", "\"\n                        title=\"", "\"\n                        @click=\"", "\"\n                      ></simple-icon-button> "])), this.editMode ? "save" : "editor:mode-edit", this.editMode ? "Edit" : "Update", this.editOp) : "", this.replyOp, this.canEdit ? (0, _index.html)(_templateObject6_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject6_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<simple-icon-button\n                        icon=\"delete\"\n                        title=\"Delete\"\n                        @click=\"", "\"\n                      ></simple-icon-button>"])), this.deleteOp) : "") : "");
      }
    }, {
      key: "emojiChanged",
      value: function emojiChanged(e) {
        this.mood = e.detail.value;
      }
    }, {
      key: "firstUpdated",
      value: function firstUpdated(changedProperties) {
        babelHelpers.get(babelHelpers.getPrototypeOf(PageFlagComment.prototype), "firstUpdated", this).call(this, changedProperties);
        this.shadowRoot.querySelector("simple-fields-field").value = this.innerHTML;
      }
    }, {
      key: "updated",
      value: function updated(changedProperties) {
        var _this2 = this;
        babelHelpers.get(babelHelpers.getPrototypeOf(PageFlagComment.prototype), "updated", this).call(this, changedProperties);
        changedProperties.forEach(function (oldValue, propName) {
          if (propName == "editMode" && oldValue !== undefined) {
            if (_this2[propName]) {
              _this2.shadowRoot.querySelector("simple-fields-field").value = _this2.innerHTML;
              setTimeout(function () {
                _this2.shadowRoot.querySelector("simple-fields-field").focus();
                _this2.shadowRoot.querySelector("simple-fields-field").select();
              }, 0);
            } else {
              _this2.innerHTML = _this2.shadowRoot.querySelector("simple-fields-field").value;
            }
          }
        });
      }
    }], [{
      key: "tag",
      get: function get() {
        return "page-flag-comment";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          seed: {
            type: String
          },
          timestamp: {
            type: Number
          },
          mood: {
            type: String
          },
          editMode: {
            type: Boolean,
            reflect: true,
            attribute: "edit-mode"
          },
          canEdit: {
            type: Boolean
          },
          reply: {
            type: Number,
            reflect: true
          },
          readOnly: {
            type: Boolean
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        return [, (0, _index.css)(_templateObject7_0e1a9ab0e51211edbf9f41986c80248c || (_templateObject7_0e1a9ab0e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n        }\n        :host slot {\n          display: block;\n        }\n        :host simple-emoji-picker,\n        :host simple-fields-field {\n          display: none;\n        }\n        :host([edit-mode]) simple-emoji-picker,\n        :host([edit-mode]) simple-fields-field {\n          display: block;\n        }\n        simple-emoji-picker {\n          float: right;\n          width: 40px;\n        }\n        .emoji {\n          float: right;\n          width: 40px;\n          height: 40px;\n          font-size: 32px;\n        }\n        :host([edit-mode]) slot {\n          display: none;\n        }\n        :host([reply=\"1\"]) .comment {\n          margin-left: 16px;\n        }\n        :host([reply=\"2\"]) .comment {\n          margin-left: 32px;\n        }\n        .comment {\n          background-color: #fff;\n          border-radius: 5px;\n          box-shadow: 0 0 5px rgba(0, 0, 0, 0.2);\n          padding: 10px;\n          margin-bottom: 10px;\n        }\n        simple-icon-button {\n          --simple-icon-height: 16px;\n          --simple-icon-width: 16px;\n          margin: 0 6px;\n        }\n        .comment__header {\n          display: flex;\n          align-items: center;\n          margin-bottom: 10px;\n        }\n        .comment__header__avatar {\n          width: 40px;\n          height: 40px;\n          border-radius: 50%;\n          overflow: hidden;\n          border: 2px solid black;\n          margin-right: 8px;\n        }\n        .comment__header__info {\n          display: flex;\n          flex-direction: column;\n        }\n        .comment__header__info__name {\n          font-weight: bold;\n          font-size: 14px;\n        }\n        .comment__header__info__date {\n          font-size: 12px;\n          color: #999;\n        }\n        .comment__body {\n          margin-bottom: 10px;\n        }\n        .comment__footer {\n          display: flex;\n          justify-content: flex-end;\n        }\n        .comment__footer__actions {\n          display: flex;\n        }\n        .comment__footer__actions__button {\n          background-color: transparent;\n          border: none;\n          color: #999;\n          font-size: 12px;\n          cursor: pointer;\n          margin-left: 10px;\n        }\n      "])))];
      }
    }]);
    return PageFlagComment;
  }(_index.LitElement);
  _exports.PageFlagComment = PageFlagComment;
  customElements.define(PageFlagComment.tag, PageFlagComment);
});
