// @lrnwebcomponents/simple-fields/lib/simple-fields-fieldset.js
define(["exports", "../../../lit-element/lit-element.js"], function (_exports, _litElement) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SimpleFieldsFieldset = void 0;

  /**
   *`simple-fields-fieldset` takes in a JSON schema of type fieldset and builds a form,
   * exposing a `value` property that represents an array described by the schema.
   * @group simple-fields
   * @demo demo/index.html
   * @customElement simple-fields-fieldset
   */
  class SimpleFieldsFieldset extends _litElement.LitElement {
    static get tag() {
      return "simple-fields-fieldset";
    }

    static get styles() {
      return [(0, _litElement.css)`
        fieldset {
          padding: 20px;
          margin: 20px;
        }
      `];
    }

    render() {
      return (0, _litElement.html)`
      <fieldset>
        ${this.legend} ${this.description} ${this.fields}
      </fieldset>
    `;
    }

    get legend() {
      return (0, _litElement.html)`
      <legend id="legend" ?hidden="${!this.schema.title}">
        ${this.schema.title}
      </legend>
    `;
    }

    get description() {
      return (0, _litElement.html)`
      <div ?hidden="${!this.schema.description}">
        ${this.schema.description}
      </div>
    `;
    }

    get fields() {
      return (0, _litElement.html)`
      <div id="item-fields">
        <slot></slot>
      </div>
    `;
    }

    static get properties() {
      return {
        schema: {
          type: Object
        }
      };
    }

    constructor() {
      super();
      this.schema = {};
    }

  }

  _exports.SimpleFieldsFieldset = SimpleFieldsFieldset;
  window.customElements.define(SimpleFieldsFieldset.tag, SimpleFieldsFieldset);
});
