// @haxtheweb/media-playlist/media-playlist.js
/**
 * Copyright 2026 haxtheweb
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as e,html as t,css as i}from"../../lit/index.js";import{DDDSuper as a}from"../d-d-d/d-d-d.js";import{I18NMixin as s}from"../i18n-manager/lib/I18NMixin.js";import{copyToClipboard as d}from"../utils/utils.js";import"../video-player/video-player.js";import"../audio-player/audio-player.js";export class MediaPlaylist extends(a(s(e))){static get tag(){return"media-playlist"}constructor(){super(),this.edit=!1,this.activeIndex=0,this.mediaItems=[],this._playTimeout=null,this.t=this.t||{},this.t={...this.t,playlist:"Playlist",play:"Play",item:"Item",editModeLabel:"Media playlist edit mode"},this.registerLocalization({context:this,localesPath:new URL("./locales/media-playlist.ar.json",import.meta.url).href+"/../"}),this._observer=new MutationObserver(()=>{clearTimeout(this._debounceMutations),this._debounceMutations=setTimeout(()=>{this._updateMediaItems()},100)}),this._onMediaStateChange=e=>{e.detail&&0===e.detail.data&&this._advanceToNext()},this._onMediaEnded=()=>{this._advanceToNext()},this._lastMediaElement=null}connectedCallback(){super.connectedCallback(),this._observer.observe(this,{childList:!0,subtree:!1}),this.addEventListener("keydown",this._handleKeydown),this._darkModeQuery=globalThis.matchMedia("(prefers-color-scheme: dark)"),this._darkModeHandler=e=>{this._setDarkMode(e.matches)},this._darkModeQuery.addEventListener("change",this._darkModeHandler),this._setDarkMode(this._darkModeQuery.matches)}disconnectedCallback(){this._observer.disconnect(),this.removeEventListener("keydown",this._handleKeydown),this._darkModeQuery&&this._darkModeHandler&&this._darkModeQuery.removeEventListener("change",this._darkModeHandler),this._resizeObserver&&this._resizeObserver.disconnect(),clearTimeout(this._attachEndedTimeout);const e=this.shadowRoot?this.shadowRoot.querySelector("#player"):null;e&&e.removeEventListener("mediastatechange",this._onMediaStateChange),this._lastMediaElement&&(this._lastMediaElement.removeEventListener("ended",this._onMediaEnded),this._lastMediaElement=null),super.disconnectedCallback()}_setDarkMode(e){this.mediaItems.forEach(t=>{t.element.toggleAttribute("dark",e)})}static get properties(){return{...super.properties,edit:{type:Boolean,reflect:!0},activeIndex:{type:Number,reflect:!0,attribute:"active-index"},mediaItems:{type:Array}}}static get styles(){return[super.styles,i`
        :host {
          display: block;
          font-family: var(--ddd-font-navigation);
        }

        /* Edit mode */
        .edit-wrapper {
          border: var(--ddd-border-md);
          border-style: dashed;
          padding: var(--ddd-spacing-4);
          background-color: var(--ddd-theme-default-limestoneLight);
          box-sizing: border-box;
        }
        .edit-wrapper ::slotted(*) {
          display: block;
          width: 100%;
          margin-bottom: var(--ddd-spacing-2);
        }

        /* Runtime layout */
        .layout {
          display: grid;
          grid-template-columns: 1fr 320px;
          gap: var(--ddd-spacing-2);
          background: var(--ddd-theme-default-black);
          border-radius: var(--ddd-radius-sm);
          overflow: hidden;
          align-items: start;
        }

        .player-area {
          position: relative;
          background: var(--ddd-theme-default-black);
          width: 100%;
          display: block;
        }

        .player-area > * {
          width: 100%;
          display: block;
        }

        /* Hide the default slot that catches non-active children */
        .slot-hidden {
          display: none;
        }

        .playlist {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-1);
          overflow-y: auto;
          max-height: var(--playlist-max-height, 100%);
          padding: var(--ddd-spacing-2);
          background: var(--ddd-theme-default-coalyGray);
        }

        .playlist-header {
          font-size: var(--ddd-font-size-s);
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-white);
          padding: var(--ddd-spacing-2);
          margin: var(--ddd-spacing-0);
        }

        .playlist-item {
          display: grid;
          grid-template-columns: var(--ddd-spacing-4) 80px 1fr;
          gap: var(--ddd-spacing-2);
          align-items: center;
          padding: var(--ddd-spacing-2);
          background: var(--ddd-theme-default-white);
          border-radius: var(--ddd-radius-sm);
          cursor: pointer;
          text-align: left;
          transition: background-color 0.1s ease;
        }

        .playlist-item:hover {
          background: var(--ddd-theme-default-limestoneLight);
        }

        .playlist-item.active {
          background: var(--ddd-theme-default-limestoneMaxLight);
          border-left: var(--ddd-spacing-1) solid var(--ddd-theme-default-link);
        }

        .playlist-number {
          font-size: var(--ddd-font-size-xs);
          color: var(--ddd-theme-default-coalyGray);
          text-align: center;
        }

        .playlist-thumbnail {
          width: 80px;
          height: 45px;
          object-fit: cover;
          border-radius: var(--ddd-radius-xs);
          background: var(--ddd-theme-default-shade);
        }

        .playlist-info {
          display: flex;
          flex-direction: column;
          gap: var(--ddd-spacing-0);
          overflow: hidden;
        }

        .playlist-title {
          font-size: var(--ddd-font-size-xs);
          font-weight: var(--ddd-font-weight-bold);
          color: var(--ddd-theme-default-coalyGray);
          white-space: nowrap;
          overflow: hidden;
          text-overflow: ellipsis;
        }

        .playlist-duration {
          font-size: var(--ddd-font-size-4xs);
          color: var(--ddd-theme-default-shade);
        }

        /* Dark mode */
        @media (prefers-color-scheme: dark) {
          .playlist {
            background: var(--ddd-theme-default-black);
          }

          .playlist-header {
            color: var(--ddd-theme-default-white);
          }

          .playlist-item {
            background: var(--ddd-theme-default-coalyGray);
          }

          .playlist-item:hover {
            background: var(--ddd-theme-default-shade);
          }

          .playlist-item.active {
            background: var(--ddd-theme-default-limestoneGray);
          }

          .playlist-title {
            color: var(--ddd-theme-default-white);
          }

          .playlist-number {
            color: var(--ddd-theme-default-limestoneLight);
          }
        }

        /* Mobile */
        @media (max-width: 768px) {
          .layout {
            grid-template-columns: 1fr;
            grid-template-rows: auto 300px;
          }
          .playlist {
            max-height: 300px;
          }
        }

        @media (prefers-reduced-motion: reduce) {
          .playlist-item {
            transition: none;
          }
        }
      `]}render(){if(this.edit)return t`
        <div class="edit-wrapper">
          <slot></slot>
        </div>
      `;const e=this.mediaItems[this.activeIndex];return t`
      <div class="layout">
        <div class="player-area">
          ${e?e.isAudio?t`
                  <audio-player
                    id="player"
                    source="${e.source}"
                    media-title="${e.mediaTitle}"
                    thumbnail-src="${e.thumbnailSrc}"
                    .tracks="${e.tracks}"
                    crossorigin="${e.crossorigin}"
                    lang="${e.lang}"
                  ></audio-player>
                `:t`
                  <video-player
                    id="player"
                    source="${e.source}"
                    media-title="${e.mediaTitle}"
                    thumbnail-src="${e.thumbnailSrc}"
                    .tracks="${e.tracks}"
                    crossorigin="${e.crossorigin}"
                    lang="${e.lang}"
                  ></video-player>
                `:t``}
        </div>
        <div
          class="playlist"
          role="list"
          aria-label="${this.t.playlist}"
        >
          <h3 class="playlist-header">${this.t.playlist}</h3>
          ${this.mediaItems.map((e,i)=>t`
              <button
                class="playlist-item ${i===this.activeIndex?"active":""}"
                role="listitem"
                @click="${()=>this._setActiveIndex(i)}"
                aria-current="${i===this.activeIndex?"true":"false"}"
                aria-label="${this.t.play} ${e.title}, ${e.duration}"
                data-index="${i}"
              >
                <span class="playlist-number">${i+1}</span>
                ${e.thumbnailSrc?t`
                      <img
                        class="playlist-thumbnail"
                        src="${e.thumbnailSrc}"
                        alt=""
                        loading="lazy"
                      />
                    `:t` <div class="playlist-thumbnail"></div> `}
                <div class="playlist-info">
                  <span class="playlist-title">${e.title}</span>
                  <span class="playlist-duration">${e.duration}</span>
                </div>
              </button>
            `)}
        </div>
      </div>
      <div class="slot-hidden"><slot></slot></div>
    `}_updateMediaItems(){const e=Array.from(this.children).filter(e=>"VIDEO-PLAYER"===e.tagName||"AUDIO-PLAYER"===e.tagName);this.activeIndex>=e.length&&(this.activeIndex=0),this.mediaItems=e.map((e,t)=>{let i=[];if(e.tracks&&Array.isArray(e.tracks))i=e.tracks;else if(e.getAttribute("tracks"))try{i=JSON.parse(e.getAttribute("tracks"))}catch(e){}const a=e.mediaTitle||e.getAttribute("media-title")||`${this.t.item} ${t+1}`;return{element:e,tagName:e.tagName,source:e.source||e.getAttribute("source")||"",title:a,mediaTitle:a,thumbnailSrc:e.thumbnailSrc||e.getAttribute("thumbnail-src")||"",duration:e.duration||e.getAttribute("duration")||"",tracks:i,crossorigin:e.crossorigin||e.getAttribute("crossorigin")||"anonymous",lang:e.lang||e.getAttribute("lang")||"en",isAudio:"AUDIO-PLAYER"===e.tagName}});const t=this._darkModeQuery&&this._darkModeQuery.matches;this.mediaItems.forEach(e=>{e.element.toggleAttribute("dark",t)})}_setActiveIndex(e){if(!(e<0||e>=this.mediaItems.length)){if(e===this.activeIndex){const e=this.shadowRoot?this.shadowRoot.querySelector("#player"):null;if(e&&e.restart&&"function"==typeof e.restart)try{e.restart()}catch(e){}return}this.activeIndex=e}}_advanceToNext(){if(this._advancing)return;this._advancing=!0;const e=this.activeIndex+1;e<this.mediaItems.length&&this._setActiveIndex(e),setTimeout(()=>{this._advancing=!1},500)}_scrollToActiveItem(){const e=this.shadowRoot.querySelector(".playlist-item.active");e&&e.scrollIntoView({behavior:"smooth",block:"nearest"})}_attachPlayerListeners(){const e=this.shadowRoot.querySelector("#player");if(!e)return;const t=e.shadowRoot?e.shadowRoot.querySelector("a11y-media-player"):null;if(e.removeEventListener("mediastatechange",this._onMediaStateChange),e.addEventListener("mediastatechange",this._onMediaStateChange),!t)return;const tryAttach=()=>{const e=t.media;return!(!e||t.isYoutube)&&(e!==this._lastMediaElement&&(this._lastMediaElement&&this._lastMediaElement.removeEventListener("ended",this._onMediaEnded),e.addEventListener("ended",this._onMediaEnded),this._lastMediaElement=e),!0)};tryAttach()||(clearTimeout(this._attachEndedTimeout),this._attachEndedTimeout=setTimeout(()=>{tryAttach()},500))}_handleKeydown(e){const t=this.shadowRoot.querySelector(".playlist");if(t&&t.contains(e.target))if("ArrowDown"===e.key){e.preventDefault();const t=Math.min(this.activeIndex+1,this.mediaItems.length-1);this._setActiveIndex(t),this._focusPlaylistItem(t)}else if("ArrowUp"===e.key){e.preventDefault();const t=Math.max(this.activeIndex-1,0);this._setActiveIndex(t),this._focusPlaylistItem(t)}else if("Enter"===e.key||" "===e.key){e.preventDefault();const t=e.composedPath().find(e=>e.classList&&e.classList.contains("playlist-item"));if(t){const e=parseInt(t.getAttribute("data-index"));this._setActiveIndex(e)}}}_focusPlaylistItem(e){const t=this.shadowRoot.querySelector(`.playlist-item[data-index="${e}"]`);t&&t.focus()}updated(e){super.updated&&super.updated(e),e.forEach((e,t)=>{"edit"===t&&(this.edit||!0===e&&(this._updateMediaItems(),this.mediaItems.forEach(e=>{try{e.element.pause&&"function"==typeof e.element.pause&&e.element.pause()}catch(e){}}))),"activeIndex"===t&&void 0!==e&&this.mediaItems[this.activeIndex]&&(clearTimeout(this._playTimeout),this._playTimeout=setTimeout(()=>{const e=this.shadowRoot?this.shadowRoot.querySelector("#player"):null;if(e&&e.restart&&"function"==typeof e.restart)try{e.restart()}catch(e){}this._attachPlayerListeners(),this._scrollToActiveItem()},100))})}firstUpdated(e){super.firstUpdated&&super.firstUpdated(e),this._updateMediaItems();const t=this.shadowRoot.querySelector(".player-area"),i=this.shadowRoot.querySelector(".playlist");t&&i&&"ResizeObserver"in globalThis&&(this._resizeObserver=new ResizeObserver(e=>{for(const a of e)a.target===t&&a.target.offsetHeight>0&&i.style.setProperty("--playlist-max-height",`${a.target.offsetHeight}px`)}),this._resizeObserver.observe(t))}static get haxProperties(){return new URL(`./lib/${this.tag}.haxProperties.json`,import.meta.url).href}haxHooks(){return{inlineContextMenu:"haxinlineContextMenu"}}haxinlineContextMenu(e){e.ceButtons=[{icon:"icons:open-in-new",callback:"haxBreakOutPlaylist",label:"Break items out of playlist"},{icon:"lrn:edit",callback:"haxToggleEdit",label:"Toggle edit mode"},{icon:"av:video-call",callback:"haxAddVideo",label:"Add video"},{icon:"image:audiotrack",callback:"haxAddAudio",label:"Add audio"},{icon:"hax:anchor",callback:"haxCopyActiveIndex",label:"Copy active index"}]}haxToggleEdit(e){return this.edit=!this.edit,!0}haxAddVideo(e){const t=globalThis.document.createElement("video-player");return t.setAttribute("source",""),t.setAttribute("media-title","New video"),this.appendChild(t),!0}haxAddAudio(e){const t=globalThis.document.createElement("audio-player");return t.setAttribute("source",""),t.setAttribute("media-title","New audio"),this.appendChild(t),!0}haxBreakOutPlaylist(e){if(!this.parentNode)return!1;let t;return Array.from(this.children).filter(e=>"VIDEO-PLAYER"===e.tagName||"AUDIO-PLAYER"===e.tagName).forEach(e=>{t=e.cloneNode(!0),this.getAttribute("slot")?t.setAttribute("slot",this.getAttribute("slot")):t.removeAttribute("slot"),this.parentNode.insertBefore(t,this)}),!!t&&(this.remove(),!0)}haxCopyActiveIndex(e){return d(this.activeIndex),!0}}globalThis.customElements.define(MediaPlaylist.tag,MediaPlaylist);
