// @haxtheweb/haxcms-elements/lib/ui-components/magic/site-ai-chat.js
import{DDDSuper as t,DDDPulseEffectSuper as e}from"../../../../d-d-d/d-d-d.js";import{html as i,css as o,LitElement as s}from"../../../../../lit/index.js";import{store as a}from"../../core/haxcms-site-store.js";import{toJS as n}from"../../../../../mobx/dist/mobx.esm.js";import{MicroFrontendRegistry as r}from"../../../../micro-frontend-registry/micro-frontend-registry.js";import{enableServices as l}from"../../../../micro-frontend-registry/lib/microServices.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";l(["haxcms"]),r.add({endpoint:"/api/apps/haxcms/aiChat",name:"@haxcms/aiChat",title:"AI Chat",description:"AI based chat agent that can answer questions about a site",params:{site:"location of the HAXcms site OR site.json data",type:"site for site.json or link for remote loading",question:"Question to ask of the AI",engine:"which engine to use as we test multiple",context:"context to query based on. Course typical"}});export class SiteAiChat extends(e(t(s))){static get tag(){return"site-ai-chat"}constructor(){super(),this.opened=!1,this.question=null,this.answers=[],this.loading=!1,this.context=n(a.manifest.metadata.site.name),this.engine="alfred",this.dataPulse="1"}askQuestion(t){t.preventDefault(),this.engine=t.target.getAttribute("name"),this.context=this.shadowRoot.querySelector("#context").value,this.question=this.shadowRoot.querySelector("#question").value,this.requestAIFeedback()}updated(t){super.updated&&super.updated(t),t.has("opened")&&(this.opened?this.shadowRoot.querySelector("dialog").showModal():this.shadowRoot.querySelector("dialog").close())}requestAIFeedback(){const t=n(a.manifest);var e="";globalThis.document.querySelector("base")&&(e=globalThis.document.querySelector("base").href);const i={site:{file:e+"site.json",metadata:t.metadata},type:"site",question:this.question,engine:this.engine,context:this.context};this.loading=!0,r.call("@haxcms/aiChat",i).then((t=>{200==t.status&&(this.answers=[...t.data.answers],this.question=t.data.question),this.loading=!1})).catch((t=>{this.loading=!1,console.error(t)}))}render(){return i`
      <simple-icon-button-lite icon="hax:wizard-hat" @click="${this.openChat}"
        >Ask a question</simple-icon-button-lite
      >
      <dialog class="chat">
        <simple-icon-button-lite
          class="close"
          icon="close"
          @click="${this.closeChat}"
          >Close</simple-icon-button-lite
        >
        <form action="#">
          <simple-icon-lite
            class="hat"
            icon="${this.loading?"hax:loading":"hax:wizard-hat"}"
          ></simple-icon-lite>
          <input id="context" value="${this.context}" type="text" />
          <input id="question" type="text" placeholder="Ask your question.." />

          <button
            id="submit"
            type="submit"
            name="alfred"
            @click="${this.askQuestion}"
          >
            Ask Alfred
          </button>
          <button
            id="submit2"
            type="submit"
            name="robin"
            @click="${this.askQuestion}"
          >
            Ask Robin
          </button>
        </form>
        ${this.question?i` ${this.loading?"":i`
                  <div>
                    ${this.answers.map(((t,e)=>i`<h3 data-primary="13" data-design-treatment="vert">
                            Answer ${e+1}
                          </h3>
                          <p>${t}</p>`))}
                  </div>
                `}`:""}
      </dialog>
    `}closeChat(){this.opened=!1}openChat(){this.opened=!0,setTimeout((()=>{this.shadowRoot.querySelector("#question").focus(),this.shadowRoot.querySelector("#question").select()}),300)}static get styles(){return[super.styles,o`
        :host {
          display: block;
          position: absolute;
          bottom: 10px;
          right: 10px;
        }
        :host([opened]) {
          z-index: 100000;
        }
        :host([loading]) .loading {
          display: block;
        }
        simple-icon-lite,
        simple-icon-button-lite {
          --simple-icon-height: 48px;
          --simple-icon-width: 48px;
          color: var(--ddd-primary-13);
        }
        .hat {
          margin: 0 var(--ddd-spacing-4);
        }
        .close {
          position: absolute;
          top: 0;
          right: 0;
          --simple-icon-height: 24px;
          --simple-icon-width: 24px;
        }
        dialog[open] {
          opacity: 1;
          position: fixed;
          left: 25%;
          right: 25%;
          top: 25%;
          bottom: 25%;
          transform: scaleY(1);
        }
        dialog {
          opacity: 0;
          padding: var(--ddd-spacing-10);
          transform: scaleY(0);
          transition:
            opacity 0.7s ease-out,
            transform 0.7s ease-out,
            overlay 0.7s ease-out allow-discrete,
            display 0.7s ease-out allow-discrete;
        }
        input,
        button {
          font-size: var(--ddd-font-size-ms);
        }
        @starting-style {
          dialog[open] {
            opacity: 0;
            transform: scaleY(0);
          }
        }

        dialog::backdrop {
          background-color: rgb(0 0 0 / 0%);
          transition:
            display 0.7s allow-discrete,
            overlay 0.7s allow-discrete,
            background-color 0.7s;
        }

        dialog[open]::backdrop {
          background-color: rgb(0 0 0 / 25%);
        }
        @starting-style {
          dialog[open]::backdrop {
            background-color: rgb(0 0 0 / 0%);
          }
        }
      `]}static get properties(){return{...super.properties,engine:{type:String},question:{type:String},context:{type:String},answers:{type:Array},opened:{type:Boolean,reflect:!0},loading:{type:Boolean,reflect:!0}}}}globalThis.customElements.define(SiteAiChat.tag,SiteAiChat);
