// lit-html/node/experimental-hydrate.js
define(["exports", "../../../buffer", "./lit-html.js", "./directive.js", "./directive-helpers.js"], function (_exports, _buffer, _litHtml, _directive, _directiveHelpers) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.hydrate = _exports.digestForTemplateResult = void 0;
  function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e6) { throw _e6; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e7) { didErr = true; err = _e7; }, f: function f() { try { if (!normalCompletion && it.return != null) it.return(); } finally { if (didErr) throw err; } } }; }
  function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
  function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
  var a = _litHtml._$LH.R,
    s = _litHtml._$LH.D,
    c = _litHtml._$LH.V,
    d = _litHtml._$LH.I,
    f = _litHtml._$LH.F,
    p = function p(e, r) {
      var t = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
      if (void 0 !== r._$litPart$) throw Error("container already contains a live render");
      var n, o;
      var l = [],
        i = document.createTreeWalker(r, NodeFilter.SHOW_COMMENT, null, !1);
      var a;
      for (; null !== (a = i.nextNode());) {
        var _r = a.data;
        if (_r.startsWith("lit-part")) {
          if (0 === l.length && void 0 !== n) throw Error("there must be only one root part per container");
          o = u(e, a, l, t), null != n || (n = o);
        } else if (_r.startsWith("lit-node")) h(a, l, t);else if (_r.startsWith("/lit-part")) {
          if (1 === l.length && o !== n) throw Error("internal error");
          o = m(a, o, l);
        }
      }
      console.assert(void 0 !== n, "there should be exactly one root part in a render container"), r._$litPart$ = n;
    },
    u = function u(e, t, n, i) {
      var f, p;
      if (0 === n.length) p = new d(t, null, void 0, i), f = e;else {
        var _e = n[n.length - 1];
        if ("template-instance" === _e.type) p = new d(t, null, _e.instance, i), _e.instance.u.push(p), f = _e.result.values[_e.instancePartIndex++], _e.templatePartIndex++;else if ("iterable" === _e.type) {
          p = new d(t, null, _e.part, i);
          var _r2 = _e.iterator.next();
          if (_r2.done) throw f = void 0, _e.done = !0, Error("Unhandled shorter than expected iterable");
          f = _r2.value, _e.part._$AH.push(p);
        } else p = new d(t, null, _e.part, i);
      }
      if (f = c(p, f), f === _litHtml.noChange) n.push({
        part: p,
        type: "leaf"
      });else if ((0, _directiveHelpers.isPrimitive)(f)) n.push({
        part: p,
        type: "leaf"
      }), p._$AH = f;else if ((0, _directiveHelpers.isTemplateResult)(f)) {
        var _e2 = "lit-part " + w(f);
        if (t.data !== _e2) throw Error("Hydration value mismatch: Unexpected TemplateResult rendered to part");
        {
          var _e3 = d.prototype._$AC(f),
            _r3 = new a(_e3, p);
          n.push({
            type: "template-instance",
            instance: _r3,
            part: p,
            templatePartIndex: 0,
            instancePartIndex: 0,
            result: f
          }), p._$AH = _r3;
        }
      } else s(f) ? (n.push({
        part: p,
        type: "iterable",
        value: f,
        iterator: f[Symbol.iterator](),
        done: !1
      }), p._$AH = []) : (n.push({
        part: p,
        type: "leaf"
      }), p._$AH = null == f ? "" : f);
      return p;
    },
    m = function m(e, r, t) {
      if (void 0 === r) throw Error("unbalanced part marker");
      r._$AB = e;
      var n = t.pop();
      if ("iterable" === n.type && !n.iterator.next().done) throw Error("unexpected longer than expected iterable");
      if (t.length > 0) return t[t.length - 1].part;
    },
    h = function h(e, r, t) {
      var o;
      var l = /lit-node (\d+)/.exec(e.data),
        a = parseInt(l[1]),
        s = null !== (o = e.previousElementSibling) && void 0 !== o ? o : e.parentElement;
      if (null === s) throw Error("could not find node for attribute parts");
      s.removeAttribute("defer-hydration");
      var d = r[r.length - 1];
      if ("template-instance" !== d.type) throw Error("internal error");
      {
        var _e4 = d.instance;
        for (;;) {
          var _r4 = _e4._$AD.parts[d.templatePartIndex];
          if (void 0 === _r4 || _r4.type !== _directive.PartType.ATTRIBUTE && _r4.type !== _directive.PartType.ELEMENT || _r4.index !== a) break;
          if (_r4.type === _directive.PartType.ATTRIBUTE) {
            var _o = new _r4.ctor(s, _r4.name, _r4.strings, d.instance, t),
              _l = (0, _directiveHelpers.isSingleExpression)(_o) ? d.result.values[d.instancePartIndex] : d.result.values,
              _a = !(_o.type === _directive.PartType.EVENT || _o.type === _directive.PartType.PROPERTY);
            _o._$AI(_l, _o, d.instancePartIndex, _a), d.instancePartIndex += _r4.strings.length - 1, _e4.u.push(_o);
          } else {
            var _r5 = new f(s, d.instance, t);
            c(_r5, d.result.values[d.instancePartIndex++]), _e4.u.push(_r5);
          }
          d.templatePartIndex++;
        }
      }
    },
    w = function w(r) {
      var t = new Uint32Array(2).fill(5381);
      var _iterator = _createForOfIteratorHelper(r.strings),
        _step;
      try {
        for (_iterator.s(); !(_step = _iterator.n()).done;) {
          var _e5 = _step.value;
          for (var _r6 = 0; _r6 < _e5.length; _r6++) t[_r6 % 2] = 33 * t[_r6 % 2] ^ _e5.charCodeAt(_r6);
        }
      } catch (err) {
        _iterator.e(err);
      } finally {
        _iterator.f();
      }
      var n = String.fromCharCode.apply(String, babelHelpers.toConsumableArray(new Uint8Array(t.buffer)));
      return _buffer.Buffer.from(n, "binary").toString("base64");
    };
  _exports.digestForTemplateResult = w;
  _exports.hydrate = p;
});
