// @haxtheweb/career-timeline/lib/career-org-item.js
/**
 * Copyright 2026 winstonwumbo
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as e,html as t,css as r}from"../../../lit/index.js";import{DDDSuper as a}from"../../d-d-d/d-d-d.js";import"./career-role-item.js";export class CareerOrgItem extends(a(e)){static get tag(){return"career-org-item"}constructor(){super(),this.organization="",this.location="",this.source="https://avatars.githubusercontent.com/u/170651362"}static get properties(){return{...super.properties,organization:{type:String},location:{type:String},source:{type:String}}}static get styles(){return[super.styles,r`
        :host {
          display: block;
          position: relative;
          padding: var(--ddd-spacing-4);
          font-family: var(--ddd-font-primary, sans-serif);
        }

        :host([data-hax-active]){
          outline: var(
            --hax-body-active-outline,
            1px solid var(--hax-ui-color-focus, #000)
          );
        }

        .item-container {
          position: relative;
          padding-left: calc(
            var(--ddd-icon-md) + var(--ddd-spacing-4)
          );
        }

        .org-info {
          display: flex;
        }

        #logo {
          width: 48px;
          height: 48px;
        }

        .org {
          padding-left: 20px;
          padding-bottom: 20px;
        }

        h3{
          margin-top: 0px;
        }

        .border-line {
          position: relative;
          border-bottom: var(--ddd-border-sm);
          width: 80%;
          margin-left: auto;
          margin-right: auto;
        }

        @media (max-width: 768px) {
          :host {
            padding: 0px;
          }
          .item-container {
            padding-left: 0px;
          }
        }
      `]}render(){return t`
      <div class="org-info">
        <img id="logo" src="${this.source}" alt=""/>
        <div class="org">
          <h3>${this.organization}</h3>
          ${this.location} · ${this._formatDate()}
        </div>
      </div>
      <div class="steps-container">
        <slot><career-role-item title="Add a Role"></career-role-item></slot>
        <div class="border-line"></div>
      </div>
    `}connectedCallback(){super.connectedCallback(),this.addEventListener("start-date-changed",this._sortByDate)}disconnectedCallback(){super.disconnectedCallback(),this.removeEventListener("start-date-changed",this._sortByDate)}get earliestDate(){const e=Array.from(this.querySelectorAll("career-role-item")).map(e=>new Date(e.startDate));return Math.min(...e)}get latestDate(){const e=Array.from(this.querySelectorAll("career-role-item")).map(e=>new Date(e.endDate));return Math.max(...e)}_formatDate(){if(!Number.isFinite(this.earliestDate)||!Number.isFinite(this.latestDate))return"0 yr 0 mo";const e=new Date(this.earliestDate),t=new Date(this.latestDate);let r=t.getFullYear()-e.getFullYear(),a=t.getMonth()-e.getMonth();return a<0&&(r--,a+=12),r=r>1?`${r} yrs`:`${r} yr`,a=a>1?`${a} mos`:`${a} mo`,`${r} ${a}`}_sortByDate(e){Array.from(this.querySelectorAll("career-role-item")).sort((e,t)=>new Date(t.startDate)-new Date(e.startDate)).forEach(e=>this.appendChild(e)),this.requestUpdate(),e.detail.node.scrollIntoView({behavior:"smooth",block:"start",inline:"nearest"})}_addItemHandler(e){const t=globalThis.document.createElement("career-role-item"),r={"data-hax-layout":!0,"data-hax-ray":"career-role-item"};for(const e in r)t.setAttribute(e,r[e]);this.prepend(t)}haxHooks(){return{inlineContextMenu:"haxinlineContextMenu"}}haxinlineContextMenu(e){e.ceButtons=[{icon:"icons:add",callback:"_addItemHandler",label:"Add role under organization"}]}static get haxProperties(){return{type:"grid",canScale:!1,designSystem:!1,canEditSource:!1,contentEditable:!1,hideDefaultSettings:!0,gizmo:{title:"Career Item",description:"An individual organization and its affiliated roles",icon:"chrome-reader-mode",color:"light-blue",tags:["Content","list","step"],handles:[],meta:{author:"HAXTheWeb core team"}},settings:{configure:[{property:"organization",title:"Organization Name",inputMethod:"textfield"},{property:"location",title:"Location",inputMethod:"textfield"},{property:"source",title:"Source",inputMethod:"haxupload"}],advanced:[],developer:[]}}}}globalThis.customElements.define(CareerOrgItem.tag,CareerOrgItem);
