// @haxtheweb/career-timeline/lib/career-role-item.js
/**
 * Copyright 2026 winstonwumbo
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as t,html as e,css as i}from"../../../lit/index.js";import{DDDSuper as r}from"../../d-d-d/d-d-d.js";export class CareerRoleItem extends(r(t)){static get tag(){return"career-role-item"}constructor(){super(),this.title="Role Title",this.startDate=new Date,this.endDate=new Date,this.skills=[]}static get properties(){return{...super.properties,title:{type:String},startDate:{type:String,attribute:"start-date"},endDate:{type:String,attribute:"end-date"},skills:{type:Array}}}static get styles(){return[super.styles,i`
        :host {
          display: flex;
          border-left: var(--ddd-border-sm);
          padding-left: 45px;
          margin-left: var(--ddd-spacing-6, 24px);
          position: relative;
          scroll-margin-top: var(--ddd-spacing-28, 112px);
        }

        :host([data-hax-active]) .role-content {
          outline: var(
            --hax-body-active-outline,
            1px solid var(--hax-ui-color-focus, #000)
          );
          outline-offset: var(--ddd-spacing-2, 8px);
        }

        .circle {
          width: var(--ddd-icon-size-xl, 8px);
          height: var(--ddd-icon-size-xl, 8px);
          border-radius: 50%;
          background-color: var(--ddd-theme-default-limestoneLight);
          margin-left: -50px;
          position: absolute;
        }

        .role-content div ::slotted(p) {
          padding: 0;
          margin: 0;
        }

        .role-content div {
          min-height: var(--ddd-spacing-9, 36px);
          padding: 0;
          margin: 0;
        }

        .skills {
          display: inline;
          font-weight: bold;
        }

        h3 {
          margin: 4px 0 0 0;
          color: var(--lowContrast-override, var(--ddd-theme-primary, #1e407c));
        }

        h5 {
          margin-top: 0;
        }

        :host(:last-of-type) {
          border-left: unset;
          margin-left: 26px;
        }

        @media (max-width: 768px) {
          :host {
            border-left: unset;
            padding-left: unset;
            display: block;
          }
          .circle {
            visibility: hidden;
          }
        }
      `]}render(){return e`
      <div class="wrapper">
        <div class="circle"></div>
        <div class="role-content">
          <h5>${this.title}</h5>
          ${this._formatDate()}
          <div><slot></slot></div>
          ${this._formatSkills()}
        </div>
      </div>
    `}updated(t){super.updated&&super.updated(t),t.forEach((t,e)=>{"startDate"===e&&void 0!==t&&(this.startDate||(this.startDate=new Date),this.dispatchEvent(new CustomEvent("start-date-changed",{bubbles:!0,composed:!0,cancelable:!1,detail:{node:this}}))),"endDate"===e&&void 0!==t&&(this.endDate||(this.endDate=new Date))})}connectedCallback(){if(super.connectedCallback(),0===this.children.length){const t=globalThis.document.createElement("p");t.textContent="Describe your role here...",this.appendChild(t)}}_formatDate(){const t=new Date(this.startDate),e=new Date(this.endDate);let i=e.getFullYear()-t.getFullYear(),r=e.getMonth()-t.getMonth();r<0&&(i--,r+=12),i=i>1?`${i} yrs`:`${i} yr`,r=r>1?`${r} mos`:`${r} mo`;const a=new Intl.DateTimeFormat("en-US",{year:"numeric",month:"short"});return`${a.format(t)} - ${this.hasAttribute("end-date")?a.format(e):"Present"} · ${i} ${r}`}_formatSkills(){if(0===this.skills.length)return;const t=this.skills.map(t=>t.name),i=t.length>5?`${t.slice(0,5).join(" · ")} and +${t.length-5} more`:t.join(" · ");return e`<div class="skills">Skills:</div> ${i}`}static get haxProperties(){return{type:"grid",canScale:!1,designSystem:!1,canEditSource:!0,contentEditable:!1,hideDefaultSettings:!0,gizmo:{title:"Career Role Item",description:"An individual role within an organization",icon:"chrome-reader-mode",color:"light-blue",tags:["Content","list","step"],handles:[],meta:{author:"winstonwumbo",hidden:!0}},settings:{configure:[{property:"title",title:"Title",inputMethod:"textfield"},{property:"startDate",title:"Start Date",inputMethod:"datepicker"},{property:"endDate",title:"End Date (Optional)",inputMethod:"datepicker"},{property:"skills",title:"Associated Skills",description:"Select the + button to start adding skills",inputMethod:"array",itemLabel:"name",hideReorder:!0,expanded:!0,hideDuplicate:!0,properties:[{property:"name",title:"Skill",inputMethod:"textfield"}]}],advanced:[],developer:[]}}}}globalThis.customElements.define(CareerRoleItem.tag,CareerRoleItem);
