// @haxtheweb/career-timeline/lib/career-role-item.js
/**
 * Copyright 2026 winstonwumbo
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as t,html as e,css as i}from"../../../lit/index.js";import{DDDSuper as r}from"../../d-d-d/d-d-d.js";export class CareerRoleItem extends(r(t)){static get tag(){return"career-role-item"}constructor(){super(),this.title="Role Title",this.startDate=(new Date).toISOString(),this.endDate=(new Date).toISOString(),this.skills=[]}static get properties(){return{...super.properties,title:{type:String},startDate:{type:String,attribute:"start-date"},endDate:{type:String,attribute:"end-date"},skills:{type:Array}}}static get styles(){return[super.styles,i`
        :host {
          display: flex;
          border-left: var(--ddd-border-sm);
          padding-left: 45px;
          margin-left: var(--ddd-spacing-6, 24px);
          position: relative;
          scroll-margin-top: var(--ddd-spacing-28, 112px);
        }

        :host([data-hax-active]) .role-content {
          outline: var(
            --hax-body-active-outline,
            1px solid var(--hax-ui-color-focus, #000)
          );
          outline-offset: var(--ddd-spacing-2, 8px);
        }

        .circle {
          width: var(--ddd-icon-size-xl, 8px);
          height: var(--ddd-icon-size-xl, 8px);
          border-radius: 50%;
          background-color: var(--ddd-theme-default-limestoneLight);
          margin-left: -50px;
          position: absolute;
        }

        .role-content div ::slotted(p) {
          padding: 0;
          margin: 0;
        }

        .role-content div {
          min-height: var(--ddd-spacing-9, 36px);
          padding: 0;
          margin: 0;
        }

        .skills {
          display: inline;
          font-weight: bold;
        }

        h4 {
          margin-top: 0;
        }

        :host(:last-of-type) {
          border-left: unset;
          margin-left: 26px;
        }

        @media (max-width: 768px) {
          :host {
            border-left: unset;
            padding-left: unset;
            display: block;
          }
          .circle {
            visibility: hidden;
          }
        }
      `]}render(){return e`
      <div class="wrapper">
        <div class="circle"></div>
        <div class="role-content">
          <h4>${this.title}</h4>
          ${this._formatDate()}
          <div><slot></slot></div>
          ${this._formatSkills()}
        </div>
      </div>
    `}updated(t){super.updated&&super.updated(t),t.forEach((t,e)=>{if("startDate"===e&&void 0!==t){const t=this._normalizeDate(this.startDate);if(!t)return void(this.startDate=(new Date).toISOString());if(this.startDate!==t)return void(this.startDate=t);this.startDate||(this.startDate=(new Date).toISOString()),this.dispatchEvent(new CustomEvent("start-date-changed",{bubbles:!0,composed:!0,cancelable:!1,detail:{node:this}}))}if("endDate"===e&&void 0!==t){const t=this._normalizeDate(this.endDate);if(!t)return void(this.endDate=(new Date).toISOString());if(this.endDate!==t)return void(this.endDate=t);this.endDate||(this.endDate=(new Date).toISOString())}})}connectedCallback(){if(super.connectedCallback(),0===this.children.length){const t=globalThis.document.createElement("p");t.textContent="Describe your role here...",this.appendChild(t)}}_stripWrappedQuotes(t){return t.replace(/(^&quot;|&quot;$)/g,"").replace(/(^"+|"+$)/g,"").replace(/(^'+|'+$)/g,"")}_normalizeDate(t){if(null==t)return"";if(t instanceof Date)return Number.isNaN(t.getTime())?"":t.toISOString();let e=String(t).trim();if(!e)return"";e=this._stripWrappedQuotes(e);const i=new Date(e);return Number.isNaN(i.getTime())?"":i.toISOString()}_formatDate(){const t=this._normalizeDate(this.startDate),e=this._normalizeDate(this.endDate);if(!t)return"";const i=this.hasAttribute("end-date")&&!!e,r=new Date(t),a=i?new Date(e):new Date;let n=a.getFullYear()-r.getFullYear(),s=a.getMonth()-r.getMonth();s<0&&(n--,s+=12),n=n>1?`${n} yrs`:`${n} yr`,s=s>1?`${s} mos`:`${s} mo`;const o=new Intl.DateTimeFormat("en-US",{year:"numeric",month:"short"});return`${o.format(r)} - ${i?o.format(a):"Present"} · ${n} ${s}`}_formatSkills(){if(0===this.skills.length)return;const t=this.skills.map(t=>t.name),i=t.length>5?`${t.slice(0,5).join(" · ")} and +${t.length-5} more`:t.join(" · ");return e`<div class="skills">Skills:</div> ${i}`}static get haxProperties(){return{type:"grid",canScale:!1,designSystem:!1,canEditSource:!0,contentEditable:!1,hideDefaultSettings:!0,gizmo:{title:"Career Role",description:"An individual role within an organization",icon:"chrome-reader-mode",color:"light-blue",tags:["Content","list","step"],handles:[],meta:{author:"HAXTheWeb core team"}},settings:{configure:[{property:"title",title:"Title",inputMethod:"textfield"},{property:"startDate",title:"Start Date",inputMethod:"datepicker"},{property:"endDate",title:"End Date (Optional)",inputMethod:"datepicker"},{property:"skills",title:"Associated Skills",description:"Select the + button to start adding skills",inputMethod:"array",itemLabel:"name",hideReorder:!0,expanded:!0,hideDuplicate:!0,properties:[{property:"name",title:"Skill",inputMethod:"textfield"}]}],advanced:[],developer:[]}}}}globalThis.customElements.define(CareerRoleItem.tag,CareerRoleItem);
