// @lrnwebcomponents/tab-list/tab-list.js
define(["exports","../../@polymer/polymer/polymer-legacy.js","../hax-body-behaviors/lib/HAXWiring.js","../../@polymer/paper-tabs/paper-tabs.js","../../@polymer/paper-tabs/paper-tab.js","../../@polymer/paper-button/paper-button.js"],function(_exports,_polymerLegacy,_HAXWiring,_paperTabs,_paperTab,_paperButton){"use strict";Object.defineProperty(_exports,"__esModule",{value:!0});_exports.TabList=void 0;/**
 * `tab-list`
 * `A simple listing of tabed links / items`
 *
 * @demo demo/index.html
 *
 * @microcopy - the mental model for this element
 * -
 */let TabList=(0,_polymerLegacy.Polymer)({_template:_polymerLegacy.html`
    <style>
      :host {
        display: block;
        margin: 0 auto;
        list-style: none;
        display: block;
        padding: 16px;
        border-bottom: 1px solid black;
      }
      paper-tabs {
        align-items: center;
        justify-items: center;
      }
      paper-tab a {
        text-decoration: none;
        flex: unset;
        height: unset;
        width: 100%;
        text-align: center;
      }
      paper-button {
        text-transform: unset;
        width: 100%;
        display: block;
        min-width: unset;
        margin: 0;
        text-overflow: ellipsis;
        overflow: hidden;
      }
      @media screen and (max-width: 600px) {
        paper-tab {
          display: block;
        }
      }
    </style>
    <paper-tabs>
      <template is="dom-repeat" items="[[tabs]]" as="tab">
        <paper-tab>
          <a target="_blank" href="[[tab.link]]" tabindex="-1">
            <paper-button raised>[[tab.label]]</paper-button>
          </a>
        </paper-tab>
      </template>
    </paper-tabs>
  `,is:"tab-list",behaviors:[HAXBehaviors.PropertiesBehaviors],observers:["_valueChanged(tabs.*)"],properties:{/**
     * List of tabs
     */tabs:{type:Array,value:[],notify:!0}},/**
   * Notice an array has changed and update the DOM.
   */_valueChanged:function(e){for(var i in e.base){for(var j in e.base[i]){this.notifyPath("tabs."+i+"."+j)}}},/**
   * Attached to the DOM, now fire.
   */attached:function(){// Establish hax property binding
let props={canScale:!1,canPosition:!1,canEditSource:!1,gizmo:{title:"Tabs",description:"A list of links as tabs.",icon:"icons:tab",color:"grey",groups:["Presentation","Links"],handles:[],meta:{author:"LRNWebComponents"}},settings:{quick:[],configure:[{property:"tabs",title:"Tabs",description:"Listing of tabs",inputMethod:"array",properties:[{property:"link",title:"Link",description:"link to go to",inputMethod:"textfield",required:!0},{property:"label",title:"Label",description:"text to place on the tab",inputMethod:"textfield",required:!0}]}],advanced:[]}};this.setHaxProperties(props)}});_exports.TabList=TabList});
