// @lrnwebcomponents/flash-card/lib/flash-card-prompt-img.js
define(["exports", "../../../lit/index.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-colors/simple-colors.js"], function (_exports, _index, _simpleIcons, _simpleIconLite, _simpleColors) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.FlashCardPromptImg = void 0;
  var _templateObject_0cef04f0e8e111ed8789699b2ad5335a, _templateObject2_0cef04f0e8e111ed8789699b2ad5335a, _templateObject3_0cef04f0e8e111ed8789699b2ad5335a, _templateObject4_0cef04f0e8e111ed8789699b2ad5335a, _templateObject5_0cef04f0e8e111ed8789699b2ad5335a;
  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  // EXPORT (so make available to other documents that reference this file) a class, that extends LitElement
  // which has the magic life-cycles and developer experience below added
  var FlashCardPromptImg = /*#__PURE__*/function (_SimpleColors) {
    babelHelpers.inherits(FlashCardPromptImg, _SimpleColors);
    var _super = _createSuper(FlashCardPromptImg);
    // overlay on div tag - wrap image in div & style div
    // HTMLElement life-cycle, built in; use this for setting defaults
    function FlashCardPromptImg() {
      var _this;
      babelHelpers.classCallCheck(this, FlashCardPromptImg);
      _this = _super.call(this);
      _this.imgSrc = "";
      _this.imgKeyword = "grey box";
      _this.status = "pending";
      _this.answerIcon = false;
      _this.icon = "";
      return _this;
    }

    // properties that you wish to use as data in HTML, CSS, and the updated life-cycle
    babelHelpers.createClass(FlashCardPromptImg, [{
      key: "updated",
      value:
      // updated fires every time a property defined above changes
      // this allows you to react to variables changing and use javascript to perform logic
      function updated(changedProperties) {
        var _this2 = this;
        changedProperties.forEach(function (oldValue, propName) {
          if (propName === "status" && _this2[propName] === "correct") {
            _this2.answerIcon = true;
            _this2.icon = "check";
          }
          if (propName === "status" && _this2[propName] === "incorrect") {
            _this2.answerIcon = true;
            _this2.icon = "cancel";
          }
          if (propName === "status" && _this2[propName] === "pending") {
            _this2.answerIcon = false;
          }
        });
      }

      // HTML - specific to Lit
    }, {
      key: "render",
      value: function render() {
        return (0, _index.html)(_templateObject_0cef04f0e8e111ed8789699b2ad5335a || (_templateObject_0cef04f0e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n      <div>\n        <div class=\"overlay\">\n          <div class=\"backgroundbox\">\n            ", "\n          </div>\n        </div>\n        ", "\n      </div>\n    "])), this.imgSrc !== "" ? (0, _index.html)(_templateObject2_0cef04f0e8e111ed8789699b2ad5335a || (_templateObject2_0cef04f0e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["<img src=\"", "\" alt=\"\" />"])), this.imgSrc) : (0, _index.html)(_templateObject3_0cef04f0e8e111ed8789699b2ad5335a || (_templateObject3_0cef04f0e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["<img\n                  src=\"https://loremflickr.com/320/240/", "?lock=1\"\n                  alt=\"\"\n                />"])), this.imgKeyword), this.answerIcon ? (0, _index.html)(_templateObject4_0cef04f0e8e111ed8789699b2ad5335a || (_templateObject4_0cef04f0e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["<simple-icon-lite icon=\"", "\"></simple-icon-lite>"])), this.icon) : "");
      }
    }], [{
      key: "tag",
      get:
      // a convention I enjoy so you can change the tag name in 1 place
      function get() {
        return "flash-card-image-prompt";
      }

      // CSS - specific to Lit
    }, {
      key: "styles",
      get: function get() {
        return (0, _index.css)(_templateObject5_0cef04f0e8e111ed8789699b2ad5335a || (_templateObject5_0cef04f0e8e111ed8789699b2ad5335a = babelHelpers.taggedTemplateLiteral(["\n      :host {\n        display: block;\n        width: 320px;\n        height: 265px;\n      }\n\n      img {\n        display: flex;\n        margin: 25px auto auto;\n        height: 200px;\n        width: 275px;\n        border: 5px solid white;\n        border-radius: 19px;\n        box-shadow: 0 0 10px black;\n      }\n\n      .backgroundbox {\n        display: flex;\n        background-color: var(--simple-colors-default-theme-accent-4);\n        border-radius: 19px 19px 0 0;\n        height: 265px;\n        width: 320px;\n      }\n\n      .overlay {\n        position: relative;\n      }\n\n      .overlay::before {\n        content: \"\";\n        width: 100%;\n        height: 100%;\n        position: absolute;\n        border: 1px;\n        border-radius: 19px 19px 0px 0px;\n      }\n\n      simple-icon-lite {\n        --simple-icon-height: 100px;\n        --simple-icon-width: 100px;\n        color: white;\n        transform: translate(-50%, -190%);\n        top: 50%;\n        left: 50%;\n        z-index: 100;\n      }\n\n      :host([status=\"pending\"]) .overlay::before {\n        display: flex;\n        background: transparent;\n      }\n\n      :host([status=\"correct\"]) .overlay::before {\n        display: flex;\n        background: green;\n        filter: opacity(0.65);\n      }\n\n      :host([status=\"incorrect\"]) .overlay::before {\n        display: flex;\n        background: red;\n        filter: opacity(0.65);\n      }\n    "])));
      }
    }, {
      key: "properties",
      get: function get() {
        return _objectSpread(_objectSpread({}, babelHelpers.get(babelHelpers.getPrototypeOf(FlashCardPromptImg), "properties", this)), {}, {
          imgSrc: {
            type: String,
            reflect: true,
            attribute: "img-src"
          },
          imgKeyword: {
            type: String,
            attribute: "img-keyword"
          },
          status: {
            type: String,
            reflect: true
          },
          // Correct, incorrect, pending
          answerIcon: {
            type: Boolean,
            reflect: true
          },
          icon: {
            type: String
          }
        });
      }
    }]);
    return FlashCardPromptImg;
  }(_simpleColors.SimpleColors);
  _exports.FlashCardPromptImg = FlashCardPromptImg;
  customElements.define(FlashCardPromptImg.tag, FlashCardPromptImg);
});
