// @haxtheweb/haxcms-elements/lib/core/micros/haxcms-page-operations.js
import{html as t,css as o}from"../../../../../lit/index.js";import{DDD as e}from"../../../../d-d-d/d-d-d.js";import{store as i}from"../haxcms-site-store.js";import{toJS as l,autorun as a}from"../../../../../mobx/dist/mobx.esm.js";import"../../../../simple-icon/lib/simple-icon-button-lite.js";import"../../../../simple-toolbar/lib/simple-toolbar-menu-item.js";import"../../../../simple-toolbar/lib/simple-toolbar-button.js";export class HAXCMSPageOperations extends e{static get tag(){return"haxcms-page-operations"}static get styles(){return[super.styles,o`
        :host {
          display: inline-block;
          position: relative;
        }
        simple-icon-button-lite.ops {
          --simple-icon-button-border-width: 1px;
          --simple-icon-button-border-color: var(--ddd-border-1, #e0e0e0);
          --simple-icon-height: var(--ddd-icon-xxs, 14px);
          --simple-icon-width: var(--ddd-icon-xxs, 14px);
          padding: var(--ddd-spacing-1) var(--ddd-spacing-1);
          background-color: var(--ddd-theme-background, #ffffff);
          color: var(--ddd-theme-text, #222);
          position: relative;
          top: -2px;
        }
        dialog {
          position: fixed;
          background: light-dark(
            var(--simple-toolbar-button-bg, white),
            var(--simple-toolbar-button-bg, #222)
          );
          border: var(--ddd-border-sm);
          border-color: light-dark(
            var(--simple-toolbar-button-border-color, var(--simple-toolbar-border-color, #e0e0e0)),
            var(--simple-toolbar-button-border-color, var(--simple-toolbar-border-color, #444))
          );
          color: light-dark(
            var(--simple-toolbar-button-color, #222),
            var(--simple-toolbar-button-color, #e0e0e0)
          );
          min-width: 180px;
          padding: 0;
          margin: 0;
          max-height: 80vh;
          overflow: auto;
          box-shadow: var(--simple-toolbar-menu-list-box-shadow, light-dark(
            0 2px 8px rgba(0,0,0,0.15),
            0 2px 8px rgba(0,0,0,0.5)
          ));
        }
        dialog::backdrop {
          background: transparent;
        }
        simple-toolbar-menu-item {
          display: block;
        }
        .menu-header {
          padding: var(--ddd-spacing-3) var(--ddd-spacing-3) var(--ddd-spacing-2);
          font-weight: var(--ddd-font-weight-bold);
          font-size: var(--ddd-font-size-xs);
          text-transform: uppercase;
          letter-spacing: 0.5px;
          border-bottom: var(--ddd-border-sm);
          margin-bottom: var(--ddd-spacing-1);
        }
      `]}static get properties(){return{actionId:{type:String,attribute:"action-id"},platformAllowsOutline:{type:Boolean},platformAllowsDelete:{type:Boolean},platformAllowsAddPage:{type:Boolean}}}constructor(){super(),this.actionId=null,this.platformAllowsOutline=!0,this.platformAllowsDelete=!0,this.platformAllowsAddPage=!0,a((()=>{const t=l(i.manifest),o=t&&t.metadata&&t.metadata.platform;this.platformAllowsOutline=!o||!1!==o.outlineDesigner,this.platformAllowsDelete=!o||!1!==o.delete,this.platformAllowsAddPage=!o||!1!==o.addPage}))}_toggleDialog(){const t=this.shadowRoot.querySelector("#menu");if(t)if(t.open)t.close();else{const o=this.shadowRoot.querySelector(".ops");if(t.showModal(),o){const e=o.getBoundingClientRect();t.style.top=`${e.bottom+4}px`,t.style.left=`${e.left}px`}}}render(){return this.platformAllowsOutline?t`
      <simple-icon-button-lite
        class="ops"
        icon="icons:more-vert"
        label="Page actions"
        title="Page actions"
        @click="${this._toggleDialog}"
      ></simple-icon-button-lite>
      <dialog id="menu" role="menu" @click=${this._handleBackdropClick}>
        <div class="menu-header">Page Actions</div>
        <simple-toolbar-menu-item>
          <simple-toolbar-button
            icon="icons:create"
            icon-position="left"
            align-horizontal="left"
            show-text-label
            label="Edit"
            @click=${this._editPage}
            autofocus
          ></simple-toolbar-button>
        </simple-toolbar-menu-item>
        ${this.platformAllowsAddPage?t`
          <simple-toolbar-menu-item>
            <haxcms-button-add
              id="addpagebutton"
              class="top-bar-button"
              icon="hax:add-page"
              icon-position="left"
              label="Add page"
              merlin
              @click="${this._closeDialog}"
              show-text-label
            ></haxcms-button-add>
          </simple-toolbar-menu-item>
        `:""}
        <simple-toolbar-menu-item>
          <simple-toolbar-button
            icon="icons:arrow-upward"
            icon-position="left"
            align-horizontal="left"
            show-text-label
            label="Move up"
            @click=${()=>this._op("moveUp")}
          ></simple-toolbar-button>
        </simple-toolbar-menu-item>
        <simple-toolbar-menu-item>
          <simple-toolbar-button
            icon="icons:arrow-downward"
            icon-position="left"
            align-horizontal="left"
            show-text-label
            label="Move down"
            @click=${()=>this._op("moveDown")}
          ></simple-toolbar-button>
        </simple-toolbar-menu-item>
        <simple-toolbar-menu-item>
          <simple-toolbar-button
            icon="editor:format-indent-increase"
            icon-position="left"
            align-horizontal="left"
            show-text-label
            label="Indent"
            @click=${()=>this._op("indent")}
          ></simple-toolbar-button>
        </simple-toolbar-menu-item>
        <simple-toolbar-menu-item>
          <simple-toolbar-button
            icon="editor:format-indent-decrease"
            icon-position="left"
            align-horizontal="left"
            show-text-label
            label="Outdent"
            @click=${()=>this._op("outdent")}
          ></simple-toolbar-button>
        </simple-toolbar-menu-item>
        <simple-toolbar-menu-item>
          <simple-toolbar-button
            icon="icons:open-with"
            icon-position="left"
            align-horizontal="left"
            show-text-label
            label="Move to.."
            @click=${this._movePageProgram}
          ></simple-toolbar-button>
        </simple-toolbar-menu-item>
        ${this.platformAllowsDelete?t`
          <simple-toolbar-menu-item>
            <simple-toolbar-button
              icon="icons:delete"
              icon-position="left"
              align-horizontal="left"
              show-text-label
              label="Delete"
              @click=${this._deletePage}
            ></simple-toolbar-button>
          </simple-toolbar-menu-item>
        `:""}
      </dialog>
    `:t``}_contextItem(){return(async()=>{let t={};return t=this.actionId&&"null"!=this.actionId?await i.findItemAsObject(this.actionId):"null"==this.actionId?l(await i.getLastChildItem(null)):l(i.activeItem),t})()}async _op(t){const o=await this._contextItem();if(!o||!o.id)return void this._closeDialog();i.playSound("click");const e={id:o.id,operation:t};this.dispatchEvent(new CustomEvent("haxcms-save-node-details",{bubbles:!0,composed:!0,cancelable:!0,detail:e})),this._closeDialog()}_editPage(){this._closeDialog();globalThis.document.querySelector("haxcms-site-editor-ui")._editButtonTap()}_deletePage(){this._closeDialog();globalThis.document.querySelector("haxcms-site-editor-ui")._deleteButtonTap()}async _movePageProgram(){this._closeDialog();const t=await this._contextItem();if(!t||!t.id)return;const o=globalThis.SuperDaemonManager.requestAvailability();i.playSound("click"),o.waveWand(["","/",{pageId:t.id},"move-page","Move Page",""])}_closeDialog(){const t=this.shadowRoot&&this.shadowRoot.querySelector("#menu");t&&t.open&&t.close()}_handleBackdropClick(t){const o=t.target;if("DIALOG"===t.target.nodeName){const e=o.getBoundingClientRect();(t.clientX<e.left||t.clientX>e.right||t.clientY<e.top||t.clientY>e.bottom)&&o.close()}}}customElements.define(HAXCMSPageOperations.tag,HAXCMSPageOperations);
