// @lrnwebcomponents/super-daemon/lib/super-daemon-row.js
define(["exports", "../../../lit/index.js", "../../simple-icon/lib/simple-icons.js", "../../simple-icon/lib/simple-icon-lite.js", "../../simple-fields/lib/simple-tag.js"], function (_exports, _index, _simpleIcons, _simpleIconLite, _simpleTag) {
  "use strict";

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.SuperDaemonRow = void 0;
  var _templateObject_10b17230e51211edbf9f41986c80248c, _templateObject2_10b17230e51211edbf9f41986c80248c, _templateObject3_10b17230e51211edbf9f41986c80248c, _templateObject4_10b17230e51211edbf9f41986c80248c, _templateObject5_10b17230e51211edbf9f41986c80248c, _templateObject6_10b17230e51211edbf9f41986c80248c, _templateObject7_10b17230e51211edbf9f41986c80248c, _templateObject8_10b17230e51211edbf9f41986c80248c, _templateObject9_10b17230e51211edbf9f41986c80248c;
  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { babelHelpers.defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
  function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = babelHelpers.getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = babelHelpers.getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return babelHelpers.possibleConstructorReturn(this, result); }; }
  function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
  var SuperDaemonRow = /*#__PURE__*/function (_LitElement) {
    babelHelpers.inherits(SuperDaemonRow, _LitElement);
    var _super = _createSuper(SuperDaemonRow);
    function SuperDaemonRow() {
      var _this;
      babelHelpers.classCallCheck(this, SuperDaemonRow);
      _this = _super.call(this);
      _this.title = null;
      _this.path = null;
      _this.icon = null;
      _this.image = null;
      _this.key = null;
      _this.eventName = null;
      _this.more = false;
      _this.showDetails = false;
      _this.value = {};
      _this.tags = [];
      _this.shadowRootOptions = _objectSpread(_objectSpread({}, _index.LitElement.shadowRootOptions), {}, {
        delegatesFocus: true
      });
      _this.active = false;
      return _this;
    }
    babelHelpers.createClass(SuperDaemonRow, [{
      key: "keyEvent",
      value: function keyEvent(e) {
        // ensure that the daemon dialog does not accidentally duplicate or get this
        // when our focus was on a specific item while using the keyboard
        if (e.code === "Enter" || e.code === "Space") {
          this.selected();
          e.preventDefault();
          e.stopPropagation();
          e.stopImmediatePropagation();
        }
      }
    }, {
      key: "clickEvent",
      value: function clickEvent(e) {
        this.selected();
      }
    }, {
      key: "selected",
      value: function selected() {
        // execute the event
        this.dispatchEvent(new CustomEvent("super-daemon-row-selected", {
          composed: true,
          bubbles: true,
          cancelable: true,
          detail: this
        }));
        this.dispatchEvent(new CustomEvent(this.eventName, {
          composed: true,
          bubbles: true,
          cancelable: true,
          detail: this.value
        }));
        // programs will run in the same window so we don't want to close the dialog
        // as every non-program running a single command would have to declare
        // when it wanted the dialog closed so we just do it automatically
        if (this.eventName !== "super-daemon-run-program") {
          // close dialog bc we executed that command
          this.dispatchEvent(new CustomEvent("super-daemon-close", {
            composed: true,
            bubbles: true,
            cancelable: true,
            detail: this
          }));
        }
      }
    }, {
      key: "_focusIn",
      value: function _focusIn(e) {
        this.active = true;
      }
    }, {
      key: "_focusOut",
      value: function _focusOut(e) {
        this.active = false;
      }
    }, {
      key: "pickColor",
      value: function pickColor(val) {
        if (val === 0) {
          return "blue";
        }
        return "orange";
      }
    }, {
      key: "keyAry",
      value: function keyAry(key) {
        if (key) {
          return key.split("+");
        }
        return [];
      }
    }, {
      key: "render",
      value: function render() {
        var _this2 = this;
        return (0, _index.html)(_templateObject_10b17230e51211edbf9f41986c80248c || (_templateObject_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["\n      <button\n        part=\"button\"\n        @click=\"", "\"\n        @keydown=\"", "\"\n        @mouseover=\"", "\"\n        @mouseout=\"", "\"\n        @focusin=\"", "\"\n        @focusout=\"", "\"\n      >\n        ", "\n        ", "\n        <div class=\"label-wrap\">\n          <div class=\"action\" part=\"action\">", "</div>\n          <div class=\"path\" part=\"path\">", "</div>\n        </div>\n        <div class=\"tags\" part=\"tags\">\n          ", "\n        </div>\n        <div class=\"key-combo\" part=\"key-combo\">\n          ", "\n        </div>\n        ", "\n      </button>\n      ", "\n    "])), this.clickEvent, this.keyEvent, this._focusIn, this._focusOut, this._focusIn, this._focusOut, this.icon ? (0, _index.html)(_templateObject2_10b17230e51211edbf9f41986c80248c || (_templateObject2_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<simple-icon-lite\n              icon=\"", "\"\n              class=\"result-icon\"\n            ></simple-icon-lite>"])), this.icon) : "", this.image ? (0, _index.html)(_templateObject3_10b17230e51211edbf9f41986c80248c || (_templateObject3_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<img src=\"", "\" class=\"result-image\" alt=\"\" />"])), this.image) : "", this.title, this.path, this.tags.map(function (tag, i) {
          return (0, _index.html)(_templateObject4_10b17230e51211edbf9f41986c80248c || (_templateObject4_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral([" <simple-tag\n              accent-color=\"", "\"\n              value=\"", "\"\n              part=\"tag tag-", "\"\n            ></simple-tag>"])), _this2.pickColor(i), tag, i);
        }), this.keyAry(this.key).map(function (key) {
          return (0, _index.html)(_templateObject5_10b17230e51211edbf9f41986c80248c || (_templateObject5_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<kbd class=\"keyboard-shortcut\">", "</kbd>"])), key);
        }), this.more ? (0, _index.html)(_templateObject6_10b17230e51211edbf9f41986c80248c || (_templateObject6_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<simple-icon-button\n              class=\"more\"\n              label=\"More details\"\n              icon=\"more-vert\"\n              aria-controls=\"details\"\n              @click=\"", "\"\n              @keydown=\"", "\"\n            ></simple-icon-button>"])), this.toggleDetailsClick, this.toggleDetailsKey) : (0, _index.html)(_templateObject7_10b17230e51211edbf9f41986c80248c || (_templateObject7_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<div class=\"more\"></div>"]))), this.more ? (0, _index.html)(_templateObject8_10b17230e51211edbf9f41986c80248c || (_templateObject8_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["<details\n            id=\"details\"\n            ?open=\"", "\"\n            @toggle=\"", "\"\n          >\n            <summary>Details</summary>\n            <div>\n              <slot></slot>\n            </div>\n          </details>"])), this.showDetails, this.openChanged) : "");
      }
      // toggle on click but prevent the event from bubbling up
    }, {
      key: "toggleDetailsKey",
      value: function toggleDetailsKey(e) {
        if (e.key === "Enter" || e.key === "Space") {
          this.showDetails = !this.showDetails;
          e.preventDefault();
          e.stopPropagation();
          e.stopImmediatePropagation();
        }
      }
      // toggle on click but prevent the event from bubbling up
    }, {
      key: "toggleDetailsClick",
      value: function toggleDetailsClick(e) {
        this.showDetails = !this.showDetails;
        e.preventDefault();
        e.stopPropagation();
        e.stopImmediatePropagation();
      }
    }, {
      key: "openChanged",
      value: function openChanged(e) {
        this.showDetails = e.target.open;
      }
    }], [{
      key: "tag",
      get: function get() {
        return "super-daemon-row";
      }
    }, {
      key: "properties",
      get: function get() {
        return {
          title: {
            type: String
          },
          path: {
            type: String
          },
          icon: {
            type: String
          },
          mini: {
            type: Boolean,
            reflect: true
          },
          image: {
            type: String
          },
          more: {
            type: Boolean
          },
          showDetails: {
            type: Boolean
          },
          key: {
            type: String
          },
          eventName: {
            type: String,
            attribute: "event-name"
          },
          value: {
            type: Object
          },
          tags: {
            type: Array
          },
          active: {
            type: Boolean,
            reflect: true
          }
        };
      }
    }, {
      key: "styles",
      get: function get() {
        var styles = [];
        if (babelHelpers.get(babelHelpers.getPrototypeOf(SuperDaemonRow), "styles", this)) {
          styles = babelHelpers.get(babelHelpers.getPrototypeOf(SuperDaemonRow), "styles", this);
        }
        return [].concat(babelHelpers.toConsumableArray(styles), [(0, _index.css)(_templateObject9_10b17230e51211edbf9f41986c80248c || (_templateObject9_10b17230e51211edbf9f41986c80248c = babelHelpers.taggedTemplateLiteral(["\n        :host {\n          display: block;\n          margin: 8px 16px;\n        }\n        :host([active]) {\n          cursor: pointer;\n          border-radius: 8px;\n          background-color: var(\n            --super-daemon-row-hover,\n            rgba(0, 100, 200, 0.1)\n          );\n          outline: 2px solid black;\n        }\n        :host([mini]) {\n          margin: 0;\n          --super-daemon-row-label: 18px;\n          --super-daemon-row-path: 10px;\n        }\n        button {\n          display: flex;\n          padding: 16px;\n          width: 100%;\n          border-radius: 0;\n          border: none;\n          align-items: middle;\n          justify-content: space-between;\n          background-color: transparent;\n        }\n        :host([mini]) button {\n          padding: 4px;\n          justify-content: flex-start;\n        }\n        .result-icon {\n          display: inline-flex;\n          --simple-icon-height: var(--super-daemon-row-icon, 50px);\n          --simple-icon-width: var(--super-daemon-row-icon, 50px);\n        }\n        :host([mini]) .result-icon {\n          --simple-icon-height: var(--super-daemon-row-icon, 20px);\n          --simple-icon-width: var(--super-daemon-row-icon, 20px);\n        }\n        .result-image {\n          display: inline-flex;\n          max-height: 64px;\n          max-width: 100px;\n        }\n        :host([mini]) .result-image {\n          max-height: 32px;\n          max-width: 50px;\n        }\n        .label-wrap {\n          min-width: 50%;\n          overflow: hidden;\n          text-align: left;\n        }\n        :host([mini]) .label-wrap {\n          width: 100%;\n          margin-left: 16px;\n        }\n        .tags {\n          width: 30%;\n          line-height: 32px;\n          height: 64px;\n          text-align: left;\n          overflow: hidden;\n        }\n        :host([mini]) .tags {\n          display: none;\n        }\n        .label-wrap .action {\n          font-size: var(--super-daemon-row-label, 32px);\n          font-weight: bold;\n          max-width: 90%;\n          word-break: break-all;\n          overflow: hidden;\n          line-height: var(--super-daemon-row-label, 32px);\n          height: var(--super-daemon-row-label, 32px);\n        }\n        :host([mini]) .label-wrap .action {\n          max-width: unset;\n        }\n        .label-wrap .path {\n          font-size: var(--super-daemon-row-path, 20px);\n          font-style: italic;\n          margin-top: 8px;\n          overflow-wrap: break-word;\n          word-break: break-all;\n          overflow: hidden;\n          max-width: 80%;\n          line-height: var(--super-daemon-row-path, 20px);\n          height: var(--super-daemon-row-path, 20px);\n        }\n        :host([mini]) .label-wrap .path {\n          max-width: unset;\n        }\n        .key-combo {\n          height: 64px;\n        }\n        :host([mini]) .key-combo {\n          display: none;\n        }\n        .keyboard-shortcut {\n          background-color: rgba(0, 0, 0, 0.1);\n          border-radius: 4px;\n          color: rgba(0, 0, 0, 0.7);\n          box-shadow: rgb(209 213 219) 0px -4px 0px inset,\n            rgb(0 0 0 / 40%) 0px 1px 1px;\n          display: block;\n          letter-spacing: 0px;\n          font-family: \"Press Start 2P\", \"Trebuchet MS\", \"Lucida Sans Unicode\",\n            \"Lucida Grande\", \"Lucida Sans\", Arial, sans-serif;\n          padding: 6px 4px;\n          margin: 4px auto;\n          font-size: 9px;\n        }\n        .more {\n          --simple-icon-width: 40px;\n          --simple-icon-height: 40px;\n          width: 40px;\n          height: 40px;\n          display: block;\n        }\n        summary {\n          display: none;\n        }\n        @keyframes details-show {\n          from {\n            opacity: 0;\n            transform: var(--details-translate, translateY(-0.5em));\n          }\n          to {\n            opacity: 1;\n            transform: var(--details-translate, translateY(0));\n          }\n        }\n\n        details[open] {\n          padding: 0 16px 16px 16px;\n          animation: details-show 100ms ease-in-out;\n        }\n      "])))]);
      }
    }]);
    return SuperDaemonRow;
  }(_index.LitElement);
  _exports.SuperDaemonRow = SuperDaemonRow;
  customElements.define(SuperDaemonRow.tag, SuperDaemonRow);
});
